// morgan/index.js
/*!
 * morgan
 * Copyright(c) 2010 Sencha Inc.
 * Copyright(c) 2011 TJ Holowaychuk
 * Copyright(c) 2014 Jonathan Ong
 * Copyright(c) 2014-2017 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module exports.
 * @public
 */

module.exports = morgan
module.exports.compile = compile
module.exports.format = format
module.exports.token = token

/**
 * Module dependencies.
 * @private
 */

var auth = require('basic-auth')
var debug = require('debug')('morgan')
var deprecate = require('depd')('morgan')
var onFinished = require('on-finished')
var onHeaders = require('on-headers')

/**
 * Array of CLF month names.
 * @private
 */

var CLF_MONTH = [
  'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
  'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'
]

/**
 * Default log buffer duration.
 * @private
 */

var DEFAULT_BUFFER_DURATION = 1000

/**
 * Create a logger middleware.
 *
 * @public
 * @param ***REMOVED***String|Function***REMOVED*** format
 * @param ***REMOVED***Object***REMOVED*** [options]
 * @return ***REMOVED***Function***REMOVED*** middleware
 */

function morgan (format, options) ***REMOVED***
  var fmt = format
  var opts = options || ***REMOVED******REMOVED***

  if (format && typeof format === 'object') ***REMOVED***
    opts = format
    fmt = opts.format || 'default'

    // smart deprecation message
    deprecate('morgan(options): use morgan(' + (typeof fmt === 'string' ? JSON.stringify(fmt) : 'format') + ', options) instead')
  ***REMOVED***

  if (fmt === undefined) ***REMOVED***
    deprecate('undefined format: specify a format')
  ***REMOVED***

  // output on request instead of response
  var immediate = opts.immediate

  // check if log entry should be skipped
  var skip = opts.skip || false

  // format function
  var formatLine = typeof fmt !== 'function'
    ? getFormatFunction(fmt)
    : fmt

  // stream
  var buffer = opts.buffer
  var stream = opts.stream || process.stdout

  // buffering support
  if (buffer) ***REMOVED***
    deprecate('buffer option')

    // flush interval
    var interval = typeof buffer !== 'number'
      ? DEFAULT_BUFFER_DURATION
      : buffer

    // swap the stream
    stream = createBufferStream(stream, interval)
  ***REMOVED***

  return function logger (req, res, next) ***REMOVED***
    // request data
    req._startAt = undefined
    req._startTime = undefined
    req._remoteAddress = getip(req)

    // response data
    res._startAt = undefined
    res._startTime = undefined

    // record request start
    recordStartTime.call(req)

    function logRequest () ***REMOVED***
      if (skip !== false && skip(req, res)) ***REMOVED***
        debug('skip request')
        return
      ***REMOVED***

      var line = formatLine(morgan, req, res)

      if (line == null) ***REMOVED***
        debug('skip line')
        return
      ***REMOVED***

      debug('log request')
      stream.write(line + '\n')
    ***REMOVED***;

    if (immediate) ***REMOVED***
      // immediate log
      logRequest()
    ***REMOVED*** else ***REMOVED***
      // record response start
      onHeaders(res, recordStartTime)

      // log when response finished
      onFinished(res, logRequest)
    ***REMOVED***

    next()
  ***REMOVED***
***REMOVED***

/**
 * Apache combined log format.
 */

morgan.format('combined', ':remote-addr - :remote-user [:date[clf]] ":method :url HTTP/:http-version" :status :res[content-length] ":referrer" ":user-agent"')

/**
 * Apache common log format.
 */

morgan.format('common', ':remote-addr - :remote-user [:date[clf]] ":method :url HTTP/:http-version" :status :res[content-length]')

/**
 * Default format.
 */

morgan.format('default', ':remote-addr - :remote-user [:date] ":method :url HTTP/:http-version" :status :res[content-length] ":referrer" ":user-agent"')
deprecate.property(morgan, 'default', 'default format: use combined format')

/**
 * Short format.
 */

morgan.format('short', ':remote-addr :remote-user :method :url HTTP/:http-version :status :res[content-length] - :response-time ms')

/**
 * Tiny format.
 */

morgan.format('tiny', ':method :url :status :res[content-length] - :response-time ms')

/**
 * dev (colored)
 */

morgan.format('dev', function developmentFormatLine (tokens, req, res) ***REMOVED***
  // get the status code if response written
  var status = headersSent(res)
    ? res.statusCode
    : undefined

  // get status color
  var color = status >= 500 ? 31 // red
    : status >= 400 ? 33 // yellow
    : status >= 300 ? 36 // cyan
    : status >= 200 ? 32 // green
    : 0 // no color

  // get colored function
  var fn = developmentFormatLine[color]

  if (!fn) ***REMOVED***
    // compile
    fn = developmentFormatLine[color] = compile('\x1b[0m:method :url \x1b[' +
      color + 'm:status \x1b[0m:response-time ms - :res[content-length]\x1b[0m')
  ***REMOVED***

  return fn(tokens, req, res)
***REMOVED***)

/**
 * request url
 */

morgan.token('url', function getUrlToken (req) ***REMOVED***
  return req.originalUrl || req.url
***REMOVED***)

/**
 * request method
 */

morgan.token('method', function getMethodToken (req) ***REMOVED***
  return req.method
***REMOVED***)

/**
 * response time in milliseconds
 */

morgan.token('response-time', function getResponseTimeToken (req, res, digits) ***REMOVED***
  if (!req._startAt || !res._startAt) ***REMOVED***
    // missing request and/or response start time
    return
  ***REMOVED***

  // calculate diff
  var ms = (res._startAt[0] - req._startAt[0]) * 1e3 +
    (res._startAt[1] - req._startAt[1]) * 1e-6

  // return truncated value
  return ms.toFixed(digits === undefined ? 3 : digits)
***REMOVED***)

/**
 * current date
 */

morgan.token('date', function getDateToken (req, res, format) ***REMOVED***
  var date = new Date()

  switch (format || 'web') ***REMOVED***
    case 'clf':
      return clfdate(date)
    case 'iso':
      return date.toISOString()
    case 'web':
      return date.toUTCString()
  ***REMOVED***
***REMOVED***)

/**
 * response status code
 */

morgan.token('status', function getStatusToken (req, res) ***REMOVED***
  return headersSent(res)
    ? String(res.statusCode)
    : undefined
***REMOVED***)

/**
 * normalized referrer
 */

morgan.token('referrer', function getReferrerToken (req) ***REMOVED***
  return req.headers['referer'] || req.headers['referrer']
***REMOVED***)

/**
 * remote address
 */

morgan.token('remote-addr', getip)

/**
 * remote user
 */

morgan.token('remote-user', function getRemoteUserToken (req) ***REMOVED***
  // parse basic credentials
  var credentials = auth(req)

  // return username
  return credentials
    ? credentials.name
    : undefined
***REMOVED***)

/**
 * HTTP version
 */

morgan.token('http-version', function getHttpVersionToken (req) ***REMOVED***
  return req.httpVersionMajor + '.' + req.httpVersionMinor
***REMOVED***)

/**
 * UA string
 */

morgan.token('user-agent', function getUserAgentToken (req) ***REMOVED***
  return req.headers['user-agent']
***REMOVED***)

/**
 * request header
 */

morgan.token('req', function getRequestToken (req, res, field) ***REMOVED***
  // get header
  var header = req.headers[field.toLowerCase()]

  return Array.isArray(header)
    ? header.join(', ')
    : header
***REMOVED***)

/**
 * response header
 */

morgan.token('res', function getResponseHeader (req, res, field) ***REMOVED***
  if (!headersSent(res)) ***REMOVED***
    return undefined
  ***REMOVED***

  // get header
  var header = res.getHeader(field)

  return Array.isArray(header)
    ? header.join(', ')
    : header
***REMOVED***)

/**
 * Format a Date in the common log format.
 *
 * @private
 * @param ***REMOVED***Date***REMOVED*** dateTime
 * @return ***REMOVED***string***REMOVED***
 */

function clfdate (dateTime) ***REMOVED***
  var date = dateTime.getUTCDate()
  var hour = dateTime.getUTCHours()
  var mins = dateTime.getUTCMinutes()
  var secs = dateTime.getUTCSeconds()
  var year = dateTime.getUTCFullYear()

  var month = CLF_MONTH[dateTime.getUTCMonth()]

  return pad2(date) + '/' + month + '/' + year +
    ':' + pad2(hour) + ':' + pad2(mins) + ':' + pad2(secs) +
    ' +0000'
***REMOVED***

/**
 * Compile a format string into a function.
 *
 * @param ***REMOVED***string***REMOVED*** format
 * @return ***REMOVED***function***REMOVED***
 * @public
 */

function compile (format) ***REMOVED***
  if (typeof format !== 'string') ***REMOVED***
    throw new TypeError('argument format must be a string')
  ***REMOVED***

  var fmt = format.replace(/"/g, '\\"')
  var js = '  "use strict"\n  return "' + fmt.replace(/:([-\w]***REMOVED***2,***REMOVED***)(?:\[([^\]]+)\])?/g, function (_, name, arg) ***REMOVED***
    var tokenArguments = 'req, res'
    var tokenFunction = 'tokens[' + String(JSON.stringify(name)) + ']'

    if (arg !== undefined) ***REMOVED***
      tokenArguments += ', ' + String(JSON.stringify(arg))
    ***REMOVED***

    return '" +\n    (' + tokenFunction + '(' + tokenArguments + ') || "-") + "'
  ***REMOVED***) + '"'

  // eslint-disable-next-line no-new-func
  return new Function('tokens, req, res', js)
***REMOVED***

/**
 * Create a basic buffering stream.
 *
 * @param ***REMOVED***object***REMOVED*** stream
 * @param ***REMOVED***number***REMOVED*** interval
 * @public
 */

function createBufferStream (stream, interval) ***REMOVED***
  var buf = []
  var timer = null

  // flush function
  function flush () ***REMOVED***
    timer = null
    stream.write(buf.join(''))
    buf.length = 0
  ***REMOVED***

  // write function
  function write (str) ***REMOVED***
    if (timer === null) ***REMOVED***
      timer = setTimeout(flush, interval)
    ***REMOVED***

    buf.push(str)
  ***REMOVED***

  // return a minimal "stream"
  return ***REMOVED*** write: write ***REMOVED***
***REMOVED***

/**
 * Define a format with the given name.
 *
 * @param ***REMOVED***string***REMOVED*** name
 * @param ***REMOVED***string|function***REMOVED*** fmt
 * @public
 */

function format (name, fmt) ***REMOVED***
  morgan[name] = fmt
  return this
***REMOVED***

/**
 * Lookup and compile a named format function.
 *
 * @param ***REMOVED***string***REMOVED*** name
 * @return ***REMOVED***function***REMOVED***
 * @public
 */

function getFormatFunction (name) ***REMOVED***
  // lookup format
  var fmt = morgan[name] || name || morgan.default

  // return compiled format
  return typeof fmt !== 'function'
    ? compile(fmt)
    : fmt
***REMOVED***

/**
 * Get request IP address.
 *
 * @private
 * @param ***REMOVED***IncomingMessage***REMOVED*** req
 * @return ***REMOVED***string***REMOVED***
 */

function getip (req) ***REMOVED***
  return req.ip ||
    req._remoteAddress ||
    (req.connection && req.connection.remoteAddress) ||
    undefined
***REMOVED***

/**
 * Determine if the response headers have been sent.
 *
 * @param ***REMOVED***object***REMOVED*** res
 * @returns ***REMOVED***boolean***REMOVED***
 * @private
 */

function headersSent (res) ***REMOVED***
  return typeof res.headersSent !== 'boolean'
    ? Boolean(res._header)
    : res.headersSent
***REMOVED***

/**
 * Pad number to two digits.
 *
 * @private
 * @param ***REMOVED***number***REMOVED*** num
 * @return ***REMOVED***string***REMOVED***
 */

function pad2 (num) ***REMOVED***
  var str = String(num)

  return (str.length === 1 ? '0' : '') + str
***REMOVED***

/**
 * Record the start time.
 * @private
 */

function recordStartTime () ***REMOVED***
  this._startAt = process.hrtime()
  this._startTime = new Date()
***REMOVED***

/**
 * Define a token function with the given name,
 * and callback fn(req, res).
 *
 * @param ***REMOVED***string***REMOVED*** name
 * @param ***REMOVED***function***REMOVED*** fn
 * @public
 */

function token (name, fn) ***REMOVED***
  morgan[name] = fn
  return this
***REMOVED***
