// request/request.js
'use strict'

var http = require('http')
var https = require('https')
var url = require('url')
var util = require('util')
var stream = require('stream')
var zlib = require('zlib')
var hawk = require('hawk')
var aws2 = require('aws-sign2')
var aws4 = require('aws4')
var httpSignature = require('http-signature')
var mime = require('mime-types')
var caseless = require('caseless')
var ForeverAgent = require('forever-agent')
var FormData = require('form-data')
var extend = require('extend')
var isstream = require('isstream')
var isTypedArray = require('is-typedarray').strict
var helpers = require('./lib/helpers')
var cookies = require('./lib/cookies')
var getProxyFromURI = require('./lib/getProxyFromURI')
var Querystring = require('./lib/querystring').Querystring
var Har = require('./lib/har').Har
var Auth = require('./lib/auth').Auth
var OAuth = require('./lib/oauth').OAuth
var Multipart = require('./lib/multipart').Multipart
var Redirect = require('./lib/redirect').Redirect
var Tunnel = require('./lib/tunnel').Tunnel
var now = require('performance-now')
var Buffer = require('safe-buffer').Buffer

var safeStringify = helpers.safeStringify
var isReadStream = helpers.isReadStream
var toBase64 = helpers.toBase64
var defer = helpers.defer
var copy = helpers.copy
var version = helpers.version
var globalCookieJar = cookies.jar()

var globalPool = ***REMOVED******REMOVED***

function filterForNonReserved (reserved, options) ***REMOVED***
  // Filter out properties that are not reserved.
  // Reserved values are passed in at call site.

  var object = ***REMOVED******REMOVED***
  for (var i in options) ***REMOVED***
    var notReserved = (reserved.indexOf(i) === -1)
    if (notReserved) ***REMOVED***
      object[i] = options[i]
    ***REMOVED***
  ***REMOVED***
  return object
***REMOVED***

function filterOutReservedFunctions (reserved, options) ***REMOVED***
  // Filter out properties that are functions and are reserved.
  // Reserved values are passed in at call site.

  var object = ***REMOVED******REMOVED***
  for (var i in options) ***REMOVED***
    var isReserved = !(reserved.indexOf(i) === -1)
    var isFunction = (typeof options[i] === 'function')
    if (!(isReserved && isFunction)) ***REMOVED***
      object[i] = options[i]
    ***REMOVED***
  ***REMOVED***
  return object
***REMOVED***

// Return a simpler request object to allow serialization
function requestToJSON () ***REMOVED***
  var self = this
  return ***REMOVED***
    uri: self.uri,
    method: self.method,
    headers: self.headers
  ***REMOVED***
***REMOVED***

// Return a simpler response object to allow serialization
function responseToJSON () ***REMOVED***
  var self = this
  return ***REMOVED***
    statusCode: self.statusCode,
    body: self.body,
    headers: self.headers,
    request: requestToJSON.call(self.request)
  ***REMOVED***
***REMOVED***

function Request (options) ***REMOVED***
  // if given the method property in options, set property explicitMethod to true

  // extend the Request instance with any non-reserved properties
  // remove any reserved functions from the options object
  // set Request instance to be readable and writable
  // call init

  var self = this

  // start with HAR, then override with additional options
  if (options.har) ***REMOVED***
    self._har = new Har(self)
    options = self._har.options(options)
  ***REMOVED***

  stream.Stream.call(self)
  var reserved = Object.keys(Request.prototype)
  var nonReserved = filterForNonReserved(reserved, options)

  extend(self, nonReserved)
  options = filterOutReservedFunctions(reserved, options)

  self.readable = true
  self.writable = true
  if (options.method) ***REMOVED***
    self.explicitMethod = true
  ***REMOVED***
  self._qs = new Querystring(self)
  self._auth = new Auth(self)
  self._oauth = new OAuth(self)
  self._multipart = new Multipart(self)
  self._redirect = new Redirect(self)
  self._tunnel = new Tunnel(self)
  self.init(options)
***REMOVED***

util.inherits(Request, stream.Stream)

// Debugging
Request.debug = process.env.NODE_DEBUG && /\brequest\b/.test(process.env.NODE_DEBUG)
function debug () ***REMOVED***
  if (Request.debug) ***REMOVED***
    console.error('REQUEST %s', util.format.apply(util, arguments))
  ***REMOVED***
***REMOVED***
Request.prototype.debug = debug

Request.prototype.init = function (options) ***REMOVED***
  // init() contains all the code to setup the request object.
  // the actual outgoing request is not started until start() is called
  // this function is called from both the constructor and on redirect.
  var self = this
  if (!options) ***REMOVED***
    options = ***REMOVED******REMOVED***
  ***REMOVED***
  self.headers = self.headers ? copy(self.headers) : ***REMOVED******REMOVED***

  // Delete headers with value undefined since they break
  // ClientRequest.OutgoingMessage.setHeader in node 0.12
  for (var headerName in self.headers) ***REMOVED***
    if (typeof self.headers[headerName] === 'undefined') ***REMOVED***
      delete self.headers[headerName]
    ***REMOVED***
  ***REMOVED***

  caseless.httpify(self, self.headers)

  if (!self.method) ***REMOVED***
    self.method = options.method || 'GET'
  ***REMOVED***
  if (!self.localAddress) ***REMOVED***
    self.localAddress = options.localAddress
  ***REMOVED***

  self._qs.init(options)

  debug(options)
  if (!self.pool && self.pool !== false) ***REMOVED***
    self.pool = globalPool
  ***REMOVED***
  self.dests = self.dests || []
  self.__isRequestRequest = true

  // Protect against double callback
  if (!self._callback && self.callback) ***REMOVED***
    self._callback = self.callback
    self.callback = function () ***REMOVED***
      if (self._callbackCalled) ***REMOVED***
        return // Print a warning maybe?
      ***REMOVED***
      self._callbackCalled = true
      self._callback.apply(self, arguments)
    ***REMOVED***
    self.on('error', self.callback.bind())
    self.on('complete', self.callback.bind(self, null))
  ***REMOVED***

  // People use this property instead all the time, so support it
  if (!self.uri && self.url) ***REMOVED***
    self.uri = self.url
    delete self.url
  ***REMOVED***

  // If there's a baseUrl, then use it as the base URL (i.e. uri must be
  // specified as a relative path and is appended to baseUrl).
  if (self.baseUrl) ***REMOVED***
    if (typeof self.baseUrl !== 'string') ***REMOVED***
      return self.emit('error', new Error('options.baseUrl must be a string'))
    ***REMOVED***

    if (typeof self.uri !== 'string') ***REMOVED***
      return self.emit('error', new Error('options.uri must be a string when using options.baseUrl'))
    ***REMOVED***

    if (self.uri.indexOf('//') === 0 || self.uri.indexOf('://') !== -1) ***REMOVED***
      return self.emit('error', new Error('options.uri must be a path when using options.baseUrl'))
    ***REMOVED***

    // Handle all cases to make sure that there's only one slash between
    // baseUrl and uri.
    var baseUrlEndsWithSlash = self.baseUrl.lastIndexOf('/') === self.baseUrl.length - 1
    var uriStartsWithSlash = self.uri.indexOf('/') === 0

    if (baseUrlEndsWithSlash && uriStartsWithSlash) ***REMOVED***
      self.uri = self.baseUrl + self.uri.slice(1)
    ***REMOVED*** else if (baseUrlEndsWithSlash || uriStartsWithSlash) ***REMOVED***
      self.uri = self.baseUrl + self.uri
    ***REMOVED*** else if (self.uri === '') ***REMOVED***
      self.uri = self.baseUrl
    ***REMOVED*** else ***REMOVED***
      self.uri = self.baseUrl + '/' + self.uri
    ***REMOVED***
    delete self.baseUrl
  ***REMOVED***

  // A URI is needed by this point, emit error if we haven't been able to get one
  if (!self.uri) ***REMOVED***
    return self.emit('error', new Error('options.uri is a required argument'))
  ***REMOVED***

  // If a string URI/URL was given, parse it into a URL object
  if (typeof self.uri === 'string') ***REMOVED***
    self.uri = url.parse(self.uri)
  ***REMOVED***

  // Some URL objects are not from a URL parsed string and need href added
  if (!self.uri.href) ***REMOVED***
    self.uri.href = url.format(self.uri)
  ***REMOVED***

  // DEPRECATED: Warning for users of the old Unix Sockets URL Scheme
  if (self.uri.protocol === 'unix:') ***REMOVED***
    return self.emit('error', new Error('`unix://` URL scheme is no longer supported. Please use the format `http://unix:SOCKET:PATH`'))
  ***REMOVED***

  // Support Unix Sockets
  if (self.uri.host === 'unix') ***REMOVED***
    self.enableUnixSocket()
  ***REMOVED***

  if (self.strictSSL === false) ***REMOVED***
    self.rejectUnauthorized = false
  ***REMOVED***

  if (!self.uri.pathname) ***REMOVED*** self.uri.pathname = '/' ***REMOVED***

  if (!(self.uri.host || (self.uri.hostname && self.uri.port)) && !self.uri.isUnix) ***REMOVED***
    // Invalid URI: it may generate lot of bad errors, like 'TypeError: Cannot call method `indexOf` of undefined' in CookieJar
    // Detect and reject it as soon as possible
    var faultyUri = url.format(self.uri)
    var message = 'Invalid URI "' + faultyUri + '"'
    if (Object.keys(options).length === 0) ***REMOVED***
      // No option ? This can be the sign of a redirect
      // As this is a case where the user cannot do anything (they didn't call request directly with this URL)
      // they should be warned that it can be caused by a redirection (can save some hair)
      message += '. This can be caused by a crappy redirection.'
    ***REMOVED***
    // This error was fatal
    self.abort()
    return self.emit('error', new Error(message))
  ***REMOVED***

  if (!self.hasOwnProperty('proxy')) ***REMOVED***
    self.proxy = getProxyFromURI(self.uri)
  ***REMOVED***

  self.tunnel = self._tunnel.isEnabled()
  if (self.proxy) ***REMOVED***
    self._tunnel.setup(options)
  ***REMOVED***

  self._redirect.onRequest(options)

  self.setHost = false
  if (!self.hasHeader('host')) ***REMOVED***
    var hostHeaderName = self.originalHostHeaderName || 'host'
    // When used with an IPv6 address, `host` will provide
    // the correct bracketed format, unlike using `hostname` and
    // optionally adding the `port` when necessary.
    self.setHeader(hostHeaderName, self.uri.host)
    self.setHost = true
  ***REMOVED***

  self.jar(self._jar || options.jar)

  if (!self.uri.port) ***REMOVED***
    if (self.uri.protocol === 'http:') ***REMOVED*** self.uri.port = 80 ***REMOVED*** else if (self.uri.protocol === 'https:') ***REMOVED*** self.uri.port = 443 ***REMOVED***
  ***REMOVED***

  if (self.proxy && !self.tunnel) ***REMOVED***
    self.port = self.proxy.port
    self.host = self.proxy.hostname
  ***REMOVED*** else ***REMOVED***
    self.port = self.uri.port
    self.host = self.uri.hostname
  ***REMOVED***

  if (options.form) ***REMOVED***
    self.form(options.form)
  ***REMOVED***

  if (options.formData) ***REMOVED***
    var formData = options.formData
    var requestForm = self.form()
    var appendFormValue = function (key, value) ***REMOVED***
      if (value && value.hasOwnProperty('value') && value.hasOwnProperty('options')) ***REMOVED***
        requestForm.append(key, value.value, value.options)
      ***REMOVED*** else ***REMOVED***
        requestForm.append(key, value)
      ***REMOVED***
    ***REMOVED***
    for (var formKey in formData) ***REMOVED***
      if (formData.hasOwnProperty(formKey)) ***REMOVED***
        var formValue = formData[formKey]
        if (formValue instanceof Array) ***REMOVED***
          for (var j = 0; j < formValue.length; j++) ***REMOVED***
            appendFormValue(formKey, formValue[j])
          ***REMOVED***
        ***REMOVED*** else ***REMOVED***
          appendFormValue(formKey, formValue)
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  if (options.qs) ***REMOVED***
    self.qs(options.qs)
  ***REMOVED***

  if (self.uri.path) ***REMOVED***
    self.path = self.uri.path
  ***REMOVED*** else ***REMOVED***
    self.path = self.uri.pathname + (self.uri.search || '')
  ***REMOVED***

  if (self.path.length === 0) ***REMOVED***
    self.path = '/'
  ***REMOVED***

  // Auth must happen last in case signing is dependent on other headers
  if (options.aws) ***REMOVED***
    self.aws(options.aws)
  ***REMOVED***

  if (options.hawk) ***REMOVED***
    self.hawk(options.hawk)
  ***REMOVED***

  if (options.httpSignature) ***REMOVED***
    self.httpSignature(options.httpSignature)
  ***REMOVED***

  if (options.auth) ***REMOVED***
    if (Object.prototype.hasOwnProperty.call(options.auth, 'username')) ***REMOVED***
      options.auth.user = options.auth.username
    ***REMOVED***
    if (Object.prototype.hasOwnProperty.call(options.auth, 'password')) ***REMOVED***
      options.auth.pass = options.auth.password
    ***REMOVED***

    self.auth(
      options.auth.user,
      options.auth.pass,
      options.auth.sendImmediately,
      options.auth.bearer
    )
  ***REMOVED***

  if (self.gzip && !self.hasHeader('accept-encoding')) ***REMOVED***
    self.setHeader('accept-encoding', 'gzip, deflate')
  ***REMOVED***

  if (self.uri.auth && !self.hasHeader('authorization')) ***REMOVED***
    var uriAuthPieces = self.uri.auth.split(':').map(function (item) ***REMOVED*** return self._qs.unescape(item) ***REMOVED***)
    self.auth(uriAuthPieces[0], uriAuthPieces.slice(1).join(':'), true)
  ***REMOVED***

  if (!self.tunnel && self.proxy && self.proxy.auth && !self.hasHeader('proxy-authorization')) ***REMOVED***
    var proxyAuthPieces = self.proxy.auth.split(':').map(function (item) ***REMOVED*** return self._qs.unescape(item) ***REMOVED***)
    var authHeader = 'Basic ' + toBase64(proxyAuthPieces.join(':'))
    self.setHeader('proxy-authorization', authHeader)
  ***REMOVED***

  if (self.proxy && !self.tunnel) ***REMOVED***
    self.path = (self.uri.protocol + '//' + self.uri.host + self.path)
  ***REMOVED***

  if (options.json) ***REMOVED***
    self.json(options.json)
  ***REMOVED***
  if (options.multipart) ***REMOVED***
    self.multipart(options.multipart)
  ***REMOVED***

  if (options.time) ***REMOVED***
    self.timing = true

    // NOTE: elapsedTime is deprecated in favor of .timings
    self.elapsedTime = self.elapsedTime || 0
  ***REMOVED***

  function setContentLength () ***REMOVED***
    if (isTypedArray(self.body)) ***REMOVED***
      self.body = Buffer.from(self.body)
    ***REMOVED***

    if (!self.hasHeader('content-length')) ***REMOVED***
      var length
      if (typeof self.body === 'string') ***REMOVED***
        length = Buffer.byteLength(self.body)
      ***REMOVED*** else if (Array.isArray(self.body)) ***REMOVED***
        length = self.body.reduce(function (a, b) ***REMOVED*** return a + b.length ***REMOVED***, 0)
      ***REMOVED*** else ***REMOVED***
        length = self.body.length
      ***REMOVED***

      if (length) ***REMOVED***
        self.setHeader('content-length', length)
      ***REMOVED*** else ***REMOVED***
        self.emit('error', new Error('Argument error, options.body.'))
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  if (self.body && !isstream(self.body)) ***REMOVED***
    setContentLength()
  ***REMOVED***

  if (options.oauth) ***REMOVED***
    self.oauth(options.oauth)
  ***REMOVED*** else if (self._oauth.params && self.hasHeader('authorization')) ***REMOVED***
    self.oauth(self._oauth.params)
  ***REMOVED***

  var protocol = self.proxy && !self.tunnel ? self.proxy.protocol : self.uri.protocol
  var defaultModules = ***REMOVED***'http:': http, 'https:': https***REMOVED***
  var httpModules = self.httpModules || ***REMOVED******REMOVED***

  self.httpModule = httpModules[protocol] || defaultModules[protocol]

  if (!self.httpModule) ***REMOVED***
    return self.emit('error', new Error('Invalid protocol: ' + protocol))
  ***REMOVED***

  if (options.ca) ***REMOVED***
    self.ca = options.ca
  ***REMOVED***

  if (!self.agent) ***REMOVED***
    if (options.agentOptions) ***REMOVED***
      self.agentOptions = options.agentOptions
    ***REMOVED***

    if (options.agentClass) ***REMOVED***
      self.agentClass = options.agentClass
    ***REMOVED*** else if (options.forever) ***REMOVED***
      var v = version()
      // use ForeverAgent in node 0.10- only
      if (v.major === 0 && v.minor <= 10) ***REMOVED***
        self.agentClass = protocol === 'http:' ? ForeverAgent : ForeverAgent.SSL
      ***REMOVED*** else ***REMOVED***
        self.agentClass = self.httpModule.Agent
        self.agentOptions = self.agentOptions || ***REMOVED******REMOVED***
        self.agentOptions.keepAlive = true
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      self.agentClass = self.httpModule.Agent
    ***REMOVED***
  ***REMOVED***

  if (self.pool === false) ***REMOVED***
    self.agent = false
  ***REMOVED*** else ***REMOVED***
    self.agent = self.agent || self.getNewAgent()
  ***REMOVED***

  self.on('pipe', function (src) ***REMOVED***
    if (self.ntick && self._started) ***REMOVED***
      self.emit('error', new Error('You cannot pipe to this stream after the outbound request has started.'))
    ***REMOVED***
    self.src = src
    if (isReadStream(src)) ***REMOVED***
      if (!self.hasHeader('content-type')) ***REMOVED***
        self.setHeader('content-type', mime.lookup(src.path))
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      if (src.headers) ***REMOVED***
        for (var i in src.headers) ***REMOVED***
          if (!self.hasHeader(i)) ***REMOVED***
            self.setHeader(i, src.headers[i])
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      if (self._json && !self.hasHeader('content-type')) ***REMOVED***
        self.setHeader('content-type', 'application/json')
      ***REMOVED***
      if (src.method && !self.explicitMethod) ***REMOVED***
        self.method = src.method
      ***REMOVED***
    ***REMOVED***

  // self.on('pipe', function () ***REMOVED***
  //   console.error('You have already piped to this stream. Pipeing twice is likely to break the request.')
  // ***REMOVED***)
  ***REMOVED***)

  defer(function () ***REMOVED***
    if (self._aborted) ***REMOVED***
      return
    ***REMOVED***

    var end = function () ***REMOVED***
      if (self._form) ***REMOVED***
        if (!self._auth.hasAuth) ***REMOVED***
          self._form.pipe(self)
        ***REMOVED*** else if (self._auth.hasAuth && self._auth.sentAuth) ***REMOVED***
          self._form.pipe(self)
        ***REMOVED***
      ***REMOVED***
      if (self._multipart && self._multipart.chunked) ***REMOVED***
        self._multipart.body.pipe(self)
      ***REMOVED***
      if (self.body) ***REMOVED***
        if (isstream(self.body)) ***REMOVED***
          self.body.pipe(self)
        ***REMOVED*** else ***REMOVED***
          setContentLength()
          if (Array.isArray(self.body)) ***REMOVED***
            self.body.forEach(function (part) ***REMOVED***
              self.write(part)
            ***REMOVED***)
          ***REMOVED*** else ***REMOVED***
            self.write(self.body)
          ***REMOVED***
          self.end()
        ***REMOVED***
      ***REMOVED*** else if (self.requestBodyStream) ***REMOVED***
        console.warn('options.requestBodyStream is deprecated, please pass the request object to stream.pipe.')
        self.requestBodyStream.pipe(self)
      ***REMOVED*** else if (!self.src) ***REMOVED***
        if (self._auth.hasAuth && !self._auth.sentAuth) ***REMOVED***
          self.end()
          return
        ***REMOVED***
        if (self.method !== 'GET' && typeof self.method !== 'undefined') ***REMOVED***
          self.setHeader('content-length', 0)
        ***REMOVED***
        self.end()
      ***REMOVED***
    ***REMOVED***

    if (self._form && !self.hasHeader('content-length')) ***REMOVED***
      // Before ending the request, we had to compute the length of the whole form, asyncly
      self.setHeader(self._form.getHeaders(), true)
      self._form.getLength(function (err, length) ***REMOVED***
        if (!err && !isNaN(length)) ***REMOVED***
          self.setHeader('content-length', length)
        ***REMOVED***
        end()
      ***REMOVED***)
    ***REMOVED*** else ***REMOVED***
      end()
    ***REMOVED***

    self.ntick = true
  ***REMOVED***)
***REMOVED***

Request.prototype.getNewAgent = function () ***REMOVED***
  var self = this
  var Agent = self.agentClass
  var options = ***REMOVED******REMOVED***
  if (self.agentOptions) ***REMOVED***
    for (var i in self.agentOptions) ***REMOVED***
      options[i] = self.agentOptions[i]
    ***REMOVED***
  ***REMOVED***
  if (self.ca) ***REMOVED***
    options.ca = self.ca
  ***REMOVED***
  if (self.ciphers) ***REMOVED***
    options.ciphers = self.ciphers
  ***REMOVED***
  if (self.secureProtocol) ***REMOVED***
    options.secureProtocol = self.secureProtocol
  ***REMOVED***
  if (self.secureOptions) ***REMOVED***
    options.secureOptions = self.secureOptions
  ***REMOVED***
  if (typeof self.rejectUnauthorized !== 'undefined') ***REMOVED***
    options.rejectUnauthorized = self.rejectUnauthorized
  ***REMOVED***

  if (self.cert && self.key) ***REMOVED***
    options.key = self.key
    options.cert = self.cert
  ***REMOVED***

  if (self.pfx) ***REMOVED***
    options.pfx = self.pfx
  ***REMOVED***

  if (self.passphrase) ***REMOVED***
    options.passphrase = self.passphrase
  ***REMOVED***

  var poolKey = ''

  // different types of agents are in different pools
  if (Agent !== self.httpModule.Agent) ***REMOVED***
    poolKey += Agent.name
  ***REMOVED***

  // ca option is only relevant if proxy or destination are https
  var proxy = self.proxy
  if (typeof proxy === 'string') ***REMOVED***
    proxy = url.parse(proxy)
  ***REMOVED***
  var isHttps = (proxy && proxy.protocol === 'https:') || this.uri.protocol === 'https:'

  if (isHttps) ***REMOVED***
    if (options.ca) ***REMOVED***
      if (poolKey) ***REMOVED***
        poolKey += ':'
      ***REMOVED***
      poolKey += options.ca
    ***REMOVED***

    if (typeof options.rejectUnauthorized !== 'undefined') ***REMOVED***
      if (poolKey) ***REMOVED***
        poolKey += ':'
      ***REMOVED***
      poolKey += options.rejectUnauthorized
    ***REMOVED***

    if (options.cert) ***REMOVED***
      if (poolKey) ***REMOVED***
        poolKey += ':'
      ***REMOVED***
      poolKey += options.cert.toString('ascii') + options.key.toString('ascii')
    ***REMOVED***

    if (options.pfx) ***REMOVED***
      if (poolKey) ***REMOVED***
        poolKey += ':'
      ***REMOVED***
      poolKey += options.pfx.toString('ascii')
    ***REMOVED***

    if (options.ciphers) ***REMOVED***
      if (poolKey) ***REMOVED***
        poolKey += ':'
      ***REMOVED***
      poolKey += options.ciphers
    ***REMOVED***

    if (options.secureProtocol) ***REMOVED***
      if (poolKey) ***REMOVED***
        poolKey += ':'
      ***REMOVED***
      poolKey += options.secureProtocol
    ***REMOVED***

    if (options.secureOptions) ***REMOVED***
      if (poolKey) ***REMOVED***
        poolKey += ':'
      ***REMOVED***
      poolKey += options.secureOptions
    ***REMOVED***
  ***REMOVED***

  if (self.pool === globalPool && !poolKey && Object.keys(options).length === 0 && self.httpModule.globalAgent) ***REMOVED***
    // not doing anything special.  Use the globalAgent
    return self.httpModule.globalAgent
  ***REMOVED***

  // we're using a stored agent.  Make sure it's protocol-specific
  poolKey = self.uri.protocol + poolKey

  // generate a new agent for this setting if none yet exists
  if (!self.pool[poolKey]) ***REMOVED***
    self.pool[poolKey] = new Agent(options)
    // properly set maxSockets on new agents
    if (self.pool.maxSockets) ***REMOVED***
      self.pool[poolKey].maxSockets = self.pool.maxSockets
    ***REMOVED***
  ***REMOVED***

  return self.pool[poolKey]
***REMOVED***

Request.prototype.start = function () ***REMOVED***
  // start() is called once we are ready to send the outgoing HTTP request.
  // this is usually called on the first write(), end() or on nextTick()
  var self = this

  if (self.timing) ***REMOVED***
    // All timings will be relative to this request's startTime.  In order to do this,
    // we need to capture the wall-clock start time (via Date), immediately followed
    // by the high-resolution timer (via now()).  While these two won't be set
    // at the _exact_ same time, they should be close enough to be able to calculate
    // high-resolution, monotonically non-decreasing timestamps relative to startTime.
    var startTime = new Date().getTime()
    var startTimeNow = now()
  ***REMOVED***

  if (self._aborted) ***REMOVED***
    return
  ***REMOVED***

  self._started = true
  self.method = self.method || 'GET'
  self.href = self.uri.href

  if (self.src && self.src.stat && self.src.stat.size && !self.hasHeader('content-length')) ***REMOVED***
    self.setHeader('content-length', self.src.stat.size)
  ***REMOVED***
  if (self._aws) ***REMOVED***
    self.aws(self._aws, true)
  ***REMOVED***

  // We have a method named auth, which is completely different from the http.request
  // auth option.  If we don't remove it, we're gonna have a bad time.
  var reqOptions = copy(self)
  delete reqOptions.auth

  debug('make request', self.uri.href)

  // node v6.8.0 now supports a `timeout` value in `http.request()`, but we
  // should delete it for now since we handle timeouts manually for better
  // consistency with node versions before v6.8.0
  delete reqOptions.timeout

  try ***REMOVED***
    self.req = self.httpModule.request(reqOptions)
  ***REMOVED*** catch (err) ***REMOVED***
    self.emit('error', err)
    return
  ***REMOVED***

  if (self.timing) ***REMOVED***
    self.startTime = startTime
    self.startTimeNow = startTimeNow

    // Timing values will all be relative to startTime (by comparing to startTimeNow
    // so we have an accurate clock)
    self.timings = ***REMOVED******REMOVED***
  ***REMOVED***

  var timeout
  if (self.timeout && !self.timeoutTimer) ***REMOVED***
    if (self.timeout < 0) ***REMOVED***
      timeout = 0
    ***REMOVED*** else if (typeof self.timeout === 'number' && isFinite(self.timeout)) ***REMOVED***
      timeout = self.timeout
    ***REMOVED***
  ***REMOVED***

  self.req.on('response', self.onRequestResponse.bind(self))
  self.req.on('error', self.onRequestError.bind(self))
  self.req.on('drain', function () ***REMOVED***
    self.emit('drain')
  ***REMOVED***)

  self.req.on('socket', function (socket) ***REMOVED***
    // `._connecting` was the old property which was made public in node v6.1.0
    var isConnecting = socket._connecting || socket.connecting
    if (self.timing) ***REMOVED***
      self.timings.socket = now() - self.startTimeNow

      if (isConnecting) ***REMOVED***
        var onLookupTiming = function () ***REMOVED***
          self.timings.lookup = now() - self.startTimeNow
        ***REMOVED***

        var onConnectTiming = function () ***REMOVED***
          self.timings.connect = now() - self.startTimeNow
        ***REMOVED***

        socket.once('lookup', onLookupTiming)
        socket.once('connect', onConnectTiming)

        // clean up timing event listeners if needed on error
        self.req.once('error', function () ***REMOVED***
          socket.removeListener('lookup', onLookupTiming)
          socket.removeListener('connect', onConnectTiming)
        ***REMOVED***)
      ***REMOVED***
    ***REMOVED***

    var setReqTimeout = function () ***REMOVED***
      // This timeout sets the amount of time to wait *between* bytes sent
      // from the server once connected.
      //
      // In particular, it's useful for erroring if the server fails to send
      // data halfway through streaming a response.
      self.req.setTimeout(timeout, function () ***REMOVED***
        if (self.req) ***REMOVED***
          self.abort()
          var e = new Error('ESOCKETTIMEDOUT')
          e.code = 'ESOCKETTIMEDOUT'
          e.connect = false
          self.emit('error', e)
        ***REMOVED***
      ***REMOVED***)
    ***REMOVED***
    if (timeout !== undefined) ***REMOVED***
      // Only start the connection timer if we're actually connecting a new
      // socket, otherwise if we're already connected (because this is a
      // keep-alive connection) do not bother. This is important since we won't
      // get a 'connect' event for an already connected socket.
      if (isConnecting) ***REMOVED***
        var onReqSockConnect = function () ***REMOVED***
          socket.removeListener('connect', onReqSockConnect)
          clearTimeout(self.timeoutTimer)
          self.timeoutTimer = null
          setReqTimeout()
        ***REMOVED***

        socket.on('connect', onReqSockConnect)

        self.req.on('error', function (err) ***REMOVED*** // eslint-disable-line handle-callback-err
          socket.removeListener('connect', onReqSockConnect)
        ***REMOVED***)

        // Set a timeout in memory - this block will throw if the server takes more
        // than `timeout` to write the HTTP status and headers (corresponding to
        // the on('response') event on the client). NB: this measures wall-clock
        // time, not the time between bytes sent by the server.
        self.timeoutTimer = setTimeout(function () ***REMOVED***
          socket.removeListener('connect', onReqSockConnect)
          self.abort()
          var e = new Error('ETIMEDOUT')
          e.code = 'ETIMEDOUT'
          e.connect = true
          self.emit('error', e)
        ***REMOVED***, timeout)
      ***REMOVED*** else ***REMOVED***
        // We're already connected
        setReqTimeout()
      ***REMOVED***
    ***REMOVED***
    self.emit('socket', socket)
  ***REMOVED***)

  self.emit('request', self.req)
***REMOVED***

Request.prototype.onRequestError = function (error) ***REMOVED***
  var self = this
  if (self._aborted) ***REMOVED***
    return
  ***REMOVED***
  if (self.req && self.req._reusedSocket && error.code === 'ECONNRESET' &&
    self.agent.addRequestNoreuse) ***REMOVED***
    self.agent = ***REMOVED*** addRequest: self.agent.addRequestNoreuse.bind(self.agent) ***REMOVED***
    self.start()
    self.req.end()
    return
  ***REMOVED***
  if (self.timeout && self.timeoutTimer) ***REMOVED***
    clearTimeout(self.timeoutTimer)
    self.timeoutTimer = null
  ***REMOVED***
  self.emit('error', error)
***REMOVED***

Request.prototype.onRequestResponse = function (response) ***REMOVED***
  var self = this

  if (self.timing) ***REMOVED***
    self.timings.response = now() - self.startTimeNow
  ***REMOVED***

  debug('onRequestResponse', self.uri.href, response.statusCode, response.headers)
  response.on('end', function () ***REMOVED***
    if (self.timing) ***REMOVED***
      self.timings.end = now() - self.startTimeNow
      response.timingStart = self.startTime

      // fill in the blanks for any periods that didn't trigger, such as
      // no lookup or connect due to keep alive
      if (!self.timings.socket) ***REMOVED***
        self.timings.socket = 0
      ***REMOVED***
      if (!self.timings.lookup) ***REMOVED***
        self.timings.lookup = self.timings.socket
      ***REMOVED***
      if (!self.timings.connect) ***REMOVED***
        self.timings.connect = self.timings.lookup
      ***REMOVED***
      if (!self.timings.response) ***REMOVED***
        self.timings.response = self.timings.connect
      ***REMOVED***

      debug('elapsed time', self.timings.end)

      // elapsedTime includes all redirects
      self.elapsedTime += Math.round(self.timings.end)

      // NOTE: elapsedTime is deprecated in favor of .timings
      response.elapsedTime = self.elapsedTime

      // timings is just for the final fetch
      response.timings = self.timings

      // pre-calculate phase timings as well
      response.timingPhases = ***REMOVED***
        wait: self.timings.socket,
        dns: self.timings.lookup - self.timings.socket,
        tcp: self.timings.connect - self.timings.lookup,
        firstByte: self.timings.response - self.timings.connect,
        download: self.timings.end - self.timings.response,
        total: self.timings.end
      ***REMOVED***
    ***REMOVED***
    debug('response end', self.uri.href, response.statusCode, response.headers)
  ***REMOVED***)

  if (self._aborted) ***REMOVED***
    debug('aborted', self.uri.href)
    response.resume()
    return
  ***REMOVED***

  self.response = response
  response.request = self
  response.toJSON = responseToJSON

  // XXX This is different on 0.10, because SSL is strict by default
  if (self.httpModule === https &&
    self.strictSSL && (!response.hasOwnProperty('socket') ||
    !response.socket.authorized)) ***REMOVED***
    debug('strict ssl error', self.uri.href)
    var sslErr = response.hasOwnProperty('socket') ? response.socket.authorizationError : self.uri.href + ' does not support SSL'
    self.emit('error', new Error('SSL Error: ' + sslErr))
    return
  ***REMOVED***

  // Save the original host before any redirect (if it changes, we need to
  // remove any authorization headers).  Also remember the case of the header
  // name because lots of broken servers expect Host instead of host and we
  // want the caller to be able to specify this.
  self.originalHost = self.getHeader('host')
  if (!self.originalHostHeaderName) ***REMOVED***
    self.originalHostHeaderName = self.hasHeader('host')
  ***REMOVED***
  if (self.setHost) ***REMOVED***
    self.removeHeader('host')
  ***REMOVED***
  if (self.timeout && self.timeoutTimer) ***REMOVED***
    clearTimeout(self.timeoutTimer)
    self.timeoutTimer = null
  ***REMOVED***

  var targetCookieJar = (self._jar && self._jar.setCookie) ? self._jar : globalCookieJar
  var addCookie = function (cookie) ***REMOVED***
    // set the cookie if it's domain in the href's domain.
    try ***REMOVED***
      targetCookieJar.setCookie(cookie, self.uri.href, ***REMOVED***ignoreError: true***REMOVED***)
    ***REMOVED*** catch (e) ***REMOVED***
      self.emit('error', e)
    ***REMOVED***
  ***REMOVED***

  response.caseless = caseless(response.headers)

  if (response.caseless.has('set-cookie') && (!self._disableCookies)) ***REMOVED***
    var headerName = response.caseless.has('set-cookie')
    if (Array.isArray(response.headers[headerName])) ***REMOVED***
      response.headers[headerName].forEach(addCookie)
    ***REMOVED*** else ***REMOVED***
      addCookie(response.headers[headerName])
    ***REMOVED***
  ***REMOVED***

  if (self._redirect.onResponse(response)) ***REMOVED***
    return // Ignore the rest of the response
  ***REMOVED*** else ***REMOVED***
    // Be a good stream and emit end when the response is finished.
    // Hack to emit end on close because of a core bug that never fires end
    response.on('close', function () ***REMOVED***
      if (!self._ended) ***REMOVED***
        self.response.emit('end')
      ***REMOVED***
    ***REMOVED***)

    response.once('end', function () ***REMOVED***
      self._ended = true
    ***REMOVED***)

    var noBody = function (code) ***REMOVED***
      return (
        self.method === 'HEAD' ||
        // Informational
        (code >= 100 && code < 200) ||
        // No Content
        code === 204 ||
        // Not Modified
        code === 304
      )
    ***REMOVED***

    var responseContent
    if (self.gzip && !noBody(response.statusCode)) ***REMOVED***
      var contentEncoding = response.headers['content-encoding'] || 'identity'
      contentEncoding = contentEncoding.trim().toLowerCase()

      // Be more lenient with decoding compressed responses, since (very rarely)
      // servers send slightly invalid gzip responses that are still accepted
      // by common browsers.
      // Always using Z_SYNC_FLUSH is what cURL does.
      var zlibOptions = ***REMOVED***
        flush: zlib.Z_SYNC_FLUSH,
        finishFlush: zlib.Z_SYNC_FLUSH
      ***REMOVED***

      if (contentEncoding === 'gzip') ***REMOVED***
        responseContent = zlib.createGunzip(zlibOptions)
        response.pipe(responseContent)
      ***REMOVED*** else if (contentEncoding === 'deflate') ***REMOVED***
        responseContent = zlib.createInflate(zlibOptions)
        response.pipe(responseContent)
      ***REMOVED*** else ***REMOVED***
        // Since previous versions didn't check for Content-Encoding header,
        // ignore any invalid values to preserve backwards-compatibility
        if (contentEncoding !== 'identity') ***REMOVED***
          debug('ignoring unrecognized Content-Encoding ' + contentEncoding)
        ***REMOVED***
        responseContent = response
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      responseContent = response
    ***REMOVED***

    if (self.encoding) ***REMOVED***
      if (self.dests.length !== 0) ***REMOVED***
        console.error('Ignoring encoding parameter as this stream is being piped to another stream which makes the encoding option invalid.')
      ***REMOVED*** else ***REMOVED***
        responseContent.setEncoding(self.encoding)
      ***REMOVED***
    ***REMOVED***

    if (self._paused) ***REMOVED***
      responseContent.pause()
    ***REMOVED***

    self.responseContent = responseContent

    self.emit('response', response)

    self.dests.forEach(function (dest) ***REMOVED***
      self.pipeDest(dest)
    ***REMOVED***)

    responseContent.on('data', function (chunk) ***REMOVED***
      if (self.timing && !self.responseStarted) ***REMOVED***
        self.responseStartTime = (new Date()).getTime()

        // NOTE: responseStartTime is deprecated in favor of .timings
        response.responseStartTime = self.responseStartTime
      ***REMOVED***
      self._destdata = true
      self.emit('data', chunk)
    ***REMOVED***)
    responseContent.once('end', function (chunk) ***REMOVED***
      self.emit('end', chunk)
    ***REMOVED***)
    responseContent.on('error', function (error) ***REMOVED***
      self.emit('error', error)
    ***REMOVED***)
    responseContent.on('close', function () ***REMOVED*** self.emit('close') ***REMOVED***)

    if (self.callback) ***REMOVED***
      self.readResponseBody(response)
    ***REMOVED*** else ***REMOVED*** // if no callback
      self.on('end', function () ***REMOVED***
        if (self._aborted) ***REMOVED***
          debug('aborted', self.uri.href)
          return
        ***REMOVED***
        self.emit('complete', response)
      ***REMOVED***)
    ***REMOVED***
  ***REMOVED***
  debug('finish init function', self.uri.href)
***REMOVED***

Request.prototype.readResponseBody = function (response) ***REMOVED***
  var self = this
  debug("reading response's body")
  var buffers = []
  var bufferLength = 0
  var strings = []

  self.on('data', function (chunk) ***REMOVED***
    if (!Buffer.isBuffer(chunk)) ***REMOVED***
      strings.push(chunk)
    ***REMOVED*** else if (chunk.length) ***REMOVED***
      bufferLength += chunk.length
      buffers.push(chunk)
    ***REMOVED***
  ***REMOVED***)
  self.on('end', function () ***REMOVED***
    debug('end event', self.uri.href)
    if (self._aborted) ***REMOVED***
      debug('aborted', self.uri.href)
      // `buffer` is defined in the parent scope and used in a closure it exists for the life of the request.
      // This can lead to leaky behavior if the user retains a reference to the request object.
      buffers = []
      bufferLength = 0
      return
    ***REMOVED***

    if (bufferLength) ***REMOVED***
      debug('has body', self.uri.href, bufferLength)
      response.body = Buffer.concat(buffers, bufferLength)
      if (self.encoding !== null) ***REMOVED***
        response.body = response.body.toString(self.encoding)
      ***REMOVED***
      // `buffer` is defined in the parent scope and used in a closure it exists for the life of the Request.
      // This can lead to leaky behavior if the user retains a reference to the request object.
      buffers = []
      bufferLength = 0
    ***REMOVED*** else if (strings.length) ***REMOVED***
      // The UTF8 BOM [0xEF,0xBB,0xBF] is converted to [0xFE,0xFF] in the JS UTC16/UCS2 representation.
      // Strip this value out when the encoding is set to 'utf8', as upstream consumers won't expect it and it breaks JSON.parse().
      if (self.encoding === 'utf8' && strings[0].length > 0 && strings[0][0] === '\uFEFF') ***REMOVED***
        strings[0] = strings[0].substring(1)
      ***REMOVED***
      response.body = strings.join('')
    ***REMOVED***

    if (self._json) ***REMOVED***
      try ***REMOVED***
        response.body = JSON.parse(response.body, self._jsonReviver)
      ***REMOVED*** catch (e) ***REMOVED***
        debug('invalid JSON received', self.uri.href)
      ***REMOVED***
    ***REMOVED***
    debug('emitting complete', self.uri.href)
    if (typeof response.body === 'undefined' && !self._json) ***REMOVED***
      response.body = self.encoding === null ? Buffer.alloc(0) : ''
    ***REMOVED***
    self.emit('complete', response, response.body)
  ***REMOVED***)
***REMOVED***

Request.prototype.abort = function () ***REMOVED***
  var self = this
  self._aborted = true

  if (self.req) ***REMOVED***
    self.req.abort()
  ***REMOVED*** else if (self.response) ***REMOVED***
    self.response.destroy()
  ***REMOVED***

  self.emit('abort')
***REMOVED***

Request.prototype.pipeDest = function (dest) ***REMOVED***
  var self = this
  var response = self.response
  // Called after the response is received
  if (dest.headers && !dest.headersSent) ***REMOVED***
    if (response.caseless.has('content-type')) ***REMOVED***
      var ctname = response.caseless.has('content-type')
      if (dest.setHeader) ***REMOVED***
        dest.setHeader(ctname, response.headers[ctname])
      ***REMOVED*** else ***REMOVED***
        dest.headers[ctname] = response.headers[ctname]
      ***REMOVED***
    ***REMOVED***

    if (response.caseless.has('content-length')) ***REMOVED***
      var clname = response.caseless.has('content-length')
      if (dest.setHeader) ***REMOVED***
        dest.setHeader(clname, response.headers[clname])
      ***REMOVED*** else ***REMOVED***
        dest.headers[clname] = response.headers[clname]
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  if (dest.setHeader && !dest.headersSent) ***REMOVED***
    for (var i in response.headers) ***REMOVED***
      // If the response content is being decoded, the Content-Encoding header
      // of the response doesn't represent the piped content, so don't pass it.
      if (!self.gzip || i !== 'content-encoding') ***REMOVED***
        dest.setHeader(i, response.headers[i])
      ***REMOVED***
    ***REMOVED***
    dest.statusCode = response.statusCode
  ***REMOVED***
  if (self.pipefilter) ***REMOVED***
    self.pipefilter(response, dest)
  ***REMOVED***
***REMOVED***

Request.prototype.qs = function (q, clobber) ***REMOVED***
  var self = this
  var base
  if (!clobber && self.uri.query) ***REMOVED***
    base = self._qs.parse(self.uri.query)
  ***REMOVED*** else ***REMOVED***
    base = ***REMOVED******REMOVED***
  ***REMOVED***

  for (var i in q) ***REMOVED***
    base[i] = q[i]
  ***REMOVED***

  var qs = self._qs.stringify(base)

  if (qs === '') ***REMOVED***
    return self
  ***REMOVED***

  self.uri = url.parse(self.uri.href.split('?')[0] + '?' + qs)
  self.url = self.uri
  self.path = self.uri.path

  if (self.uri.host === 'unix') ***REMOVED***
    self.enableUnixSocket()
  ***REMOVED***

  return self
***REMOVED***
Request.prototype.form = function (form) ***REMOVED***
  var self = this
  if (form) ***REMOVED***
    if (!/^application\/x-www-form-urlencoded\b/.test(self.getHeader('content-type'))) ***REMOVED***
      self.setHeader('content-type', 'application/x-www-form-urlencoded')
    ***REMOVED***
    self.body = (typeof form === 'string')
      ? self._qs.rfc3986(form.toString('utf8'))
      : self._qs.stringify(form).toString('utf8')
    return self
  ***REMOVED***
  // create form-data object
  self._form = new FormData()
  self._form.on('error', function (err) ***REMOVED***
    err.message = 'form-data: ' + err.message
    self.emit('error', err)
    self.abort()
  ***REMOVED***)
  return self._form
***REMOVED***
Request.prototype.multipart = function (multipart) ***REMOVED***
  var self = this

  self._multipart.onRequest(multipart)

  if (!self._multipart.chunked) ***REMOVED***
    self.body = self._multipart.body
  ***REMOVED***

  return self
***REMOVED***
Request.prototype.json = function (val) ***REMOVED***
  var self = this

  if (!self.hasHeader('accept')) ***REMOVED***
    self.setHeader('accept', 'application/json')
  ***REMOVED***

  if (typeof self.jsonReplacer === 'function') ***REMOVED***
    self._jsonReplacer = self.jsonReplacer
  ***REMOVED***

  self._json = true
  if (typeof val === 'boolean') ***REMOVED***
    if (self.body !== undefined) ***REMOVED***
      if (!/^application\/x-www-form-urlencoded\b/.test(self.getHeader('content-type'))) ***REMOVED***
        self.body = safeStringify(self.body, self._jsonReplacer)
      ***REMOVED*** else ***REMOVED***
        self.body = self._qs.rfc3986(self.body)
      ***REMOVED***
      if (!self.hasHeader('content-type')) ***REMOVED***
        self.setHeader('content-type', 'application/json')
      ***REMOVED***
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    self.body = safeStringify(val, self._jsonReplacer)
    if (!self.hasHeader('content-type')) ***REMOVED***
      self.setHeader('content-type', 'application/json')
    ***REMOVED***
  ***REMOVED***

  if (typeof self.jsonReviver === 'function') ***REMOVED***
    self._jsonReviver = self.jsonReviver
  ***REMOVED***

  return self
***REMOVED***
Request.prototype.getHeader = function (name, headers) ***REMOVED***
  var self = this
  var result, re, match
  if (!headers) ***REMOVED***
    headers = self.headers
  ***REMOVED***
  Object.keys(headers).forEach(function (key) ***REMOVED***
    if (key.length !== name.length) ***REMOVED***
      return
    ***REMOVED***
    re = new RegExp(name, 'i')
    match = key.match(re)
    if (match) ***REMOVED***
      result = headers[key]
    ***REMOVED***
  ***REMOVED***)
  return result
***REMOVED***
Request.prototype.enableUnixSocket = function () ***REMOVED***
  // Get the socket & request paths from the URL
  var unixParts = this.uri.path.split(':')
  var host = unixParts[0]
  var path = unixParts[1]
  // Apply unix properties to request
  this.socketPath = host
  this.uri.pathname = path
  this.uri.path = path
  this.uri.host = host
  this.uri.hostname = host
  this.uri.isUnix = true
***REMOVED***

Request.prototype.auth = function (user, pass, sendImmediately, bearer) ***REMOVED***
  var self = this

  self._auth.onRequest(user, pass, sendImmediately, bearer)

  return self
***REMOVED***
Request.prototype.aws = function (opts, now) ***REMOVED***
  var self = this

  if (!now) ***REMOVED***
    self._aws = opts
    return self
  ***REMOVED***

  if (opts.sign_version === 4 || opts.sign_version === '4') ***REMOVED***
    // use aws4
    var options = ***REMOVED***
      host: self.uri.host,
      path: self.uri.path,
      method: self.method,
      headers: ***REMOVED***
        'content-type': self.getHeader('content-type') || ''
      ***REMOVED***,
      body: self.body
    ***REMOVED***
    var signRes = aws4.sign(options, ***REMOVED***
      accessKeyId: opts.key,
      secretAccessKey: opts.secret,
      sessionToken: opts.session
    ***REMOVED***)
    self.setHeader('authorization', signRes.headers.Authorization)
    self.setHeader('x-amz-date', signRes.headers['X-Amz-Date'])
    if (signRes.headers['X-Amz-Security-Token']) ***REMOVED***
      self.setHeader('x-amz-security-token', signRes.headers['X-Amz-Security-Token'])
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    // default: use aws-sign2
    var date = new Date()
    self.setHeader('date', date.toUTCString())
    var auth = ***REMOVED***
      key: opts.key,
      secret: opts.secret,
      verb: self.method.toUpperCase(),
      date: date,
      contentType: self.getHeader('content-type') || '',
      md5: self.getHeader('content-md5') || '',
      amazonHeaders: aws2.canonicalizeHeaders(self.headers)
    ***REMOVED***
    var path = self.uri.path
    if (opts.bucket && path) ***REMOVED***
      auth.resource = '/' + opts.bucket + path
    ***REMOVED*** else if (opts.bucket && !path) ***REMOVED***
      auth.resource = '/' + opts.bucket
    ***REMOVED*** else if (!opts.bucket && path) ***REMOVED***
      auth.resource = path
    ***REMOVED*** else if (!opts.bucket && !path) ***REMOVED***
      auth.resource = '/'
    ***REMOVED***
    auth.resource = aws2.canonicalizeResource(auth.resource)
    self.setHeader('authorization', aws2.authorization(auth))
  ***REMOVED***

  return self
***REMOVED***
Request.prototype.httpSignature = function (opts) ***REMOVED***
  var self = this
  httpSignature.signRequest(***REMOVED***
    getHeader: function (header) ***REMOVED***
      return self.getHeader(header, self.headers)
    ***REMOVED***,
    setHeader: function (header, value) ***REMOVED***
      self.setHeader(header, value)
    ***REMOVED***,
    method: self.method,
    path: self.path
  ***REMOVED***, opts)
  debug('httpSignature authorization', self.getHeader('authorization'))

  return self
***REMOVED***
Request.prototype.hawk = function (opts) ***REMOVED***
  var self = this
  self.setHeader('Authorization', hawk.client.header(self.uri, self.method, opts).field)
***REMOVED***
Request.prototype.oauth = function (_oauth) ***REMOVED***
  var self = this

  self._oauth.onRequest(_oauth)

  return self
***REMOVED***

Request.prototype.jar = function (jar) ***REMOVED***
  var self = this
  var cookies

  if (self._redirect.redirectsFollowed === 0) ***REMOVED***
    self.originalCookieHeader = self.getHeader('cookie')
  ***REMOVED***

  if (!jar) ***REMOVED***
    // disable cookies
    cookies = false
    self._disableCookies = true
  ***REMOVED*** else ***REMOVED***
    var targetCookieJar = (jar && jar.getCookieString) ? jar : globalCookieJar
    var urihref = self.uri.href
    // fetch cookie in the Specified host
    if (targetCookieJar) ***REMOVED***
      cookies = targetCookieJar.getCookieString(urihref)
    ***REMOVED***
  ***REMOVED***

  // if need cookie and cookie is not empty
  if (cookies && cookies.length) ***REMOVED***
    if (self.originalCookieHeader) ***REMOVED***
      // Don't overwrite existing Cookie header
      self.setHeader('cookie', self.originalCookieHeader + '; ' + cookies)
    ***REMOVED*** else ***REMOVED***
      self.setHeader('cookie', cookies)
    ***REMOVED***
  ***REMOVED***
  self._jar = jar
  return self
***REMOVED***

// Stream API
Request.prototype.pipe = function (dest, opts) ***REMOVED***
  var self = this

  if (self.response) ***REMOVED***
    if (self._destdata) ***REMOVED***
      self.emit('error', new Error('You cannot pipe after data has been emitted from the response.'))
    ***REMOVED*** else if (self._ended) ***REMOVED***
      self.emit('error', new Error('You cannot pipe after the response has been ended.'))
    ***REMOVED*** else ***REMOVED***
      stream.Stream.prototype.pipe.call(self, dest, opts)
      self.pipeDest(dest)
      return dest
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    self.dests.push(dest)
    stream.Stream.prototype.pipe.call(self, dest, opts)
    return dest
  ***REMOVED***
***REMOVED***
Request.prototype.write = function () ***REMOVED***
  var self = this
  if (self._aborted) ***REMOVED*** return ***REMOVED***

  if (!self._started) ***REMOVED***
    self.start()
  ***REMOVED***
  if (self.req) ***REMOVED***
    return self.req.write.apply(self.req, arguments)
  ***REMOVED***
***REMOVED***
Request.prototype.end = function (chunk) ***REMOVED***
  var self = this
  if (self._aborted) ***REMOVED*** return ***REMOVED***

  if (chunk) ***REMOVED***
    self.write(chunk)
  ***REMOVED***
  if (!self._started) ***REMOVED***
    self.start()
  ***REMOVED***
  if (self.req) ***REMOVED***
    self.req.end()
  ***REMOVED***
***REMOVED***
Request.prototype.pause = function () ***REMOVED***
  var self = this
  if (!self.responseContent) ***REMOVED***
    self._paused = true
  ***REMOVED*** else ***REMOVED***
    self.responseContent.pause.apply(self.responseContent, arguments)
  ***REMOVED***
***REMOVED***
Request.prototype.resume = function () ***REMOVED***
  var self = this
  if (!self.responseContent) ***REMOVED***
    self._paused = false
  ***REMOVED*** else ***REMOVED***
    self.responseContent.resume.apply(self.responseContent, arguments)
  ***REMOVED***
***REMOVED***
Request.prototype.destroy = function () ***REMOVED***
  var self = this
  if (!self._ended) ***REMOVED***
    self.end()
  ***REMOVED*** else if (self.response) ***REMOVED***
    self.response.destroy()
  ***REMOVED***
***REMOVED***

Request.defaultProxyHeaderWhiteList =
  Tunnel.defaultProxyHeaderWhiteList.slice()

Request.defaultProxyHeaderExclusiveList =
  Tunnel.defaultProxyHeaderExclusiveList.slice()

// Exports

Request.prototype.toJSON = requestToJSON
module.exports = Request
