// mongoose/lib/virtualtype.js

/**
 * VirtualType constructor
 *
 * This is what mongoose uses to define virtual attributes via `Schema.prototype.virtual`.
 *
 * ####Example:
 *
 *     const fullname = schema.virtual('fullname');
 *     fullname instanceof mongoose.VirtualType // true
 *
 * @param ***REMOVED***Object***REMOVED*** options
 * @param ***REMOVED***string|function***REMOVED*** [options.ref] if `ref` is not nullish, this becomes a [populated virtual](/docs/populate.html#populate-virtuals)
 * @param ***REMOVED***string|function***REMOVED*** [options.localField] the local field to populate on if this is a populated virtual.
 * @param ***REMOVED***string|function***REMOVED*** [options.foreignField] the foreign field to populate on if this is a populated virtual.
 * @param ***REMOVED***boolean***REMOVED*** [options.justOne=false] by default, a populated virtual is an array. If you set `justOne`, the populated virtual will be a single doc or `null`.
 * @api public
 */

function VirtualType(options, name) ***REMOVED***
  this.path = name;
  this.getters = [];
  this.setters = [];
  this.options = options || ***REMOVED******REMOVED***;
***REMOVED***

/**
 * If no getters/getters, add a default
 *
 * @param ***REMOVED***Function***REMOVED*** fn
 * @return ***REMOVED***VirtualType***REMOVED*** this
 * @api public
 */

VirtualType.prototype._applyDefaultGetters = function() ***REMOVED***
  if (this.getters.length > 0 || this.setters.length > 0) ***REMOVED***
    return;
  ***REMOVED***

  const path = this.path;
  const internalProperty = '$' + path;
  this.getters.push(function() ***REMOVED***
    return this[internalProperty];
  ***REMOVED***);
  this.setters.push(function(v) ***REMOVED***
    this[internalProperty] = v;
  ***REMOVED***);
***REMOVED***;

/**
 * Defines a getter.
 *
 * ####Example:
 *
 *     var virtual = schema.virtual('fullname');
 *     virtual.get(function () ***REMOVED***
 *       return this.name.first + ' ' + this.name.last;
 *     ***REMOVED***);
 *
 * @param ***REMOVED***Function***REMOVED*** fn
 * @return ***REMOVED***VirtualType***REMOVED*** this
 * @api public
 */

VirtualType.prototype.get = function(fn) ***REMOVED***
  this.getters.push(fn);
  return this;
***REMOVED***;

/**
 * Defines a setter.
 *
 * ####Example:
 *
 *     var virtual = schema.virtual('fullname');
 *     virtual.set(function (v) ***REMOVED***
 *       var parts = v.split(' ');
 *       this.name.first = parts[0];
 *       this.name.last = parts[1];
 *     ***REMOVED***);
 *
 * @param ***REMOVED***Function***REMOVED*** fn
 * @return ***REMOVED***VirtualType***REMOVED*** this
 * @api public
 */

VirtualType.prototype.set = function(fn) ***REMOVED***
  this.setters.push(fn);
  return this;
***REMOVED***;

/**
 * Applies getters to `value` using optional `scope`.
 *
 * @param ***REMOVED***Object***REMOVED*** value
 * @param ***REMOVED***Object***REMOVED*** scope
 * @return ***REMOVED***any***REMOVED*** the value after applying all getters
 * @api public
 */

VirtualType.prototype.applyGetters = function(value, scope) ***REMOVED***
  var v = value;
  for (var l = this.getters.length - 1; l >= 0; l--) ***REMOVED***
    v = this.getters[l].call(scope, v, this);
  ***REMOVED***
  return v;
***REMOVED***;

/**
 * Applies setters to `value` using optional `scope`.
 *
 * @param ***REMOVED***Object***REMOVED*** value
 * @param ***REMOVED***Object***REMOVED*** scope
 * @return ***REMOVED***any***REMOVED*** the value after applying all setters
 * @api public
 */

VirtualType.prototype.applySetters = function(value, scope) ***REMOVED***
  var v = value;
  for (var l = this.setters.length - 1; l >= 0; l--) ***REMOVED***
    v = this.setters[l].call(scope, v, this);
  ***REMOVED***
  return v;
***REMOVED***;

/*!
 * exports
 */

module.exports = VirtualType;
