// ajv/lib/patternGroups.js
'use strict';

var META_SCHEMA_ID = 'http://json-schema.org/draft-06/schema';

module.exports = function (ajv) ***REMOVED***
  var defaultMeta = ajv._opts.defaultMeta;
  var metaSchemaRef = typeof defaultMeta == 'string'
                      ? ***REMOVED*** $ref: defaultMeta ***REMOVED***
                      : ajv.getSchema(META_SCHEMA_ID)
                        ? ***REMOVED*** $ref: META_SCHEMA_ID ***REMOVED***
                        : ***REMOVED******REMOVED***;

  ajv.addKeyword('patternGroups', ***REMOVED***
    // implemented in properties.jst
    metaSchema: ***REMOVED***
      type: 'object',
      additionalProperties: ***REMOVED***
        type: 'object',
        required: [ 'schema' ],
        properties: ***REMOVED***
          maximum: ***REMOVED***
            type: 'integer',
            minimum: 0
          ***REMOVED***,
          minimum: ***REMOVED***
            type: 'integer',
            minimum: 0
          ***REMOVED***,
          schema: metaSchemaRef
        ***REMOVED***,
        additionalProperties: false
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***);
  ajv.RULES.all.properties.implements.push('patternGroups');
***REMOVED***;
