// lodash.get/index.js
/**
 * lodash (Custom Build) <https://lodash.com/>
 * Build: `lodash modularize exports="npm" -o ./`
 * Copyright jQuery Foundation and other contributors <https://jquery.org/>
 * Released under MIT license <https://lodash.com/license>
 * Based on Underscore.js 1.8.3 <http://underscorejs.org/LICENSE>
 * Copyright Jeremy Ashkenas, DocumentCloud and Investigative Reporters & Editors
 */

/** Used as the `TypeError` message for "Functions" methods. */
var FUNC_ERROR_TEXT = 'Expected a function';

/** Used to stand-in for `undefined` hash values. */
var HASH_UNDEFINED = '__lodash_hash_undefined__';

/** Used as references for various `Number` constants. */
var INFINITY = 1 / 0;

/** `Object#toString` result references. */
var funcTag = '[object Function]',
    genTag = '[object GeneratorFunction]',
    symbolTag = '[object Symbol]';

/** Used to match property names within property paths. */
var reIsDeepProp = /\.|\[(?:[^[\]]*|(["'])(?:(?!\1)[^\\]|\\.)*?\1)\]/,
    reIsPlainProp = /^\w*$/,
    reLeadingDot = /^\./,
    rePropName = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g;

/**
 * Used to match `RegExp`
 * [syntax characters](http://ecma-international.org/ecma-262/7.0/#sec-patterns).
 */
var reRegExpChar = /[\\^$.*+?()[\]***REMOVED******REMOVED***|]/g;

/** Used to match backslashes in property paths. */
var reEscapeChar = /\\(\\)?/g;

/** Used to detect host constructors (Safari). */
var reIsHostCtor = /^\[object .+?Constructor\]$/;

/** Detect free variable `global` from Node.js. */
var freeGlobal = typeof global == 'object' && global && global.Object === Object && global;

/** Detect free variable `self`. */
var freeSelf = typeof self == 'object' && self && self.Object === Object && self;

/** Used as a reference to the global object. */
var root = freeGlobal || freeSelf || Function('return this')();

/**
 * Gets the value at `key` of `object`.
 *
 * @private
 * @param ***REMOVED***Object***REMOVED*** [object] The object to query.
 * @param ***REMOVED***string***REMOVED*** key The key of the property to get.
 * @returns ***REMOVED*******REMOVED*** Returns the property value.
 */
function getValue(object, key) ***REMOVED***
  return object == null ? undefined : object[key];
***REMOVED***

/**
 * Checks if `value` is a host object in IE < 9.
 *
 * @private
 * @param ***REMOVED*******REMOVED*** value The value to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a host object, else `false`.
 */
function isHostObject(value) ***REMOVED***
  // Many host objects are `Object` objects that can coerce to strings
  // despite having improperly defined `toString` methods.
  var result = false;
  if (value != null && typeof value.toString != 'function') ***REMOVED***
    try ***REMOVED***
      result = !!(value + '');
    ***REMOVED*** catch (e) ***REMOVED******REMOVED***
  ***REMOVED***
  return result;
***REMOVED***

/** Used for built-in method references. */
var arrayProto = Array.prototype,
    funcProto = Function.prototype,
    objectProto = Object.prototype;

/** Used to detect overreaching core-js shims. */
var coreJsData = root['__core-js_shared__'];

/** Used to detect methods masquerading as native. */
var maskSrcKey = (function() ***REMOVED***
  var uid = /[^.]+$/.exec(coreJsData && coreJsData.keys && coreJsData.keys.IE_PROTO || '');
  return uid ? ('Symbol(src)_1.' + uid) : '';
***REMOVED***());

/** Used to resolve the decompiled source of functions. */
var funcToString = funcProto.toString;

/** Used to check objects for own properties. */
var hasOwnProperty = objectProto.hasOwnProperty;

/**
 * Used to resolve the
 * [`toStringTag`](http://ecma-international.org/ecma-262/7.0/#sec-object.prototype.tostring)
 * of values.
 */
var objectToString = objectProto.toString;

/** Used to detect if a method is native. */
var reIsNative = RegExp('^' +
  funcToString.call(hasOwnProperty).replace(reRegExpChar, '\\$&')
  .replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, '$1.*?') + '$'
);

/** Built-in value references. */
var Symbol = root.Symbol,
    splice = arrayProto.splice;

/* Built-in method references that are verified to be native. */
var Map = getNative(root, 'Map'),
    nativeCreate = getNative(Object, 'create');

/** Used to convert symbols to primitives and strings. */
var symbolProto = Symbol ? Symbol.prototype : undefined,
    symbolToString = symbolProto ? symbolProto.toString : undefined;

/**
 * Creates a hash object.
 *
 * @private
 * @constructor
 * @param ***REMOVED***Array***REMOVED*** [entries] The key-value pairs to cache.
 */
function Hash(entries) ***REMOVED***
  var index = -1,
      length = entries ? entries.length : 0;

  this.clear();
  while (++index < length) ***REMOVED***
    var entry = entries[index];
    this.set(entry[0], entry[1]);
  ***REMOVED***
***REMOVED***

/**
 * Removes all key-value entries from the hash.
 *
 * @private
 * @name clear
 * @memberOf Hash
 */
function hashClear() ***REMOVED***
  this.__data__ = nativeCreate ? nativeCreate(null) : ***REMOVED******REMOVED***;
***REMOVED***

/**
 * Removes `key` and its value from the hash.
 *
 * @private
 * @name delete
 * @memberOf Hash
 * @param ***REMOVED***Object***REMOVED*** hash The hash to modify.
 * @param ***REMOVED***string***REMOVED*** key The key of the value to remove.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if the entry was removed, else `false`.
 */
function hashDelete(key) ***REMOVED***
  return this.has(key) && delete this.__data__[key];
***REMOVED***

/**
 * Gets the hash value for `key`.
 *
 * @private
 * @name get
 * @memberOf Hash
 * @param ***REMOVED***string***REMOVED*** key The key of the value to get.
 * @returns ***REMOVED*******REMOVED*** Returns the entry value.
 */
function hashGet(key) ***REMOVED***
  var data = this.__data__;
  if (nativeCreate) ***REMOVED***
    var result = data[key];
    return result === HASH_UNDEFINED ? undefined : result;
  ***REMOVED***
  return hasOwnProperty.call(data, key) ? data[key] : undefined;
***REMOVED***

/**
 * Checks if a hash value for `key` exists.
 *
 * @private
 * @name has
 * @memberOf Hash
 * @param ***REMOVED***string***REMOVED*** key The key of the entry to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if an entry for `key` exists, else `false`.
 */
function hashHas(key) ***REMOVED***
  var data = this.__data__;
  return nativeCreate ? data[key] !== undefined : hasOwnProperty.call(data, key);
***REMOVED***

/**
 * Sets the hash `key` to `value`.
 *
 * @private
 * @name set
 * @memberOf Hash
 * @param ***REMOVED***string***REMOVED*** key The key of the value to set.
 * @param ***REMOVED*******REMOVED*** value The value to set.
 * @returns ***REMOVED***Object***REMOVED*** Returns the hash instance.
 */
function hashSet(key, value) ***REMOVED***
  var data = this.__data__;
  data[key] = (nativeCreate && value === undefined) ? HASH_UNDEFINED : value;
  return this;
***REMOVED***

// Add methods to `Hash`.
Hash.prototype.clear = hashClear;
Hash.prototype['delete'] = hashDelete;
Hash.prototype.get = hashGet;
Hash.prototype.has = hashHas;
Hash.prototype.set = hashSet;

/**
 * Creates an list cache object.
 *
 * @private
 * @constructor
 * @param ***REMOVED***Array***REMOVED*** [entries] The key-value pairs to cache.
 */
function ListCache(entries) ***REMOVED***
  var index = -1,
      length = entries ? entries.length : 0;

  this.clear();
  while (++index < length) ***REMOVED***
    var entry = entries[index];
    this.set(entry[0], entry[1]);
  ***REMOVED***
***REMOVED***

/**
 * Removes all key-value entries from the list cache.
 *
 * @private
 * @name clear
 * @memberOf ListCache
 */
function listCacheClear() ***REMOVED***
  this.__data__ = [];
***REMOVED***

/**
 * Removes `key` and its value from the list cache.
 *
 * @private
 * @name delete
 * @memberOf ListCache
 * @param ***REMOVED***string***REMOVED*** key The key of the value to remove.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if the entry was removed, else `false`.
 */
function listCacheDelete(key) ***REMOVED***
  var data = this.__data__,
      index = assocIndexOf(data, key);

  if (index < 0) ***REMOVED***
    return false;
  ***REMOVED***
  var lastIndex = data.length - 1;
  if (index == lastIndex) ***REMOVED***
    data.pop();
  ***REMOVED*** else ***REMOVED***
    splice.call(data, index, 1);
  ***REMOVED***
  return true;
***REMOVED***

/**
 * Gets the list cache value for `key`.
 *
 * @private
 * @name get
 * @memberOf ListCache
 * @param ***REMOVED***string***REMOVED*** key The key of the value to get.
 * @returns ***REMOVED*******REMOVED*** Returns the entry value.
 */
function listCacheGet(key) ***REMOVED***
  var data = this.__data__,
      index = assocIndexOf(data, key);

  return index < 0 ? undefined : data[index][1];
***REMOVED***

/**
 * Checks if a list cache value for `key` exists.
 *
 * @private
 * @name has
 * @memberOf ListCache
 * @param ***REMOVED***string***REMOVED*** key The key of the entry to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if an entry for `key` exists, else `false`.
 */
function listCacheHas(key) ***REMOVED***
  return assocIndexOf(this.__data__, key) > -1;
***REMOVED***

/**
 * Sets the list cache `key` to `value`.
 *
 * @private
 * @name set
 * @memberOf ListCache
 * @param ***REMOVED***string***REMOVED*** key The key of the value to set.
 * @param ***REMOVED*******REMOVED*** value The value to set.
 * @returns ***REMOVED***Object***REMOVED*** Returns the list cache instance.
 */
function listCacheSet(key, value) ***REMOVED***
  var data = this.__data__,
      index = assocIndexOf(data, key);

  if (index < 0) ***REMOVED***
    data.push([key, value]);
  ***REMOVED*** else ***REMOVED***
    data[index][1] = value;
  ***REMOVED***
  return this;
***REMOVED***

// Add methods to `ListCache`.
ListCache.prototype.clear = listCacheClear;
ListCache.prototype['delete'] = listCacheDelete;
ListCache.prototype.get = listCacheGet;
ListCache.prototype.has = listCacheHas;
ListCache.prototype.set = listCacheSet;

/**
 * Creates a map cache object to store key-value pairs.
 *
 * @private
 * @constructor
 * @param ***REMOVED***Array***REMOVED*** [entries] The key-value pairs to cache.
 */
function MapCache(entries) ***REMOVED***
  var index = -1,
      length = entries ? entries.length : 0;

  this.clear();
  while (++index < length) ***REMOVED***
    var entry = entries[index];
    this.set(entry[0], entry[1]);
  ***REMOVED***
***REMOVED***

/**
 * Removes all key-value entries from the map.
 *
 * @private
 * @name clear
 * @memberOf MapCache
 */
function mapCacheClear() ***REMOVED***
  this.__data__ = ***REMOVED***
    'hash': new Hash,
    'map': new (Map || ListCache),
    'string': new Hash
  ***REMOVED***;
***REMOVED***

/**
 * Removes `key` and its value from the map.
 *
 * @private
 * @name delete
 * @memberOf MapCache
 * @param ***REMOVED***string***REMOVED*** key The key of the value to remove.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if the entry was removed, else `false`.
 */
function mapCacheDelete(key) ***REMOVED***
  return getMapData(this, key)['delete'](key);
***REMOVED***

/**
 * Gets the map value for `key`.
 *
 * @private
 * @name get
 * @memberOf MapCache
 * @param ***REMOVED***string***REMOVED*** key The key of the value to get.
 * @returns ***REMOVED*******REMOVED*** Returns the entry value.
 */
function mapCacheGet(key) ***REMOVED***
  return getMapData(this, key).get(key);
***REMOVED***

/**
 * Checks if a map value for `key` exists.
 *
 * @private
 * @name has
 * @memberOf MapCache
 * @param ***REMOVED***string***REMOVED*** key The key of the entry to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if an entry for `key` exists, else `false`.
 */
function mapCacheHas(key) ***REMOVED***
  return getMapData(this, key).has(key);
***REMOVED***

/**
 * Sets the map `key` to `value`.
 *
 * @private
 * @name set
 * @memberOf MapCache
 * @param ***REMOVED***string***REMOVED*** key The key of the value to set.
 * @param ***REMOVED*******REMOVED*** value The value to set.
 * @returns ***REMOVED***Object***REMOVED*** Returns the map cache instance.
 */
function mapCacheSet(key, value) ***REMOVED***
  getMapData(this, key).set(key, value);
  return this;
***REMOVED***

// Add methods to `MapCache`.
MapCache.prototype.clear = mapCacheClear;
MapCache.prototype['delete'] = mapCacheDelete;
MapCache.prototype.get = mapCacheGet;
MapCache.prototype.has = mapCacheHas;
MapCache.prototype.set = mapCacheSet;

/**
 * Gets the index at which the `key` is found in `array` of key-value pairs.
 *
 * @private
 * @param ***REMOVED***Array***REMOVED*** array The array to inspect.
 * @param ***REMOVED*******REMOVED*** key The key to search for.
 * @returns ***REMOVED***number***REMOVED*** Returns the index of the matched value, else `-1`.
 */
function assocIndexOf(array, key) ***REMOVED***
  var length = array.length;
  while (length--) ***REMOVED***
    if (eq(array[length][0], key)) ***REMOVED***
      return length;
    ***REMOVED***
  ***REMOVED***
  return -1;
***REMOVED***

/**
 * The base implementation of `_.get` without support for default values.
 *
 * @private
 * @param ***REMOVED***Object***REMOVED*** object The object to query.
 * @param ***REMOVED***Array|string***REMOVED*** path The path of the property to get.
 * @returns ***REMOVED*******REMOVED*** Returns the resolved value.
 */
function baseGet(object, path) ***REMOVED***
  path = isKey(path, object) ? [path] : castPath(path);

  var index = 0,
      length = path.length;

  while (object != null && index < length) ***REMOVED***
    object = object[toKey(path[index++])];
  ***REMOVED***
  return (index && index == length) ? object : undefined;
***REMOVED***

/**
 * The base implementation of `_.isNative` without bad shim checks.
 *
 * @private
 * @param ***REMOVED*******REMOVED*** value The value to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a native function,
 *  else `false`.
 */
function baseIsNative(value) ***REMOVED***
  if (!isObject(value) || isMasked(value)) ***REMOVED***
    return false;
  ***REMOVED***
  var pattern = (isFunction(value) || isHostObject(value)) ? reIsNative : reIsHostCtor;
  return pattern.test(toSource(value));
***REMOVED***

/**
 * The base implementation of `_.toString` which doesn't convert nullish
 * values to empty strings.
 *
 * @private
 * @param ***REMOVED*******REMOVED*** value The value to process.
 * @returns ***REMOVED***string***REMOVED*** Returns the string.
 */
function baseToString(value) ***REMOVED***
  // Exit early for strings to avoid a performance hit in some environments.
  if (typeof value == 'string') ***REMOVED***
    return value;
  ***REMOVED***
  if (isSymbol(value)) ***REMOVED***
    return symbolToString ? symbolToString.call(value) : '';
  ***REMOVED***
  var result = (value + '');
  return (result == '0' && (1 / value) == -INFINITY) ? '-0' : result;
***REMOVED***

/**
 * Casts `value` to a path array if it's not one.
 *
 * @private
 * @param ***REMOVED*******REMOVED*** value The value to inspect.
 * @returns ***REMOVED***Array***REMOVED*** Returns the cast property path array.
 */
function castPath(value) ***REMOVED***
  return isArray(value) ? value : stringToPath(value);
***REMOVED***

/**
 * Gets the data for `map`.
 *
 * @private
 * @param ***REMOVED***Object***REMOVED*** map The map to query.
 * @param ***REMOVED***string***REMOVED*** key The reference key.
 * @returns ***REMOVED*******REMOVED*** Returns the map data.
 */
function getMapData(map, key) ***REMOVED***
  var data = map.__data__;
  return isKeyable(key)
    ? data[typeof key == 'string' ? 'string' : 'hash']
    : data.map;
***REMOVED***

/**
 * Gets the native function at `key` of `object`.
 *
 * @private
 * @param ***REMOVED***Object***REMOVED*** object The object to query.
 * @param ***REMOVED***string***REMOVED*** key The key of the method to get.
 * @returns ***REMOVED*******REMOVED*** Returns the function if it's native, else `undefined`.
 */
function getNative(object, key) ***REMOVED***
  var value = getValue(object, key);
  return baseIsNative(value) ? value : undefined;
***REMOVED***

/**
 * Checks if `value` is a property name and not a property path.
 *
 * @private
 * @param ***REMOVED*******REMOVED*** value The value to check.
 * @param ***REMOVED***Object***REMOVED*** [object] The object to query keys on.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a property name, else `false`.
 */
function isKey(value, object) ***REMOVED***
  if (isArray(value)) ***REMOVED***
    return false;
  ***REMOVED***
  var type = typeof value;
  if (type == 'number' || type == 'symbol' || type == 'boolean' ||
      value == null || isSymbol(value)) ***REMOVED***
    return true;
  ***REMOVED***
  return reIsPlainProp.test(value) || !reIsDeepProp.test(value) ||
    (object != null && value in Object(object));
***REMOVED***

/**
 * Checks if `value` is suitable for use as unique object key.
 *
 * @private
 * @param ***REMOVED*******REMOVED*** value The value to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is suitable, else `false`.
 */
function isKeyable(value) ***REMOVED***
  var type = typeof value;
  return (type == 'string' || type == 'number' || type == 'symbol' || type == 'boolean')
    ? (value !== '__proto__')
    : (value === null);
***REMOVED***

/**
 * Checks if `func` has its source masked.
 *
 * @private
 * @param ***REMOVED***Function***REMOVED*** func The function to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `func` is masked, else `false`.
 */
function isMasked(func) ***REMOVED***
  return !!maskSrcKey && (maskSrcKey in func);
***REMOVED***

/**
 * Converts `string` to a property path array.
 *
 * @private
 * @param ***REMOVED***string***REMOVED*** string The string to convert.
 * @returns ***REMOVED***Array***REMOVED*** Returns the property path array.
 */
var stringToPath = memoize(function(string) ***REMOVED***
  string = toString(string);

  var result = [];
  if (reLeadingDot.test(string)) ***REMOVED***
    result.push('');
  ***REMOVED***
  string.replace(rePropName, function(match, number, quote, string) ***REMOVED***
    result.push(quote ? string.replace(reEscapeChar, '$1') : (number || match));
  ***REMOVED***);
  return result;
***REMOVED***);

/**
 * Converts `value` to a string key if it's not a string or symbol.
 *
 * @private
 * @param ***REMOVED*******REMOVED*** value The value to inspect.
 * @returns ***REMOVED***string|symbol***REMOVED*** Returns the key.
 */
function toKey(value) ***REMOVED***
  if (typeof value == 'string' || isSymbol(value)) ***REMOVED***
    return value;
  ***REMOVED***
  var result = (value + '');
  return (result == '0' && (1 / value) == -INFINITY) ? '-0' : result;
***REMOVED***

/**
 * Converts `func` to its source code.
 *
 * @private
 * @param ***REMOVED***Function***REMOVED*** func The function to process.
 * @returns ***REMOVED***string***REMOVED*** Returns the source code.
 */
function toSource(func) ***REMOVED***
  if (func != null) ***REMOVED***
    try ***REMOVED***
      return funcToString.call(func);
    ***REMOVED*** catch (e) ***REMOVED******REMOVED***
    try ***REMOVED***
      return (func + '');
    ***REMOVED*** catch (e) ***REMOVED******REMOVED***
  ***REMOVED***
  return '';
***REMOVED***

/**
 * Creates a function that memoizes the result of `func`. If `resolver` is
 * provided, it determines the cache key for storing the result based on the
 * arguments provided to the memoized function. By default, the first argument
 * provided to the memoized function is used as the map cache key. The `func`
 * is invoked with the `this` binding of the memoized function.
 *
 * **Note:** The cache is exposed as the `cache` property on the memoized
 * function. Its creation may be customized by replacing the `_.memoize.Cache`
 * constructor with one whose instances implement the
 * [`Map`](http://ecma-international.org/ecma-262/7.0/#sec-properties-of-the-map-prototype-object)
 * method interface of `delete`, `get`, `has`, and `set`.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Function
 * @param ***REMOVED***Function***REMOVED*** func The function to have its output memoized.
 * @param ***REMOVED***Function***REMOVED*** [resolver] The function to resolve the cache key.
 * @returns ***REMOVED***Function***REMOVED*** Returns the new memoized function.
 * @example
 *
 * var object = ***REMOVED*** 'a': 1, 'b': 2 ***REMOVED***;
 * var other = ***REMOVED*** 'c': 3, 'd': 4 ***REMOVED***;
 *
 * var values = _.memoize(_.values);
 * values(object);
 * // => [1, 2]
 *
 * values(other);
 * // => [3, 4]
 *
 * object.a = 2;
 * values(object);
 * // => [1, 2]
 *
 * // Modify the result cache.
 * values.cache.set(object, ['a', 'b']);
 * values(object);
 * // => ['a', 'b']
 *
 * // Replace `_.memoize.Cache`.
 * _.memoize.Cache = WeakMap;
 */
function memoize(func, resolver) ***REMOVED***
  if (typeof func != 'function' || (resolver && typeof resolver != 'function')) ***REMOVED***
    throw new TypeError(FUNC_ERROR_TEXT);
  ***REMOVED***
  var memoized = function() ***REMOVED***
    var args = arguments,
        key = resolver ? resolver.apply(this, args) : args[0],
        cache = memoized.cache;

    if (cache.has(key)) ***REMOVED***
      return cache.get(key);
    ***REMOVED***
    var result = func.apply(this, args);
    memoized.cache = cache.set(key, result);
    return result;
  ***REMOVED***;
  memoized.cache = new (memoize.Cache || MapCache);
  return memoized;
***REMOVED***

// Assign cache to `_.memoize`.
memoize.Cache = MapCache;

/**
 * Performs a
 * [`SameValueZero`](http://ecma-international.org/ecma-262/7.0/#sec-samevaluezero)
 * comparison between two values to determine if they are equivalent.
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Lang
 * @param ***REMOVED*******REMOVED*** value The value to compare.
 * @param ***REMOVED*******REMOVED*** other The other value to compare.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if the values are equivalent, else `false`.
 * @example
 *
 * var object = ***REMOVED*** 'a': 1 ***REMOVED***;
 * var other = ***REMOVED*** 'a': 1 ***REMOVED***;
 *
 * _.eq(object, object);
 * // => true
 *
 * _.eq(object, other);
 * // => false
 *
 * _.eq('a', 'a');
 * // => true
 *
 * _.eq('a', Object('a'));
 * // => false
 *
 * _.eq(NaN, NaN);
 * // => true
 */
function eq(value, other) ***REMOVED***
  return value === other || (value !== value && other !== other);
***REMOVED***

/**
 * Checks if `value` is classified as an `Array` object.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Lang
 * @param ***REMOVED*******REMOVED*** value The value to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is an array, else `false`.
 * @example
 *
 * _.isArray([1, 2, 3]);
 * // => true
 *
 * _.isArray(document.body.children);
 * // => false
 *
 * _.isArray('abc');
 * // => false
 *
 * _.isArray(_.noop);
 * // => false
 */
var isArray = Array.isArray;

/**
 * Checks if `value` is classified as a `Function` object.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Lang
 * @param ***REMOVED*******REMOVED*** value The value to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a function, else `false`.
 * @example
 *
 * _.isFunction(_);
 * // => true
 *
 * _.isFunction(/abc/);
 * // => false
 */
function isFunction(value) ***REMOVED***
  // The use of `Object#toString` avoids issues with the `typeof` operator
  // in Safari 8-9 which returns 'object' for typed array and other constructors.
  var tag = isObject(value) ? objectToString.call(value) : '';
  return tag == funcTag || tag == genTag;
***REMOVED***

/**
 * Checks if `value` is the
 * [language type](http://www.ecma-international.org/ecma-262/7.0/#sec-ecmascript-language-types)
 * of `Object`. (e.g. arrays, functions, objects, regexes, `new Number(0)`, and `new String('')`)
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Lang
 * @param ***REMOVED*******REMOVED*** value The value to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is an object, else `false`.
 * @example
 *
 * _.isObject(***REMOVED******REMOVED***);
 * // => true
 *
 * _.isObject([1, 2, 3]);
 * // => true
 *
 * _.isObject(_.noop);
 * // => true
 *
 * _.isObject(null);
 * // => false
 */
function isObject(value) ***REMOVED***
  var type = typeof value;
  return !!value && (type == 'object' || type == 'function');
***REMOVED***

/**
 * Checks if `value` is object-like. A value is object-like if it's not `null`
 * and has a `typeof` result of "object".
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Lang
 * @param ***REMOVED*******REMOVED*** value The value to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is object-like, else `false`.
 * @example
 *
 * _.isObjectLike(***REMOVED******REMOVED***);
 * // => true
 *
 * _.isObjectLike([1, 2, 3]);
 * // => true
 *
 * _.isObjectLike(_.noop);
 * // => false
 *
 * _.isObjectLike(null);
 * // => false
 */
function isObjectLike(value) ***REMOVED***
  return !!value && typeof value == 'object';
***REMOVED***

/**
 * Checks if `value` is classified as a `Symbol` primitive or object.
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Lang
 * @param ***REMOVED*******REMOVED*** value The value to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a symbol, else `false`.
 * @example
 *
 * _.isSymbol(Symbol.iterator);
 * // => true
 *
 * _.isSymbol('abc');
 * // => false
 */
function isSymbol(value) ***REMOVED***
  return typeof value == 'symbol' ||
    (isObjectLike(value) && objectToString.call(value) == symbolTag);
***REMOVED***

/**
 * Converts `value` to a string. An empty string is returned for `null`
 * and `undefined` values. The sign of `-0` is preserved.
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Lang
 * @param ***REMOVED*******REMOVED*** value The value to process.
 * @returns ***REMOVED***string***REMOVED*** Returns the string.
 * @example
 *
 * _.toString(null);
 * // => ''
 *
 * _.toString(-0);
 * // => '-0'
 *
 * _.toString([1, 2, 3]);
 * // => '1,2,3'
 */
function toString(value) ***REMOVED***
  return value == null ? '' : baseToString(value);
***REMOVED***

/**
 * Gets the value at `path` of `object`. If the resolved value is
 * `undefined`, the `defaultValue` is returned in its place.
 *
 * @static
 * @memberOf _
 * @since 3.7.0
 * @category Object
 * @param ***REMOVED***Object***REMOVED*** object The object to query.
 * @param ***REMOVED***Array|string***REMOVED*** path The path of the property to get.
 * @param ***REMOVED*******REMOVED*** [defaultValue] The value returned for `undefined` resolved values.
 * @returns ***REMOVED*******REMOVED*** Returns the resolved value.
 * @example
 *
 * var object = ***REMOVED*** 'a': [***REMOVED*** 'b': ***REMOVED*** 'c': 3 ***REMOVED*** ***REMOVED***] ***REMOVED***;
 *
 * _.get(object, 'a[0].b.c');
 * // => 3
 *
 * _.get(object, ['a', '0', 'b', 'c']);
 * // => 3
 *
 * _.get(object, 'a.b.c', 'default');
 * // => 'default'
 */
function get(object, path, defaultValue) ***REMOVED***
  var result = object == null ? undefined : baseGet(object, path);
  return result === undefined ? defaultValue : result;
***REMOVED***

module.exports = get;
