// aws4/aws4.js
var aws4 = exports,
    url = require('url'),
    querystring = require('querystring'),
    crypto = require('crypto'),
    lru = require('./lru'),
    credentialsCache = lru(1000)

// http://docs.amazonwebservices.com/general/latest/gr/signature-version-4.html

function hmac(key, string, encoding) ***REMOVED***
  return crypto.createHmac('sha256', key).update(string, 'utf8').digest(encoding)
***REMOVED***

function hash(string, encoding) ***REMOVED***
  return crypto.createHash('sha256').update(string, 'utf8').digest(encoding)
***REMOVED***

// This function assumes the string has already been percent encoded
function encodeRfc3986(urlEncodedString) ***REMOVED***
  return urlEncodedString.replace(/[!'()*]/g, function(c) ***REMOVED***
    return '%' + c.charCodeAt(0).toString(16).toUpperCase()
  ***REMOVED***)
***REMOVED***

// request: ***REMOVED*** path | body, [host], [method], [headers], [service], [region] ***REMOVED***
// credentials: ***REMOVED*** accessKeyId, secretAccessKey, [sessionToken] ***REMOVED***
function RequestSigner(request, credentials) ***REMOVED***

  if (typeof request === 'string') request = url.parse(request)

  var headers = request.headers = (request.headers || ***REMOVED******REMOVED***),
      hostParts = this.matchHost(request.hostname || request.host || headers.Host || headers.host)

  this.request = request
  this.credentials = credentials || this.defaultCredentials()

  this.service = request.service || hostParts[0] || ''
  this.region = request.region || hostParts[1] || 'us-east-1'

  // SES uses a different domain from the service name
  if (this.service === 'email') this.service = 'ses'

  if (!request.method && request.body)
    request.method = 'POST'

  if (!headers.Host && !headers.host) ***REMOVED***
    headers.Host = request.hostname || request.host || this.createHost()

    // If a port is specified explicitly, use it as is
    if (request.port)
      headers.Host += ':' + request.port
  ***REMOVED***
  if (!request.hostname && !request.host)
    request.hostname = headers.Host || headers.host

  this.isCodeCommitGit = this.service === 'codecommit' && request.method === 'GIT'
***REMOVED***

RequestSigner.prototype.matchHost = function(host) ***REMOVED***
  var match = (host || '').match(/([^\.]+)\.(?:([^\.]*)\.)?amazonaws\.com$/)
  var hostParts = (match || []).slice(1, 3)

  // ES's hostParts are sometimes the other way round, if the value that is expected
  // to be region equals ‘es’ switch them back
  // e.g. search-cluster-name-aaaa00aaaa0aaa0aaaaaaa0aaa.us-east-1.es.amazonaws.com
  if (hostParts[1] === 'es')
    hostParts = hostParts.reverse()

  return hostParts
***REMOVED***

// http://docs.aws.amazon.com/general/latest/gr/rande.html
RequestSigner.prototype.isSingleRegion = function() ***REMOVED***
  // Special case for S3 and SimpleDB in us-east-1
  if (['s3', 'sdb'].indexOf(this.service) >= 0 && this.region === 'us-east-1') return true

  return ['cloudfront', 'ls', 'route53', 'iam', 'importexport', 'sts']
    .indexOf(this.service) >= 0
***REMOVED***

RequestSigner.prototype.createHost = function() ***REMOVED***
  var region = this.isSingleRegion() ? '' :
        (this.service === 's3' && this.region !== 'us-east-1' ? '-' : '.') + this.region,
      service = this.service === 'ses' ? 'email' : this.service
  return service + region + '.amazonaws.com'
***REMOVED***

RequestSigner.prototype.prepareRequest = function() ***REMOVED***
  this.parsePath()

  var request = this.request, headers = request.headers, query

  if (request.signQuery) ***REMOVED***

    this.parsedPath.query = query = this.parsedPath.query || ***REMOVED******REMOVED***

    if (this.credentials.sessionToken)
      query['X-Amz-Security-Token'] = this.credentials.sessionToken

    if (this.service === 's3' && !query['X-Amz-Expires'])
      query['X-Amz-Expires'] = 86400

    if (query['X-Amz-Date'])
      this.datetime = query['X-Amz-Date']
    else
      query['X-Amz-Date'] = this.getDateTime()

    query['X-Amz-Algorithm'] = 'AWS4-HMAC-SHA256'
    query['X-Amz-Credential'] = this.credentials.accessKeyId + '/' + this.credentialString()
    query['X-Amz-SignedHeaders'] = this.signedHeaders()

  ***REMOVED*** else ***REMOVED***

    if (!request.doNotModifyHeaders && !this.isCodeCommitGit) ***REMOVED***
      if (request.body && !headers['Content-Type'] && !headers['content-type'])
        headers['Content-Type'] = 'application/x-www-form-urlencoded; charset=utf-8'

      if (request.body && !headers['Content-Length'] && !headers['content-length'])
        headers['Content-Length'] = Buffer.byteLength(request.body)

      if (this.credentials.sessionToken && !headers['X-Amz-Security-Token'] && !headers['x-amz-security-token'])
        headers['X-Amz-Security-Token'] = this.credentials.sessionToken

      if (this.service === 's3' && !headers['X-Amz-Content-Sha256'] && !headers['x-amz-content-sha256'])
        headers['X-Amz-Content-Sha256'] = hash(this.request.body || '', 'hex')

      if (headers['X-Amz-Date'] || headers['x-amz-date'])
        this.datetime = headers['X-Amz-Date'] || headers['x-amz-date']
      else
        headers['X-Amz-Date'] = this.getDateTime()
    ***REMOVED***

    delete headers.Authorization
    delete headers.authorization
  ***REMOVED***
***REMOVED***

RequestSigner.prototype.sign = function() ***REMOVED***
  if (!this.parsedPath) this.prepareRequest()

  if (this.request.signQuery) ***REMOVED***
    this.parsedPath.query['X-Amz-Signature'] = this.signature()
  ***REMOVED*** else ***REMOVED***
    this.request.headers.Authorization = this.authHeader()
  ***REMOVED***

  this.request.path = this.formatPath()

  return this.request
***REMOVED***

RequestSigner.prototype.getDateTime = function() ***REMOVED***
  if (!this.datetime) ***REMOVED***
    var headers = this.request.headers,
      date = new Date(headers.Date || headers.date || new Date)

    this.datetime = date.toISOString().replace(/[:\-]|\.\d***REMOVED***3***REMOVED***/g, '')

    // Remove the trailing 'Z' on the timestamp string for CodeCommit git access
    if (this.isCodeCommitGit) this.datetime = this.datetime.slice(0, -1)
  ***REMOVED***
  return this.datetime
***REMOVED***

RequestSigner.prototype.getDate = function() ***REMOVED***
  return this.getDateTime().substr(0, 8)
***REMOVED***

RequestSigner.prototype.authHeader = function() ***REMOVED***
  return [
    'AWS4-HMAC-SHA256 Credential=' + this.credentials.accessKeyId + '/' + this.credentialString(),
    'SignedHeaders=' + this.signedHeaders(),
    'Signature=' + this.signature(),
  ].join(', ')
***REMOVED***

RequestSigner.prototype.signature = function() ***REMOVED***
  var date = this.getDate(),
      cacheKey = [this.credentials.secretAccessKey, date, this.region, this.service].join(),
      kDate, kRegion, kService, kCredentials = credentialsCache.get(cacheKey)
  if (!kCredentials) ***REMOVED***
    kDate = hmac('AWS4' + this.credentials.secretAccessKey, date)
    kRegion = hmac(kDate, this.region)
    kService = hmac(kRegion, this.service)
    kCredentials = hmac(kService, 'aws4_request')
    credentialsCache.set(cacheKey, kCredentials)
  ***REMOVED***
  return hmac(kCredentials, this.stringToSign(), 'hex')
***REMOVED***

RequestSigner.prototype.stringToSign = function() ***REMOVED***
  return [
    'AWS4-HMAC-SHA256',
    this.getDateTime(),
    this.credentialString(),
    hash(this.canonicalString(), 'hex'),
  ].join('\n')
***REMOVED***

RequestSigner.prototype.canonicalString = function() ***REMOVED***
  if (!this.parsedPath) this.prepareRequest()

  var pathStr = this.parsedPath.path,
      query = this.parsedPath.query,
      headers = this.request.headers,
      queryStr = '',
      normalizePath = this.service !== 's3',
      decodePath = this.service === 's3' || this.request.doNotEncodePath,
      decodeSlashesInPath = this.service === 's3',
      firstValOnly = this.service === 's3',
      bodyHash

  if (this.service === 's3' && this.request.signQuery) ***REMOVED***
    bodyHash = 'UNSIGNED-PAYLOAD'
  ***REMOVED*** else if (this.isCodeCommitGit) ***REMOVED***
    bodyHash = ''
  ***REMOVED*** else ***REMOVED***
    bodyHash = headers['X-Amz-Content-Sha256'] || headers['x-amz-content-sha256'] ||
      hash(this.request.body || '', 'hex')
  ***REMOVED***

  if (query) ***REMOVED***
    queryStr = encodeRfc3986(querystring.stringify(Object.keys(query).sort().reduce(function(obj, key) ***REMOVED***
      if (!key) return obj
      obj[key] = !Array.isArray(query[key]) ? query[key] :
        (firstValOnly ? query[key][0] : query[key].slice().sort())
      return obj
    ***REMOVED***, ***REMOVED******REMOVED***)))
  ***REMOVED***
  if (pathStr !== '/') ***REMOVED***
    if (normalizePath) pathStr = pathStr.replace(/\/***REMOVED***2,***REMOVED***/g, '/')
    pathStr = pathStr.split('/').reduce(function(path, piece) ***REMOVED***
      if (normalizePath && piece === '..') ***REMOVED***
        path.pop()
      ***REMOVED*** else if (!normalizePath || piece !== '.') ***REMOVED***
        if (decodePath) piece = decodeURIComponent(piece)
        path.push(encodeRfc3986(encodeURIComponent(piece)))
      ***REMOVED***
      return path
    ***REMOVED***, []).join('/')
    if (pathStr[0] !== '/') pathStr = '/' + pathStr
    if (decodeSlashesInPath) pathStr = pathStr.replace(/%2F/g, '/')
  ***REMOVED***

  return [
    this.request.method || 'GET',
    pathStr,
    queryStr,
    this.canonicalHeaders() + '\n',
    this.signedHeaders(),
    bodyHash,
  ].join('\n')
***REMOVED***

RequestSigner.prototype.canonicalHeaders = function() ***REMOVED***
  var headers = this.request.headers
  function trimAll(header) ***REMOVED***
    return header.toString().trim().replace(/\s+/g, ' ')
  ***REMOVED***
  return Object.keys(headers)
    .sort(function(a, b) ***REMOVED*** return a.toLowerCase() < b.toLowerCase() ? -1 : 1 ***REMOVED***)
    .map(function(key) ***REMOVED*** return key.toLowerCase() + ':' + trimAll(headers[key]) ***REMOVED***)
    .join('\n')
***REMOVED***

RequestSigner.prototype.signedHeaders = function() ***REMOVED***
  return Object.keys(this.request.headers)
    .map(function(key) ***REMOVED*** return key.toLowerCase() ***REMOVED***)
    .sort()
    .join(';')
***REMOVED***

RequestSigner.prototype.credentialString = function() ***REMOVED***
  return [
    this.getDate(),
    this.region,
    this.service,
    'aws4_request',
  ].join('/')
***REMOVED***

RequestSigner.prototype.defaultCredentials = function() ***REMOVED***
  var env = process.env
  return ***REMOVED***
    accessKeyId: env.AWS_ACCESS_KEY_ID || env.AWS_ACCESS_KEY,
    secretAccessKey: env.AWS_SECRET_ACCESS_KEY || env.AWS_SECRET_KEY,
    sessionToken: env.AWS_SESSION_TOKEN,
  ***REMOVED***
***REMOVED***

RequestSigner.prototype.parsePath = function() ***REMOVED***
  var path = this.request.path || '/',
      queryIx = path.indexOf('?'),
      query = null

  if (queryIx >= 0) ***REMOVED***
    query = querystring.parse(path.slice(queryIx + 1))
    path = path.slice(0, queryIx)
  ***REMOVED***

  // S3 doesn't always encode characters > 127 correctly and
  // all services don't encode characters > 255 correctly
  // So if there are non-reserved chars (and it's not already all % encoded), just encode them all
  if (/[^0-9A-Za-z!'()*\-._~%/]/.test(path)) ***REMOVED***
    path = path.split('/').map(function(piece) ***REMOVED***
      return encodeURIComponent(decodeURIComponent(piece))
    ***REMOVED***).join('/')
  ***REMOVED***

  this.parsedPath = ***REMOVED***
    path: path,
    query: query,
  ***REMOVED***
***REMOVED***

RequestSigner.prototype.formatPath = function() ***REMOVED***
  var path = this.parsedPath.path,
      query = this.parsedPath.query

  if (!query) return path

  // Services don't support empty query string keys
  if (query[''] != null) delete query['']

  return path + '?' + encodeRfc3986(querystring.stringify(query))
***REMOVED***

aws4.RequestSigner = RequestSigner

aws4.sign = function(request, credentials) ***REMOVED***
  return new RequestSigner(request, credentials).sign()
***REMOVED***
