// aws4/lru.js
module.exports = function(size) ***REMOVED***
  return new LruCache(size)
***REMOVED***

function LruCache(size) ***REMOVED***
  this.capacity = size | 0
  this.map = Object.create(null)
  this.list = new DoublyLinkedList()
***REMOVED***

LruCache.prototype.get = function(key) ***REMOVED***
  var node = this.map[key]
  if (node == null) return undefined
  this.used(node)
  return node.val
***REMOVED***

LruCache.prototype.set = function(key, val) ***REMOVED***
  var node = this.map[key]
  if (node != null) ***REMOVED***
    node.val = val
  ***REMOVED*** else ***REMOVED***
    if (!this.capacity) this.prune()
    if (!this.capacity) return false
    node = new DoublyLinkedNode(key, val)
    this.map[key] = node
    this.capacity--
  ***REMOVED***
  this.used(node)
  return true
***REMOVED***

LruCache.prototype.used = function(node) ***REMOVED***
  this.list.moveToFront(node)
***REMOVED***

LruCache.prototype.prune = function() ***REMOVED***
  var node = this.list.pop()
  if (node != null) ***REMOVED***
    delete this.map[node.key]
    this.capacity++
  ***REMOVED***
***REMOVED***


function DoublyLinkedList() ***REMOVED***
  this.firstNode = null
  this.lastNode = null
***REMOVED***

DoublyLinkedList.prototype.moveToFront = function(node) ***REMOVED***
  if (this.firstNode == node) return

  this.remove(node)

  if (this.firstNode == null) ***REMOVED***
    this.firstNode = node
    this.lastNode = node
    node.prev = null
    node.next = null
  ***REMOVED*** else ***REMOVED***
    node.prev = null
    node.next = this.firstNode
    node.next.prev = node
    this.firstNode = node
  ***REMOVED***
***REMOVED***

DoublyLinkedList.prototype.pop = function() ***REMOVED***
  var lastNode = this.lastNode
  if (lastNode != null) ***REMOVED***
    this.remove(lastNode)
  ***REMOVED***
  return lastNode
***REMOVED***

DoublyLinkedList.prototype.remove = function(node) ***REMOVED***
  if (this.firstNode == node) ***REMOVED***
    this.firstNode = node.next
  ***REMOVED*** else if (node.prev != null) ***REMOVED***
    node.prev.next = node.next
  ***REMOVED***
  if (this.lastNode == node) ***REMOVED***
    this.lastNode = node.prev
  ***REMOVED*** else if (node.next != null) ***REMOVED***
    node.next.prev = node.prev
  ***REMOVED***
***REMOVED***


function DoublyLinkedNode(key, val) ***REMOVED***
  this.key = key
  this.val = val
  this.prev = null
  this.next = null
***REMOVED***
