// uglify-js/lib/mozilla-ast.js
/***********************************************************************

  A JavaScript tokenizer / parser / beautifier / compressor.
  https://github.com/mishoo/UglifyJS2

  -------------------------------- (C) ---------------------------------

                           Author: Mihai Bazon
                         <mihai.bazon@gmail.com>
                       http://mihai.bazon.net/blog

  Distributed under the BSD license:

    Copyright 2012 (c) Mihai Bazon <mihai.bazon@gmail.com>

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:

        * Redistributions of source code must retain the above
          copyright notice, this list of conditions and the following
          disclaimer.

        * Redistributions in binary form must reproduce the above
          copyright notice, this list of conditions and the following
          disclaimer in the documentation and/or other materials
          provided with the distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER “AS IS” AND ANY
    EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
    PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE
    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
    PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
    TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
    THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
    SUCH DAMAGE.

 ***********************************************************************/

"use strict";

(function()***REMOVED***

    var normalize_directives = function(body) ***REMOVED***
        var in_directive = true;

        for (var i = 0; i < body.length; i++) ***REMOVED***
            if (in_directive && body[i] instanceof AST_Statement && body[i].body instanceof AST_String) ***REMOVED***
                body[i] = new AST_Directive(***REMOVED***
                    start: body[i].start,
                    end: body[i].end,
                    value: body[i].body.value
                ***REMOVED***);
            ***REMOVED*** else if (in_directive && !(body[i] instanceof AST_Statement && body[i].body instanceof AST_String)) ***REMOVED***
                in_directive = false;
            ***REMOVED***
        ***REMOVED***

        return body;
    ***REMOVED***;

    var MOZ_TO_ME = ***REMOVED***
        Program: function(M) ***REMOVED***
            return new AST_Toplevel(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                body: normalize_directives(M.body.map(from_moz))
            ***REMOVED***);
        ***REMOVED***,
        FunctionDeclaration: function(M) ***REMOVED***
            return new AST_Defun(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                name: from_moz(M.id),
                argnames: M.params.map(from_moz),
                body: normalize_directives(from_moz(M.body).body)
            ***REMOVED***);
        ***REMOVED***,
        FunctionExpression: function(M) ***REMOVED***
            return new AST_Function(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                name: from_moz(M.id),
                argnames: M.params.map(from_moz),
                body: normalize_directives(from_moz(M.body).body)
            ***REMOVED***);
        ***REMOVED***,
        ExpressionStatement: function(M) ***REMOVED***
            return new AST_SimpleStatement(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                body: from_moz(M.expression)
            ***REMOVED***);
        ***REMOVED***,
        TryStatement: function(M) ***REMOVED***
            var handlers = M.handlers || [M.handler];
            if (handlers.length > 1 || M.guardedHandlers && M.guardedHandlers.length) ***REMOVED***
                throw new Error("Multiple catch clauses are not supported.");
            ***REMOVED***
            return new AST_Try(***REMOVED***
                start    : my_start_token(M),
                end      : my_end_token(M),
                body     : from_moz(M.block).body,
                bcatch   : from_moz(handlers[0]),
                bfinally : M.finalizer ? new AST_Finally(from_moz(M.finalizer)) : null
            ***REMOVED***);
        ***REMOVED***,
        Property: function(M) ***REMOVED***
            var key = M.key;
            var args = ***REMOVED***
                start    : my_start_token(key),
                end      : my_end_token(M.value),
                key      : key.type == "Identifier" ? key.name : key.value,
                value    : from_moz(M.value)
            ***REMOVED***;
            if (M.kind == "init") return new AST_ObjectKeyVal(args);
            args.key = new AST_SymbolAccessor(***REMOVED***
                name: args.key
            ***REMOVED***);
            args.value = new AST_Accessor(args.value);
            if (M.kind == "get") return new AST_ObjectGetter(args);
            if (M.kind == "set") return new AST_ObjectSetter(args);
        ***REMOVED***,
        ArrayExpression: function(M) ***REMOVED***
            return new AST_Array(***REMOVED***
                start    : my_start_token(M),
                end      : my_end_token(M),
                elements : M.elements.map(function(elem)***REMOVED***
                    return elem === null ? new AST_Hole() : from_moz(elem);
                ***REMOVED***)
            ***REMOVED***);
        ***REMOVED***,
        ObjectExpression: function(M) ***REMOVED***
            return new AST_Object(***REMOVED***
                start      : my_start_token(M),
                end        : my_end_token(M),
                properties : M.properties.map(function(prop)***REMOVED***
                    prop.type = "Property";
                    return from_moz(prop)
                ***REMOVED***)
            ***REMOVED***);
        ***REMOVED***,
        SequenceExpression: function(M) ***REMOVED***
            return AST_Seq.from_array(M.expressions.map(from_moz));
        ***REMOVED***,
        MemberExpression: function(M) ***REMOVED***
            return new (M.computed ? AST_Sub : AST_Dot)(***REMOVED***
                start      : my_start_token(M),
                end        : my_end_token(M),
                property   : M.computed ? from_moz(M.property) : M.property.name,
                expression : from_moz(M.object)
            ***REMOVED***);
        ***REMOVED***,
        SwitchCase: function(M) ***REMOVED***
            return new (M.test ? AST_Case : AST_Default)(***REMOVED***
                start      : my_start_token(M),
                end        : my_end_token(M),
                expression : from_moz(M.test),
                body       : M.consequent.map(from_moz)
            ***REMOVED***);
        ***REMOVED***,
        VariableDeclaration: function(M) ***REMOVED***
            return new (M.kind === "const" ? AST_Const : AST_Var)(***REMOVED***
                start       : my_start_token(M),
                end         : my_end_token(M),
                definitions : M.declarations.map(from_moz)
            ***REMOVED***);
        ***REMOVED***,
        Literal: function(M) ***REMOVED***
            var val = M.value, args = ***REMOVED***
                start  : my_start_token(M),
                end    : my_end_token(M)
            ***REMOVED***;
            if (val === null) return new AST_Null(args);
            switch (typeof val) ***REMOVED***
              case "string":
                args.value = val;
                return new AST_String(args);
              case "number":
                args.value = val;
                return new AST_Number(args);
              case "boolean":
                return new (val ? AST_True : AST_False)(args);
              default:
                var rx = M.regex;
                if (rx && rx.pattern) ***REMOVED***
                    // RegExpLiteral as per ESTree AST spec
                    args.value = new RegExp(rx.pattern, rx.flags).toString();
                ***REMOVED*** else ***REMOVED***
                    // support legacy RegExp
                    args.value = M.regex && M.raw ? M.raw : val;
                ***REMOVED***
                return new AST_RegExp(args);
            ***REMOVED***
        ***REMOVED***,
        Identifier: function(M) ***REMOVED***
            var p = FROM_MOZ_STACK[FROM_MOZ_STACK.length - 2];
            return new (  p.type == "LabeledStatement" ? AST_Label
                        : p.type == "VariableDeclarator" && p.id === M ? (p.kind == "const" ? AST_SymbolConst : AST_SymbolVar)
                        : p.type == "FunctionExpression" ? (p.id === M ? AST_SymbolLambda : AST_SymbolFunarg)
                        : p.type == "FunctionDeclaration" ? (p.id === M ? AST_SymbolDefun : AST_SymbolFunarg)
                        : p.type == "CatchClause" ? AST_SymbolCatch
                        : p.type == "BreakStatement" || p.type == "ContinueStatement" ? AST_LabelRef
                        : AST_SymbolRef)(***REMOVED***
                            start : my_start_token(M),
                            end   : my_end_token(M),
                            name  : M.name
                        ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;

    MOZ_TO_ME.UpdateExpression =
    MOZ_TO_ME.UnaryExpression = function To_Moz_Unary(M) ***REMOVED***
        var prefix = "prefix" in M ? M.prefix
            : M.type == "UnaryExpression" ? true : false;
        return new (prefix ? AST_UnaryPrefix : AST_UnaryPostfix)(***REMOVED***
            start      : my_start_token(M),
            end        : my_end_token(M),
            operator   : M.operator,
            expression : from_moz(M.argument)
        ***REMOVED***);
    ***REMOVED***;

    map("EmptyStatement", AST_EmptyStatement);
    map("BlockStatement", AST_BlockStatement, "body@body");
    map("IfStatement", AST_If, "test>condition, consequent>body, alternate>alternative");
    map("LabeledStatement", AST_LabeledStatement, "label>label, body>body");
    map("BreakStatement", AST_Break, "label>label");
    map("ContinueStatement", AST_Continue, "label>label");
    map("WithStatement", AST_With, "object>expression, body>body");
    map("SwitchStatement", AST_Switch, "discriminant>expression, cases@body");
    map("ReturnStatement", AST_Return, "argument>value");
    map("ThrowStatement", AST_Throw, "argument>value");
    map("WhileStatement", AST_While, "test>condition, body>body");
    map("DoWhileStatement", AST_Do, "test>condition, body>body");
    map("ForStatement", AST_For, "init>init, test>condition, update>step, body>body");
    map("ForInStatement", AST_ForIn, "left>init, right>object, body>body");
    map("DebuggerStatement", AST_Debugger);
    map("VariableDeclarator", AST_VarDef, "id>name, init>value");
    map("CatchClause", AST_Catch, "param>argname, body%body");

    map("ThisExpression", AST_This);
    map("BinaryExpression", AST_Binary, "operator=operator, left>left, right>right");
    map("LogicalExpression", AST_Binary, "operator=operator, left>left, right>right");
    map("AssignmentExpression", AST_Assign, "operator=operator, left>left, right>right");
    map("ConditionalExpression", AST_Conditional, "test>condition, consequent>consequent, alternate>alternative");
    map("NewExpression", AST_New, "callee>expression, arguments@args");
    map("CallExpression", AST_Call, "callee>expression, arguments@args");

    def_to_moz(AST_Toplevel, function To_Moz_Program(M) ***REMOVED***
        return to_moz_scope("Program", M);
    ***REMOVED***);

    def_to_moz(AST_Defun, function To_Moz_FunctionDeclaration(M) ***REMOVED***
        return ***REMOVED***
            type: "FunctionDeclaration",
            id: to_moz(M.name),
            params: M.argnames.map(to_moz),
            body: to_moz_scope("BlockStatement", M)
        ***REMOVED***
    ***REMOVED***);

    def_to_moz(AST_Function, function To_Moz_FunctionExpression(M) ***REMOVED***
        return ***REMOVED***
            type: "FunctionExpression",
            id: to_moz(M.name),
            params: M.argnames.map(to_moz),
            body: to_moz_scope("BlockStatement", M)
        ***REMOVED***
    ***REMOVED***);

    def_to_moz(AST_Directive, function To_Moz_Directive(M) ***REMOVED***
        return ***REMOVED***
            type: "ExpressionStatement",
            expression: ***REMOVED***
                type: "Literal",
                value: M.value
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_SimpleStatement, function To_Moz_ExpressionStatement(M) ***REMOVED***
        return ***REMOVED***
            type: "ExpressionStatement",
            expression: to_moz(M.body)
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_SwitchBranch, function To_Moz_SwitchCase(M) ***REMOVED***
        return ***REMOVED***
            type: "SwitchCase",
            test: to_moz(M.expression),
            consequent: M.body.map(to_moz)
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_Try, function To_Moz_TryStatement(M) ***REMOVED***
        return ***REMOVED***
            type: "TryStatement",
            block: to_moz_block(M),
            handler: to_moz(M.bcatch),
            guardedHandlers: [],
            finalizer: to_moz(M.bfinally)
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_Catch, function To_Moz_CatchClause(M) ***REMOVED***
        return ***REMOVED***
            type: "CatchClause",
            param: to_moz(M.argname),
            guard: null,
            body: to_moz_block(M)
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_Definitions, function To_Moz_VariableDeclaration(M) ***REMOVED***
        return ***REMOVED***
            type: "VariableDeclaration",
            kind: M instanceof AST_Const ? "const" : "var",
            declarations: M.definitions.map(to_moz)
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_Seq, function To_Moz_SequenceExpression(M) ***REMOVED***
        return ***REMOVED***
            type: "SequenceExpression",
            expressions: M.to_array().map(to_moz)
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_PropAccess, function To_Moz_MemberExpression(M) ***REMOVED***
        var isComputed = M instanceof AST_Sub;
        return ***REMOVED***
            type: "MemberExpression",
            object: to_moz(M.expression),
            computed: isComputed,
            property: isComputed ? to_moz(M.property) : ***REMOVED***type: "Identifier", name: M.property***REMOVED***
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_Unary, function To_Moz_Unary(M) ***REMOVED***
        return ***REMOVED***
            type: M.operator == "++" || M.operator == "--" ? "UpdateExpression" : "UnaryExpression",
            operator: M.operator,
            prefix: M instanceof AST_UnaryPrefix,
            argument: to_moz(M.expression)
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_Binary, function To_Moz_BinaryExpression(M) ***REMOVED***
        return ***REMOVED***
            type: M.operator == "&&" || M.operator == "||" ? "LogicalExpression" : "BinaryExpression",
            left: to_moz(M.left),
            operator: M.operator,
            right: to_moz(M.right)
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_Array, function To_Moz_ArrayExpression(M) ***REMOVED***
        return ***REMOVED***
            type: "ArrayExpression",
            elements: M.elements.map(to_moz)
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_Object, function To_Moz_ObjectExpression(M) ***REMOVED***
        return ***REMOVED***
            type: "ObjectExpression",
            properties: M.properties.map(to_moz)
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_ObjectProperty, function To_Moz_Property(M) ***REMOVED***
        var key = ***REMOVED***
            type: "Literal",
            value: M.key instanceof AST_SymbolAccessor ? M.key.name : M.key
        ***REMOVED***;
        var kind;
        if (M instanceof AST_ObjectKeyVal) ***REMOVED***
            kind = "init";
        ***REMOVED*** else
        if (M instanceof AST_ObjectGetter) ***REMOVED***
            kind = "get";
        ***REMOVED*** else
        if (M instanceof AST_ObjectSetter) ***REMOVED***
            kind = "set";
        ***REMOVED***
        return ***REMOVED***
            type: "Property",
            kind: kind,
            key: key,
            value: to_moz(M.value)
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_Symbol, function To_Moz_Identifier(M) ***REMOVED***
        var def = M.definition();
        return ***REMOVED***
            type: "Identifier",
            name: def ? def.mangled_name || def.name : M.name
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_RegExp, function To_Moz_RegExpLiteral(M) ***REMOVED***
        var value = M.value;
        return ***REMOVED***
            type: "Literal",
            value: value,
            raw: value.toString(),
            regex: ***REMOVED***
                pattern: value.source,
                flags: value.toString().match(/[gimuy]*$/)[0]
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_Constant, function To_Moz_Literal(M) ***REMOVED***
        var value = M.value;
        if (typeof value === 'number' && (value < 0 || (value === 0 && 1 / value < 0))) ***REMOVED***
            return ***REMOVED***
                type: "UnaryExpression",
                operator: "-",
                prefix: true,
                argument: ***REMOVED***
                    type: "Literal",
                    value: -value,
                    raw: M.start.raw
                ***REMOVED***
            ***REMOVED***;
        ***REMOVED***
        return ***REMOVED***
            type: "Literal",
            value: value,
            raw: M.start.raw
        ***REMOVED***;
    ***REMOVED***);

    def_to_moz(AST_Atom, function To_Moz_Atom(M) ***REMOVED***
        return ***REMOVED***
            type: "Identifier",
            name: String(M.value)
        ***REMOVED***;
    ***REMOVED***);

    AST_Boolean.DEFMETHOD("to_mozilla_ast", AST_Constant.prototype.to_mozilla_ast);
    AST_Null.DEFMETHOD("to_mozilla_ast", AST_Constant.prototype.to_mozilla_ast);
    AST_Hole.DEFMETHOD("to_mozilla_ast", function To_Moz_ArrayHole() ***REMOVED*** return null ***REMOVED***);

    AST_Block.DEFMETHOD("to_mozilla_ast", AST_BlockStatement.prototype.to_mozilla_ast);
    AST_Lambda.DEFMETHOD("to_mozilla_ast", AST_Function.prototype.to_mozilla_ast);

    /* -----[ tools ]----- */

    function raw_token(moznode) ***REMOVED***
        if (moznode.type == "Literal") ***REMOVED***
            return moznode.raw != null ? moznode.raw : moznode.value + "";
        ***REMOVED***
    ***REMOVED***

    function my_start_token(moznode) ***REMOVED***
        var loc = moznode.loc, start = loc && loc.start;
        var range = moznode.range;
        return new AST_Token(***REMOVED***
            file    : loc && loc.source,
            line    : start && start.line,
            col     : start && start.column,
            pos     : range ? range[0] : moznode.start,
            endline : start && start.line,
            endcol  : start && start.column,
            endpos  : range ? range[0] : moznode.start,
            raw     : raw_token(moznode),
        ***REMOVED***);
    ***REMOVED***;

    function my_end_token(moznode) ***REMOVED***
        var loc = moznode.loc, end = loc && loc.end;
        var range = moznode.range;
        return new AST_Token(***REMOVED***
            file    : loc && loc.source,
            line    : end && end.line,
            col     : end && end.column,
            pos     : range ? range[1] : moznode.end,
            endline : end && end.line,
            endcol  : end && end.column,
            endpos  : range ? range[1] : moznode.end,
            raw     : raw_token(moznode),
        ***REMOVED***);
    ***REMOVED***;

    function map(moztype, mytype, propmap) ***REMOVED***
        var moz_to_me = "function From_Moz_" + moztype + "(M)***REMOVED***\n";
        moz_to_me += "return new U2." + mytype.name + "(***REMOVED***\n" +
            "start: my_start_token(M),\n" +
            "end: my_end_token(M)";

        var me_to_moz = "function To_Moz_" + moztype + "(M)***REMOVED***\n";
        me_to_moz += "return ***REMOVED***\n" +
            "type: " + JSON.stringify(moztype);

        if (propmap) propmap.split(/\s*,\s*/).forEach(function(prop)***REMOVED***
            var m = /([a-z0-9$_]+)(=|@|>|%)([a-z0-9$_]+)/i.exec(prop);
            if (!m) throw new Error("Can't understand property map: " + prop);
            var moz = m[1], how = m[2], my = m[3];
            moz_to_me += ",\n" + my + ": ";
            me_to_moz += ",\n" + moz + ": ";
            switch (how) ***REMOVED***
                case "@":
                    moz_to_me += "M." + moz + ".map(from_moz)";
                    me_to_moz += "M." +  my + ".map(to_moz)";
                    break;
                case ">":
                    moz_to_me += "from_moz(M." + moz + ")";
                    me_to_moz += "to_moz(M." + my + ")";
                    break;
                case "=":
                    moz_to_me += "M." + moz;
                    me_to_moz += "M." + my;
                    break;
                case "%":
                    moz_to_me += "from_moz(M." + moz + ").body";
                    me_to_moz += "to_moz_block(M)";
                    break;
                default:
                    throw new Error("Can't understand operator in propmap: " + prop);
            ***REMOVED***
        ***REMOVED***);

        moz_to_me += "\n***REMOVED***)\n***REMOVED***";
        me_to_moz += "\n***REMOVED***\n***REMOVED***";

        //moz_to_me = parse(moz_to_me).print_to_string(***REMOVED*** beautify: true ***REMOVED***);
        //me_to_moz = parse(me_to_moz).print_to_string(***REMOVED*** beautify: true ***REMOVED***);
        //console.log(moz_to_me);

        moz_to_me = new Function("U2", "my_start_token", "my_end_token", "from_moz", "return(" + moz_to_me + ")")(
            exports, my_start_token, my_end_token, from_moz
        );
        me_to_moz = new Function("to_moz", "to_moz_block", "to_moz_scope", "return(" + me_to_moz + ")")(
            to_moz, to_moz_block, to_moz_scope
        );
        MOZ_TO_ME[moztype] = moz_to_me;
        def_to_moz(mytype, me_to_moz);
    ***REMOVED***;

    var FROM_MOZ_STACK = null;

    function from_moz(node) ***REMOVED***
        FROM_MOZ_STACK.push(node);
        var ret = node != null ? MOZ_TO_ME[node.type](node) : null;
        FROM_MOZ_STACK.pop();
        return ret;
    ***REMOVED***;

    AST_Node.from_mozilla_ast = function(node)***REMOVED***
        var save_stack = FROM_MOZ_STACK;
        FROM_MOZ_STACK = [];
        var ast = from_moz(node);
        FROM_MOZ_STACK = save_stack;
        return ast;
    ***REMOVED***;

    function set_moz_loc(mynode, moznode, myparent) ***REMOVED***
        var start = mynode.start;
        var end = mynode.end;
        if (start.pos != null && end.endpos != null) ***REMOVED***
            moznode.range = [start.pos, end.endpos];
        ***REMOVED***
        if (start.line) ***REMOVED***
            moznode.loc = ***REMOVED***
                start: ***REMOVED***line: start.line, column: start.col***REMOVED***,
                end: end.endline ? ***REMOVED***line: end.endline, column: end.endcol***REMOVED*** : null
            ***REMOVED***;
            if (start.file) ***REMOVED***
                moznode.loc.source = start.file;
            ***REMOVED***
        ***REMOVED***
        return moznode;
    ***REMOVED***;

    function def_to_moz(mytype, handler) ***REMOVED***
        mytype.DEFMETHOD("to_mozilla_ast", function() ***REMOVED***
            return set_moz_loc(this, handler(this));
        ***REMOVED***);
    ***REMOVED***;

    function to_moz(node) ***REMOVED***
        return node != null ? node.to_mozilla_ast() : null;
    ***REMOVED***;

    function to_moz_block(node) ***REMOVED***
        return ***REMOVED***
            type: "BlockStatement",
            body: node.body.map(to_moz)
        ***REMOVED***;
    ***REMOVED***;

    function to_moz_scope(type, node) ***REMOVED***
        var body = node.body.map(to_moz);
        if (node.body[0] instanceof AST_SimpleStatement && node.body[0].body instanceof AST_String) ***REMOVED***
            body.unshift(to_moz(new AST_EmptyStatement(node.body[0])));
        ***REMOVED***
        return ***REMOVED***
            type: type,
            body: body
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***)();
