// mongodb-core/lib/error.js
'use strict';

var util = require('util');

/**
 * Creates a new MongoError
 * @class
 * @augments Error
 * @param ***REMOVED***Error|string|object***REMOVED*** message The error message
 * @property ***REMOVED***string***REMOVED*** message The error message
 * @property ***REMOVED***string***REMOVED*** stack The error call stack
 * @return ***REMOVED***MongoError***REMOVED*** A MongoError instance
 */
function MongoError(message) ***REMOVED***
  var tmp = Error.apply(this, arguments);
  tmp.name = this.name = 'MongoError';

  if (message instanceof Error) ***REMOVED***
    this.message = message.message;
    this.stack = message.stack;
  ***REMOVED*** else ***REMOVED***
    if (typeof message === 'string') ***REMOVED***
      this.message = message;
    ***REMOVED*** else ***REMOVED***
      this.message = message.message || message.errmsg || message.$err || 'n/a';
      for (var name in message) ***REMOVED***
        this[name] = message[name];
      ***REMOVED***
    ***REMOVED***
    if (Error.captureStackTrace) ***REMOVED***
      Error.captureStackTrace(this, this.constructor);
    ***REMOVED***
  ***REMOVED***
***REMOVED***
util.inherits(MongoError, Error);

/**
 * Creates a new MongoError object
 * @method
 * @param ***REMOVED***Error|string|object***REMOVED*** options The options used to create the error.
 * @return ***REMOVED***MongoError***REMOVED*** A MongoError instance
 * @deprecated Use `new MongoError()` instead.
 */
MongoError.create = function(options) ***REMOVED***
  return new MongoError(options);
***REMOVED***;

/**
 * Creates a new MongoNetworkError
 * @class
 * @param ***REMOVED***Error|string|object***REMOVED*** message The error message
 * @property ***REMOVED***string***REMOVED*** message The error message
 * @property ***REMOVED***string***REMOVED*** stack The error call stack
 * @return ***REMOVED***MongoNetworkError***REMOVED*** A MongoNetworkError instance
 * @extends ***REMOVED***MongoError***REMOVED***
 */
var MongoNetworkError = function(message) ***REMOVED***
  MongoError.call(this, message);
  this.name = 'MongoNetworkError';
***REMOVED***;
util.inherits(MongoNetworkError, MongoError);

/**
 * An error used when attempting to parse a value (like a connection string)
 *
 * @class
 * @param ***REMOVED***Error|string|object***REMOVED*** message The error message
 * @property ***REMOVED***string***REMOVED*** message The error message
 * @return ***REMOVED***MongoParseError***REMOVED*** A MongoNetworkError instance
 * @extends ***REMOVED***MongoError***REMOVED***
 */
const MongoParseError = function(message) ***REMOVED***
  MongoError.call(this, message);
  this.name = 'MongoParseError';
***REMOVED***;
util.inherits(MongoParseError, MongoError);

module.exports = ***REMOVED***
  MongoError: MongoError,
  MongoNetworkError: MongoNetworkError,
  MongoParseError: MongoParseError
***REMOVED***;
