// bluebird/js/release/promise.js
"use strict";
module.exports = function() ***REMOVED***
var makeSelfResolutionError = function () ***REMOVED***
    return new TypeError("circular promise resolution chain\u000a\u000a    See http://goo.gl/MqrFmX\u000a");
***REMOVED***;
var reflectHandler = function() ***REMOVED***
    return new Promise.PromiseInspection(this._target());
***REMOVED***;
var apiRejection = function(msg) ***REMOVED***
    return Promise.reject(new TypeError(msg));
***REMOVED***;
function Proxyable() ***REMOVED******REMOVED***
var UNDEFINED_BINDING = ***REMOVED******REMOVED***;
var util = require("./util");

var getDomain;
if (util.isNode) ***REMOVED***
    getDomain = function() ***REMOVED***
        var ret = process.domain;
        if (ret === undefined) ret = null;
        return ret;
    ***REMOVED***;
***REMOVED*** else ***REMOVED***
    getDomain = function() ***REMOVED***
        return null;
    ***REMOVED***;
***REMOVED***
util.notEnumerableProp(Promise, "_getDomain", getDomain);

var es5 = require("./es5");
var Async = require("./async");
var async = new Async();
es5.defineProperty(Promise, "_async", ***REMOVED***value: async***REMOVED***);
var errors = require("./errors");
var TypeError = Promise.TypeError = errors.TypeError;
Promise.RangeError = errors.RangeError;
var CancellationError = Promise.CancellationError = errors.CancellationError;
Promise.TimeoutError = errors.TimeoutError;
Promise.OperationalError = errors.OperationalError;
Promise.RejectionError = errors.OperationalError;
Promise.AggregateError = errors.AggregateError;
var INTERNAL = function()***REMOVED******REMOVED***;
var APPLY = ***REMOVED******REMOVED***;
var NEXT_FILTER = ***REMOVED******REMOVED***;
var tryConvertToPromise = require("./thenables")(Promise, INTERNAL);
var PromiseArray =
    require("./promise_array")(Promise, INTERNAL,
                               tryConvertToPromise, apiRejection, Proxyable);
var Context = require("./context")(Promise);
 /*jshint unused:false*/
var createContext = Context.create;
var debug = require("./debuggability")(Promise, Context);
var CapturedTrace = debug.CapturedTrace;
var PassThroughHandlerContext =
    require("./finally")(Promise, tryConvertToPromise, NEXT_FILTER);
var catchFilter = require("./catch_filter")(NEXT_FILTER);
var nodebackForPromise = require("./nodeback");
var errorObj = util.errorObj;
var tryCatch = util.tryCatch;
function check(self, executor) ***REMOVED***
    if (self == null || self.constructor !== Promise) ***REMOVED***
        throw new TypeError("the promise constructor cannot be invoked directly\u000a\u000a    See http://goo.gl/MqrFmX\u000a");
    ***REMOVED***
    if (typeof executor !== "function") ***REMOVED***
        throw new TypeError("expecting a function but got " + util.classString(executor));
    ***REMOVED***

***REMOVED***

function Promise(executor) ***REMOVED***
    if (executor !== INTERNAL) ***REMOVED***
        check(this, executor);
    ***REMOVED***
    this._bitField = 0;
    this._fulfillmentHandler0 = undefined;
    this._rejectionHandler0 = undefined;
    this._promise0 = undefined;
    this._receiver0 = undefined;
    this._resolveFromExecutor(executor);
    this._promiseCreated();
    this._fireEvent("promiseCreated", this);
***REMOVED***

Promise.prototype.toString = function () ***REMOVED***
    return "[object Promise]";
***REMOVED***;

Promise.prototype.caught = Promise.prototype["catch"] = function (fn) ***REMOVED***
    var len = arguments.length;
    if (len > 1) ***REMOVED***
        var catchInstances = new Array(len - 1),
            j = 0, i;
        for (i = 0; i < len - 1; ++i) ***REMOVED***
            var item = arguments[i];
            if (util.isObject(item)) ***REMOVED***
                catchInstances[j++] = item;
            ***REMOVED*** else ***REMOVED***
                return apiRejection("Catch statement predicate: " +
                    "expecting an object but got " + util.classString(item));
            ***REMOVED***
        ***REMOVED***
        catchInstances.length = j;
        fn = arguments[i];
        return this.then(undefined, catchFilter(catchInstances, fn, this));
    ***REMOVED***
    return this.then(undefined, fn);
***REMOVED***;

Promise.prototype.reflect = function () ***REMOVED***
    return this._then(reflectHandler,
        reflectHandler, undefined, this, undefined);
***REMOVED***;

Promise.prototype.then = function (didFulfill, didReject) ***REMOVED***
    if (debug.warnings() && arguments.length > 0 &&
        typeof didFulfill !== "function" &&
        typeof didReject !== "function") ***REMOVED***
        var msg = ".then() only accepts functions but was passed: " +
                util.classString(didFulfill);
        if (arguments.length > 1) ***REMOVED***
            msg += ", " + util.classString(didReject);
        ***REMOVED***
        this._warn(msg);
    ***REMOVED***
    return this._then(didFulfill, didReject, undefined, undefined, undefined);
***REMOVED***;

Promise.prototype.done = function (didFulfill, didReject) ***REMOVED***
    var promise =
        this._then(didFulfill, didReject, undefined, undefined, undefined);
    promise._setIsFinal();
***REMOVED***;

Promise.prototype.spread = function (fn) ***REMOVED***
    if (typeof fn !== "function") ***REMOVED***
        return apiRejection("expecting a function but got " + util.classString(fn));
    ***REMOVED***
    return this.all()._then(fn, undefined, undefined, APPLY, undefined);
***REMOVED***;

Promise.prototype.toJSON = function () ***REMOVED***
    var ret = ***REMOVED***
        isFulfilled: false,
        isRejected: false,
        fulfillmentValue: undefined,
        rejectionReason: undefined
    ***REMOVED***;
    if (this.isFulfilled()) ***REMOVED***
        ret.fulfillmentValue = this.value();
        ret.isFulfilled = true;
    ***REMOVED*** else if (this.isRejected()) ***REMOVED***
        ret.rejectionReason = this.reason();
        ret.isRejected = true;
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.prototype.all = function () ***REMOVED***
    if (arguments.length > 0) ***REMOVED***
        this._warn(".all() was passed arguments but it does not take any");
    ***REMOVED***
    return new PromiseArray(this).promise();
***REMOVED***;

Promise.prototype.error = function (fn) ***REMOVED***
    return this.caught(util.originatesFromRejection, fn);
***REMOVED***;

Promise.getNewLibraryCopy = module.exports;

Promise.is = function (val) ***REMOVED***
    return val instanceof Promise;
***REMOVED***;

Promise.fromNode = Promise.fromCallback = function(fn) ***REMOVED***
    var ret = new Promise(INTERNAL);
    ret._captureStackTrace();
    var multiArgs = arguments.length > 1 ? !!Object(arguments[1]).multiArgs
                                         : false;
    var result = tryCatch(fn)(nodebackForPromise(ret, multiArgs));
    if (result === errorObj) ***REMOVED***
        ret._rejectCallback(result.e, true);
    ***REMOVED***
    if (!ret._isFateSealed()) ret._setAsyncGuaranteed();
    return ret;
***REMOVED***;

Promise.all = function (promises) ***REMOVED***
    return new PromiseArray(promises).promise();
***REMOVED***;

Promise.cast = function (obj) ***REMOVED***
    var ret = tryConvertToPromise(obj);
    if (!(ret instanceof Promise)) ***REMOVED***
        ret = new Promise(INTERNAL);
        ret._captureStackTrace();
        ret._setFulfilled();
        ret._rejectionHandler0 = obj;
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.resolve = Promise.fulfilled = Promise.cast;

Promise.reject = Promise.rejected = function (reason) ***REMOVED***
    var ret = new Promise(INTERNAL);
    ret._captureStackTrace();
    ret._rejectCallback(reason, true);
    return ret;
***REMOVED***;

Promise.setScheduler = function(fn) ***REMOVED***
    if (typeof fn !== "function") ***REMOVED***
        throw new TypeError("expecting a function but got " + util.classString(fn));
    ***REMOVED***
    return async.setScheduler(fn);
***REMOVED***;

Promise.prototype._then = function (
    didFulfill,
    didReject,
    _,    receiver,
    internalData
) ***REMOVED***
    var haveInternalData = internalData !== undefined;
    var promise = haveInternalData ? internalData : new Promise(INTERNAL);
    var target = this._target();
    var bitField = target._bitField;

    if (!haveInternalData) ***REMOVED***
        promise._propagateFrom(this, 3);
        promise._captureStackTrace();
        if (receiver === undefined &&
            ((this._bitField & 2097152) !== 0)) ***REMOVED***
            if (!((bitField & 50397184) === 0)) ***REMOVED***
                receiver = this._boundValue();
            ***REMOVED*** else ***REMOVED***
                receiver = target === this ? undefined : this._boundTo;
            ***REMOVED***
        ***REMOVED***
        this._fireEvent("promiseChained", this, promise);
    ***REMOVED***

    var domain = getDomain();
    if (!((bitField & 50397184) === 0)) ***REMOVED***
        var handler, value, settler = target._settlePromiseCtx;
        if (((bitField & 33554432) !== 0)) ***REMOVED***
            value = target._rejectionHandler0;
            handler = didFulfill;
        ***REMOVED*** else if (((bitField & 16777216) !== 0)) ***REMOVED***
            value = target._fulfillmentHandler0;
            handler = didReject;
            target._unsetRejectionIsUnhandled();
        ***REMOVED*** else ***REMOVED***
            settler = target._settlePromiseLateCancellationObserver;
            value = new CancellationError("late cancellation observer");
            target._attachExtraTrace(value);
            handler = didReject;
        ***REMOVED***

        async.invoke(settler, target, ***REMOVED***
            handler: domain === null ? handler
                : (typeof handler === "function" &&
                    util.domainBind(domain, handler)),
            promise: promise,
            receiver: receiver,
            value: value
        ***REMOVED***);
    ***REMOVED*** else ***REMOVED***
        target._addCallbacks(didFulfill, didReject, promise, receiver, domain);
    ***REMOVED***

    return promise;
***REMOVED***;

Promise.prototype._length = function () ***REMOVED***
    return this._bitField & 65535;
***REMOVED***;

Promise.prototype._isFateSealed = function () ***REMOVED***
    return (this._bitField & 117506048) !== 0;
***REMOVED***;

Promise.prototype._isFollowing = function () ***REMOVED***
    return (this._bitField & 67108864) === 67108864;
***REMOVED***;

Promise.prototype._setLength = function (len) ***REMOVED***
    this._bitField = (this._bitField & -65536) |
        (len & 65535);
***REMOVED***;

Promise.prototype._setFulfilled = function () ***REMOVED***
    this._bitField = this._bitField | 33554432;
    this._fireEvent("promiseFulfilled", this);
***REMOVED***;

Promise.prototype._setRejected = function () ***REMOVED***
    this._bitField = this._bitField | 16777216;
    this._fireEvent("promiseRejected", this);
***REMOVED***;

Promise.prototype._setFollowing = function () ***REMOVED***
    this._bitField = this._bitField | 67108864;
    this._fireEvent("promiseResolved", this);
***REMOVED***;

Promise.prototype._setIsFinal = function () ***REMOVED***
    this._bitField = this._bitField | 4194304;
***REMOVED***;

Promise.prototype._isFinal = function () ***REMOVED***
    return (this._bitField & 4194304) > 0;
***REMOVED***;

Promise.prototype._unsetCancelled = function() ***REMOVED***
    this._bitField = this._bitField & (~65536);
***REMOVED***;

Promise.prototype._setCancelled = function() ***REMOVED***
    this._bitField = this._bitField | 65536;
    this._fireEvent("promiseCancelled", this);
***REMOVED***;

Promise.prototype._setWillBeCancelled = function() ***REMOVED***
    this._bitField = this._bitField | 8388608;
***REMOVED***;

Promise.prototype._setAsyncGuaranteed = function() ***REMOVED***
    if (async.hasCustomScheduler()) return;
    this._bitField = this._bitField | 134217728;
***REMOVED***;

Promise.prototype._receiverAt = function (index) ***REMOVED***
    var ret = index === 0 ? this._receiver0 : this[
            index * 4 - 4 + 3];
    if (ret === UNDEFINED_BINDING) ***REMOVED***
        return undefined;
    ***REMOVED*** else if (ret === undefined && this._isBound()) ***REMOVED***
        return this._boundValue();
    ***REMOVED***
    return ret;
***REMOVED***;

Promise.prototype._promiseAt = function (index) ***REMOVED***
    return this[
            index * 4 - 4 + 2];
***REMOVED***;

Promise.prototype._fulfillmentHandlerAt = function (index) ***REMOVED***
    return this[
            index * 4 - 4 + 0];
***REMOVED***;

Promise.prototype._rejectionHandlerAt = function (index) ***REMOVED***
    return this[
            index * 4 - 4 + 1];
***REMOVED***;

Promise.prototype._boundValue = function() ***REMOVED******REMOVED***;

Promise.prototype._migrateCallback0 = function (follower) ***REMOVED***
    var bitField = follower._bitField;
    var fulfill = follower._fulfillmentHandler0;
    var reject = follower._rejectionHandler0;
    var promise = follower._promise0;
    var receiver = follower._receiverAt(0);
    if (receiver === undefined) receiver = UNDEFINED_BINDING;
    this._addCallbacks(fulfill, reject, promise, receiver, null);
***REMOVED***;

Promise.prototype._migrateCallbackAt = function (follower, index) ***REMOVED***
    var fulfill = follower._fulfillmentHandlerAt(index);
    var reject = follower._rejectionHandlerAt(index);
    var promise = follower._promiseAt(index);
    var receiver = follower._receiverAt(index);
    if (receiver === undefined) receiver = UNDEFINED_BINDING;
    this._addCallbacks(fulfill, reject, promise, receiver, null);
***REMOVED***;

Promise.prototype._addCallbacks = function (
    fulfill,
    reject,
    promise,
    receiver,
    domain
) ***REMOVED***
    var index = this._length();

    if (index >= 65535 - 4) ***REMOVED***
        index = 0;
        this._setLength(0);
    ***REMOVED***

    if (index === 0) ***REMOVED***
        this._promise0 = promise;
        this._receiver0 = receiver;
        if (typeof fulfill === "function") ***REMOVED***
            this._fulfillmentHandler0 =
                domain === null ? fulfill : util.domainBind(domain, fulfill);
        ***REMOVED***
        if (typeof reject === "function") ***REMOVED***
            this._rejectionHandler0 =
                domain === null ? reject : util.domainBind(domain, reject);
        ***REMOVED***
    ***REMOVED*** else ***REMOVED***
        var base = index * 4 - 4;
        this[base + 2] = promise;
        this[base + 3] = receiver;
        if (typeof fulfill === "function") ***REMOVED***
            this[base + 0] =
                domain === null ? fulfill : util.domainBind(domain, fulfill);
        ***REMOVED***
        if (typeof reject === "function") ***REMOVED***
            this[base + 1] =
                domain === null ? reject : util.domainBind(domain, reject);
        ***REMOVED***
    ***REMOVED***
    this._setLength(index + 1);
    return index;
***REMOVED***;

Promise.prototype._proxy = function (proxyable, arg) ***REMOVED***
    this._addCallbacks(undefined, undefined, arg, proxyable, null);
***REMOVED***;

Promise.prototype._resolveCallback = function(value, shouldBind) ***REMOVED***
    if (((this._bitField & 117506048) !== 0)) return;
    if (value === this)
        return this._rejectCallback(makeSelfResolutionError(), false);
    var maybePromise = tryConvertToPromise(value, this);
    if (!(maybePromise instanceof Promise)) return this._fulfill(value);

    if (shouldBind) this._propagateFrom(maybePromise, 2);

    var promise = maybePromise._target();

    if (promise === this) ***REMOVED***
        this._reject(makeSelfResolutionError());
        return;
    ***REMOVED***

    var bitField = promise._bitField;
    if (((bitField & 50397184) === 0)) ***REMOVED***
        var len = this._length();
        if (len > 0) promise._migrateCallback0(this);
        for (var i = 1; i < len; ++i) ***REMOVED***
            promise._migrateCallbackAt(this, i);
        ***REMOVED***
        this._setFollowing();
        this._setLength(0);
        this._setFollowee(promise);
    ***REMOVED*** else if (((bitField & 33554432) !== 0)) ***REMOVED***
        this._fulfill(promise._value());
    ***REMOVED*** else if (((bitField & 16777216) !== 0)) ***REMOVED***
        this._reject(promise._reason());
    ***REMOVED*** else ***REMOVED***
        var reason = new CancellationError("late cancellation observer");
        promise._attachExtraTrace(reason);
        this._reject(reason);
    ***REMOVED***
***REMOVED***;

Promise.prototype._rejectCallback =
function(reason, synchronous, ignoreNonErrorWarnings) ***REMOVED***
    var trace = util.ensureErrorObject(reason);
    var hasStack = trace === reason;
    if (!hasStack && !ignoreNonErrorWarnings && debug.warnings()) ***REMOVED***
        var message = "a promise was rejected with a non-error: " +
            util.classString(reason);
        this._warn(message, true);
    ***REMOVED***
    this._attachExtraTrace(trace, synchronous ? hasStack : false);
    this._reject(reason);
***REMOVED***;

Promise.prototype._resolveFromExecutor = function (executor) ***REMOVED***
    if (executor === INTERNAL) return;
    var promise = this;
    this._captureStackTrace();
    this._pushContext();
    var synchronous = true;
    var r = this._execute(executor, function(value) ***REMOVED***
        promise._resolveCallback(value);
    ***REMOVED***, function (reason) ***REMOVED***
        promise._rejectCallback(reason, synchronous);
    ***REMOVED***);
    synchronous = false;
    this._popContext();

    if (r !== undefined) ***REMOVED***
        promise._rejectCallback(r, true);
    ***REMOVED***
***REMOVED***;

Promise.prototype._settlePromiseFromHandler = function (
    handler, receiver, value, promise
) ***REMOVED***
    var bitField = promise._bitField;
    if (((bitField & 65536) !== 0)) return;
    promise._pushContext();
    var x;
    if (receiver === APPLY) ***REMOVED***
        if (!value || typeof value.length !== "number") ***REMOVED***
            x = errorObj;
            x.e = new TypeError("cannot .spread() a non-array: " +
                                    util.classString(value));
        ***REMOVED*** else ***REMOVED***
            x = tryCatch(handler).apply(this._boundValue(), value);
        ***REMOVED***
    ***REMOVED*** else ***REMOVED***
        x = tryCatch(handler).call(receiver, value);
    ***REMOVED***
    var promiseCreated = promise._popContext();
    bitField = promise._bitField;
    if (((bitField & 65536) !== 0)) return;

    if (x === NEXT_FILTER) ***REMOVED***
        promise._reject(value);
    ***REMOVED*** else if (x === errorObj) ***REMOVED***
        promise._rejectCallback(x.e, false);
    ***REMOVED*** else ***REMOVED***
        debug.checkForgottenReturns(x, promiseCreated, "",  promise, this);
        promise._resolveCallback(x);
    ***REMOVED***
***REMOVED***;

Promise.prototype._target = function() ***REMOVED***
    var ret = this;
    while (ret._isFollowing()) ret = ret._followee();
    return ret;
***REMOVED***;

Promise.prototype._followee = function() ***REMOVED***
    return this._rejectionHandler0;
***REMOVED***;

Promise.prototype._setFollowee = function(promise) ***REMOVED***
    this._rejectionHandler0 = promise;
***REMOVED***;

Promise.prototype._settlePromise = function(promise, handler, receiver, value) ***REMOVED***
    var isPromise = promise instanceof Promise;
    var bitField = this._bitField;
    var asyncGuaranteed = ((bitField & 134217728) !== 0);
    if (((bitField & 65536) !== 0)) ***REMOVED***
        if (isPromise) promise._invokeInternalOnCancel();

        if (receiver instanceof PassThroughHandlerContext &&
            receiver.isFinallyHandler()) ***REMOVED***
            receiver.cancelPromise = promise;
            if (tryCatch(handler).call(receiver, value) === errorObj) ***REMOVED***
                promise._reject(errorObj.e);
            ***REMOVED***
        ***REMOVED*** else if (handler === reflectHandler) ***REMOVED***
            promise._fulfill(reflectHandler.call(receiver));
        ***REMOVED*** else if (receiver instanceof Proxyable) ***REMOVED***
            receiver._promiseCancelled(promise);
        ***REMOVED*** else if (isPromise || promise instanceof PromiseArray) ***REMOVED***
            promise._cancel();
        ***REMOVED*** else ***REMOVED***
            receiver.cancel();
        ***REMOVED***
    ***REMOVED*** else if (typeof handler === "function") ***REMOVED***
        if (!isPromise) ***REMOVED***
            handler.call(receiver, value, promise);
        ***REMOVED*** else ***REMOVED***
            if (asyncGuaranteed) promise._setAsyncGuaranteed();
            this._settlePromiseFromHandler(handler, receiver, value, promise);
        ***REMOVED***
    ***REMOVED*** else if (receiver instanceof Proxyable) ***REMOVED***
        if (!receiver._isResolved()) ***REMOVED***
            if (((bitField & 33554432) !== 0)) ***REMOVED***
                receiver._promiseFulfilled(value, promise);
            ***REMOVED*** else ***REMOVED***
                receiver._promiseRejected(value, promise);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED*** else if (isPromise) ***REMOVED***
        if (asyncGuaranteed) promise._setAsyncGuaranteed();
        if (((bitField & 33554432) !== 0)) ***REMOVED***
            promise._fulfill(value);
        ***REMOVED*** else ***REMOVED***
            promise._reject(value);
        ***REMOVED***
    ***REMOVED***
***REMOVED***;

Promise.prototype._settlePromiseLateCancellationObserver = function(ctx) ***REMOVED***
    var handler = ctx.handler;
    var promise = ctx.promise;
    var receiver = ctx.receiver;
    var value = ctx.value;
    if (typeof handler === "function") ***REMOVED***
        if (!(promise instanceof Promise)) ***REMOVED***
            handler.call(receiver, value, promise);
        ***REMOVED*** else ***REMOVED***
            this._settlePromiseFromHandler(handler, receiver, value, promise);
        ***REMOVED***
    ***REMOVED*** else if (promise instanceof Promise) ***REMOVED***
        promise._reject(value);
    ***REMOVED***
***REMOVED***;

Promise.prototype._settlePromiseCtx = function(ctx) ***REMOVED***
    this._settlePromise(ctx.promise, ctx.handler, ctx.receiver, ctx.value);
***REMOVED***;

Promise.prototype._settlePromise0 = function(handler, value, bitField) ***REMOVED***
    var promise = this._promise0;
    var receiver = this._receiverAt(0);
    this._promise0 = undefined;
    this._receiver0 = undefined;
    this._settlePromise(promise, handler, receiver, value);
***REMOVED***;

Promise.prototype._clearCallbackDataAtIndex = function(index) ***REMOVED***
    var base = index * 4 - 4;
    this[base + 2] =
    this[base + 3] =
    this[base + 0] =
    this[base + 1] = undefined;
***REMOVED***;

Promise.prototype._fulfill = function (value) ***REMOVED***
    var bitField = this._bitField;
    if (((bitField & 117506048) >>> 16)) return;
    if (value === this) ***REMOVED***
        var err = makeSelfResolutionError();
        this._attachExtraTrace(err);
        return this._reject(err);
    ***REMOVED***
    this._setFulfilled();
    this._rejectionHandler0 = value;

    if ((bitField & 65535) > 0) ***REMOVED***
        if (((bitField & 134217728) !== 0)) ***REMOVED***
            this._settlePromises();
        ***REMOVED*** else ***REMOVED***
            async.settlePromises(this);
        ***REMOVED***
    ***REMOVED***
***REMOVED***;

Promise.prototype._reject = function (reason) ***REMOVED***
    var bitField = this._bitField;
    if (((bitField & 117506048) >>> 16)) return;
    this._setRejected();
    this._fulfillmentHandler0 = reason;

    if (this._isFinal()) ***REMOVED***
        return async.fatalError(reason, util.isNode);
    ***REMOVED***

    if ((bitField & 65535) > 0) ***REMOVED***
        async.settlePromises(this);
    ***REMOVED*** else ***REMOVED***
        this._ensurePossibleRejectionHandled();
    ***REMOVED***
***REMOVED***;

Promise.prototype._fulfillPromises = function (len, value) ***REMOVED***
    for (var i = 1; i < len; i++) ***REMOVED***
        var handler = this._fulfillmentHandlerAt(i);
        var promise = this._promiseAt(i);
        var receiver = this._receiverAt(i);
        this._clearCallbackDataAtIndex(i);
        this._settlePromise(promise, handler, receiver, value);
    ***REMOVED***
***REMOVED***;

Promise.prototype._rejectPromises = function (len, reason) ***REMOVED***
    for (var i = 1; i < len; i++) ***REMOVED***
        var handler = this._rejectionHandlerAt(i);
        var promise = this._promiseAt(i);
        var receiver = this._receiverAt(i);
        this._clearCallbackDataAtIndex(i);
        this._settlePromise(promise, handler, receiver, reason);
    ***REMOVED***
***REMOVED***;

Promise.prototype._settlePromises = function () ***REMOVED***
    var bitField = this._bitField;
    var len = (bitField & 65535);

    if (len > 0) ***REMOVED***
        if (((bitField & 16842752) !== 0)) ***REMOVED***
            var reason = this._fulfillmentHandler0;
            this._settlePromise0(this._rejectionHandler0, reason, bitField);
            this._rejectPromises(len, reason);
        ***REMOVED*** else ***REMOVED***
            var value = this._rejectionHandler0;
            this._settlePromise0(this._fulfillmentHandler0, value, bitField);
            this._fulfillPromises(len, value);
        ***REMOVED***
        this._setLength(0);
    ***REMOVED***
    this._clearCancellationData();
***REMOVED***;

Promise.prototype._settledValue = function() ***REMOVED***
    var bitField = this._bitField;
    if (((bitField & 33554432) !== 0)) ***REMOVED***
        return this._rejectionHandler0;
    ***REMOVED*** else if (((bitField & 16777216) !== 0)) ***REMOVED***
        return this._fulfillmentHandler0;
    ***REMOVED***
***REMOVED***;

function deferResolve(v) ***REMOVED***this.promise._resolveCallback(v);***REMOVED***
function deferReject(v) ***REMOVED***this.promise._rejectCallback(v, false);***REMOVED***

Promise.defer = Promise.pending = function() ***REMOVED***
    debug.deprecated("Promise.defer", "new Promise");
    var promise = new Promise(INTERNAL);
    return ***REMOVED***
        promise: promise,
        resolve: deferResolve,
        reject: deferReject
    ***REMOVED***;
***REMOVED***;

util.notEnumerableProp(Promise,
                       "_makeSelfResolutionError",
                       makeSelfResolutionError);

require("./method")(Promise, INTERNAL, tryConvertToPromise, apiRejection,
    debug);
require("./bind")(Promise, INTERNAL, tryConvertToPromise, debug);
require("./cancel")(Promise, PromiseArray, apiRejection, debug);
require("./direct_resolve")(Promise);
require("./synchronous_inspection")(Promise);
require("./join")(
    Promise, PromiseArray, tryConvertToPromise, INTERNAL, async, getDomain);
Promise.Promise = Promise;
Promise.version = "3.5.0";
require('./map.js')(Promise, PromiseArray, apiRejection, tryConvertToPromise, INTERNAL, debug);
require('./call_get.js')(Promise);
require('./using.js')(Promise, apiRejection, tryConvertToPromise, createContext, INTERNAL, debug);
require('./timers.js')(Promise, INTERNAL, debug);
require('./generators.js')(Promise, apiRejection, INTERNAL, tryConvertToPromise, Proxyable, debug);
require('./nodeify.js')(Promise);
require('./promisify.js')(Promise, INTERNAL);
require('./props.js')(Promise, PromiseArray, tryConvertToPromise, apiRejection);
require('./race.js')(Promise, INTERNAL, tryConvertToPromise, apiRejection);
require('./reduce.js')(Promise, PromiseArray, apiRejection, tryConvertToPromise, INTERNAL, debug);
require('./settle.js')(Promise, PromiseArray, debug);
require('./some.js')(Promise, PromiseArray, apiRejection);
require('./filter.js')(Promise, INTERNAL);
require('./each.js')(Promise, INTERNAL);
require('./any.js')(Promise);
                                                         
    util.toFastProperties(Promise);                                          
    util.toFastProperties(Promise.prototype);                                
    function fillTypes(value) ***REMOVED***                                              
        var p = new Promise(INTERNAL);                                       
        p._fulfillmentHandler0 = value;                                      
        p._rejectionHandler0 = value;                                        
        p._promise0 = value;                                                 
        p._receiver0 = value;                                                
    ***REMOVED***                                                                        
    // Complete slack tracking, opt out of field-type tracking and           
    // stabilize map                                                         
    fillTypes(***REMOVED***a: 1***REMOVED***);                                                       
    fillTypes(***REMOVED***b: 2***REMOVED***);                                                       
    fillTypes(***REMOVED***c: 3***REMOVED***);                                                       
    fillTypes(1);                                                            
    fillTypes(function()***REMOVED******REMOVED***);                                                 
    fillTypes(undefined);                                                    
    fillTypes(false);                                                        
    fillTypes(new Promise(INTERNAL));                                        
    debug.setBounds(Async.firstLineError, util.lastLineError);               
    return Promise;                                                          

***REMOVED***;
