// boom/lib/index.js
'use strict';

// Load modules

const Hoek = require('hoek');


// Declare internals

const internals = ***REMOVED***
    STATUS_CODES: Object.setPrototypeOf(***REMOVED***
        '100': 'Continue',
        '101': 'Switching Protocols',
        '102': 'Processing',
        '200': 'OK',
        '201': 'Created',
        '202': 'Accepted',
        '203': 'Non-Authoritative Information',
        '204': 'No Content',
        '205': 'Reset Content',
        '206': 'Partial Content',
        '207': 'Multi-Status',
        '300': 'Multiple Choices',
        '301': 'Moved Permanently',
        '302': 'Moved Temporarily',
        '303': 'See Other',
        '304': 'Not Modified',
        '305': 'Use Proxy',
        '307': 'Temporary Redirect',
        '400': 'Bad Request',
        '401': 'Unauthorized',
        '402': 'Payment Required',
        '403': 'Forbidden',
        '404': 'Not Found',
        '405': 'Method Not Allowed',
        '406': 'Not Acceptable',
        '407': 'Proxy Authentication Required',
        '408': 'Request Time-out',
        '409': 'Conflict',
        '410': 'Gone',
        '411': 'Length Required',
        '412': 'Precondition Failed',
        '413': 'Request Entity Too Large',
        '414': 'Request-URI Too Large',
        '415': 'Unsupported Media Type',
        '416': 'Requested Range Not Satisfiable',
        '417': 'Expectation Failed',
        '418': 'I\'m a teapot',
        '422': 'Unprocessable Entity',
        '423': 'Locked',
        '424': 'Failed Dependency',
        '425': 'Unordered Collection',
        '426': 'Upgrade Required',
        '428': 'Precondition Required',
        '429': 'Too Many Requests',
        '431': 'Request Header Fields Too Large',
        '451': 'Unavailable For Legal Reasons',
        '500': 'Internal Server Error',
        '501': 'Not Implemented',
        '502': 'Bad Gateway',
        '503': 'Service Unavailable',
        '504': 'Gateway Time-out',
        '505': 'HTTP Version Not Supported',
        '506': 'Variant Also Negotiates',
        '507': 'Insufficient Storage',
        '509': 'Bandwidth Limit Exceeded',
        '510': 'Not Extended',
        '511': 'Network Authentication Required'
    ***REMOVED***, null)
***REMOVED***;


exports.wrap = function (error, statusCode, message) ***REMOVED***

    Hoek.assert(error instanceof Error, 'Cannot wrap non-Error object');
    Hoek.assert(!error.isBoom || (!statusCode && !message), 'Cannot provide statusCode or message with boom error');

    return (error.isBoom ? error : internals.initialize(error, statusCode || 500, message));
***REMOVED***;


exports.create = function (statusCode, message, data) ***REMOVED***

    return internals.create(statusCode, message, data, exports.create);
***REMOVED***;


internals.create = function (statusCode, message, data, ctor) ***REMOVED***

    if (message instanceof Error) ***REMOVED***
        if (data) ***REMOVED***
            message.data = data;
        ***REMOVED***
        return exports.wrap(message, statusCode);
    ***REMOVED***

    const error = new Error(message ? message : undefined);       // Avoids settings null message
    Error.captureStackTrace(error, ctor);                       // Filter the stack to our external API
    error.data = data || null;
    internals.initialize(error, statusCode);
    return error;
***REMOVED***;


internals.initialize = function (error, statusCode, message) ***REMOVED***

    const numberCode = parseInt(statusCode, 10);
    Hoek.assert(!isNaN(numberCode) && numberCode >= 400, 'First argument must be a number (400+):', statusCode);

    error.isBoom = true;
    error.isServer = numberCode >= 500;

    if (!error.hasOwnProperty('data')) ***REMOVED***
        error.data = null;
    ***REMOVED***

    error.output = ***REMOVED***
        statusCode: numberCode,
        payload: ***REMOVED******REMOVED***,
        headers: ***REMOVED******REMOVED***
    ***REMOVED***;

    error.reformat = internals.reformat;
    error.reformat();

    if (!message &&
        !error.message) ***REMOVED***

        message = error.output.payload.error;
    ***REMOVED***

    if (message) ***REMOVED***
        error.message = (message + (error.message ? ': ' + error.message : ''));
    ***REMOVED***

    return error;
***REMOVED***;


internals.reformat = function () ***REMOVED***

    this.output.payload.statusCode = this.output.statusCode;
    this.output.payload.error = internals.STATUS_CODES[this.output.statusCode] || 'Unknown';

    if (this.output.statusCode === 500) ***REMOVED***
        this.output.payload.message = 'An internal server error occurred';              // Hide actual error from user
    ***REMOVED***
    else if (this.message) ***REMOVED***
        this.output.payload.message = this.message;
    ***REMOVED***
***REMOVED***;


// 4xx Client Errors

exports.badRequest = function (message, data) ***REMOVED***

    return internals.create(400, message, data, exports.badRequest);
***REMOVED***;


exports.unauthorized = function (message, scheme, attributes) ***REMOVED***          // Or function (message, wwwAuthenticate[])

    const err = internals.create(401, message, undefined, exports.unauthorized);

    if (!scheme) ***REMOVED***
        return err;
    ***REMOVED***

    let wwwAuthenticate = '';

    if (typeof scheme === 'string') ***REMOVED***

        // function (message, scheme, attributes)

        wwwAuthenticate = scheme;

        if (attributes || message) ***REMOVED***
            err.output.payload.attributes = ***REMOVED******REMOVED***;
        ***REMOVED***

        if (attributes) ***REMOVED***
            if (typeof attributes === 'string') ***REMOVED***
                wwwAuthenticate = wwwAuthenticate + ' ' + Hoek.escapeHeaderAttribute(attributes);
                err.output.payload.attributes = attributes;
            ***REMOVED***
            else ***REMOVED***
                const names = Object.keys(attributes);
                for (let i = 0; i < names.length; ++i) ***REMOVED***
                    const name = names[i];
                    if (i) ***REMOVED***
                        wwwAuthenticate = wwwAuthenticate + ',';
                    ***REMOVED***

                    let value = attributes[name];
                    if (value === null ||
                        value === undefined) ***REMOVED***              // Value can be zero

                        value = '';
                    ***REMOVED***
                    wwwAuthenticate = wwwAuthenticate + ' ' + name + '="' + Hoek.escapeHeaderAttribute(value.toString()) + '"';
                    err.output.payload.attributes[name] = value;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***


        if (message) ***REMOVED***
            if (attributes) ***REMOVED***
                wwwAuthenticate = wwwAuthenticate + ',';
            ***REMOVED***
            wwwAuthenticate = wwwAuthenticate + ' error="' + Hoek.escapeHeaderAttribute(message) + '"';
            err.output.payload.attributes.error = message;
        ***REMOVED***
        else ***REMOVED***
            err.isMissing = true;
        ***REMOVED***
    ***REMOVED***
    else ***REMOVED***

        // function (message, wwwAuthenticate[])

        const wwwArray = scheme;
        for (let i = 0; i < wwwArray.length; ++i) ***REMOVED***
            if (i) ***REMOVED***
                wwwAuthenticate = wwwAuthenticate + ', ';
            ***REMOVED***

            wwwAuthenticate = wwwAuthenticate + wwwArray[i];
        ***REMOVED***
    ***REMOVED***

    err.output.headers['WWW-Authenticate'] = wwwAuthenticate;

    return err;
***REMOVED***;


exports.paymentRequired = function (message, data) ***REMOVED***

    return internals.create(402, message, data, exports.paymentRequired);
***REMOVED***;


exports.forbidden = function (message, data) ***REMOVED***

    return internals.create(403, message, data, exports.forbidden);
***REMOVED***;


exports.notFound = function (message, data) ***REMOVED***

    return internals.create(404, message, data, exports.notFound);
***REMOVED***;


exports.methodNotAllowed = function (message, data, allow) ***REMOVED***

    const err = internals.create(405, message, data, exports.methodNotAllowed);

    if (typeof allow === 'string') ***REMOVED***
        allow = [allow];
    ***REMOVED***

    if (Array.isArray(allow)) ***REMOVED***
        err.output.headers.Allow = allow.join(', ');
    ***REMOVED***

    return err;
***REMOVED***;


exports.notAcceptable = function (message, data) ***REMOVED***

    return internals.create(406, message, data, exports.notAcceptable);
***REMOVED***;


exports.proxyAuthRequired = function (message, data) ***REMOVED***

    return internals.create(407, message, data, exports.proxyAuthRequired);
***REMOVED***;


exports.clientTimeout = function (message, data) ***REMOVED***

    return internals.create(408, message, data, exports.clientTimeout);
***REMOVED***;


exports.conflict = function (message, data) ***REMOVED***

    return internals.create(409, message, data, exports.conflict);
***REMOVED***;


exports.resourceGone = function (message, data) ***REMOVED***

    return internals.create(410, message, data, exports.resourceGone);
***REMOVED***;


exports.lengthRequired = function (message, data) ***REMOVED***

    return internals.create(411, message, data, exports.lengthRequired);
***REMOVED***;


exports.preconditionFailed = function (message, data) ***REMOVED***

    return internals.create(412, message, data, exports.preconditionFailed);
***REMOVED***;


exports.entityTooLarge = function (message, data) ***REMOVED***

    return internals.create(413, message, data, exports.entityTooLarge);
***REMOVED***;


exports.uriTooLong = function (message, data) ***REMOVED***

    return internals.create(414, message, data, exports.uriTooLong);
***REMOVED***;


exports.unsupportedMediaType = function (message, data) ***REMOVED***

    return internals.create(415, message, data, exports.unsupportedMediaType);
***REMOVED***;


exports.rangeNotSatisfiable = function (message, data) ***REMOVED***

    return internals.create(416, message, data, exports.rangeNotSatisfiable);
***REMOVED***;


exports.expectationFailed = function (message, data) ***REMOVED***

    return internals.create(417, message, data, exports.expectationFailed);
***REMOVED***;


exports.teapot = function (message, data) ***REMOVED***

    return internals.create(418, message, data, exports.teapot);
***REMOVED***;


exports.badData = function (message, data) ***REMOVED***

    return internals.create(422, message, data, exports.badData);
***REMOVED***;


exports.locked = function (message, data) ***REMOVED***

    return internals.create(423, message, data, exports.locked);
***REMOVED***;


exports.preconditionRequired = function (message, data) ***REMOVED***

    return internals.create(428, message, data, exports.preconditionRequired);
***REMOVED***;


exports.tooManyRequests = function (message, data) ***REMOVED***

    return internals.create(429, message, data, exports.tooManyRequests);
***REMOVED***;


exports.illegal = function (message, data) ***REMOVED***

    return internals.create(451, message, data, exports.illegal);
***REMOVED***;


// 5xx Server Errors

exports.internal = function (message, data, statusCode) ***REMOVED***

    return internals.serverError(message, data, statusCode, exports.internal);
***REMOVED***;


internals.serverError = function (message, data, statusCode, ctor) ***REMOVED***

    let error;
    if (data instanceof Error) ***REMOVED***
        error = exports.wrap(data, statusCode, message);
    ***REMOVED***
    else ***REMOVED***
        error = internals.create(statusCode || 500, message, undefined, ctor);
        error.data = data;
    ***REMOVED***

    return error;
***REMOVED***;


exports.notImplemented = function (message, data) ***REMOVED***

    return internals.serverError(message, data, 501, exports.notImplemented);
***REMOVED***;


exports.badGateway = function (message, data) ***REMOVED***

    return internals.serverError(message, data, 502, exports.badGateway);
***REMOVED***;


exports.serverUnavailable = function (message, data) ***REMOVED***

    return internals.serverError(message, data, 503, exports.serverUnavailable);
***REMOVED***;


exports.gatewayTimeout = function (message, data) ***REMOVED***

    return internals.serverError(message, data, 504, exports.gatewayTimeout);
***REMOVED***;


exports.badImplementation = function (message, data) ***REMOVED***

    const err = internals.serverError(message, data, 500, exports.badImplementation);
    err.isDeveloperError = true;
    return err;
***REMOVED***;
