// async/reflect.js
'use strict';

Object.defineProperty(exports, "__esModule", ***REMOVED***
    value: true
***REMOVED***);
exports.default = reflect;

var _initialParams = require('./internal/initialParams');

var _initialParams2 = _interopRequireDefault(_initialParams);

var _slice = require('./internal/slice');

var _slice2 = _interopRequireDefault(_slice);

var _wrapAsync = require('./internal/wrapAsync');

var _wrapAsync2 = _interopRequireDefault(_wrapAsync);

function _interopRequireDefault(obj) ***REMOVED*** return obj && obj.__esModule ? obj : ***REMOVED*** default: obj ***REMOVED***; ***REMOVED***

/**
 * Wraps the async function in another function that always completes with a
 * result object, even when it errors.
 *
 * The result object has either the property `error` or `value`.
 *
 * @name reflect
 * @static
 * @memberOf module:Utils
 * @method
 * @category Util
 * @param ***REMOVED***AsyncFunction***REMOVED*** fn - The async function you want to wrap
 * @returns ***REMOVED***Function***REMOVED*** - A function that always passes null to it's callback as
 * the error. The second argument to the callback will be an `object` with
 * either an `error` or a `value` property.
 * @example
 *
 * async.parallel([
 *     async.reflect(function(callback) ***REMOVED***
 *         // do some stuff ...
 *         callback(null, 'one');
 *     ***REMOVED***),
 *     async.reflect(function(callback) ***REMOVED***
 *         // do some more stuff but error ...
 *         callback('bad stuff happened');
 *     ***REMOVED***),
 *     async.reflect(function(callback) ***REMOVED***
 *         // do some more stuff ...
 *         callback(null, 'two');
 *     ***REMOVED***)
 * ],
 * // optional callback
 * function(err, results) ***REMOVED***
 *     // values
 *     // results[0].value = 'one'
 *     // results[1].error = 'bad stuff happened'
 *     // results[2].value = 'two'
 * ***REMOVED***);
 */
function reflect(fn) ***REMOVED***
    var _fn = (0, _wrapAsync2.default)(fn);
    return (0, _initialParams2.default)(function reflectOn(args, reflectCallback) ***REMOVED***
        args.push(function callback(error, cbArg) ***REMOVED***
            if (error) ***REMOVED***
                reflectCallback(null, ***REMOVED*** error: error ***REMOVED***);
            ***REMOVED*** else ***REMOVED***
                var value;
                if (arguments.length <= 2) ***REMOVED***
                    value = cbArg;
                ***REMOVED*** else ***REMOVED***
                    value = (0, _slice2.default)(arguments, 1);
                ***REMOVED***
                reflectCallback(null, ***REMOVED*** value: value ***REMOVED***);
            ***REMOVED***
        ***REMOVED***);

        return _fn.apply(this, args);
    ***REMOVED***);
***REMOVED***
module.exports = exports['default'];
