// ajv/lib/compile/async.js
'use strict';

var MissingRefError = require('./error_classes').MissingRef;

module.exports = compileAsync;


/**
 * Creates validating function for passed schema with asynchronous loading of missing schemas.
 * `loadSchema` option should be a function that accepts schema uri and returns promise that resolves with the schema.
 * @this  Ajv
 * @param ***REMOVED***Object***REMOVED***   schema schema object
 * @param ***REMOVED***Boolean***REMOVED***  meta optional true to compile meta-schema; this parameter can be skipped
 * @param ***REMOVED***Function***REMOVED*** callback an optional node-style callback, it is called with 2 parameters: error (or null) and validating function.
 * @return ***REMOVED***Promise***REMOVED*** promise that resolves with a validating function.
 */
function compileAsync(schema, meta, callback) ***REMOVED***
  /* eslint no-shadow: 0 */
  /* global Promise */
  /* jshint validthis: true */
  var self = this;
  if (typeof this._opts.loadSchema != 'function')
    throw new Error('options.loadSchema should be a function');

  if (typeof meta == 'function') ***REMOVED***
    callback = meta;
    meta = undefined;
  ***REMOVED***

  var p = loadMetaSchemaOf(schema).then(function () ***REMOVED***
    var schemaObj = self._addSchema(schema, undefined, meta);
    return schemaObj.validate || _compileAsync(schemaObj);
  ***REMOVED***);

  if (callback) ***REMOVED***
    p.then(
      function(v) ***REMOVED*** callback(null, v); ***REMOVED***,
      callback
    );
  ***REMOVED***

  return p;


  function loadMetaSchemaOf(sch) ***REMOVED***
    var $schema = sch.$schema;
    return $schema && !self.getSchema($schema)
            ? compileAsync.call(self, ***REMOVED*** $ref: $schema ***REMOVED***, true)
            : Promise.resolve();
  ***REMOVED***


  function _compileAsync(schemaObj) ***REMOVED***
    try ***REMOVED*** return self._compile(schemaObj); ***REMOVED***
    catch(e) ***REMOVED***
      if (e instanceof MissingRefError) return loadMissingSchema(e);
      throw e;
    ***REMOVED***


    function loadMissingSchema(e) ***REMOVED***
      var ref = e.missingSchema;
      if (added(ref)) throw new Error('Schema ' + ref + ' is loaded but ' + e.missingRef + ' cannot be resolved');

      var schemaPromise = self._loadingSchemas[ref];
      if (!schemaPromise) ***REMOVED***
        schemaPromise = self._loadingSchemas[ref] = self._opts.loadSchema(ref);
        schemaPromise.then(removePromise, removePromise);
      ***REMOVED***

      return schemaPromise.then(function (sch) ***REMOVED***
        if (!added(ref)) ***REMOVED***
          return loadMetaSchemaOf(sch).then(function () ***REMOVED***
            if (!added(ref)) self.addSchema(sch, ref, undefined, meta);
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***).then(function() ***REMOVED***
        return _compileAsync(schemaObj);
      ***REMOVED***);

      function removePromise() ***REMOVED***
        delete self._loadingSchemas[ref];
      ***REMOVED***

      function added(ref) ***REMOVED***
        return self._refs[ref] || self._schemas[ref];
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***
