// mongodb-core/lib/wireprotocol/2_6_support.js
'use strict';

var copy = require('../connection/utils').copy,
  retrieveBSON = require('../connection/utils').retrieveBSON,
  KillCursor = require('../connection/commands').KillCursor,
  GetMore = require('../connection/commands').GetMore,
  Query = require('../connection/commands').Query,
  f = require('util').format,
  MongoError = require('../error').MongoError,
  getReadPreference = require('./shared').getReadPreference;

var BSON = retrieveBSON(),
  Long = BSON.Long;

var WireProtocol = function() ***REMOVED******REMOVED***;

//
// Execute a write operation
var executeWrite = function(pool, bson, type, opsField, ns, ops, options, callback) ***REMOVED***
  if (ops.length === 0) throw new MongoError('insert must contain at least one document');
  if (typeof options === 'function') ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
    options = options || ***REMOVED******REMOVED***;
  ***REMOVED***

  // Split the ns up to get db and collection
  var p = ns.split('.');
  var d = p.shift();
  // Options
  var ordered = typeof options.ordered === 'boolean' ? options.ordered : true;
  var writeConcern = options.writeConcern;

  // return skeleton
  var writeCommand = ***REMOVED******REMOVED***;
  writeCommand[type] = p.join('.');
  writeCommand[opsField] = ops;
  writeCommand.ordered = ordered;

  // Did we specify a write concern
  if (writeConcern && Object.keys(writeConcern).length > 0) ***REMOVED***
    writeCommand.writeConcern = writeConcern;
  ***REMOVED***

  // Do we have bypassDocumentValidation set, then enable it on the write command
  if (options.bypassDocumentValidation === true) ***REMOVED***
    writeCommand.bypassDocumentValidation = options.bypassDocumentValidation;
  ***REMOVED***

  // Options object
  var opts = ***REMOVED*** command: true ***REMOVED***;
  if (typeof options.session !== 'undefined') opts.session = options.session;
  var queryOptions = ***REMOVED*** checkKeys: false, numberToSkip: 0, numberToReturn: 1 ***REMOVED***;
  if (type === 'insert') queryOptions.checkKeys = false;
  if (typeof options.checkKeys === 'boolean') queryOptions.checkKeys = options.checkKeys;
  // Ensure we support serialization of functions
  if (options.serializeFunctions) queryOptions.serializeFunctions = options.serializeFunctions;
  // Do not serialize the undefined fields
  if (options.ignoreUndefined) queryOptions.ignoreUndefined = options.ignoreUndefined;

  try ***REMOVED***
    // Create write command
    var cmd = new Query(bson, f('%s.$cmd', d), writeCommand, queryOptions);
    // Execute command
    pool.write(cmd, opts, callback);
  ***REMOVED*** catch (err) ***REMOVED***
    callback(err);
  ***REMOVED***
***REMOVED***;

//
// Needs to support legacy mass insert as well as ordered/unordered legacy
// emulation
//
WireProtocol.prototype.insert = function(pool, ns, bson, ops, options, callback) ***REMOVED***
  executeWrite(pool, bson, 'insert', 'documents', ns, ops, options, callback);
***REMOVED***;

WireProtocol.prototype.update = function(pool, ns, bson, ops, options, callback) ***REMOVED***
  executeWrite(pool, bson, 'update', 'updates', ns, ops, options, callback);
***REMOVED***;

WireProtocol.prototype.remove = function(pool, ns, bson, ops, options, callback) ***REMOVED***
  executeWrite(pool, bson, 'delete', 'deletes', ns, ops, options, callback);
***REMOVED***;

WireProtocol.prototype.killCursor = function(bson, ns, cursorState, pool, callback) ***REMOVED***
  var cursorId = cursorState.cursorId;
  // Create a kill cursor command
  var killCursor = new KillCursor(bson, ns, [cursorId]);

  // Build killCursor options
  const options = ***REMOVED***
    immediateRelease: true,
    noResponse: true
  ***REMOVED***;

  if (typeof cursorState.session === 'object') ***REMOVED***
    options.session = cursorState.session;
  ***REMOVED***

  // Execute the kill cursor command
  if (pool && pool.isConnected()) ***REMOVED***
    try ***REMOVED***
      pool.write(killCursor, options, callback);
    ***REMOVED*** catch (err) ***REMOVED***
      if (typeof callback === 'function') ***REMOVED***
        callback(err, null);
      ***REMOVED*** else ***REMOVED***
        console.warn(err);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

WireProtocol.prototype.getMore = function(
  bson,
  ns,
  cursorState,
  batchSize,
  raw,
  connection,
  options,
  callback
) ***REMOVED***
  // Create getMore command
  var getMore = new GetMore(bson, ns, cursorState.cursorId, ***REMOVED*** numberToReturn: batchSize ***REMOVED***);

  // Query callback
  var queryCallback = function(err, result) ***REMOVED***
    if (err) return callback(err);
    // Get the raw message
    var r = result.message;

    // If we have a timed out query or a cursor that was killed
    if ((r.responseFlags & (1 << 0)) !== 0) ***REMOVED***
      return callback(new MongoError('cursor does not exist, was killed or timed out'), null);
    ***REMOVED***

    // Ensure we have a Long valie cursor id
    var cursorId = typeof r.cursorId === 'number' ? Long.fromNumber(r.cursorId) : r.cursorId;

    // Set all the values
    cursorState.documents = r.documents;
    cursorState.cursorId = cursorId;

    // Return
    callback(null, null, r.connection);
  ***REMOVED***;

  // Contains any query options
  var queryOptions = ***REMOVED******REMOVED***;

  // If we have a raw query decorate the function
  if (raw) ***REMOVED***
    queryOptions.raw = raw;
  ***REMOVED***

  // Check if we need to promote longs
  if (typeof cursorState.promoteLongs === 'boolean') ***REMOVED***
    queryOptions.promoteLongs = cursorState.promoteLongs;
  ***REMOVED***

  if (typeof cursorState.promoteValues === 'boolean') ***REMOVED***
    queryOptions.promoteValues = cursorState.promoteValues;
  ***REMOVED***

  if (typeof cursorState.promoteBuffers === 'boolean') ***REMOVED***
    queryOptions.promoteBuffers = cursorState.promoteBuffers;
  ***REMOVED***

  if (typeof cursorState.session === 'object') ***REMOVED***
    queryOptions.session = cursorState.session;
  ***REMOVED***

  // Write out the getMore command
  connection.write(getMore, queryOptions, queryCallback);
***REMOVED***;

WireProtocol.prototype.command = function(bson, ns, cmd, cursorState, topology, options) ***REMOVED***
  // Establish type of command
  if (cmd.find) ***REMOVED***
    return setupClassicFind(bson, ns, cmd, cursorState, topology, options);
  ***REMOVED*** else if (cursorState.cursorId != null) ***REMOVED***
    return;
  ***REMOVED*** else if (cmd) ***REMOVED***
    return setupCommand(bson, ns, cmd, cursorState, topology, options);
  ***REMOVED*** else ***REMOVED***
    throw new MongoError(f('command %s does not return a cursor', JSON.stringify(cmd)));
  ***REMOVED***
***REMOVED***;

//
// Execute a find command
var setupClassicFind = function(bson, ns, cmd, cursorState, topology, options) ***REMOVED***
  // Ensure we have at least some options
  options = options || ***REMOVED******REMOVED***;
  // Get the readPreference
  var readPreference = getReadPreference(cmd, options);
  // Set the optional batchSize
  cursorState.batchSize = cmd.batchSize || cursorState.batchSize;
  var numberToReturn = 0;

  // Unpack the limit and batchSize values
  if (cursorState.limit === 0) ***REMOVED***
    numberToReturn = cursorState.batchSize;
  ***REMOVED*** else if (
    cursorState.limit < 0 ||
    cursorState.limit < cursorState.batchSize ||
    (cursorState.limit > 0 && cursorState.batchSize === 0)
  ) ***REMOVED***
    numberToReturn = cursorState.limit;
  ***REMOVED*** else ***REMOVED***
    numberToReturn = cursorState.batchSize;
  ***REMOVED***

  var numberToSkip = cursorState.skip || 0;
  // Build actual find command
  var findCmd = ***REMOVED******REMOVED***;

  // We have a Mongos topology, check if we need to add a readPreference
  if (topology.type === 'mongos' && readPreference) ***REMOVED***
    findCmd['$readPreference'] = readPreference.toJSON();
  ***REMOVED***

  // Add special modifiers to the query
  if (cmd.sort) findCmd['$orderby'] = cmd.sort;
  if (cmd.hint) findCmd['$hint'] = cmd.hint;
  if (cmd.snapshot) findCmd['$snapshot'] = cmd.snapshot;
  if (typeof cmd.returnKey !== 'undefined') findCmd['$returnKey'] = cmd.returnKey;
  if (cmd.maxScan) findCmd['$maxScan'] = cmd.maxScan;
  if (cmd.min) findCmd['$min'] = cmd.min;
  if (cmd.max) findCmd['$max'] = cmd.max;
  if (typeof cmd.showDiskLoc !== 'undefined') findCmd['$showDiskLoc'] = cmd.showDiskLoc;
  if (cmd.comment) findCmd['$comment'] = cmd.comment;
  if (cmd.maxTimeMS) findCmd['$maxTimeMS'] = cmd.maxTimeMS;

  if (cmd.explain) ***REMOVED***
    // nToReturn must be 0 (match all) or negative (match N and close cursor)
    // nToReturn > 0 will give explain results equivalent to limit(0)
    numberToReturn = -Math.abs(cmd.limit || 0);
    findCmd['$explain'] = true;
  ***REMOVED***

  // Add the query
  findCmd['$query'] = cmd.query;

  // Throw on majority readConcern passed in
  if (cmd.readConcern && cmd.readConcern.level !== 'local') ***REMOVED***
    throw new MongoError(
      f('server find command does not support a readConcern level of %s', cmd.readConcern.level)
    );
  ***REMOVED***

  // Remove readConcern, ensure no failing commands
  if (cmd.readConcern) ***REMOVED***
    cmd = copy(cmd);
    delete cmd['readConcern'];
  ***REMOVED***

  // Serialize functions
  var serializeFunctions =
    typeof options.serializeFunctions === 'boolean' ? options.serializeFunctions : false;
  var ignoreUndefined =
    typeof options.ignoreUndefined === 'boolean' ? options.ignoreUndefined : false;

  // Build Query object
  var query = new Query(bson, ns, findCmd, ***REMOVED***
    numberToSkip: numberToSkip,
    numberToReturn: numberToReturn,
    pre32Limit: typeof cmd.limit !== 'undefined' ? cmd.limit : undefined,
    checkKeys: false,
    returnFieldSelector: cmd.fields,
    serializeFunctions: serializeFunctions,
    ignoreUndefined: ignoreUndefined
  ***REMOVED***);

  // Set query flags
  query.slaveOk = readPreference.slaveOk();

  // Set up the option bits for wire protocol
  if (typeof cmd.tailable === 'boolean') ***REMOVED***
    query.tailable = cmd.tailable;
  ***REMOVED***

  if (typeof cmd.oplogReplay === 'boolean') ***REMOVED***
    query.oplogReplay = cmd.oplogReplay;
  ***REMOVED***

  if (typeof cmd.noCursorTimeout === 'boolean') ***REMOVED***
    query.noCursorTimeout = cmd.noCursorTimeout;
  ***REMOVED***

  if (typeof cmd.awaitData === 'boolean') ***REMOVED***
    query.awaitData = cmd.awaitData;
  ***REMOVED***

  if (typeof cmd.partial === 'boolean') ***REMOVED***
    query.partial = cmd.partial;
  ***REMOVED***

  // Return the query
  return query;
***REMOVED***;

//
// Set up a command cursor
var setupCommand = function(bson, ns, cmd, cursorState, topology, options) ***REMOVED***
  // Set empty options object
  options = options || ***REMOVED******REMOVED***;
  // Get the readPreference
  var readPreference = getReadPreference(cmd, options);

  // Final query
  var finalCmd = ***REMOVED******REMOVED***;
  for (var name in cmd) ***REMOVED***
    finalCmd[name] = cmd[name];
  ***REMOVED***

  // Build command namespace
  var parts = ns.split(/\./);

  // Serialize functions
  var serializeFunctions =
    typeof options.serializeFunctions === 'boolean' ? options.serializeFunctions : false;

  var ignoreUndefined =
    typeof options.ignoreUndefined === 'boolean' ? options.ignoreUndefined : false;

  // Throw on majority readConcern passed in
  if (cmd.readConcern && cmd.readConcern.level !== 'local') ***REMOVED***
    throw new MongoError(
      f(
        'server %s command does not support a readConcern level of %s',
        JSON.stringify(cmd),
        cmd.readConcern.level
      )
    );
  ***REMOVED***

  // Remove readConcern, ensure no failing commands
  if (cmd.readConcern) delete cmd['readConcern'];

  // We have a Mongos topology, check if we need to add a readPreference
  if (topology.type === 'mongos' && readPreference && readPreference.preference !== 'primary') ***REMOVED***
    finalCmd = ***REMOVED***
      $query: finalCmd,
      $readPreference: readPreference.toJSON()
    ***REMOVED***;
  ***REMOVED***

  // Build Query object
  var query = new Query(bson, f('%s.$cmd', parts.shift()), finalCmd, ***REMOVED***
    numberToSkip: 0,
    numberToReturn: -1,
    checkKeys: false,
    serializeFunctions: serializeFunctions,
    ignoreUndefined: ignoreUndefined
  ***REMOVED***);

  // Set query flags
  query.slaveOk = readPreference.slaveOk();

  // Return the query
  return query;
***REMOVED***;

module.exports = WireProtocol;
