// mongodb-core/lib/wireprotocol/shared.js
'use strict';

var ReadPreference = require('../topologies/read_preference'),
  MongoError = require('../error').MongoError;

var MESSAGE_HEADER_SIZE = 16;

// OPCODE Numbers
// Defined at https://docs.mongodb.com/manual/reference/mongodb-wire-protocol/#request-opcodes
var opcodes = ***REMOVED***
  OP_REPLY: 1,
  OP_UPDATE: 2001,
  OP_INSERT: 2002,
  OP_QUERY: 2004,
  OP_GETMORE: 2005,
  OP_DELETE: 2006,
  OP_KILL_CURSORS: 2007,
  OP_COMPRESSED: 2012
***REMOVED***;

var getReadPreference = function(cmd, options) ***REMOVED***
  // Default to command version of the readPreference
  var readPreference = cmd.readPreference || new ReadPreference('primary');
  // If we have an option readPreference override the command one
  if (options.readPreference) ***REMOVED***
    readPreference = options.readPreference;
  ***REMOVED***

  if (typeof readPreference === 'string') ***REMOVED***
    readPreference = new ReadPreference(readPreference);
  ***REMOVED***

  if (!(readPreference instanceof ReadPreference)) ***REMOVED***
    throw new MongoError('read preference must be a ReadPreference instance');
  ***REMOVED***

  return readPreference;
***REMOVED***;

// Parses the header of a wire protocol message
var parseHeader = function(message) ***REMOVED***
  return ***REMOVED***
    length: message.readInt32LE(0),
    requestId: message.readInt32LE(4),
    responseTo: message.readInt32LE(8),
    opCode: message.readInt32LE(12)
  ***REMOVED***;
***REMOVED***;

module.exports = ***REMOVED***
  getReadPreference: getReadPreference,
  MESSAGE_HEADER_SIZE: MESSAGE_HEADER_SIZE,
  opcodes: opcodes,
  parseHeader: parseHeader
***REMOVED***;
