// mongoose/lib/error/validation.js
/*!
 * Module requirements
 */

const MongooseError = require('./');

/**
 * Document Validation Error
 *
 * @api private
 * @param ***REMOVED***Document***REMOVED*** instance
 * @inherits MongooseError
 */

function ValidationError(instance) ***REMOVED***
  this.errors = ***REMOVED******REMOVED***;
  this._message = '';
  if (instance && instance.constructor.name === 'model') ***REMOVED***
    this._message = instance.constructor.modelName + ' validation failed';
    MongooseError.call(this, this._message);
  ***REMOVED*** else ***REMOVED***
    this._message = 'Validation failed';
    MongooseError.call(this, this._message);
  ***REMOVED***
  this.name = 'ValidationError';
  if (Error.captureStackTrace) ***REMOVED***
    Error.captureStackTrace(this);
  ***REMOVED*** else ***REMOVED***
    this.stack = new Error().stack;
  ***REMOVED***
  if (instance) ***REMOVED***
    instance.errors = this.errors;
  ***REMOVED***
***REMOVED***

/*!
 * Inherits from MongooseError.
 */

ValidationError.prototype = Object.create(MongooseError.prototype);
ValidationError.prototype.constructor = MongooseError;

/**
 * Console.log helper
 */

ValidationError.prototype.toString = function() ***REMOVED***
  return this.name + ': ' + _generateMessage(this);
***REMOVED***;

/*!
 * inspect helper
 */

ValidationError.prototype.inspect = function() ***REMOVED***
  return Object.assign(new Error(this.message), this);
***REMOVED***;

/*!
 * Helper for JSON.stringify
 */

ValidationError.prototype.toJSON = function() ***REMOVED***
  return Object.assign(***REMOVED******REMOVED***, this, ***REMOVED*** message: this.message ***REMOVED***);
***REMOVED***;

/*!
 * add message
 */

ValidationError.prototype.addError = function(path, error) ***REMOVED***
  this.errors[path] = error;
  this.message = this._message + ': ' + _generateMessage(this);
***REMOVED***;

/*!
 * ignore
 */

function _generateMessage(err) ***REMOVED***
  var keys = Object.keys(err.errors || ***REMOVED******REMOVED***);
  var len = keys.length;
  var msgs = [];
  var key;

  for (var i = 0; i < len; ++i) ***REMOVED***
    key = keys[i];
    if (err === err.errors[key]) ***REMOVED***
      continue;
    ***REMOVED***
    msgs.push(key + ': ' + err.errors[key].message);
  ***REMOVED***

  return msgs.join(', ');
***REMOVED***

/*!
 * Module exports
 */

module.exports = exports = ValidationError;
