// jsbn/index.js
(function()***REMOVED***

    // Copyright (c) 2005  Tom Wu
    // All Rights Reserved.
    // See "LICENSE" for details.

    // Basic JavaScript BN library - subset useful for RSA encryption.

    // Bits per digit
    var dbits;

    // JavaScript engine analysis
    var canary = 0xdeadbeefcafe;
    var j_lm = ((canary&0xffffff)==0xefcafe);

    // (public) Constructor
    function BigInteger(a,b,c) ***REMOVED***
      if(a != null)
        if("number" == typeof a) this.fromNumber(a,b,c);
        else if(b == null && "string" != typeof a) this.fromString(a,256);
        else this.fromString(a,b);
    ***REMOVED***

    // return new, unset BigInteger
    function nbi() ***REMOVED*** return new BigInteger(null); ***REMOVED***

    // am: Compute w_j += (x*this_i), propagate carries,
    // c is initial carry, returns final carry.
    // c < 3*dvalue, x < 2*dvalue, this_i < dvalue
    // We need to select the fastest one that works in this environment.

    // am1: use a single mult and divide to get the high bits,
    // max digit bits should be 26 because
    // max internal value = 2*dvalue^2-2*dvalue (< 2^53)
    function am1(i,x,w,j,c,n) ***REMOVED***
      while(--n >= 0) ***REMOVED***
        var v = x*this[i++]+w[j]+c;
        c = Math.floor(v/0x4000000);
        w[j++] = v&0x3ffffff;
      ***REMOVED***
      return c;
    ***REMOVED***
    // am2 avoids a big mult-and-extract completely.
    // Max digit bits should be <= 30 because we do bitwise ops
    // on values up to 2*hdvalue^2-hdvalue-1 (< 2^31)
    function am2(i,x,w,j,c,n) ***REMOVED***
      var xl = x&0x7fff, xh = x>>15;
      while(--n >= 0) ***REMOVED***
        var l = this[i]&0x7fff;
        var h = this[i++]>>15;
        var m = xh*l+h*xl;
        l = xl*l+((m&0x7fff)<<15)+w[j]+(c&0x3fffffff);
        c = (l>>>30)+(m>>>15)+xh*h+(c>>>30);
        w[j++] = l&0x3fffffff;
      ***REMOVED***
      return c;
    ***REMOVED***
    // Alternately, set max digit bits to 28 since some
    // browsers slow down when dealing with 32-bit numbers.
    function am3(i,x,w,j,c,n) ***REMOVED***
      var xl = x&0x3fff, xh = x>>14;
      while(--n >= 0) ***REMOVED***
        var l = this[i]&0x3fff;
        var h = this[i++]>>14;
        var m = xh*l+h*xl;
        l = xl*l+((m&0x3fff)<<14)+w[j]+c;
        c = (l>>28)+(m>>14)+xh*h;
        w[j++] = l&0xfffffff;
      ***REMOVED***
      return c;
    ***REMOVED***
    var inBrowser = typeof navigator !== "undefined";
    if(inBrowser && j_lm && (navigator.appName == "Microsoft Internet Explorer")) ***REMOVED***
      BigInteger.prototype.am = am2;
      dbits = 30;
    ***REMOVED***
    else if(inBrowser && j_lm && (navigator.appName != "Netscape")) ***REMOVED***
      BigInteger.prototype.am = am1;
      dbits = 26;
    ***REMOVED***
    else ***REMOVED*** // Mozilla/Netscape seems to prefer am3
      BigInteger.prototype.am = am3;
      dbits = 28;
    ***REMOVED***

    BigInteger.prototype.DB = dbits;
    BigInteger.prototype.DM = ((1<<dbits)-1);
    BigInteger.prototype.DV = (1<<dbits);

    var BI_FP = 52;
    BigInteger.prototype.FV = Math.pow(2,BI_FP);
    BigInteger.prototype.F1 = BI_FP-dbits;
    BigInteger.prototype.F2 = 2*dbits-BI_FP;

    // Digit conversions
    var BI_RM = "0123456789abcdefghijklmnopqrstuvwxyz";
    var BI_RC = new Array();
    var rr,vv;
    rr = "0".charCodeAt(0);
    for(vv = 0; vv <= 9; ++vv) BI_RC[rr++] = vv;
    rr = "a".charCodeAt(0);
    for(vv = 10; vv < 36; ++vv) BI_RC[rr++] = vv;
    rr = "A".charCodeAt(0);
    for(vv = 10; vv < 36; ++vv) BI_RC[rr++] = vv;

    function int2char(n) ***REMOVED*** return BI_RM.charAt(n); ***REMOVED***
    function intAt(s,i) ***REMOVED***
      var c = BI_RC[s.charCodeAt(i)];
      return (c==null)?-1:c;
    ***REMOVED***

    // (protected) copy this to r
    function bnpCopyTo(r) ***REMOVED***
      for(var i = this.t-1; i >= 0; --i) r[i] = this[i];
      r.t = this.t;
      r.s = this.s;
    ***REMOVED***

    // (protected) set from integer value x, -DV <= x < DV
    function bnpFromInt(x) ***REMOVED***
      this.t = 1;
      this.s = (x<0)?-1:0;
      if(x > 0) this[0] = x;
      else if(x < -1) this[0] = x+this.DV;
      else this.t = 0;
    ***REMOVED***

    // return bigint initialized to value
    function nbv(i) ***REMOVED*** var r = nbi(); r.fromInt(i); return r; ***REMOVED***

    // (protected) set from string and radix
    function bnpFromString(s,b) ***REMOVED***
      var k;
      if(b == 16) k = 4;
      else if(b == 8) k = 3;
      else if(b == 256) k = 8; // byte array
      else if(b == 2) k = 1;
      else if(b == 32) k = 5;
      else if(b == 4) k = 2;
      else ***REMOVED*** this.fromRadix(s,b); return; ***REMOVED***
      this.t = 0;
      this.s = 0;
      var i = s.length, mi = false, sh = 0;
      while(--i >= 0) ***REMOVED***
        var x = (k==8)?s[i]&0xff:intAt(s,i);
        if(x < 0) ***REMOVED***
          if(s.charAt(i) == "-") mi = true;
          continue;
        ***REMOVED***
        mi = false;
        if(sh == 0)
          this[this.t++] = x;
        else if(sh+k > this.DB) ***REMOVED***
          this[this.t-1] |= (x&((1<<(this.DB-sh))-1))<<sh;
          this[this.t++] = (x>>(this.DB-sh));
        ***REMOVED***
        else
          this[this.t-1] |= x<<sh;
        sh += k;
        if(sh >= this.DB) sh -= this.DB;
      ***REMOVED***
      if(k == 8 && (s[0]&0x80) != 0) ***REMOVED***
        this.s = -1;
        if(sh > 0) this[this.t-1] |= ((1<<(this.DB-sh))-1)<<sh;
      ***REMOVED***
      this.clamp();
      if(mi) BigInteger.ZERO.subTo(this,this);
    ***REMOVED***

    // (protected) clamp off excess high words
    function bnpClamp() ***REMOVED***
      var c = this.s&this.DM;
      while(this.t > 0 && this[this.t-1] == c) --this.t;
    ***REMOVED***

    // (public) return string representation in given radix
    function bnToString(b) ***REMOVED***
      if(this.s < 0) return "-"+this.negate().toString(b);
      var k;
      if(b == 16) k = 4;
      else if(b == 8) k = 3;
      else if(b == 2) k = 1;
      else if(b == 32) k = 5;
      else if(b == 4) k = 2;
      else return this.toRadix(b);
      var km = (1<<k)-1, d, m = false, r = "", i = this.t;
      var p = this.DB-(i*this.DB)%k;
      if(i-- > 0) ***REMOVED***
        if(p < this.DB && (d = this[i]>>p) > 0) ***REMOVED*** m = true; r = int2char(d); ***REMOVED***
        while(i >= 0) ***REMOVED***
          if(p < k) ***REMOVED***
            d = (this[i]&((1<<p)-1))<<(k-p);
            d |= this[--i]>>(p+=this.DB-k);
          ***REMOVED***
          else ***REMOVED***
            d = (this[i]>>(p-=k))&km;
            if(p <= 0) ***REMOVED*** p += this.DB; --i; ***REMOVED***
          ***REMOVED***
          if(d > 0) m = true;
          if(m) r += int2char(d);
        ***REMOVED***
      ***REMOVED***
      return m?r:"0";
    ***REMOVED***

    // (public) -this
    function bnNegate() ***REMOVED*** var r = nbi(); BigInteger.ZERO.subTo(this,r); return r; ***REMOVED***

    // (public) |this|
    function bnAbs() ***REMOVED*** return (this.s<0)?this.negate():this; ***REMOVED***

    // (public) return + if this > a, - if this < a, 0 if equal
    function bnCompareTo(a) ***REMOVED***
      var r = this.s-a.s;
      if(r != 0) return r;
      var i = this.t;
      r = i-a.t;
      if(r != 0) return (this.s<0)?-r:r;
      while(--i >= 0) if((r=this[i]-a[i]) != 0) return r;
      return 0;
    ***REMOVED***

    // returns bit length of the integer x
    function nbits(x) ***REMOVED***
      var r = 1, t;
      if((t=x>>>16) != 0) ***REMOVED*** x = t; r += 16; ***REMOVED***
      if((t=x>>8) != 0) ***REMOVED*** x = t; r += 8; ***REMOVED***
      if((t=x>>4) != 0) ***REMOVED*** x = t; r += 4; ***REMOVED***
      if((t=x>>2) != 0) ***REMOVED*** x = t; r += 2; ***REMOVED***
      if((t=x>>1) != 0) ***REMOVED*** x = t; r += 1; ***REMOVED***
      return r;
    ***REMOVED***

    // (public) return the number of bits in "this"
    function bnBitLength() ***REMOVED***
      if(this.t <= 0) return 0;
      return this.DB*(this.t-1)+nbits(this[this.t-1]^(this.s&this.DM));
    ***REMOVED***

    // (protected) r = this << n*DB
    function bnpDLShiftTo(n,r) ***REMOVED***
      var i;
      for(i = this.t-1; i >= 0; --i) r[i+n] = this[i];
      for(i = n-1; i >= 0; --i) r[i] = 0;
      r.t = this.t+n;
      r.s = this.s;
    ***REMOVED***

    // (protected) r = this >> n*DB
    function bnpDRShiftTo(n,r) ***REMOVED***
      for(var i = n; i < this.t; ++i) r[i-n] = this[i];
      r.t = Math.max(this.t-n,0);
      r.s = this.s;
    ***REMOVED***

    // (protected) r = this << n
    function bnpLShiftTo(n,r) ***REMOVED***
      var bs = n%this.DB;
      var cbs = this.DB-bs;
      var bm = (1<<cbs)-1;
      var ds = Math.floor(n/this.DB), c = (this.s<<bs)&this.DM, i;
      for(i = this.t-1; i >= 0; --i) ***REMOVED***
        r[i+ds+1] = (this[i]>>cbs)|c;
        c = (this[i]&bm)<<bs;
      ***REMOVED***
      for(i = ds-1; i >= 0; --i) r[i] = 0;
      r[ds] = c;
      r.t = this.t+ds+1;
      r.s = this.s;
      r.clamp();
    ***REMOVED***

    // (protected) r = this >> n
    function bnpRShiftTo(n,r) ***REMOVED***
      r.s = this.s;
      var ds = Math.floor(n/this.DB);
      if(ds >= this.t) ***REMOVED*** r.t = 0; return; ***REMOVED***
      var bs = n%this.DB;
      var cbs = this.DB-bs;
      var bm = (1<<bs)-1;
      r[0] = this[ds]>>bs;
      for(var i = ds+1; i < this.t; ++i) ***REMOVED***
        r[i-ds-1] |= (this[i]&bm)<<cbs;
        r[i-ds] = this[i]>>bs;
      ***REMOVED***
      if(bs > 0) r[this.t-ds-1] |= (this.s&bm)<<cbs;
      r.t = this.t-ds;
      r.clamp();
    ***REMOVED***

    // (protected) r = this - a
    function bnpSubTo(a,r) ***REMOVED***
      var i = 0, c = 0, m = Math.min(a.t,this.t);
      while(i < m) ***REMOVED***
        c += this[i]-a[i];
        r[i++] = c&this.DM;
        c >>= this.DB;
      ***REMOVED***
      if(a.t < this.t) ***REMOVED***
        c -= a.s;
        while(i < this.t) ***REMOVED***
          c += this[i];
          r[i++] = c&this.DM;
          c >>= this.DB;
        ***REMOVED***
        c += this.s;
      ***REMOVED***
      else ***REMOVED***
        c += this.s;
        while(i < a.t) ***REMOVED***
          c -= a[i];
          r[i++] = c&this.DM;
          c >>= this.DB;
        ***REMOVED***
        c -= a.s;
      ***REMOVED***
      r.s = (c<0)?-1:0;
      if(c < -1) r[i++] = this.DV+c;
      else if(c > 0) r[i++] = c;
      r.t = i;
      r.clamp();
    ***REMOVED***

    // (protected) r = this * a, r != this,a (HAC 14.12)
    // "this" should be the larger one if appropriate.
    function bnpMultiplyTo(a,r) ***REMOVED***
      var x = this.abs(), y = a.abs();
      var i = x.t;
      r.t = i+y.t;
      while(--i >= 0) r[i] = 0;
      for(i = 0; i < y.t; ++i) r[i+x.t] = x.am(0,y[i],r,i,0,x.t);
      r.s = 0;
      r.clamp();
      if(this.s != a.s) BigInteger.ZERO.subTo(r,r);
    ***REMOVED***

    // (protected) r = this^2, r != this (HAC 14.16)
    function bnpSquareTo(r) ***REMOVED***
      var x = this.abs();
      var i = r.t = 2*x.t;
      while(--i >= 0) r[i] = 0;
      for(i = 0; i < x.t-1; ++i) ***REMOVED***
        var c = x.am(i,x[i],r,2*i,0,1);
        if((r[i+x.t]+=x.am(i+1,2*x[i],r,2*i+1,c,x.t-i-1)) >= x.DV) ***REMOVED***
          r[i+x.t] -= x.DV;
          r[i+x.t+1] = 1;
        ***REMOVED***
      ***REMOVED***
      if(r.t > 0) r[r.t-1] += x.am(i,x[i],r,2*i,0,1);
      r.s = 0;
      r.clamp();
    ***REMOVED***

    // (protected) divide this by m, quotient and remainder to q, r (HAC 14.20)
    // r != q, this != m.  q or r may be null.
    function bnpDivRemTo(m,q,r) ***REMOVED***
      var pm = m.abs();
      if(pm.t <= 0) return;
      var pt = this.abs();
      if(pt.t < pm.t) ***REMOVED***
        if(q != null) q.fromInt(0);
        if(r != null) this.copyTo(r);
        return;
      ***REMOVED***
      if(r == null) r = nbi();
      var y = nbi(), ts = this.s, ms = m.s;
      var nsh = this.DB-nbits(pm[pm.t-1]);   // normalize modulus
      if(nsh > 0) ***REMOVED*** pm.lShiftTo(nsh,y); pt.lShiftTo(nsh,r); ***REMOVED***
      else ***REMOVED*** pm.copyTo(y); pt.copyTo(r); ***REMOVED***
      var ys = y.t;
      var y0 = y[ys-1];
      if(y0 == 0) return;
      var yt = y0*(1<<this.F1)+((ys>1)?y[ys-2]>>this.F2:0);
      var d1 = this.FV/yt, d2 = (1<<this.F1)/yt, e = 1<<this.F2;
      var i = r.t, j = i-ys, t = (q==null)?nbi():q;
      y.dlShiftTo(j,t);
      if(r.compareTo(t) >= 0) ***REMOVED***
        r[r.t++] = 1;
        r.subTo(t,r);
      ***REMOVED***
      BigInteger.ONE.dlShiftTo(ys,t);
      t.subTo(y,y);  // "negative" y so we can replace sub with am later
      while(y.t < ys) y[y.t++] = 0;
      while(--j >= 0) ***REMOVED***
        // Estimate quotient digit
        var qd = (r[--i]==y0)?this.DM:Math.floor(r[i]*d1+(r[i-1]+e)*d2);
        if((r[i]+=y.am(0,qd,r,j,0,ys)) < qd) ***REMOVED***   // Try it out
          y.dlShiftTo(j,t);
          r.subTo(t,r);
          while(r[i] < --qd) r.subTo(t,r);
        ***REMOVED***
      ***REMOVED***
      if(q != null) ***REMOVED***
        r.drShiftTo(ys,q);
        if(ts != ms) BigInteger.ZERO.subTo(q,q);
      ***REMOVED***
      r.t = ys;
      r.clamp();
      if(nsh > 0) r.rShiftTo(nsh,r); // Denormalize remainder
      if(ts < 0) BigInteger.ZERO.subTo(r,r);
    ***REMOVED***

    // (public) this mod a
    function bnMod(a) ***REMOVED***
      var r = nbi();
      this.abs().divRemTo(a,null,r);
      if(this.s < 0 && r.compareTo(BigInteger.ZERO) > 0) a.subTo(r,r);
      return r;
    ***REMOVED***

    // Modular reduction using "classic" algorithm
    function Classic(m) ***REMOVED*** this.m = m; ***REMOVED***
    function cConvert(x) ***REMOVED***
      if(x.s < 0 || x.compareTo(this.m) >= 0) return x.mod(this.m);
      else return x;
    ***REMOVED***
    function cRevert(x) ***REMOVED*** return x; ***REMOVED***
    function cReduce(x) ***REMOVED*** x.divRemTo(this.m,null,x); ***REMOVED***
    function cMulTo(x,y,r) ***REMOVED*** x.multiplyTo(y,r); this.reduce(r); ***REMOVED***
    function cSqrTo(x,r) ***REMOVED*** x.squareTo(r); this.reduce(r); ***REMOVED***

    Classic.prototype.convert = cConvert;
    Classic.prototype.revert = cRevert;
    Classic.prototype.reduce = cReduce;
    Classic.prototype.mulTo = cMulTo;
    Classic.prototype.sqrTo = cSqrTo;

    // (protected) return "-1/this % 2^DB"; useful for Mont. reduction
    // justification:
    //         xy == 1 (mod m)
    //         xy =  1+km
    //   xy(2-xy) = (1+km)(1-km)
    // x[y(2-xy)] = 1-k^2m^2
    // x[y(2-xy)] == 1 (mod m^2)
    // if y is 1/x mod m, then y(2-xy) is 1/x mod m^2
    // should reduce x and y(2-xy) by m^2 at each step to keep size bounded.
    // JS multiply "overflows" differently from C/C++, so care is needed here.
    function bnpInvDigit() ***REMOVED***
      if(this.t < 1) return 0;
      var x = this[0];
      if((x&1) == 0) return 0;
      var y = x&3;       // y == 1/x mod 2^2
      y = (y*(2-(x&0xf)*y))&0xf; // y == 1/x mod 2^4
      y = (y*(2-(x&0xff)*y))&0xff;   // y == 1/x mod 2^8
      y = (y*(2-(((x&0xffff)*y)&0xffff)))&0xffff;    // y == 1/x mod 2^16
      // last step - calculate inverse mod DV directly;
      // assumes 16 < DB <= 32 and assumes ability to handle 48-bit ints
      y = (y*(2-x*y%this.DV))%this.DV;       // y == 1/x mod 2^dbits
      // we really want the negative inverse, and -DV < y < DV
      return (y>0)?this.DV-y:-y;
    ***REMOVED***

    // Montgomery reduction
    function Montgomery(m) ***REMOVED***
      this.m = m;
      this.mp = m.invDigit();
      this.mpl = this.mp&0x7fff;
      this.mph = this.mp>>15;
      this.um = (1<<(m.DB-15))-1;
      this.mt2 = 2*m.t;
    ***REMOVED***

    // xR mod m
    function montConvert(x) ***REMOVED***
      var r = nbi();
      x.abs().dlShiftTo(this.m.t,r);
      r.divRemTo(this.m,null,r);
      if(x.s < 0 && r.compareTo(BigInteger.ZERO) > 0) this.m.subTo(r,r);
      return r;
    ***REMOVED***

    // x/R mod m
    function montRevert(x) ***REMOVED***
      var r = nbi();
      x.copyTo(r);
      this.reduce(r);
      return r;
    ***REMOVED***

    // x = x/R mod m (HAC 14.32)
    function montReduce(x) ***REMOVED***
      while(x.t <= this.mt2) // pad x so am has enough room later
        x[x.t++] = 0;
      for(var i = 0; i < this.m.t; ++i) ***REMOVED***
        // faster way of calculating u0 = x[i]*mp mod DV
        var j = x[i]&0x7fff;
        var u0 = (j*this.mpl+(((j*this.mph+(x[i]>>15)*this.mpl)&this.um)<<15))&x.DM;
        // use am to combine the multiply-shift-add into one call
        j = i+this.m.t;
        x[j] += this.m.am(0,u0,x,i,0,this.m.t);
        // propagate carry
        while(x[j] >= x.DV) ***REMOVED*** x[j] -= x.DV; x[++j]++; ***REMOVED***
      ***REMOVED***
      x.clamp();
      x.drShiftTo(this.m.t,x);
      if(x.compareTo(this.m) >= 0) x.subTo(this.m,x);
    ***REMOVED***

    // r = "x^2/R mod m"; x != r
    function montSqrTo(x,r) ***REMOVED*** x.squareTo(r); this.reduce(r); ***REMOVED***

    // r = "xy/R mod m"; x,y != r
    function montMulTo(x,y,r) ***REMOVED*** x.multiplyTo(y,r); this.reduce(r); ***REMOVED***

    Montgomery.prototype.convert = montConvert;
    Montgomery.prototype.revert = montRevert;
    Montgomery.prototype.reduce = montReduce;
    Montgomery.prototype.mulTo = montMulTo;
    Montgomery.prototype.sqrTo = montSqrTo;

    // (protected) true iff this is even
    function bnpIsEven() ***REMOVED*** return ((this.t>0)?(this[0]&1):this.s) == 0; ***REMOVED***

    // (protected) this^e, e < 2^32, doing sqr and mul with "r" (HAC 14.79)
    function bnpExp(e,z) ***REMOVED***
      if(e > 0xffffffff || e < 1) return BigInteger.ONE;
      var r = nbi(), r2 = nbi(), g = z.convert(this), i = nbits(e)-1;
      g.copyTo(r);
      while(--i >= 0) ***REMOVED***
        z.sqrTo(r,r2);
        if((e&(1<<i)) > 0) z.mulTo(r2,g,r);
        else ***REMOVED*** var t = r; r = r2; r2 = t; ***REMOVED***
      ***REMOVED***
      return z.revert(r);
    ***REMOVED***

    // (public) this^e % m, 0 <= e < 2^32
    function bnModPowInt(e,m) ***REMOVED***
      var z;
      if(e < 256 || m.isEven()) z = new Classic(m); else z = new Montgomery(m);
      return this.exp(e,z);
    ***REMOVED***

    // protected
    BigInteger.prototype.copyTo = bnpCopyTo;
    BigInteger.prototype.fromInt = bnpFromInt;
    BigInteger.prototype.fromString = bnpFromString;
    BigInteger.prototype.clamp = bnpClamp;
    BigInteger.prototype.dlShiftTo = bnpDLShiftTo;
    BigInteger.prototype.drShiftTo = bnpDRShiftTo;
    BigInteger.prototype.lShiftTo = bnpLShiftTo;
    BigInteger.prototype.rShiftTo = bnpRShiftTo;
    BigInteger.prototype.subTo = bnpSubTo;
    BigInteger.prototype.multiplyTo = bnpMultiplyTo;
    BigInteger.prototype.squareTo = bnpSquareTo;
    BigInteger.prototype.divRemTo = bnpDivRemTo;
    BigInteger.prototype.invDigit = bnpInvDigit;
    BigInteger.prototype.isEven = bnpIsEven;
    BigInteger.prototype.exp = bnpExp;

    // public
    BigInteger.prototype.toString = bnToString;
    BigInteger.prototype.negate = bnNegate;
    BigInteger.prototype.abs = bnAbs;
    BigInteger.prototype.compareTo = bnCompareTo;
    BigInteger.prototype.bitLength = bnBitLength;
    BigInteger.prototype.mod = bnMod;
    BigInteger.prototype.modPowInt = bnModPowInt;

    // "constants"
    BigInteger.ZERO = nbv(0);
    BigInteger.ONE = nbv(1);

    // Copyright (c) 2005-2009  Tom Wu
    // All Rights Reserved.
    // See "LICENSE" for details.

    // Extended JavaScript BN functions, required for RSA private ops.

    // Version 1.1: new BigInteger("0", 10) returns "proper" zero
    // Version 1.2: square() API, isProbablePrime fix

    // (public)
    function bnClone() ***REMOVED*** var r = nbi(); this.copyTo(r); return r; ***REMOVED***

    // (public) return value as integer
    function bnIntValue() ***REMOVED***
      if(this.s < 0) ***REMOVED***
        if(this.t == 1) return this[0]-this.DV;
        else if(this.t == 0) return -1;
      ***REMOVED***
      else if(this.t == 1) return this[0];
      else if(this.t == 0) return 0;
      // assumes 16 < DB < 32
      return ((this[1]&((1<<(32-this.DB))-1))<<this.DB)|this[0];
    ***REMOVED***

    // (public) return value as byte
    function bnByteValue() ***REMOVED*** return (this.t==0)?this.s:(this[0]<<24)>>24; ***REMOVED***

    // (public) return value as short (assumes DB>=16)
    function bnShortValue() ***REMOVED*** return (this.t==0)?this.s:(this[0]<<16)>>16; ***REMOVED***

    // (protected) return x s.t. r^x < DV
    function bnpChunkSize(r) ***REMOVED*** return Math.floor(Math.LN2*this.DB/Math.log(r)); ***REMOVED***

    // (public) 0 if this == 0, 1 if this > 0
    function bnSigNum() ***REMOVED***
      if(this.s < 0) return -1;
      else if(this.t <= 0 || (this.t == 1 && this[0] <= 0)) return 0;
      else return 1;
    ***REMOVED***

    // (protected) convert to radix string
    function bnpToRadix(b) ***REMOVED***
      if(b == null) b = 10;
      if(this.signum() == 0 || b < 2 || b > 36) return "0";
      var cs = this.chunkSize(b);
      var a = Math.pow(b,cs);
      var d = nbv(a), y = nbi(), z = nbi(), r = "";
      this.divRemTo(d,y,z);
      while(y.signum() > 0) ***REMOVED***
        r = (a+z.intValue()).toString(b).substr(1) + r;
        y.divRemTo(d,y,z);
      ***REMOVED***
      return z.intValue().toString(b) + r;
    ***REMOVED***

    // (protected) convert from radix string
    function bnpFromRadix(s,b) ***REMOVED***
      this.fromInt(0);
      if(b == null) b = 10;
      var cs = this.chunkSize(b);
      var d = Math.pow(b,cs), mi = false, j = 0, w = 0;
      for(var i = 0; i < s.length; ++i) ***REMOVED***
        var x = intAt(s,i);
        if(x < 0) ***REMOVED***
          if(s.charAt(i) == "-" && this.signum() == 0) mi = true;
          continue;
        ***REMOVED***
        w = b*w+x;
        if(++j >= cs) ***REMOVED***
          this.dMultiply(d);
          this.dAddOffset(w,0);
          j = 0;
          w = 0;
        ***REMOVED***
      ***REMOVED***
      if(j > 0) ***REMOVED***
        this.dMultiply(Math.pow(b,j));
        this.dAddOffset(w,0);
      ***REMOVED***
      if(mi) BigInteger.ZERO.subTo(this,this);
    ***REMOVED***

    // (protected) alternate constructor
    function bnpFromNumber(a,b,c) ***REMOVED***
      if("number" == typeof b) ***REMOVED***
        // new BigInteger(int,int,RNG)
        if(a < 2) this.fromInt(1);
        else ***REMOVED***
          this.fromNumber(a,c);
          if(!this.testBit(a-1))	// force MSB set
            this.bitwiseTo(BigInteger.ONE.shiftLeft(a-1),op_or,this);
          if(this.isEven()) this.dAddOffset(1,0); // force odd
          while(!this.isProbablePrime(b)) ***REMOVED***
            this.dAddOffset(2,0);
            if(this.bitLength() > a) this.subTo(BigInteger.ONE.shiftLeft(a-1),this);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      else ***REMOVED***
        // new BigInteger(int,RNG)
        var x = new Array(), t = a&7;
        x.length = (a>>3)+1;
        b.nextBytes(x);
        if(t > 0) x[0] &= ((1<<t)-1); else x[0] = 0;
        this.fromString(x,256);
      ***REMOVED***
    ***REMOVED***

    // (public) convert to bigendian byte array
    function bnToByteArray() ***REMOVED***
      var i = this.t, r = new Array();
      r[0] = this.s;
      var p = this.DB-(i*this.DB)%8, d, k = 0;
      if(i-- > 0) ***REMOVED***
        if(p < this.DB && (d = this[i]>>p) != (this.s&this.DM)>>p)
          r[k++] = d|(this.s<<(this.DB-p));
        while(i >= 0) ***REMOVED***
          if(p < 8) ***REMOVED***
            d = (this[i]&((1<<p)-1))<<(8-p);
            d |= this[--i]>>(p+=this.DB-8);
          ***REMOVED***
          else ***REMOVED***
            d = (this[i]>>(p-=8))&0xff;
            if(p <= 0) ***REMOVED*** p += this.DB; --i; ***REMOVED***
          ***REMOVED***
          if((d&0x80) != 0) d |= -256;
          if(k == 0 && (this.s&0x80) != (d&0x80)) ++k;
          if(k > 0 || d != this.s) r[k++] = d;
        ***REMOVED***
      ***REMOVED***
      return r;
    ***REMOVED***

    function bnEquals(a) ***REMOVED*** return(this.compareTo(a)==0); ***REMOVED***
    function bnMin(a) ***REMOVED*** return(this.compareTo(a)<0)?this:a; ***REMOVED***
    function bnMax(a) ***REMOVED*** return(this.compareTo(a)>0)?this:a; ***REMOVED***

    // (protected) r = this op a (bitwise)
    function bnpBitwiseTo(a,op,r) ***REMOVED***
      var i, f, m = Math.min(a.t,this.t);
      for(i = 0; i < m; ++i) r[i] = op(this[i],a[i]);
      if(a.t < this.t) ***REMOVED***
        f = a.s&this.DM;
        for(i = m; i < this.t; ++i) r[i] = op(this[i],f);
        r.t = this.t;
      ***REMOVED***
      else ***REMOVED***
        f = this.s&this.DM;
        for(i = m; i < a.t; ++i) r[i] = op(f,a[i]);
        r.t = a.t;
      ***REMOVED***
      r.s = op(this.s,a.s);
      r.clamp();
    ***REMOVED***

    // (public) this & a
    function op_and(x,y) ***REMOVED*** return x&y; ***REMOVED***
    function bnAnd(a) ***REMOVED*** var r = nbi(); this.bitwiseTo(a,op_and,r); return r; ***REMOVED***

    // (public) this | a
    function op_or(x,y) ***REMOVED*** return x|y; ***REMOVED***
    function bnOr(a) ***REMOVED*** var r = nbi(); this.bitwiseTo(a,op_or,r); return r; ***REMOVED***

    // (public) this ^ a
    function op_xor(x,y) ***REMOVED*** return x^y; ***REMOVED***
    function bnXor(a) ***REMOVED*** var r = nbi(); this.bitwiseTo(a,op_xor,r); return r; ***REMOVED***

    // (public) this & ~a
    function op_andnot(x,y) ***REMOVED*** return x&~y; ***REMOVED***
    function bnAndNot(a) ***REMOVED*** var r = nbi(); this.bitwiseTo(a,op_andnot,r); return r; ***REMOVED***

    // (public) ~this
    function bnNot() ***REMOVED***
      var r = nbi();
      for(var i = 0; i < this.t; ++i) r[i] = this.DM&~this[i];
      r.t = this.t;
      r.s = ~this.s;
      return r;
    ***REMOVED***

    // (public) this << n
    function bnShiftLeft(n) ***REMOVED***
      var r = nbi();
      if(n < 0) this.rShiftTo(-n,r); else this.lShiftTo(n,r);
      return r;
    ***REMOVED***

    // (public) this >> n
    function bnShiftRight(n) ***REMOVED***
      var r = nbi();
      if(n < 0) this.lShiftTo(-n,r); else this.rShiftTo(n,r);
      return r;
    ***REMOVED***

    // return index of lowest 1-bit in x, x < 2^31
    function lbit(x) ***REMOVED***
      if(x == 0) return -1;
      var r = 0;
      if((x&0xffff) == 0) ***REMOVED*** x >>= 16; r += 16; ***REMOVED***
      if((x&0xff) == 0) ***REMOVED*** x >>= 8; r += 8; ***REMOVED***
      if((x&0xf) == 0) ***REMOVED*** x >>= 4; r += 4; ***REMOVED***
      if((x&3) == 0) ***REMOVED*** x >>= 2; r += 2; ***REMOVED***
      if((x&1) == 0) ++r;
      return r;
    ***REMOVED***

    // (public) returns index of lowest 1-bit (or -1 if none)
    function bnGetLowestSetBit() ***REMOVED***
      for(var i = 0; i < this.t; ++i)
        if(this[i] != 0) return i*this.DB+lbit(this[i]);
      if(this.s < 0) return this.t*this.DB;
      return -1;
    ***REMOVED***

    // return number of 1 bits in x
    function cbit(x) ***REMOVED***
      var r = 0;
      while(x != 0) ***REMOVED*** x &= x-1; ++r; ***REMOVED***
      return r;
    ***REMOVED***

    // (public) return number of set bits
    function bnBitCount() ***REMOVED***
      var r = 0, x = this.s&this.DM;
      for(var i = 0; i < this.t; ++i) r += cbit(this[i]^x);
      return r;
    ***REMOVED***

    // (public) true iff nth bit is set
    function bnTestBit(n) ***REMOVED***
      var j = Math.floor(n/this.DB);
      if(j >= this.t) return(this.s!=0);
      return((this[j]&(1<<(n%this.DB)))!=0);
    ***REMOVED***

    // (protected) this op (1<<n)
    function bnpChangeBit(n,op) ***REMOVED***
      var r = BigInteger.ONE.shiftLeft(n);
      this.bitwiseTo(r,op,r);
      return r;
    ***REMOVED***

    // (public) this | (1<<n)
    function bnSetBit(n) ***REMOVED*** return this.changeBit(n,op_or); ***REMOVED***

    // (public) this & ~(1<<n)
    function bnClearBit(n) ***REMOVED*** return this.changeBit(n,op_andnot); ***REMOVED***

    // (public) this ^ (1<<n)
    function bnFlipBit(n) ***REMOVED*** return this.changeBit(n,op_xor); ***REMOVED***

    // (protected) r = this + a
    function bnpAddTo(a,r) ***REMOVED***
      var i = 0, c = 0, m = Math.min(a.t,this.t);
      while(i < m) ***REMOVED***
        c += this[i]+a[i];
        r[i++] = c&this.DM;
        c >>= this.DB;
      ***REMOVED***
      if(a.t < this.t) ***REMOVED***
        c += a.s;
        while(i < this.t) ***REMOVED***
          c += this[i];
          r[i++] = c&this.DM;
          c >>= this.DB;
        ***REMOVED***
        c += this.s;
      ***REMOVED***
      else ***REMOVED***
        c += this.s;
        while(i < a.t) ***REMOVED***
          c += a[i];
          r[i++] = c&this.DM;
          c >>= this.DB;
        ***REMOVED***
        c += a.s;
      ***REMOVED***
      r.s = (c<0)?-1:0;
      if(c > 0) r[i++] = c;
      else if(c < -1) r[i++] = this.DV+c;
      r.t = i;
      r.clamp();
    ***REMOVED***

    // (public) this + a
    function bnAdd(a) ***REMOVED*** var r = nbi(); this.addTo(a,r); return r; ***REMOVED***

    // (public) this - a
    function bnSubtract(a) ***REMOVED*** var r = nbi(); this.subTo(a,r); return r; ***REMOVED***

    // (public) this * a
    function bnMultiply(a) ***REMOVED*** var r = nbi(); this.multiplyTo(a,r); return r; ***REMOVED***

    // (public) this^2
    function bnSquare() ***REMOVED*** var r = nbi(); this.squareTo(r); return r; ***REMOVED***

    // (public) this / a
    function bnDivide(a) ***REMOVED*** var r = nbi(); this.divRemTo(a,r,null); return r; ***REMOVED***

    // (public) this % a
    function bnRemainder(a) ***REMOVED*** var r = nbi(); this.divRemTo(a,null,r); return r; ***REMOVED***

    // (public) [this/a,this%a]
    function bnDivideAndRemainder(a) ***REMOVED***
      var q = nbi(), r = nbi();
      this.divRemTo(a,q,r);
      return new Array(q,r);
    ***REMOVED***

    // (protected) this *= n, this >= 0, 1 < n < DV
    function bnpDMultiply(n) ***REMOVED***
      this[this.t] = this.am(0,n-1,this,0,0,this.t);
      ++this.t;
      this.clamp();
    ***REMOVED***

    // (protected) this += n << w words, this >= 0
    function bnpDAddOffset(n,w) ***REMOVED***
      if(n == 0) return;
      while(this.t <= w) this[this.t++] = 0;
      this[w] += n;
      while(this[w] >= this.DV) ***REMOVED***
        this[w] -= this.DV;
        if(++w >= this.t) this[this.t++] = 0;
        ++this[w];
      ***REMOVED***
    ***REMOVED***

    // A "null" reducer
    function NullExp() ***REMOVED******REMOVED***
    function nNop(x) ***REMOVED*** return x; ***REMOVED***
    function nMulTo(x,y,r) ***REMOVED*** x.multiplyTo(y,r); ***REMOVED***
    function nSqrTo(x,r) ***REMOVED*** x.squareTo(r); ***REMOVED***

    NullExp.prototype.convert = nNop;
    NullExp.prototype.revert = nNop;
    NullExp.prototype.mulTo = nMulTo;
    NullExp.prototype.sqrTo = nSqrTo;

    // (public) this^e
    function bnPow(e) ***REMOVED*** return this.exp(e,new NullExp()); ***REMOVED***

    // (protected) r = lower n words of "this * a", a.t <= n
    // "this" should be the larger one if appropriate.
    function bnpMultiplyLowerTo(a,n,r) ***REMOVED***
      var i = Math.min(this.t+a.t,n);
      r.s = 0; // assumes a,this >= 0
      r.t = i;
      while(i > 0) r[--i] = 0;
      var j;
      for(j = r.t-this.t; i < j; ++i) r[i+this.t] = this.am(0,a[i],r,i,0,this.t);
      for(j = Math.min(a.t,n); i < j; ++i) this.am(0,a[i],r,i,0,n-i);
      r.clamp();
    ***REMOVED***

    // (protected) r = "this * a" without lower n words, n > 0
    // "this" should be the larger one if appropriate.
    function bnpMultiplyUpperTo(a,n,r) ***REMOVED***
      --n;
      var i = r.t = this.t+a.t-n;
      r.s = 0; // assumes a,this >= 0
      while(--i >= 0) r[i] = 0;
      for(i = Math.max(n-this.t,0); i < a.t; ++i)
        r[this.t+i-n] = this.am(n-i,a[i],r,0,0,this.t+i-n);
      r.clamp();
      r.drShiftTo(1,r);
    ***REMOVED***

    // Barrett modular reduction
    function Barrett(m) ***REMOVED***
      // setup Barrett
      this.r2 = nbi();
      this.q3 = nbi();
      BigInteger.ONE.dlShiftTo(2*m.t,this.r2);
      this.mu = this.r2.divide(m);
      this.m = m;
    ***REMOVED***

    function barrettConvert(x) ***REMOVED***
      if(x.s < 0 || x.t > 2*this.m.t) return x.mod(this.m);
      else if(x.compareTo(this.m) < 0) return x;
      else ***REMOVED*** var r = nbi(); x.copyTo(r); this.reduce(r); return r; ***REMOVED***
    ***REMOVED***

    function barrettRevert(x) ***REMOVED*** return x; ***REMOVED***

    // x = x mod m (HAC 14.42)
    function barrettReduce(x) ***REMOVED***
      x.drShiftTo(this.m.t-1,this.r2);
      if(x.t > this.m.t+1) ***REMOVED*** x.t = this.m.t+1; x.clamp(); ***REMOVED***
      this.mu.multiplyUpperTo(this.r2,this.m.t+1,this.q3);
      this.m.multiplyLowerTo(this.q3,this.m.t+1,this.r2);
      while(x.compareTo(this.r2) < 0) x.dAddOffset(1,this.m.t+1);
      x.subTo(this.r2,x);
      while(x.compareTo(this.m) >= 0) x.subTo(this.m,x);
    ***REMOVED***

    // r = x^2 mod m; x != r
    function barrettSqrTo(x,r) ***REMOVED*** x.squareTo(r); this.reduce(r); ***REMOVED***

    // r = x*y mod m; x,y != r
    function barrettMulTo(x,y,r) ***REMOVED*** x.multiplyTo(y,r); this.reduce(r); ***REMOVED***

    Barrett.prototype.convert = barrettConvert;
    Barrett.prototype.revert = barrettRevert;
    Barrett.prototype.reduce = barrettReduce;
    Barrett.prototype.mulTo = barrettMulTo;
    Barrett.prototype.sqrTo = barrettSqrTo;

    // (public) this^e % m (HAC 14.85)
    function bnModPow(e,m) ***REMOVED***
      var i = e.bitLength(), k, r = nbv(1), z;
      if(i <= 0) return r;
      else if(i < 18) k = 1;
      else if(i < 48) k = 3;
      else if(i < 144) k = 4;
      else if(i < 768) k = 5;
      else k = 6;
      if(i < 8)
        z = new Classic(m);
      else if(m.isEven())
        z = new Barrett(m);
      else
        z = new Montgomery(m);

      // precomputation
      var g = new Array(), n = 3, k1 = k-1, km = (1<<k)-1;
      g[1] = z.convert(this);
      if(k > 1) ***REMOVED***
        var g2 = nbi();
        z.sqrTo(g[1],g2);
        while(n <= km) ***REMOVED***
          g[n] = nbi();
          z.mulTo(g2,g[n-2],g[n]);
          n += 2;
        ***REMOVED***
      ***REMOVED***

      var j = e.t-1, w, is1 = true, r2 = nbi(), t;
      i = nbits(e[j])-1;
      while(j >= 0) ***REMOVED***
        if(i >= k1) w = (e[j]>>(i-k1))&km;
        else ***REMOVED***
          w = (e[j]&((1<<(i+1))-1))<<(k1-i);
          if(j > 0) w |= e[j-1]>>(this.DB+i-k1);
        ***REMOVED***

        n = k;
        while((w&1) == 0) ***REMOVED*** w >>= 1; --n; ***REMOVED***
        if((i -= n) < 0) ***REMOVED*** i += this.DB; --j; ***REMOVED***
        if(is1) ***REMOVED***	// ret == 1, don't bother squaring or multiplying it
          g[w].copyTo(r);
          is1 = false;
        ***REMOVED***
        else ***REMOVED***
          while(n > 1) ***REMOVED*** z.sqrTo(r,r2); z.sqrTo(r2,r); n -= 2; ***REMOVED***
          if(n > 0) z.sqrTo(r,r2); else ***REMOVED*** t = r; r = r2; r2 = t; ***REMOVED***
          z.mulTo(r2,g[w],r);
        ***REMOVED***

        while(j >= 0 && (e[j]&(1<<i)) == 0) ***REMOVED***
          z.sqrTo(r,r2); t = r; r = r2; r2 = t;
          if(--i < 0) ***REMOVED*** i = this.DB-1; --j; ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      return z.revert(r);
    ***REMOVED***

    // (public) gcd(this,a) (HAC 14.54)
    function bnGCD(a) ***REMOVED***
      var x = (this.s<0)?this.negate():this.clone();
      var y = (a.s<0)?a.negate():a.clone();
      if(x.compareTo(y) < 0) ***REMOVED*** var t = x; x = y; y = t; ***REMOVED***
      var i = x.getLowestSetBit(), g = y.getLowestSetBit();
      if(g < 0) return x;
      if(i < g) g = i;
      if(g > 0) ***REMOVED***
        x.rShiftTo(g,x);
        y.rShiftTo(g,y);
      ***REMOVED***
      while(x.signum() > 0) ***REMOVED***
        if((i = x.getLowestSetBit()) > 0) x.rShiftTo(i,x);
        if((i = y.getLowestSetBit()) > 0) y.rShiftTo(i,y);
        if(x.compareTo(y) >= 0) ***REMOVED***
          x.subTo(y,x);
          x.rShiftTo(1,x);
        ***REMOVED***
        else ***REMOVED***
          y.subTo(x,y);
          y.rShiftTo(1,y);
        ***REMOVED***
      ***REMOVED***
      if(g > 0) y.lShiftTo(g,y);
      return y;
    ***REMOVED***

    // (protected) this % n, n < 2^26
    function bnpModInt(n) ***REMOVED***
      if(n <= 0) return 0;
      var d = this.DV%n, r = (this.s<0)?n-1:0;
      if(this.t > 0)
        if(d == 0) r = this[0]%n;
        else for(var i = this.t-1; i >= 0; --i) r = (d*r+this[i])%n;
      return r;
    ***REMOVED***

    // (public) 1/this % m (HAC 14.61)
    function bnModInverse(m) ***REMOVED***
      var ac = m.isEven();
      if((this.isEven() && ac) || m.signum() == 0) return BigInteger.ZERO;
      var u = m.clone(), v = this.clone();
      var a = nbv(1), b = nbv(0), c = nbv(0), d = nbv(1);
      while(u.signum() != 0) ***REMOVED***
        while(u.isEven()) ***REMOVED***
          u.rShiftTo(1,u);
          if(ac) ***REMOVED***
            if(!a.isEven() || !b.isEven()) ***REMOVED*** a.addTo(this,a); b.subTo(m,b); ***REMOVED***
            a.rShiftTo(1,a);
          ***REMOVED***
          else if(!b.isEven()) b.subTo(m,b);
          b.rShiftTo(1,b);
        ***REMOVED***
        while(v.isEven()) ***REMOVED***
          v.rShiftTo(1,v);
          if(ac) ***REMOVED***
            if(!c.isEven() || !d.isEven()) ***REMOVED*** c.addTo(this,c); d.subTo(m,d); ***REMOVED***
            c.rShiftTo(1,c);
          ***REMOVED***
          else if(!d.isEven()) d.subTo(m,d);
          d.rShiftTo(1,d);
        ***REMOVED***
        if(u.compareTo(v) >= 0) ***REMOVED***
          u.subTo(v,u);
          if(ac) a.subTo(c,a);
          b.subTo(d,b);
        ***REMOVED***
        else ***REMOVED***
          v.subTo(u,v);
          if(ac) c.subTo(a,c);
          d.subTo(b,d);
        ***REMOVED***
      ***REMOVED***
      if(v.compareTo(BigInteger.ONE) != 0) return BigInteger.ZERO;
      if(d.compareTo(m) >= 0) return d.subtract(m);
      if(d.signum() < 0) d.addTo(m,d); else return d;
      if(d.signum() < 0) return d.add(m); else return d;
    ***REMOVED***

    var lowprimes = [2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,79,83,89,97,101,103,107,109,113,127,131,137,139,149,151,157,163,167,173,179,181,191,193,197,199,211,223,227,229,233,239,241,251,257,263,269,271,277,281,283,293,307,311,313,317,331,337,347,349,353,359,367,373,379,383,389,397,401,409,419,421,431,433,439,443,449,457,461,463,467,479,487,491,499,503,509,521,523,541,547,557,563,569,571,577,587,593,599,601,607,613,617,619,631,641,643,647,653,659,661,673,677,683,691,701,709,719,727,733,739,743,751,757,761,769,773,787,797,809,811,821,823,827,829,839,853,857,859,863,877,881,883,887,907,911,919,929,937,941,947,953,967,971,977,983,991,997];
    var lplim = (1<<26)/lowprimes[lowprimes.length-1];

    // (public) test primality with certainty >= 1-.5^t
    function bnIsProbablePrime(t) ***REMOVED***
      var i, x = this.abs();
      if(x.t == 1 && x[0] <= lowprimes[lowprimes.length-1]) ***REMOVED***
        for(i = 0; i < lowprimes.length; ++i)
          if(x[0] == lowprimes[i]) return true;
        return false;
      ***REMOVED***
      if(x.isEven()) return false;
      i = 1;
      while(i < lowprimes.length) ***REMOVED***
        var m = lowprimes[i], j = i+1;
        while(j < lowprimes.length && m < lplim) m *= lowprimes[j++];
        m = x.modInt(m);
        while(i < j) if(m%lowprimes[i++] == 0) return false;
      ***REMOVED***
      return x.millerRabin(t);
    ***REMOVED***

    // (protected) true if probably prime (HAC 4.24, Miller-Rabin)
    function bnpMillerRabin(t) ***REMOVED***
      var n1 = this.subtract(BigInteger.ONE);
      var k = n1.getLowestSetBit();
      if(k <= 0) return false;
      var r = n1.shiftRight(k);
      t = (t+1)>>1;
      if(t > lowprimes.length) t = lowprimes.length;
      var a = nbi();
      for(var i = 0; i < t; ++i) ***REMOVED***
        //Pick bases at random, instead of starting at 2
        a.fromInt(lowprimes[Math.floor(Math.random()*lowprimes.length)]);
        var y = a.modPow(r,this);
        if(y.compareTo(BigInteger.ONE) != 0 && y.compareTo(n1) != 0) ***REMOVED***
          var j = 1;
          while(j++ < k && y.compareTo(n1) != 0) ***REMOVED***
            y = y.modPowInt(2,this);
            if(y.compareTo(BigInteger.ONE) == 0) return false;
          ***REMOVED***
          if(y.compareTo(n1) != 0) return false;
        ***REMOVED***
      ***REMOVED***
      return true;
    ***REMOVED***

    // protected
    BigInteger.prototype.chunkSize = bnpChunkSize;
    BigInteger.prototype.toRadix = bnpToRadix;
    BigInteger.prototype.fromRadix = bnpFromRadix;
    BigInteger.prototype.fromNumber = bnpFromNumber;
    BigInteger.prototype.bitwiseTo = bnpBitwiseTo;
    BigInteger.prototype.changeBit = bnpChangeBit;
    BigInteger.prototype.addTo = bnpAddTo;
    BigInteger.prototype.dMultiply = bnpDMultiply;
    BigInteger.prototype.dAddOffset = bnpDAddOffset;
    BigInteger.prototype.multiplyLowerTo = bnpMultiplyLowerTo;
    BigInteger.prototype.multiplyUpperTo = bnpMultiplyUpperTo;
    BigInteger.prototype.modInt = bnpModInt;
    BigInteger.prototype.millerRabin = bnpMillerRabin;

    // public
    BigInteger.prototype.clone = bnClone;
    BigInteger.prototype.intValue = bnIntValue;
    BigInteger.prototype.byteValue = bnByteValue;
    BigInteger.prototype.shortValue = bnShortValue;
    BigInteger.prototype.signum = bnSigNum;
    BigInteger.prototype.toByteArray = bnToByteArray;
    BigInteger.prototype.equals = bnEquals;
    BigInteger.prototype.min = bnMin;
    BigInteger.prototype.max = bnMax;
    BigInteger.prototype.and = bnAnd;
    BigInteger.prototype.or = bnOr;
    BigInteger.prototype.xor = bnXor;
    BigInteger.prototype.andNot = bnAndNot;
    BigInteger.prototype.not = bnNot;
    BigInteger.prototype.shiftLeft = bnShiftLeft;
    BigInteger.prototype.shiftRight = bnShiftRight;
    BigInteger.prototype.getLowestSetBit = bnGetLowestSetBit;
    BigInteger.prototype.bitCount = bnBitCount;
    BigInteger.prototype.testBit = bnTestBit;
    BigInteger.prototype.setBit = bnSetBit;
    BigInteger.prototype.clearBit = bnClearBit;
    BigInteger.prototype.flipBit = bnFlipBit;
    BigInteger.prototype.add = bnAdd;
    BigInteger.prototype.subtract = bnSubtract;
    BigInteger.prototype.multiply = bnMultiply;
    BigInteger.prototype.divide = bnDivide;
    BigInteger.prototype.remainder = bnRemainder;
    BigInteger.prototype.divideAndRemainder = bnDivideAndRemainder;
    BigInteger.prototype.modPow = bnModPow;
    BigInteger.prototype.modInverse = bnModInverse;
    BigInteger.prototype.pow = bnPow;
    BigInteger.prototype.gcd = bnGCD;
    BigInteger.prototype.isProbablePrime = bnIsProbablePrime;

    // JSBN-specific extension
    BigInteger.prototype.square = bnSquare;

    // Expose the Barrett function
    BigInteger.prototype.Barrett = Barrett

    // BigInteger interfaces not implemented in jsbn:

    // BigInteger(int signum, byte[] magnitude)
    // double doubleValue()
    // float floatValue()
    // int hashCode()
    // long longValue()
    // static BigInteger valueOf(long val)

	// Random number generator - requires a PRNG backend, e.g. prng4.js

	// For best results, put code like
	// <body onClick='rng_seed_time();' onKeyPress='rng_seed_time();'>
	// in your main HTML document.

	var rng_state;
	var rng_pool;
	var rng_pptr;

	// Mix in a 32-bit integer into the pool
	function rng_seed_int(x) ***REMOVED***
	  rng_pool[rng_pptr++] ^= x & 255;
	  rng_pool[rng_pptr++] ^= (x >> 8) & 255;
	  rng_pool[rng_pptr++] ^= (x >> 16) & 255;
	  rng_pool[rng_pptr++] ^= (x >> 24) & 255;
	  if(rng_pptr >= rng_psize) rng_pptr -= rng_psize;
	***REMOVED***

	// Mix in the current time (w/milliseconds) into the pool
	function rng_seed_time() ***REMOVED***
	  rng_seed_int(new Date().getTime());
	***REMOVED***

	// Initialize the pool with junk if needed.
	if(rng_pool == null) ***REMOVED***
	  rng_pool = new Array();
	  rng_pptr = 0;
	  var t;
	  if(typeof window !== "undefined" && window.crypto) ***REMOVED***
		if (window.crypto.getRandomValues) ***REMOVED***
		  // Use webcrypto if available
		  var ua = new Uint8Array(32);
		  window.crypto.getRandomValues(ua);
		  for(t = 0; t < 32; ++t)
			rng_pool[rng_pptr++] = ua[t];
		***REMOVED***
		else if(navigator.appName == "Netscape" && navigator.appVersion < "5") ***REMOVED***
		  // Extract entropy (256 bits) from NS4 RNG if available
		  var z = window.crypto.random(32);
		  for(t = 0; t < z.length; ++t)
			rng_pool[rng_pptr++] = z.charCodeAt(t) & 255;
		***REMOVED***
	  ***REMOVED***
	  while(rng_pptr < rng_psize) ***REMOVED***  // extract some randomness from Math.random()
		t = Math.floor(65536 * Math.random());
		rng_pool[rng_pptr++] = t >>> 8;
		rng_pool[rng_pptr++] = t & 255;
	  ***REMOVED***
	  rng_pptr = 0;
	  rng_seed_time();
	  //rng_seed_int(window.screenX);
	  //rng_seed_int(window.screenY);
	***REMOVED***

	function rng_get_byte() ***REMOVED***
	  if(rng_state == null) ***REMOVED***
		rng_seed_time();
		rng_state = prng_newstate();
		rng_state.init(rng_pool);
		for(rng_pptr = 0; rng_pptr < rng_pool.length; ++rng_pptr)
		  rng_pool[rng_pptr] = 0;
		rng_pptr = 0;
		//rng_pool = null;
	  ***REMOVED***
	  // TODO: allow reseeding after first request
	  return rng_state.next();
	***REMOVED***

	function rng_get_bytes(ba) ***REMOVED***
	  var i;
	  for(i = 0; i < ba.length; ++i) ba[i] = rng_get_byte();
	***REMOVED***

	function SecureRandom() ***REMOVED******REMOVED***

	SecureRandom.prototype.nextBytes = rng_get_bytes;

	// prng4.js - uses Arcfour as a PRNG

	function Arcfour() ***REMOVED***
	  this.i = 0;
	  this.j = 0;
	  this.S = new Array();
	***REMOVED***

	// Initialize arcfour context from key, an array of ints, each from [0..255]
	function ARC4init(key) ***REMOVED***
	  var i, j, t;
	  for(i = 0; i < 256; ++i)
		this.S[i] = i;
	  j = 0;
	  for(i = 0; i < 256; ++i) ***REMOVED***
		j = (j + this.S[i] + key[i % key.length]) & 255;
		t = this.S[i];
		this.S[i] = this.S[j];
		this.S[j] = t;
	  ***REMOVED***
	  this.i = 0;
	  this.j = 0;
	***REMOVED***

	function ARC4next() ***REMOVED***
	  var t;
	  this.i = (this.i + 1) & 255;
	  this.j = (this.j + this.S[this.i]) & 255;
	  t = this.S[this.i];
	  this.S[this.i] = this.S[this.j];
	  this.S[this.j] = t;
	  return this.S[(t + this.S[this.i]) & 255];
	***REMOVED***

	Arcfour.prototype.init = ARC4init;
	Arcfour.prototype.next = ARC4next;

	// Plug in your RNG constructor here
	function prng_newstate() ***REMOVED***
	  return new Arcfour();
	***REMOVED***

	// Pool size must be a multiple of 4 and greater than 32.
	// An array of bytes the size of the pool will be passed to init()
	var rng_psize = 256;

  BigInteger.SecureRandom = SecureRandom;
  BigInteger.BigInteger = BigInteger;
  if (typeof exports !== 'undefined') ***REMOVED***
    exports = module.exports = BigInteger;
  ***REMOVED*** else ***REMOVED***
    this.BigInteger = BigInteger;
    this.SecureRandom = SecureRandom;
  ***REMOVED***

***REMOVED***).call(this);
