// tough-cookie/lib/cookie.js
/*!
 * Copyright (c) 2015, Salesforce.com, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of Salesforce.com nor the names of its contributors may
 * be used to endorse or promote products derived from this software without
 * specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
'use strict';
var net = require('net');
var urlParse = require('url').parse;
var pubsuffix = require('./pubsuffix');
var Store = require('./store').Store;
var MemoryCookieStore = require('./memstore').MemoryCookieStore;
var pathMatch = require('./pathMatch').pathMatch;
var VERSION = require('../package.json').version;

var punycode;
try ***REMOVED***
  punycode = require('punycode');
***REMOVED*** catch(e) ***REMOVED***
  console.warn("cookie: can't load punycode; won't use punycode for domain normalization");
***REMOVED***

// From RFC6265 S4.1.1
// note that it excludes \x3B ";"
var COOKIE_OCTETS = /^[\x21\x23-\x2B\x2D-\x3A\x3C-\x5B\x5D-\x7E]+$/;

var CONTROL_CHARS = /[\x00-\x1F]/;

// From Chromium // '\r', '\n' and '\0' should be treated as a terminator in
// the "relaxed" mode, see:
// https://github.com/ChromiumWebApps/chromium/blob/b3d3b4da8bb94c1b2e061600df106d590fda3620/net/cookies/parsed_cookie.cc#L60
var TERMINATORS = ['\n', '\r', '\0'];

// RFC6265 S4.1.1 defines path value as 'any CHAR except CTLs or ";"'
// Note ';' is \x3B
var PATH_VALUE = /[\x20-\x3A\x3C-\x7E]+/;

// date-time parsing constants (RFC6265 S5.1.1)

var DATE_DELIM = /[\x09\x20-\x2F\x3B-\x40\x5B-\x60\x7B-\x7E]/;

var MONTH_TO_NUM = ***REMOVED***
  jan:0, feb:1, mar:2, apr:3, may:4, jun:5,
  jul:6, aug:7, sep:8, oct:9, nov:10, dec:11
***REMOVED***;
var NUM_TO_MONTH = [
  'Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'
];
var NUM_TO_DAY = [
  'Sun','Mon','Tue','Wed','Thu','Fri','Sat'
];

var MAX_TIME = 2147483647000; // 31-bit max
var MIN_TIME = 0; // 31-bit min

/*
 * Parses a Natural number (i.e., non-negative integer) with either the
 *    <min>*<max>DIGIT ( non-digit *OCTET )
 * or
 *    <min>*<max>DIGIT
 * grammar (RFC6265 S5.1.1).
 *
 * The "trailingOK" boolean controls if the grammar accepts a
 * "( non-digit *OCTET )" trailer.
 */
function parseDigits(token, minDigits, maxDigits, trailingOK) ***REMOVED***
  var count = 0;
  while (count < token.length) ***REMOVED***
    var c = token.charCodeAt(count);
    // "non-digit = %x00-2F / %x3A-FF"
    if (c <= 0x2F || c >= 0x3A) ***REMOVED***
      break;
    ***REMOVED***
    count++;
  ***REMOVED***

  // constrain to a minimum and maximum number of digits.
  if (count < minDigits || count > maxDigits) ***REMOVED***
    return null;
  ***REMOVED***

  if (!trailingOK && count != token.length) ***REMOVED***
    return null;
  ***REMOVED***

  return parseInt(token.substr(0,count), 10);
***REMOVED***

function parseTime(token) ***REMOVED***
  var parts = token.split(':');
  var result = [0,0,0];

  /* RF6256 S5.1.1:
   *      time            = hms-time ( non-digit *OCTET )
   *      hms-time        = time-field ":" time-field ":" time-field
   *      time-field      = 1*2DIGIT
   */

  if (parts.length !== 3) ***REMOVED***
    return null;
  ***REMOVED***

  for (var i = 0; i < 3; i++) ***REMOVED***
    // "time-field" must be strictly "1*2DIGIT", HOWEVER, "hms-time" can be
    // followed by "( non-digit *OCTET )" so therefore the last time-field can
    // have a trailer
    var trailingOK = (i == 2);
    var num = parseDigits(parts[i], 1, 2, trailingOK);
    if (num === null) ***REMOVED***
      return null;
    ***REMOVED***
    result[i] = num;
  ***REMOVED***

  return result;
***REMOVED***

function parseMonth(token) ***REMOVED***
  token = String(token).substr(0,3).toLowerCase();
  var num = MONTH_TO_NUM[token];
  return num >= 0 ? num : null;
***REMOVED***

/*
 * RFC6265 S5.1.1 date parser (see RFC for full grammar)
 */
function parseDate(str) ***REMOVED***
  if (!str) ***REMOVED***
    return;
  ***REMOVED***

  /* RFC6265 S5.1.1:
   * 2. Process each date-token sequentially in the order the date-tokens
   * appear in the cookie-date
   */
  var tokens = str.split(DATE_DELIM);
  if (!tokens) ***REMOVED***
    return;
  ***REMOVED***

  var hour = null;
  var minute = null;
  var second = null;
  var dayOfMonth = null;
  var month = null;
  var year = null;

  for (var i=0; i<tokens.length; i++) ***REMOVED***
    var token = tokens[i].trim();
    if (!token.length) ***REMOVED***
      continue;
    ***REMOVED***

    var result;

    /* 2.1. If the found-time flag is not set and the token matches the time
     * production, set the found-time flag and set the hour- value,
     * minute-value, and second-value to the numbers denoted by the digits in
     * the date-token, respectively.  Skip the remaining sub-steps and continue
     * to the next date-token.
     */
    if (second === null) ***REMOVED***
      result = parseTime(token);
      if (result) ***REMOVED***
        hour = result[0];
        minute = result[1];
        second = result[2];
        continue;
      ***REMOVED***
    ***REMOVED***

    /* 2.2. If the found-day-of-month flag is not set and the date-token matches
     * the day-of-month production, set the found-day-of- month flag and set
     * the day-of-month-value to the number denoted by the date-token.  Skip
     * the remaining sub-steps and continue to the next date-token.
     */
    if (dayOfMonth === null) ***REMOVED***
      // "day-of-month = 1*2DIGIT ( non-digit *OCTET )"
      result = parseDigits(token, 1, 2, true);
      if (result !== null) ***REMOVED***
        dayOfMonth = result;
        continue;
      ***REMOVED***
    ***REMOVED***

    /* 2.3. If the found-month flag is not set and the date-token matches the
     * month production, set the found-month flag and set the month-value to
     * the month denoted by the date-token.  Skip the remaining sub-steps and
     * continue to the next date-token.
     */
    if (month === null) ***REMOVED***
      result = parseMonth(token);
      if (result !== null) ***REMOVED***
        month = result;
        continue;
      ***REMOVED***
    ***REMOVED***

    /* 2.4. If the found-year flag is not set and the date-token matches the
     * year production, set the found-year flag and set the year-value to the
     * number denoted by the date-token.  Skip the remaining sub-steps and
     * continue to the next date-token.
     */
    if (year === null) ***REMOVED***
      // "year = 2*4DIGIT ( non-digit *OCTET )"
      result = parseDigits(token, 2, 4, true);
      if (result !== null) ***REMOVED***
        year = result;
        /* From S5.1.1:
         * 3.  If the year-value is greater than or equal to 70 and less
         * than or equal to 99, increment the year-value by 1900.
         * 4.  If the year-value is greater than or equal to 0 and less
         * than or equal to 69, increment the year-value by 2000.
         */
        if (year >= 70 && year <= 99) ***REMOVED***
          year += 1900;
        ***REMOVED*** else if (year >= 0 && year <= 69) ***REMOVED***
          year += 2000;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  /* RFC 6265 S5.1.1
   * "5. Abort these steps and fail to parse the cookie-date if:
   *     *  at least one of the found-day-of-month, found-month, found-
   *        year, or found-time flags is not set,
   *     *  the day-of-month-value is less than 1 or greater than 31,
   *     *  the year-value is less than 1601,
   *     *  the hour-value is greater than 23,
   *     *  the minute-value is greater than 59, or
   *     *  the second-value is greater than 59.
   *     (Note that leap seconds cannot be represented in this syntax.)"
   *
   * So, in order as above:
   */
  if (
    dayOfMonth === null || month === null || year === null || second === null ||
    dayOfMonth < 1 || dayOfMonth > 31 ||
    year < 1601 ||
    hour > 23 ||
    minute > 59 ||
    second > 59
  ) ***REMOVED***
    return;
  ***REMOVED***

  return new Date(Date.UTC(year, month, dayOfMonth, hour, minute, second));
***REMOVED***

function formatDate(date) ***REMOVED***
  var d = date.getUTCDate(); d = d >= 10 ? d : '0'+d;
  var h = date.getUTCHours(); h = h >= 10 ? h : '0'+h;
  var m = date.getUTCMinutes(); m = m >= 10 ? m : '0'+m;
  var s = date.getUTCSeconds(); s = s >= 10 ? s : '0'+s;
  return NUM_TO_DAY[date.getUTCDay()] + ', ' +
    d+' '+ NUM_TO_MONTH[date.getUTCMonth()] +' '+ date.getUTCFullYear() +' '+
    h+':'+m+':'+s+' GMT';
***REMOVED***

// S5.1.2 Canonicalized Host Names
function canonicalDomain(str) ***REMOVED***
  if (str == null) ***REMOVED***
    return null;
  ***REMOVED***
  str = str.trim().replace(/^\./,''); // S4.1.2.3 & S5.2.3: ignore leading .

  // convert to IDN if any non-ASCII characters
  if (punycode && /[^\u0001-\u007f]/.test(str)) ***REMOVED***
    str = punycode.toASCII(str);
  ***REMOVED***

  return str.toLowerCase();
***REMOVED***

// S5.1.3 Domain Matching
function domainMatch(str, domStr, canonicalize) ***REMOVED***
  if (str == null || domStr == null) ***REMOVED***
    return null;
  ***REMOVED***
  if (canonicalize !== false) ***REMOVED***
    str = canonicalDomain(str);
    domStr = canonicalDomain(domStr);
  ***REMOVED***

  /*
   * "The domain string and the string are identical. (Note that both the
   * domain string and the string will have been canonicalized to lower case at
   * this point)"
   */
  if (str == domStr) ***REMOVED***
    return true;
  ***REMOVED***

  /* "All of the following [three] conditions hold:" (order adjusted from the RFC) */

  /* "* The string is a host name (i.e., not an IP address)." */
  if (net.isIP(str)) ***REMOVED***
    return false;
  ***REMOVED***

  /* "* The domain string is a suffix of the string" */
  var idx = str.indexOf(domStr);
  if (idx <= 0) ***REMOVED***
    return false; // it's a non-match (-1) or prefix (0)
  ***REMOVED***

  // e.g "a.b.c".indexOf("b.c") === 2
  // 5 === 3+2
  if (str.length !== domStr.length + idx) ***REMOVED*** // it's not a suffix
    return false;
  ***REMOVED***

  /* "* The last character of the string that is not included in the domain
  * string is a %x2E (".") character." */
  if (str.substr(idx-1,1) !== '.') ***REMOVED***
    return false;
  ***REMOVED***

  return true;
***REMOVED***


// RFC6265 S5.1.4 Paths and Path-Match

/*
 * "The user agent MUST use an algorithm equivalent to the following algorithm
 * to compute the default-path of a cookie:"
 *
 * Assumption: the path (and not query part or absolute uri) is passed in.
 */
function defaultPath(path) ***REMOVED***
  // "2. If the uri-path is empty or if the first character of the uri-path is not
  // a %x2F ("/") character, output %x2F ("/") and skip the remaining steps.
  if (!path || path.substr(0,1) !== "/") ***REMOVED***
    return "/";
  ***REMOVED***

  // "3. If the uri-path contains no more than one %x2F ("/") character, output
  // %x2F ("/") and skip the remaining step."
  if (path === "/") ***REMOVED***
    return path;
  ***REMOVED***

  var rightSlash = path.lastIndexOf("/");
  if (rightSlash === 0) ***REMOVED***
    return "/";
  ***REMOVED***

  // "4. Output the characters of the uri-path from the first character up to,
  // but not including, the right-most %x2F ("/")."
  return path.slice(0, rightSlash);
***REMOVED***

function trimTerminator(str) ***REMOVED***
  for (var t = 0; t < TERMINATORS.length; t++) ***REMOVED***
    var terminatorIdx = str.indexOf(TERMINATORS[t]);
    if (terminatorIdx !== -1) ***REMOVED***
      str = str.substr(0,terminatorIdx);
    ***REMOVED***
  ***REMOVED***

  return str;
***REMOVED***

function parseCookiePair(cookiePair, looseMode) ***REMOVED***
  cookiePair = trimTerminator(cookiePair);

  var firstEq = cookiePair.indexOf('=');
  if (looseMode) ***REMOVED***
    if (firstEq === 0) ***REMOVED*** // '=' is immediately at start
      cookiePair = cookiePair.substr(1);
      firstEq = cookiePair.indexOf('='); // might still need to split on '='
    ***REMOVED***
  ***REMOVED*** else ***REMOVED*** // non-loose mode
    if (firstEq <= 0) ***REMOVED*** // no '=' or is at start
      return; // needs to have non-empty "cookie-name"
    ***REMOVED***
  ***REMOVED***

  var cookieName, cookieValue;
  if (firstEq <= 0) ***REMOVED***
    cookieName = "";
    cookieValue = cookiePair.trim();
  ***REMOVED*** else ***REMOVED***
    cookieName = cookiePair.substr(0, firstEq).trim();
    cookieValue = cookiePair.substr(firstEq+1).trim();
  ***REMOVED***

  if (CONTROL_CHARS.test(cookieName) || CONTROL_CHARS.test(cookieValue)) ***REMOVED***
    return;
  ***REMOVED***

  var c = new Cookie();
  c.key = cookieName;
  c.value = cookieValue;
  return c;
***REMOVED***

function parse(str, options) ***REMOVED***
  if (!options || typeof options !== 'object') ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  str = str.trim();

  // We use a regex to parse the "name-value-pair" part of S5.2
  var firstSemi = str.indexOf(';'); // S5.2 step 1
  var cookiePair = (firstSemi === -1) ? str : str.substr(0, firstSemi);
  var c = parseCookiePair(cookiePair, !!options.loose);
  if (!c) ***REMOVED***
    return;
  ***REMOVED***

  if (firstSemi === -1) ***REMOVED***
    return c;
  ***REMOVED***

  // S5.2.3 "unparsed-attributes consist of the remainder of the set-cookie-string
  // (including the %x3B (";") in question)." plus later on in the same section
  // "discard the first ";" and trim".
  var unparsed = str.slice(firstSemi + 1).trim();

  // "If the unparsed-attributes string is empty, skip the rest of these
  // steps."
  if (unparsed.length === 0) ***REMOVED***
    return c;
  ***REMOVED***

  /*
   * S5.2 says that when looping over the items "[p]rocess the attribute-name
   * and attribute-value according to the requirements in the following
   * subsections" for every item.  Plus, for many of the individual attributes
   * in S5.3 it says to use the "attribute-value of the last attribute in the
   * cookie-attribute-list".  Therefore, in this implementation, we overwrite
   * the previous value.
   */
  var cookie_avs = unparsed.split(';');
  while (cookie_avs.length) ***REMOVED***
    var av = cookie_avs.shift().trim();
    if (av.length === 0) ***REMOVED*** // happens if ";;" appears
      continue;
    ***REMOVED***
    var av_sep = av.indexOf('=');
    var av_key, av_value;

    if (av_sep === -1) ***REMOVED***
      av_key = av;
      av_value = null;
    ***REMOVED*** else ***REMOVED***
      av_key = av.substr(0,av_sep);
      av_value = av.substr(av_sep+1);
    ***REMOVED***

    av_key = av_key.trim().toLowerCase();

    if (av_value) ***REMOVED***
      av_value = av_value.trim();
    ***REMOVED***

    switch(av_key) ***REMOVED***
    case 'expires': // S5.2.1
      if (av_value) ***REMOVED***
        var exp = parseDate(av_value);
        // "If the attribute-value failed to parse as a cookie date, ignore the
        // cookie-av."
        if (exp) ***REMOVED***
          // over and underflow not realistically a concern: V8's getTime() seems to
          // store something larger than a 32-bit time_t (even with 32-bit node)
          c.expires = exp;
        ***REMOVED***
      ***REMOVED***
      break;

    case 'max-age': // S5.2.2
      if (av_value) ***REMOVED***
        // "If the first character of the attribute-value is not a DIGIT or a "-"
        // character ...[or]... If the remainder of attribute-value contains a
        // non-DIGIT character, ignore the cookie-av."
        if (/^-?[0-9]+$/.test(av_value)) ***REMOVED***
          var delta = parseInt(av_value, 10);
          // "If delta-seconds is less than or equal to zero (0), let expiry-time
          // be the earliest representable date and time."
          c.setMaxAge(delta);
        ***REMOVED***
      ***REMOVED***
      break;

    case 'domain': // S5.2.3
      // "If the attribute-value is empty, the behavior is undefined.  However,
      // the user agent SHOULD ignore the cookie-av entirely."
      if (av_value) ***REMOVED***
        // S5.2.3 "Let cookie-domain be the attribute-value without the leading %x2E
        // (".") character."
        var domain = av_value.trim().replace(/^\./, '');
        if (domain) ***REMOVED***
          // "Convert the cookie-domain to lower case."
          c.domain = domain.toLowerCase();
        ***REMOVED***
      ***REMOVED***
      break;

    case 'path': // S5.2.4
      /*
       * "If the attribute-value is empty or if the first character of the
       * attribute-value is not %x2F ("/"):
       *   Let cookie-path be the default-path.
       * Otherwise:
       *   Let cookie-path be the attribute-value."
       *
       * We'll represent the default-path as null since it depends on the
       * context of the parsing.
       */
      c.path = av_value && av_value[0] === "/" ? av_value : null;
      break;

    case 'secure': // S5.2.5
      /*
       * "If the attribute-name case-insensitively matches the string "Secure",
       * the user agent MUST append an attribute to the cookie-attribute-list
       * with an attribute-name of Secure and an empty attribute-value."
       */
      c.secure = true;
      break;

    case 'httponly': // S5.2.6 -- effectively the same as 'secure'
      c.httpOnly = true;
      break;

    default:
      c.extensions = c.extensions || [];
      c.extensions.push(av);
      break;
    ***REMOVED***
  ***REMOVED***

  return c;
***REMOVED***

// avoid the V8 deoptimization monster!
function jsonParse(str) ***REMOVED***
  var obj;
  try ***REMOVED***
    obj = JSON.parse(str);
  ***REMOVED*** catch (e) ***REMOVED***
    return e;
  ***REMOVED***
  return obj;
***REMOVED***

function fromJSON(str) ***REMOVED***
  if (!str) ***REMOVED***
    return null;
  ***REMOVED***

  var obj;
  if (typeof str === 'string') ***REMOVED***
    obj = jsonParse(str);
    if (obj instanceof Error) ***REMOVED***
      return null;
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    // assume it's an Object
    obj = str;
  ***REMOVED***

  var c = new Cookie();
  for (var i=0; i<Cookie.serializableProperties.length; i++) ***REMOVED***
    var prop = Cookie.serializableProperties[i];
    if (obj[prop] === undefined ||
        obj[prop] === Cookie.prototype[prop])
    ***REMOVED***
      continue; // leave as prototype default
    ***REMOVED***

    if (prop === 'expires' ||
        prop === 'creation' ||
        prop === 'lastAccessed')
    ***REMOVED***
      if (obj[prop] === null) ***REMOVED***
        c[prop] = null;
      ***REMOVED*** else ***REMOVED***
        c[prop] = obj[prop] == "Infinity" ?
          "Infinity" : new Date(obj[prop]);
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      c[prop] = obj[prop];
    ***REMOVED***
  ***REMOVED***

  return c;
***REMOVED***

/* Section 5.4 part 2:
 * "*  Cookies with longer paths are listed before cookies with
 *     shorter paths.
 *
 *  *  Among cookies that have equal-length path fields, cookies with
 *     earlier creation-times are listed before cookies with later
 *     creation-times."
 */

function cookieCompare(a,b) ***REMOVED***
  var cmp = 0;

  // descending for length: b CMP a
  var aPathLen = a.path ? a.path.length : 0;
  var bPathLen = b.path ? b.path.length : 0;
  cmp = bPathLen - aPathLen;
  if (cmp !== 0) ***REMOVED***
    return cmp;
  ***REMOVED***

  // ascending for time: a CMP b
  var aTime = a.creation ? a.creation.getTime() : MAX_TIME;
  var bTime = b.creation ? b.creation.getTime() : MAX_TIME;
  cmp = aTime - bTime;
  if (cmp !== 0) ***REMOVED***
    return cmp;
  ***REMOVED***

  // break ties for the same millisecond (precision of JavaScript's clock)
  cmp = a.creationIndex - b.creationIndex;

  return cmp;
***REMOVED***

// Gives the permutation of all possible pathMatch()es of a given path. The
// array is in longest-to-shortest order.  Handy for indexing.
function permutePath(path) ***REMOVED***
  if (path === '/') ***REMOVED***
    return ['/'];
  ***REMOVED***
  if (path.lastIndexOf('/') === path.length-1) ***REMOVED***
    path = path.substr(0,path.length-1);
  ***REMOVED***
  var permutations = [path];
  while (path.length > 1) ***REMOVED***
    var lindex = path.lastIndexOf('/');
    if (lindex === 0) ***REMOVED***
      break;
    ***REMOVED***
    path = path.substr(0,lindex);
    permutations.push(path);
  ***REMOVED***
  permutations.push('/');
  return permutations;
***REMOVED***

function getCookieContext(url) ***REMOVED***
  if (url instanceof Object) ***REMOVED***
    return url;
  ***REMOVED***
  // NOTE: decodeURI will throw on malformed URIs (see GH-32).
  // Therefore, we will just skip decoding for such URIs.
  try ***REMOVED***
    url = decodeURI(url);
  ***REMOVED***
  catch(err) ***REMOVED***
    // Silently swallow error
  ***REMOVED***

  return urlParse(url);
***REMOVED***

function Cookie(options) ***REMOVED***
  options = options || ***REMOVED******REMOVED***;

  Object.keys(options).forEach(function(prop) ***REMOVED***
    if (Cookie.prototype.hasOwnProperty(prop) &&
        Cookie.prototype[prop] !== options[prop] &&
        prop.substr(0,1) !== '_')
    ***REMOVED***
      this[prop] = options[prop];
    ***REMOVED***
  ***REMOVED***, this);

  this.creation = this.creation || new Date();

  // used to break creation ties in cookieCompare():
  Object.defineProperty(this, 'creationIndex', ***REMOVED***
    configurable: false,
    enumerable: false, // important for assert.deepEqual checks
    writable: true,
    value: ++Cookie.cookiesCreated
  ***REMOVED***);
***REMOVED***

Cookie.cookiesCreated = 0; // incremented each time a cookie is created

Cookie.parse = parse;
Cookie.fromJSON = fromJSON;

Cookie.prototype.key = "";
Cookie.prototype.value = "";

// the order in which the RFC has them:
Cookie.prototype.expires = "Infinity"; // coerces to literal Infinity
Cookie.prototype.maxAge = null; // takes precedence over expires for TTL
Cookie.prototype.domain = null;
Cookie.prototype.path = null;
Cookie.prototype.secure = false;
Cookie.prototype.httpOnly = false;
Cookie.prototype.extensions = null;

// set by the CookieJar:
Cookie.prototype.hostOnly = null; // boolean when set
Cookie.prototype.pathIsDefault = null; // boolean when set
Cookie.prototype.creation = null; // Date when set; defaulted by Cookie.parse
Cookie.prototype.lastAccessed = null; // Date when set
Object.defineProperty(Cookie.prototype, 'creationIndex', ***REMOVED***
  configurable: true,
  enumerable: false,
  writable: true,
  value: 0
***REMOVED***);

Cookie.serializableProperties = Object.keys(Cookie.prototype)
  .filter(function(prop) ***REMOVED***
    return !(
      Cookie.prototype[prop] instanceof Function ||
      prop === 'creationIndex' ||
      prop.substr(0,1) === '_'
    );
  ***REMOVED***);

Cookie.prototype.inspect = function inspect() ***REMOVED***
  var now = Date.now();
  return 'Cookie="'+this.toString() +
    '; hostOnly='+(this.hostOnly != null ? this.hostOnly : '?') +
    '; aAge='+(this.lastAccessed ? (now-this.lastAccessed.getTime())+'ms' : '?') +
    '; cAge='+(this.creation ? (now-this.creation.getTime())+'ms' : '?') +
    '"';
***REMOVED***;

Cookie.prototype.toJSON = function() ***REMOVED***
  var obj = ***REMOVED******REMOVED***;

  var props = Cookie.serializableProperties;
  for (var i=0; i<props.length; i++) ***REMOVED***
    var prop = props[i];
    if (this[prop] === Cookie.prototype[prop]) ***REMOVED***
      continue; // leave as prototype default
    ***REMOVED***

    if (prop === 'expires' ||
        prop === 'creation' ||
        prop === 'lastAccessed')
    ***REMOVED***
      if (this[prop] === null) ***REMOVED***
        obj[prop] = null;
      ***REMOVED*** else ***REMOVED***
        obj[prop] = this[prop] == "Infinity" ? // intentionally not ===
          "Infinity" : this[prop].toISOString();
      ***REMOVED***
    ***REMOVED*** else if (prop === 'maxAge') ***REMOVED***
      if (this[prop] !== null) ***REMOVED***
        // again, intentionally not ===
        obj[prop] = (this[prop] == Infinity || this[prop] == -Infinity) ?
          this[prop].toString() : this[prop];
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      if (this[prop] !== Cookie.prototype[prop]) ***REMOVED***
        obj[prop] = this[prop];
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  return obj;
***REMOVED***;

Cookie.prototype.clone = function() ***REMOVED***
  return fromJSON(this.toJSON());
***REMOVED***;

Cookie.prototype.validate = function validate() ***REMOVED***
  if (!COOKIE_OCTETS.test(this.value)) ***REMOVED***
    return false;
  ***REMOVED***
  if (this.expires != Infinity && !(this.expires instanceof Date) && !parseDate(this.expires)) ***REMOVED***
    return false;
  ***REMOVED***
  if (this.maxAge != null && this.maxAge <= 0) ***REMOVED***
    return false; // "Max-Age=" non-zero-digit *DIGIT
  ***REMOVED***
  if (this.path != null && !PATH_VALUE.test(this.path)) ***REMOVED***
    return false;
  ***REMOVED***

  var cdomain = this.cdomain();
  if (cdomain) ***REMOVED***
    if (cdomain.match(/\.$/)) ***REMOVED***
      return false; // S4.1.2.3 suggests that this is bad. domainMatch() tests confirm this
    ***REMOVED***
    var suffix = pubsuffix.getPublicSuffix(cdomain);
    if (suffix == null) ***REMOVED*** // it's a public suffix
      return false;
    ***REMOVED***
  ***REMOVED***
  return true;
***REMOVED***;

Cookie.prototype.setExpires = function setExpires(exp) ***REMOVED***
  if (exp instanceof Date) ***REMOVED***
    this.expires = exp;
  ***REMOVED*** else ***REMOVED***
    this.expires = parseDate(exp) || "Infinity";
  ***REMOVED***
***REMOVED***;

Cookie.prototype.setMaxAge = function setMaxAge(age) ***REMOVED***
  if (age === Infinity || age === -Infinity) ***REMOVED***
    this.maxAge = age.toString(); // so JSON.stringify() works
  ***REMOVED*** else ***REMOVED***
    this.maxAge = age;
  ***REMOVED***
***REMOVED***;

// gives Cookie header format
Cookie.prototype.cookieString = function cookieString() ***REMOVED***
  var val = this.value;
  if (val == null) ***REMOVED***
    val = '';
  ***REMOVED***
  if (this.key === '') ***REMOVED***
    return val;
  ***REMOVED***
  return this.key+'='+val;
***REMOVED***;

// gives Set-Cookie header format
Cookie.prototype.toString = function toString() ***REMOVED***
  var str = this.cookieString();

  if (this.expires != Infinity) ***REMOVED***
    if (this.expires instanceof Date) ***REMOVED***
      str += '; Expires='+formatDate(this.expires);
    ***REMOVED*** else ***REMOVED***
      str += '; Expires='+this.expires;
    ***REMOVED***
  ***REMOVED***

  if (this.maxAge != null && this.maxAge != Infinity) ***REMOVED***
    str += '; Max-Age='+this.maxAge;
  ***REMOVED***

  if (this.domain && !this.hostOnly) ***REMOVED***
    str += '; Domain='+this.domain;
  ***REMOVED***
  if (this.path) ***REMOVED***
    str += '; Path='+this.path;
  ***REMOVED***

  if (this.secure) ***REMOVED***
    str += '; Secure';
  ***REMOVED***
  if (this.httpOnly) ***REMOVED***
    str += '; HttpOnly';
  ***REMOVED***
  if (this.extensions) ***REMOVED***
    this.extensions.forEach(function(ext) ***REMOVED***
      str += '; '+ext;
    ***REMOVED***);
  ***REMOVED***

  return str;
***REMOVED***;

// TTL() partially replaces the "expiry-time" parts of S5.3 step 3 (setCookie()
// elsewhere)
// S5.3 says to give the "latest representable date" for which we use Infinity
// For "expired" we use 0
Cookie.prototype.TTL = function TTL(now) ***REMOVED***
  /* RFC6265 S4.1.2.2 If a cookie has both the Max-Age and the Expires
   * attribute, the Max-Age attribute has precedence and controls the
   * expiration date of the cookie.
   * (Concurs with S5.3 step 3)
   */
  if (this.maxAge != null) ***REMOVED***
    return this.maxAge<=0 ? 0 : this.maxAge*1000;
  ***REMOVED***

  var expires = this.expires;
  if (expires != Infinity) ***REMOVED***
    if (!(expires instanceof Date)) ***REMOVED***
      expires = parseDate(expires) || Infinity;
    ***REMOVED***

    if (expires == Infinity) ***REMOVED***
      return Infinity;
    ***REMOVED***

    return expires.getTime() - (now || Date.now());
  ***REMOVED***

  return Infinity;
***REMOVED***;

// expiryTime() replaces the "expiry-time" parts of S5.3 step 3 (setCookie()
// elsewhere)
Cookie.prototype.expiryTime = function expiryTime(now) ***REMOVED***
  if (this.maxAge != null) ***REMOVED***
    var relativeTo = now || this.creation || new Date();
    var age = (this.maxAge <= 0) ? -Infinity : this.maxAge*1000;
    return relativeTo.getTime() + age;
  ***REMOVED***

  if (this.expires == Infinity) ***REMOVED***
    return Infinity;
  ***REMOVED***
  return this.expires.getTime();
***REMOVED***;

// expiryDate() replaces the "expiry-time" parts of S5.3 step 3 (setCookie()
// elsewhere), except it returns a Date
Cookie.prototype.expiryDate = function expiryDate(now) ***REMOVED***
  var millisec = this.expiryTime(now);
  if (millisec == Infinity) ***REMOVED***
    return new Date(MAX_TIME);
  ***REMOVED*** else if (millisec == -Infinity) ***REMOVED***
    return new Date(MIN_TIME);
  ***REMOVED*** else ***REMOVED***
    return new Date(millisec);
  ***REMOVED***
***REMOVED***;

// This replaces the "persistent-flag" parts of S5.3 step 3
Cookie.prototype.isPersistent = function isPersistent() ***REMOVED***
  return (this.maxAge != null || this.expires != Infinity);
***REMOVED***;

// Mostly S5.1.2 and S5.2.3:
Cookie.prototype.cdomain =
Cookie.prototype.canonicalizedDomain = function canonicalizedDomain() ***REMOVED***
  if (this.domain == null) ***REMOVED***
    return null;
  ***REMOVED***
  return canonicalDomain(this.domain);
***REMOVED***;

function CookieJar(store, options) ***REMOVED***
  if (typeof options === "boolean") ***REMOVED***
    options = ***REMOVED***rejectPublicSuffixes: options***REMOVED***;
  ***REMOVED*** else if (options == null) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options.rejectPublicSuffixes != null) ***REMOVED***
    this.rejectPublicSuffixes = options.rejectPublicSuffixes;
  ***REMOVED***
  if (options.looseMode != null) ***REMOVED***
    this.enableLooseMode = options.looseMode;
  ***REMOVED***

  if (!store) ***REMOVED***
    store = new MemoryCookieStore();
  ***REMOVED***
  this.store = store;
***REMOVED***
CookieJar.prototype.store = null;
CookieJar.prototype.rejectPublicSuffixes = true;
CookieJar.prototype.enableLooseMode = false;
var CAN_BE_SYNC = [];

CAN_BE_SYNC.push('setCookie');
CookieJar.prototype.setCookie = function(cookie, url, options, cb) ***REMOVED***
  var err;
  var context = getCookieContext(url);
  if (options instanceof Function) ***REMOVED***
    cb = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  var host = canonicalDomain(context.hostname);
  var loose = this.enableLooseMode;
  if (options.loose != null) ***REMOVED***
    loose = options.loose;
  ***REMOVED***

  // S5.3 step 1
  if (!(cookie instanceof Cookie)) ***REMOVED***
    cookie = Cookie.parse(cookie, ***REMOVED*** loose: loose ***REMOVED***);
  ***REMOVED***
  if (!cookie) ***REMOVED***
    err = new Error("Cookie failed to parse");
    return cb(options.ignoreError ? null : err);
  ***REMOVED***

  // S5.3 step 2
  var now = options.now || new Date(); // will assign later to save effort in the face of errors

  // S5.3 step 3: NOOP; persistent-flag and expiry-time is handled by getCookie()

  // S5.3 step 4: NOOP; domain is null by default

  // S5.3 step 5: public suffixes
  if (this.rejectPublicSuffixes && cookie.domain) ***REMOVED***
    var suffix = pubsuffix.getPublicSuffix(cookie.cdomain());
    if (suffix == null) ***REMOVED*** // e.g. "com"
      err = new Error("Cookie has domain set to a public suffix");
      return cb(options.ignoreError ? null : err);
    ***REMOVED***
  ***REMOVED***

  // S5.3 step 6:
  if (cookie.domain) ***REMOVED***
    if (!domainMatch(host, cookie.cdomain(), false)) ***REMOVED***
      err = new Error("Cookie not in this host's domain. Cookie:"+cookie.cdomain()+" Request:"+host);
      return cb(options.ignoreError ? null : err);
    ***REMOVED***

    if (cookie.hostOnly == null) ***REMOVED*** // don't reset if already set
      cookie.hostOnly = false;
    ***REMOVED***

  ***REMOVED*** else ***REMOVED***
    cookie.hostOnly = true;
    cookie.domain = host;
  ***REMOVED***

  //S5.2.4 If the attribute-value is empty or if the first character of the
  //attribute-value is not %x2F ("/"):
  //Let cookie-path be the default-path.
  if (!cookie.path || cookie.path[0] !== '/') ***REMOVED***
    cookie.path = defaultPath(context.pathname);
    cookie.pathIsDefault = true;
  ***REMOVED***

  // S5.3 step 8: NOOP; secure attribute
  // S5.3 step 9: NOOP; httpOnly attribute

  // S5.3 step 10
  if (options.http === false && cookie.httpOnly) ***REMOVED***
    err = new Error("Cookie is HttpOnly and this isn't an HTTP API");
    return cb(options.ignoreError ? null : err);
  ***REMOVED***

  var store = this.store;

  if (!store.updateCookie) ***REMOVED***
    store.updateCookie = function(oldCookie, newCookie, cb) ***REMOVED***
      this.putCookie(newCookie, cb);
    ***REMOVED***;
  ***REMOVED***

  function withCookie(err, oldCookie) ***REMOVED***
    if (err) ***REMOVED***
      return cb(err);
    ***REMOVED***

    var next = function(err) ***REMOVED***
      if (err) ***REMOVED***
        return cb(err);
      ***REMOVED*** else ***REMOVED***
        cb(null, cookie);
      ***REMOVED***
    ***REMOVED***;

    if (oldCookie) ***REMOVED***
      // S5.3 step 11 - "If the cookie store contains a cookie with the same name,
      // domain, and path as the newly created cookie:"
      if (options.http === false && oldCookie.httpOnly) ***REMOVED*** // step 11.2
        err = new Error("old Cookie is HttpOnly and this isn't an HTTP API");
        return cb(options.ignoreError ? null : err);
      ***REMOVED***
      cookie.creation = oldCookie.creation; // step 11.3
      cookie.creationIndex = oldCookie.creationIndex; // preserve tie-breaker
      cookie.lastAccessed = now;
      // Step 11.4 (delete cookie) is implied by just setting the new one:
      store.updateCookie(oldCookie, cookie, next); // step 12

    ***REMOVED*** else ***REMOVED***
      cookie.creation = cookie.lastAccessed = now;
      store.putCookie(cookie, next); // step 12
    ***REMOVED***
  ***REMOVED***

  store.findCookie(cookie.domain, cookie.path, cookie.key, withCookie);
***REMOVED***;

// RFC6365 S5.4
CAN_BE_SYNC.push('getCookies');
CookieJar.prototype.getCookies = function(url, options, cb) ***REMOVED***
  var context = getCookieContext(url);
  if (options instanceof Function) ***REMOVED***
    cb = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  var host = canonicalDomain(context.hostname);
  var path = context.pathname || '/';

  var secure = options.secure;
  if (secure == null && context.protocol &&
      (context.protocol == 'https:' || context.protocol == 'wss:'))
  ***REMOVED***
    secure = true;
  ***REMOVED***

  var http = options.http;
  if (http == null) ***REMOVED***
    http = true;
  ***REMOVED***

  var now = options.now || Date.now();
  var expireCheck = options.expire !== false;
  var allPaths = !!options.allPaths;
  var store = this.store;

  function matchingCookie(c) ***REMOVED***
    // "Either:
    //   The cookie's host-only-flag is true and the canonicalized
    //   request-host is identical to the cookie's domain.
    // Or:
    //   The cookie's host-only-flag is false and the canonicalized
    //   request-host domain-matches the cookie's domain."
    if (c.hostOnly) ***REMOVED***
      if (c.domain != host) ***REMOVED***
        return false;
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      if (!domainMatch(host, c.domain, false)) ***REMOVED***
        return false;
      ***REMOVED***
    ***REMOVED***

    // "The request-uri's path path-matches the cookie's path."
    if (!allPaths && !pathMatch(path, c.path)) ***REMOVED***
      return false;
    ***REMOVED***

    // "If the cookie's secure-only-flag is true, then the request-uri's
    // scheme must denote a "secure" protocol"
    if (c.secure && !secure) ***REMOVED***
      return false;
    ***REMOVED***

    // "If the cookie's http-only-flag is true, then exclude the cookie if the
    // cookie-string is being generated for a "non-HTTP" API"
    if (c.httpOnly && !http) ***REMOVED***
      return false;
    ***REMOVED***

    // deferred from S5.3
    // non-RFC: allow retention of expired cookies by choice
    if (expireCheck && c.expiryTime() <= now) ***REMOVED***
      store.removeCookie(c.domain, c.path, c.key, function()***REMOVED******REMOVED***); // result ignored
      return false;
    ***REMOVED***

    return true;
  ***REMOVED***

  store.findCookies(host, allPaths ? null : path, function(err,cookies) ***REMOVED***
    if (err) ***REMOVED***
      return cb(err);
    ***REMOVED***

    cookies = cookies.filter(matchingCookie);

    // sorting of S5.4 part 2
    if (options.sort !== false) ***REMOVED***
      cookies = cookies.sort(cookieCompare);
    ***REMOVED***

    // S5.4 part 3
    var now = new Date();
    cookies.forEach(function(c) ***REMOVED***
      c.lastAccessed = now;
    ***REMOVED***);
    // TODO persist lastAccessed

    cb(null,cookies);
  ***REMOVED***);
***REMOVED***;

CAN_BE_SYNC.push('getCookieString');
CookieJar.prototype.getCookieString = function(/*..., cb*/) ***REMOVED***
  var args = Array.prototype.slice.call(arguments,0);
  var cb = args.pop();
  var next = function(err,cookies) ***REMOVED***
    if (err) ***REMOVED***
      cb(err);
    ***REMOVED*** else ***REMOVED***
      cb(null, cookies
        .sort(cookieCompare)
        .map(function(c)***REMOVED***
          return c.cookieString();
        ***REMOVED***)
        .join('; '));
    ***REMOVED***
  ***REMOVED***;
  args.push(next);
  this.getCookies.apply(this,args);
***REMOVED***;

CAN_BE_SYNC.push('getSetCookieStrings');
CookieJar.prototype.getSetCookieStrings = function(/*..., cb*/) ***REMOVED***
  var args = Array.prototype.slice.call(arguments,0);
  var cb = args.pop();
  var next = function(err,cookies) ***REMOVED***
    if (err) ***REMOVED***
      cb(err);
    ***REMOVED*** else ***REMOVED***
      cb(null, cookies.map(function(c)***REMOVED***
        return c.toString();
      ***REMOVED***));
    ***REMOVED***
  ***REMOVED***;
  args.push(next);
  this.getCookies.apply(this,args);
***REMOVED***;

CAN_BE_SYNC.push('serialize');
CookieJar.prototype.serialize = function(cb) ***REMOVED***
  var type = this.store.constructor.name;
  if (type === 'Object') ***REMOVED***
    type = null;
  ***REMOVED***

  // update README.md "Serialization Format" if you change this, please!
  var serialized = ***REMOVED***
    // The version of tough-cookie that serialized this jar. Generally a good
    // practice since future versions can make data import decisions based on
    // known past behavior. When/if this matters, use `semver`.
    version: 'tough-cookie@'+VERSION,

    // add the store type, to make humans happy:
    storeType: type,

    // CookieJar configuration:
    rejectPublicSuffixes: !!this.rejectPublicSuffixes,

    // this gets filled from getAllCookies:
    cookies: []
  ***REMOVED***;

  if (!(this.store.getAllCookies &&
        typeof this.store.getAllCookies === 'function'))
  ***REMOVED***
    return cb(new Error('store does not support getAllCookies and cannot be serialized'));
  ***REMOVED***

  this.store.getAllCookies(function(err,cookies) ***REMOVED***
    if (err) ***REMOVED***
      return cb(err);
    ***REMOVED***

    serialized.cookies = cookies.map(function(cookie) ***REMOVED***
      // convert to serialized 'raw' cookies
      cookie = (cookie instanceof Cookie) ? cookie.toJSON() : cookie;

      // Remove the index so new ones get assigned during deserialization
      delete cookie.creationIndex;

      return cookie;
    ***REMOVED***);

    return cb(null, serialized);
  ***REMOVED***);
***REMOVED***;

// well-known name that JSON.stringify calls
CookieJar.prototype.toJSON = function() ***REMOVED***
  return this.serializeSync();
***REMOVED***;

// use the class method CookieJar.deserialize instead of calling this directly
CAN_BE_SYNC.push('_importCookies');
CookieJar.prototype._importCookies = function(serialized, cb) ***REMOVED***
  var jar = this;
  var cookies = serialized.cookies;
  if (!cookies || !Array.isArray(cookies)) ***REMOVED***
    return cb(new Error('serialized jar has no cookies array'));
  ***REMOVED***
  cookies = cookies.slice(); // do not modify the original

  function putNext(err) ***REMOVED***
    if (err) ***REMOVED***
      return cb(err);
    ***REMOVED***

    if (!cookies.length) ***REMOVED***
      return cb(err, jar);
    ***REMOVED***

    var cookie;
    try ***REMOVED***
      cookie = fromJSON(cookies.shift());
    ***REMOVED*** catch (e) ***REMOVED***
      return cb(e);
    ***REMOVED***

    if (cookie === null) ***REMOVED***
      return putNext(null); // skip this cookie
    ***REMOVED***

    jar.store.putCookie(cookie, putNext);
  ***REMOVED***

  putNext();
***REMOVED***;

CookieJar.deserialize = function(strOrObj, store, cb) ***REMOVED***
  if (arguments.length !== 3) ***REMOVED***
    // store is optional
    cb = store;
    store = null;
  ***REMOVED***

  var serialized;
  if (typeof strOrObj === 'string') ***REMOVED***
    serialized = jsonParse(strOrObj);
    if (serialized instanceof Error) ***REMOVED***
      return cb(serialized);
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    serialized = strOrObj;
  ***REMOVED***

  var jar = new CookieJar(store, serialized.rejectPublicSuffixes);
  jar._importCookies(serialized, function(err) ***REMOVED***
    if (err) ***REMOVED***
      return cb(err);
    ***REMOVED***
    cb(null, jar);
  ***REMOVED***);
***REMOVED***;

CookieJar.deserializeSync = function(strOrObj, store) ***REMOVED***
  var serialized = typeof strOrObj === 'string' ?
    JSON.parse(strOrObj) : strOrObj;
  var jar = new CookieJar(store, serialized.rejectPublicSuffixes);

  // catch this mistake early:
  if (!jar.store.synchronous) ***REMOVED***
    throw new Error('CookieJar store is not synchronous; use async API instead.');
  ***REMOVED***

  jar._importCookiesSync(serialized);
  return jar;
***REMOVED***;
CookieJar.fromJSON = CookieJar.deserializeSync;

CAN_BE_SYNC.push('clone');
CookieJar.prototype.clone = function(newStore, cb) ***REMOVED***
  if (arguments.length === 1) ***REMOVED***
    cb = newStore;
    newStore = null;
  ***REMOVED***

  this.serialize(function(err,serialized) ***REMOVED***
    if (err) ***REMOVED***
      return cb(err);
    ***REMOVED***
    CookieJar.deserialize(newStore, serialized, cb);
  ***REMOVED***);
***REMOVED***;

// Use a closure to provide a true imperative API for synchronous stores.
function syncWrap(method) ***REMOVED***
  return function() ***REMOVED***
    if (!this.store.synchronous) ***REMOVED***
      throw new Error('CookieJar store is not synchronous; use async API instead.');
    ***REMOVED***

    var args = Array.prototype.slice.call(arguments);
    var syncErr, syncResult;
    args.push(function syncCb(err, result) ***REMOVED***
      syncErr = err;
      syncResult = result;
    ***REMOVED***);
    this[method].apply(this, args);

    if (syncErr) ***REMOVED***
      throw syncErr;
    ***REMOVED***
    return syncResult;
  ***REMOVED***;
***REMOVED***

// wrap all declared CAN_BE_SYNC methods in the sync wrapper
CAN_BE_SYNC.forEach(function(method) ***REMOVED***
  CookieJar.prototype[method+'Sync'] = syncWrap(method);
***REMOVED***);

module.exports = ***REMOVED***
  CookieJar: CookieJar,
  Cookie: Cookie,
  Store: Store,
  MemoryCookieStore: MemoryCookieStore,
  parseDate: parseDate,
  formatDate: formatDate,
  parse: parse,
  fromJSON: fromJSON,
  domainMatch: domainMatch,
  defaultPath: defaultPath,
  pathMatch: pathMatch,
  getPublicSuffix: pubsuffix.getPublicSuffix,
  cookieCompare: cookieCompare,
  permuteDomain: require('./permuteDomain').permuteDomain,
  permutePath: permutePath,
  canonicalDomain: canonicalDomain
***REMOVED***;
