// tough-cookie/lib/memstore.js
/*!
 * Copyright (c) 2015, Salesforce.com, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of Salesforce.com nor the names of its contributors may
 * be used to endorse or promote products derived from this software without
 * specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
'use strict';
var Store = require('./store').Store;
var permuteDomain = require('./permuteDomain').permuteDomain;
var pathMatch = require('./pathMatch').pathMatch;
var util = require('util');

function MemoryCookieStore() ***REMOVED***
  Store.call(this);
  this.idx = ***REMOVED******REMOVED***;
***REMOVED***
util.inherits(MemoryCookieStore, Store);
exports.MemoryCookieStore = MemoryCookieStore;
MemoryCookieStore.prototype.idx = null;

// Since it's just a struct in RAM, this Store is synchronous
MemoryCookieStore.prototype.synchronous = true;

// force a default depth:
MemoryCookieStore.prototype.inspect = function() ***REMOVED***
  return "***REMOVED*** idx: "+util.inspect(this.idx, false, 2)+' ***REMOVED***';
***REMOVED***;

MemoryCookieStore.prototype.findCookie = function(domain, path, key, cb) ***REMOVED***
  if (!this.idx[domain]) ***REMOVED***
    return cb(null,undefined);
  ***REMOVED***
  if (!this.idx[domain][path]) ***REMOVED***
    return cb(null,undefined);
  ***REMOVED***
  return cb(null,this.idx[domain][path][key]||null);
***REMOVED***;

MemoryCookieStore.prototype.findCookies = function(domain, path, cb) ***REMOVED***
  var results = [];
  if (!domain) ***REMOVED***
    return cb(null,[]);
  ***REMOVED***

  var pathMatcher;
  if (!path) ***REMOVED***
    // null means "all paths"
    pathMatcher = function matchAll(domainIndex) ***REMOVED***
      for (var curPath in domainIndex) ***REMOVED***
        var pathIndex = domainIndex[curPath];
        for (var key in pathIndex) ***REMOVED***
          results.push(pathIndex[key]);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;

  ***REMOVED*** else ***REMOVED***
    pathMatcher = function matchRFC(domainIndex) ***REMOVED***
       //NOTE: we should use path-match algorithm from S5.1.4 here
       //(see : https://github.com/ChromiumWebApps/chromium/blob/b3d3b4da8bb94c1b2e061600df106d590fda3620/net/cookies/canonical_cookie.cc#L299)
       Object.keys(domainIndex).forEach(function (cookiePath) ***REMOVED***
         if (pathMatch(path, cookiePath)) ***REMOVED***
           var pathIndex = domainIndex[cookiePath];

           for (var key in pathIndex) ***REMOVED***
             results.push(pathIndex[key]);
           ***REMOVED***
         ***REMOVED***
       ***REMOVED***);
     ***REMOVED***;
  ***REMOVED***

  var domains = permuteDomain(domain) || [domain];
  var idx = this.idx;
  domains.forEach(function(curDomain) ***REMOVED***
    var domainIndex = idx[curDomain];
    if (!domainIndex) ***REMOVED***
      return;
    ***REMOVED***
    pathMatcher(domainIndex);
  ***REMOVED***);

  cb(null,results);
***REMOVED***;

MemoryCookieStore.prototype.putCookie = function(cookie, cb) ***REMOVED***
  if (!this.idx[cookie.domain]) ***REMOVED***
    this.idx[cookie.domain] = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (!this.idx[cookie.domain][cookie.path]) ***REMOVED***
    this.idx[cookie.domain][cookie.path] = ***REMOVED******REMOVED***;
  ***REMOVED***
  this.idx[cookie.domain][cookie.path][cookie.key] = cookie;
  cb(null);
***REMOVED***;

MemoryCookieStore.prototype.updateCookie = function(oldCookie, newCookie, cb) ***REMOVED***
  // updateCookie() may avoid updating cookies that are identical.  For example,
  // lastAccessed may not be important to some stores and an equality
  // comparison could exclude that field.
  this.putCookie(newCookie,cb);
***REMOVED***;

MemoryCookieStore.prototype.removeCookie = function(domain, path, key, cb) ***REMOVED***
  if (this.idx[domain] && this.idx[domain][path] && this.idx[domain][path][key]) ***REMOVED***
    delete this.idx[domain][path][key];
  ***REMOVED***
  cb(null);
***REMOVED***;

MemoryCookieStore.prototype.removeCookies = function(domain, path, cb) ***REMOVED***
  if (this.idx[domain]) ***REMOVED***
    if (path) ***REMOVED***
      delete this.idx[domain][path];
    ***REMOVED*** else ***REMOVED***
      delete this.idx[domain];
    ***REMOVED***
  ***REMOVED***
  return cb(null);
***REMOVED***;

MemoryCookieStore.prototype.getAllCookies = function(cb) ***REMOVED***
  var cookies = [];
  var idx = this.idx;

  var domains = Object.keys(idx);
  domains.forEach(function(domain) ***REMOVED***
    var paths = Object.keys(idx[domain]);
    paths.forEach(function(path) ***REMOVED***
      var keys = Object.keys(idx[domain][path]);
      keys.forEach(function(key) ***REMOVED***
        if (key !== null) ***REMOVED***
          cookies.push(idx[domain][path][key]);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***);

  // Sort by creationIndex so deserializing retains the creation order.
  // When implementing your own store, this SHOULD retain the order too
  cookies.sort(function(a,b) ***REMOVED***
    return (a.creationIndex||0) - (b.creationIndex||0);
  ***REMOVED***);

  cb(null, cookies);
***REMOVED***;
