// mongodb/lib/admin.js
'use strict';

const executeOperation = require('./utils').executeOperation;
const applyWriteConcern = require('./utils').applyWriteConcern;

const addUser = require('./operations/db_ops').addUser;
const executeDbAdminCommand = require('./operations/db_ops').executeDbAdminCommand;
const removeUser = require('./operations/db_ops').removeUser;
const replSetGetStatus = require('./operations/admin_ops').replSetGetStatus;
const serverStatus = require('./operations/admin_ops').serverStatus;
const validateCollection = require('./operations/admin_ops').validateCollection;

/**
 * @fileOverview The **Admin** class is an internal class that allows convenient access to
 * the admin functionality and commands for MongoDB.
 *
 * **ADMIN Cannot directly be instantiated**
 * @example
 * const MongoClient = require('mongodb').MongoClient;
 * const test = require('assert');
 * // Connection url
 * const url = 'mongodb://localhost:27017';
 * // Database Name
 * const dbName = 'test';
 *
 * // Connect using MongoClient
 * MongoClient.connect(url, function(err, client) ***REMOVED***
 *   // Use the admin database for the operation
 *   const adminDb = client.db(dbName).admin();
 *
 *   // List all the available databases
 *   adminDb.listDatabases(function(err, dbs) ***REMOVED***
 *     test.equal(null, err);
 *     test.ok(dbs.databases.length > 0);
 *     client.close();
 *   ***REMOVED***);
 * ***REMOVED***);
 */

/**
 * Create a new Admin instance (INTERNAL TYPE, do not instantiate directly)
 * @class
 * @return ***REMOVED***Admin***REMOVED*** a collection instance.
 */
function Admin(db, topology, promiseLibrary) ***REMOVED***
  if (!(this instanceof Admin)) return new Admin(db, topology);

  // Internal state
  this.s = ***REMOVED***
    db: db,
    topology: topology,
    promiseLibrary: promiseLibrary
  ***REMOVED***;
***REMOVED***

/**
 * The callback format for results
 * @callback Admin~resultCallback
 * @param ***REMOVED***MongoError***REMOVED*** error An error instance representing the error during the execution.
 * @param ***REMOVED***object***REMOVED*** result The result object if the command was executed successfully.
 */

/**
 * Execute a command
 * @method
 * @param ***REMOVED***object***REMOVED*** command The command hash
 * @param ***REMOVED***object***REMOVED*** [options] Optional settings.
 * @param ***REMOVED***(ReadPreference|string)***REMOVED*** [options.readPreference] The preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST).
 * @param ***REMOVED***number***REMOVED*** [options.maxTimeMS] Number of milliseconds to wait before aborting the query.
 * @param ***REMOVED***Admin~resultCallback***REMOVED*** [callback] The command result callback
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Admin.prototype.command = function(command, options, callback) ***REMOVED***
  const args = Array.prototype.slice.call(arguments, 1);
  callback = typeof args[args.length - 1] === 'function' ? args.pop() : undefined;
  options = args.length ? args.shift() : ***REMOVED******REMOVED***;

  return executeOperation(this.s.db.s.topology, executeDbAdminCommand.bind(this.s.db), [
    this.s.db,
    command,
    options,
    callback
  ]);
***REMOVED***;

/**
 * Retrieve the server information for the current
 * instance of the db client
 *
 * @param ***REMOVED***Object***REMOVED*** [options] optional parameters for this operation
 * @param ***REMOVED***ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param ***REMOVED***Admin~resultCallback***REMOVED*** [callback] The command result callback
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Admin.prototype.buildInfo = function(options, callback) ***REMOVED***
  if (typeof options === 'function') (callback = options), (options = ***REMOVED******REMOVED***);
  options = options || ***REMOVED******REMOVED***;

  const cmd = ***REMOVED*** buildinfo: 1 ***REMOVED***;
  return executeOperation(this.s.db.s.topology, executeDbAdminCommand.bind(this.s.db), [
    this.s.db,
    cmd,
    options,
    callback
  ]);
***REMOVED***;

/**
 * Retrieve the server information for the current
 * instance of the db client
 *
 * @param ***REMOVED***Object***REMOVED*** [options] optional parameters for this operation
 * @param ***REMOVED***ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param ***REMOVED***Admin~resultCallback***REMOVED*** [callback] The command result callback
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Admin.prototype.serverInfo = function(options, callback) ***REMOVED***
  if (typeof options === 'function') (callback = options), (options = ***REMOVED******REMOVED***);
  options = options || ***REMOVED******REMOVED***;

  const cmd = ***REMOVED*** buildinfo: 1 ***REMOVED***;
  return executeOperation(this.s.db.s.topology, executeDbAdminCommand.bind(this.s.db), [
    this.s.db,
    cmd,
    options,
    callback
  ]);
***REMOVED***;

/**
 * Retrieve this db's server status.
 *
 * @param ***REMOVED***Object***REMOVED*** [options] optional parameters for this operation
 * @param ***REMOVED***ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param ***REMOVED***Admin~resultCallback***REMOVED*** [callback] The command result callback
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Admin.prototype.serverStatus = function(options, callback) ***REMOVED***
  if (typeof options === 'function') (callback = options), (options = ***REMOVED******REMOVED***);
  options = options || ***REMOVED******REMOVED***;

  return executeOperation(this.s.db.s.topology, serverStatus, [this, options, callback]);
***REMOVED***;

/**
 * Ping the MongoDB server and retrieve results
 *
 * @param ***REMOVED***Object***REMOVED*** [options] optional parameters for this operation
 * @param ***REMOVED***ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param ***REMOVED***Admin~resultCallback***REMOVED*** [callback] The command result callback
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Admin.prototype.ping = function(options, callback) ***REMOVED***
  if (typeof options === 'function') (callback = options), (options = ***REMOVED******REMOVED***);
  options = options || ***REMOVED******REMOVED***;

  const cmd = ***REMOVED*** ping: 1 ***REMOVED***;
  return executeOperation(this.s.db.s.topology, executeDbAdminCommand.bind(this.s.db), [
    this.s.db,
    cmd,
    options,
    callback
  ]);
***REMOVED***;

/**
 * Add a user to the database.
 * @method
 * @param ***REMOVED***string***REMOVED*** username The username.
 * @param ***REMOVED***string***REMOVED*** password The password.
 * @param ***REMOVED***object***REMOVED*** [options] Optional settings.
 * @param ***REMOVED***(number|string)***REMOVED*** [options.w] The write concern.
 * @param ***REMOVED***number***REMOVED*** [options.wtimeout] The write concern timeout.
 * @param ***REMOVED***boolean***REMOVED*** [options.j=false] Specify a journal write concern.
 * @param ***REMOVED***boolean***REMOVED*** [options.fsync=false] Specify a file sync write concern.
 * @param ***REMOVED***object***REMOVED*** [options.customData] Custom data associated with the user (only Mongodb 2.6 or higher)
 * @param ***REMOVED***object[]***REMOVED*** [options.roles] Roles associated with the created user (only Mongodb 2.6 or higher)
 * @param ***REMOVED***ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param ***REMOVED***Admin~resultCallback***REMOVED*** [callback] The command result callback
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Admin.prototype.addUser = function(username, password, options, callback) ***REMOVED***
  const args = Array.prototype.slice.call(arguments, 2);
  callback = typeof args[args.length - 1] === 'function' ? args.pop() : undefined;

  options = args.length ? args.shift() : ***REMOVED******REMOVED***;
  options = Object.assign(***REMOVED******REMOVED***, options);
  // Get the options
  options = applyWriteConcern(options, ***REMOVED*** db: this.s.db ***REMOVED***);
  // Set the db name to admin
  options.dbName = 'admin';

  return executeOperation(this.s.db.s.topology, addUser, [
    this.s.db,
    username,
    password,
    options,
    callback
  ]);
***REMOVED***;

/**
 * Remove a user from a database
 * @method
 * @param ***REMOVED***string***REMOVED*** username The username.
 * @param ***REMOVED***object***REMOVED*** [options] Optional settings.
 * @param ***REMOVED***(number|string)***REMOVED*** [options.w] The write concern.
 * @param ***REMOVED***number***REMOVED*** [options.wtimeout] The write concern timeout.
 * @param ***REMOVED***boolean***REMOVED*** [options.j=false] Specify a journal write concern.
 * @param ***REMOVED***boolean***REMOVED*** [options.fsync=false] Specify a file sync write concern.
 * @param ***REMOVED***ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param ***REMOVED***Admin~resultCallback***REMOVED*** [callback] The command result callback
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Admin.prototype.removeUser = function(username, options, callback) ***REMOVED***
  const args = Array.prototype.slice.call(arguments, 1);
  callback = typeof args[args.length - 1] === 'function' ? args.pop() : undefined;

  options = args.length ? args.shift() : ***REMOVED******REMOVED***;
  options = Object.assign(***REMOVED******REMOVED***, options);
  // Get the options
  options = applyWriteConcern(options, ***REMOVED*** db: this.s.db ***REMOVED***);
  // Set the db name
  options.dbName = 'admin';

  return executeOperation(this.s.db.s.topology, removeUser, [
    this.s.db,
    username,
    options,
    callback
  ]);
***REMOVED***;

/**
 * Validate an existing collection
 *
 * @param ***REMOVED***string***REMOVED*** collectionName The name of the collection to validate.
 * @param ***REMOVED***object***REMOVED*** [options] Optional settings.
 * @param ***REMOVED***ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param ***REMOVED***Admin~resultCallback***REMOVED*** [callback] The command result callback.
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Admin.prototype.validateCollection = function(collectionName, options, callback) ***REMOVED***
  if (typeof options === 'function') (callback = options), (options = ***REMOVED******REMOVED***);
  options = options || ***REMOVED******REMOVED***;

  return executeOperation(this.s.db.s.topology, validateCollection, [
    this,
    collectionName,
    options,
    callback
  ]);
***REMOVED***;

/**
 * List the available databases
 *
 * @param ***REMOVED***object***REMOVED*** [options] Optional settings.
 * @param ***REMOVED***boolean***REMOVED*** [options.nameOnly=false] Whether the command should return only db names, or names and size info.
 * @param ***REMOVED***ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param ***REMOVED***Admin~resultCallback***REMOVED*** [callback] The command result callback.
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Admin.prototype.listDatabases = function(options, callback) ***REMOVED***
  if (typeof options === 'function') (callback = options), (options = ***REMOVED******REMOVED***);
  options = options || ***REMOVED******REMOVED***;

  const cmd = ***REMOVED*** listDatabases: 1 ***REMOVED***;
  if (options.nameOnly) cmd.nameOnly = Number(cmd.nameOnly);
  return executeOperation(this.s.db.s.topology, executeDbAdminCommand.bind(this.s.db), [
    this.s.db,
    cmd,
    options,
    callback
  ]);
***REMOVED***;

/**
 * Get ReplicaSet status
 *
 * @param ***REMOVED***Object***REMOVED*** [options] optional parameters for this operation
 * @param ***REMOVED***ClientSession***REMOVED*** [options.session] optional session to use for this operation
 * @param ***REMOVED***Admin~resultCallback***REMOVED*** [callback] The command result callback.
 * @return ***REMOVED***Promise***REMOVED*** returns Promise if no callback passed
 */
Admin.prototype.replSetGetStatus = function(options, callback) ***REMOVED***
  if (typeof options === 'function') (callback = options), (options = ***REMOVED******REMOVED***);
  options = options || ***REMOVED******REMOVED***;

  return executeOperation(this.s.db.s.topology, replSetGetStatus, [this, options, callback]);
***REMOVED***;

module.exports = Admin;
