// mquery/lib/collection/node.js
'use strict';

/**
 * Module dependencies
 */

var Collection = require('./collection');
var utils = require('../utils');

function NodeCollection (col) ***REMOVED***
  this.collection = col;
  this.collectionName = col.collectionName;
***REMOVED***

/**
 * inherit from collection base class
 */

utils.inherits(NodeCollection, Collection);

/**
 * find(match, options, function(err, docs))
 */

NodeCollection.prototype.find = function (match, options, cb) ***REMOVED***
  this.collection.find(match, options, function (err, cursor) ***REMOVED***
    if (err) return cb(err);

    try ***REMOVED***
      cursor.toArray(cb);
    ***REMOVED*** catch (error) ***REMOVED***
      cb(error);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***

/**
 * findOne(match, options, function(err, doc))
 */

NodeCollection.prototype.findOne = function (match, options, cb) ***REMOVED***
  this.collection.findOne(match, options, cb);
***REMOVED***

/**
 * count(match, options, function(err, count))
 */

NodeCollection.prototype.count = function (match, options, cb) ***REMOVED***
  this.collection.count(match, options, cb);
***REMOVED***

/**
 * distinct(prop, match, options, function(err, count))
 */

NodeCollection.prototype.distinct  = function (prop, match, options, cb) ***REMOVED***
  this.collection.distinct(prop, match, options, cb);
***REMOVED***

/**
 * update(match, update, options, function(err[, result]))
 */

NodeCollection.prototype.update = function (match, update, options, cb) ***REMOVED***
  this.collection.update(match, update, options, cb);
***REMOVED***

/**
 * update(match, update, options, function(err[, result]))
 */

NodeCollection.prototype.updateMany = function (match, update, options, cb) ***REMOVED***
  this.collection.updateMany(match, update, options, cb);
***REMOVED***

/**
 * update(match, update, options, function(err[, result]))
 */

NodeCollection.prototype.updateOne = function (match, update, options, cb) ***REMOVED***
  this.collection.updateOne(match, update, options, cb);
***REMOVED***

/**
 * replaceOne(match, update, options, function(err[, result]))
 */

NodeCollection.prototype.replaceOne = function (match, update, options, cb) ***REMOVED***
  this.collection.replaceOne(match, update, options, cb);
***REMOVED***

/**
 * deleteOne(match, options, function(err[, result])
 */

NodeCollection.prototype.deleteOne = function (match, options, cb) ***REMOVED***
  this.collection.deleteOne(match, options, cb);
***REMOVED***

/**
 * deleteMany(match, options, function(err[, result])
 */

NodeCollection.prototype.deleteMany = function (match, options, cb) ***REMOVED***
  this.collection.deleteMany(match, options, cb);
***REMOVED***

/**
 * remove(match, options, function(err[, result])
 */

NodeCollection.prototype.remove = function (match, options, cb) ***REMOVED***
  this.collection.remove(match, options, cb);
***REMOVED***

/**
 * findAndModify(match, update, options, function(err, doc))
 */

NodeCollection.prototype.findAndModify = function (match, update, options, cb) ***REMOVED***
  var sort = Array.isArray(options.sort) ? options.sort : [];
  this.collection.findAndModify(match, sort, update, options, cb);
***REMOVED***

/**
 * var stream = findStream(match, findOptions, streamOptions)
 */

NodeCollection.prototype.findStream = function(match, findOptions, streamOptions) ***REMOVED***
  return this.collection.find(match, findOptions);
***REMOVED***

/**
 * var cursor = findCursor(match, findOptions)
 */

NodeCollection.prototype.findCursor = function(match, findOptions) ***REMOVED***
  return this.collection.find(match, findOptions);
***REMOVED***

/**
 * aggregation(operators..., function(err, doc))
 * TODO
 */

/**
 * Expose
 */

module.exports = exports = NodeCollection;
