// source-map/lib/mapping-list.js
/* -*- Mode: js; js-indent-level: 2; -*- */
/*
 * Copyright 2014 Mozilla Foundation and contributors
 * Licensed under the New BSD license. See LICENSE or:
 * http://opensource.org/licenses/BSD-3-Clause
 */

var util = require('./util');

/**
 * Determine whether mappingB is after mappingA with respect to generated
 * position.
 */
function generatedPositionAfter(mappingA, mappingB) ***REMOVED***
  // Optimized for most common case
  var lineA = mappingA.generatedLine;
  var lineB = mappingB.generatedLine;
  var columnA = mappingA.generatedColumn;
  var columnB = mappingB.generatedColumn;
  return lineB > lineA || lineB == lineA && columnB >= columnA ||
         util.compareByGeneratedPositionsInflated(mappingA, mappingB) <= 0;
***REMOVED***

/**
 * A data structure to provide a sorted view of accumulated mappings in a
 * performance conscious manner. It trades a neglibable overhead in general
 * case for a large speedup in case of mappings being added in order.
 */
function MappingList() ***REMOVED***
  this._array = [];
  this._sorted = true;
  // Serves as infimum
  this._last = ***REMOVED***generatedLine: -1, generatedColumn: 0***REMOVED***;
***REMOVED***

/**
 * Iterate through internal items. This method takes the same arguments that
 * `Array.prototype.forEach` takes.
 *
 * NOTE: The order of the mappings is NOT guaranteed.
 */
MappingList.prototype.unsortedForEach =
  function MappingList_forEach(aCallback, aThisArg) ***REMOVED***
    this._array.forEach(aCallback, aThisArg);
  ***REMOVED***;

/**
 * Add the given source mapping.
 *
 * @param Object aMapping
 */
MappingList.prototype.add = function MappingList_add(aMapping) ***REMOVED***
  if (generatedPositionAfter(this._last, aMapping)) ***REMOVED***
    this._last = aMapping;
    this._array.push(aMapping);
  ***REMOVED*** else ***REMOVED***
    this._sorted = false;
    this._array.push(aMapping);
  ***REMOVED***
***REMOVED***;

/**
 * Returns the flat, sorted array of mappings. The mappings are sorted by
 * generated position.
 *
 * WARNING: This method returns internal data without copying, for
 * performance. The return value must NOT be mutated, and should be treated as
 * an immutable borrow. If you want to take ownership, you must make your own
 * copy.
 */
MappingList.prototype.toArray = function MappingList_toArray() ***REMOVED***
  if (!this._sorted) ***REMOVED***
    this._array.sort(util.compareByGeneratedPositionsInflated);
    this._sorted = true;
  ***REMOVED***
  return this._array;
***REMOVED***;

exports.MappingList = MappingList;
