// sshpk/lib/dhe.js
// Copyright 2017 Joyent, Inc.

module.exports = ***REMOVED***
	DiffieHellman: DiffieHellman,
	generateECDSA: generateECDSA,
	generateED25519: generateED25519
***REMOVED***;

var assert = require('assert-plus');
var crypto = require('crypto');
var Buffer = require('safer-buffer').Buffer;
var algs = require('./algs');
var utils = require('./utils');
var nacl;

var Key = require('./key');
var PrivateKey = require('./private-key');

var CRYPTO_HAVE_ECDH = (crypto.createECDH !== undefined);

var ecdh, ec, jsbn;

function DiffieHellman(key) ***REMOVED***
	utils.assertCompatible(key, Key, [1, 4], 'key');
	this._isPriv = PrivateKey.isPrivateKey(key, [1, 3]);
	this._algo = key.type;
	this._curve = key.curve;
	this._key = key;
	if (key.type === 'dsa') ***REMOVED***
		if (!CRYPTO_HAVE_ECDH) ***REMOVED***
			throw (new Error('Due to bugs in the node 0.10 ' +
			    'crypto API, node 0.12.x or later is required ' +
			    'to use DH'));
		***REMOVED***
		this._dh = crypto.createDiffieHellman(
		    key.part.p.data, undefined,
		    key.part.g.data, undefined);
		this._p = key.part.p;
		this._g = key.part.g;
		if (this._isPriv)
			this._dh.setPrivateKey(key.part.x.data);
		this._dh.setPublicKey(key.part.y.data);

	***REMOVED*** else if (key.type === 'ecdsa') ***REMOVED***
		if (!CRYPTO_HAVE_ECDH) ***REMOVED***
			if (ecdh === undefined)
				ecdh = require('ecc-jsbn');
			if (ec === undefined)
				ec = require('ecc-jsbn/lib/ec');
			if (jsbn === undefined)
				jsbn = require('jsbn').BigInteger;

			this._ecParams = new X9ECParameters(this._curve);

			if (this._isPriv) ***REMOVED***
				this._priv = new ECPrivate(
				    this._ecParams, key.part.d.data);
			***REMOVED***
			return;
		***REMOVED***

		var curve = ***REMOVED***
			'nistp256': 'prime256v1',
			'nistp384': 'secp384r1',
			'nistp521': 'secp521r1'
		***REMOVED***[key.curve];
		this._dh = crypto.createECDH(curve);
		if (typeof (this._dh) !== 'object' ||
		    typeof (this._dh.setPrivateKey) !== 'function') ***REMOVED***
			CRYPTO_HAVE_ECDH = false;
			DiffieHellman.call(this, key);
			return;
		***REMOVED***
		if (this._isPriv)
			this._dh.setPrivateKey(key.part.d.data);
		this._dh.setPublicKey(key.part.Q.data);

	***REMOVED*** else if (key.type === 'curve25519') ***REMOVED***
		if (nacl === undefined)
			nacl = require('tweetnacl');

		if (this._isPriv) ***REMOVED***
			utils.assertCompatible(key, PrivateKey, [1, 5], 'key');
			this._priv = key.part.k.data;
		***REMOVED***

	***REMOVED*** else ***REMOVED***
		throw (new Error('DH not supported for ' + key.type + ' keys'));
	***REMOVED***
***REMOVED***

DiffieHellman.prototype.getPublicKey = function () ***REMOVED***
	if (this._isPriv)
		return (this._key.toPublic());
	return (this._key);
***REMOVED***;

DiffieHellman.prototype.getPrivateKey = function () ***REMOVED***
	if (this._isPriv)
		return (this._key);
	else
		return (undefined);
***REMOVED***;
DiffieHellman.prototype.getKey = DiffieHellman.prototype.getPrivateKey;

DiffieHellman.prototype._keyCheck = function (pk, isPub) ***REMOVED***
	assert.object(pk, 'key');
	if (!isPub)
		utils.assertCompatible(pk, PrivateKey, [1, 3], 'key');
	utils.assertCompatible(pk, Key, [1, 4], 'key');

	if (pk.type !== this._algo) ***REMOVED***
		throw (new Error('A ' + pk.type + ' key cannot be used in ' +
		    this._algo + ' Diffie-Hellman'));
	***REMOVED***

	if (pk.curve !== this._curve) ***REMOVED***
		throw (new Error('A key from the ' + pk.curve + ' curve ' +
		    'cannot be used with a ' + this._curve +
		    ' Diffie-Hellman'));
	***REMOVED***

	if (pk.type === 'dsa') ***REMOVED***
		assert.deepEqual(pk.part.p, this._p,
		    'DSA key prime does not match');
		assert.deepEqual(pk.part.g, this._g,
		    'DSA key generator does not match');
	***REMOVED***
***REMOVED***;

DiffieHellman.prototype.setKey = function (pk) ***REMOVED***
	this._keyCheck(pk);

	if (pk.type === 'dsa') ***REMOVED***
		this._dh.setPrivateKey(pk.part.x.data);
		this._dh.setPublicKey(pk.part.y.data);

	***REMOVED*** else if (pk.type === 'ecdsa') ***REMOVED***
		if (CRYPTO_HAVE_ECDH) ***REMOVED***
			this._dh.setPrivateKey(pk.part.d.data);
			this._dh.setPublicKey(pk.part.Q.data);
		***REMOVED*** else ***REMOVED***
			this._priv = new ECPrivate(
			    this._ecParams, pk.part.d.data);
		***REMOVED***

	***REMOVED*** else if (pk.type === 'curve25519') ***REMOVED***
		var k = pk.part.k;
		if (!pk.part.k)
			k = pk.part.r;
		this._priv = k.data;
		if (this._priv[0] === 0x00)
			this._priv = this._priv.slice(1);
		this._priv = this._priv.slice(0, 32);
	***REMOVED***
	this._key = pk;
	this._isPriv = true;
***REMOVED***;
DiffieHellman.prototype.setPrivateKey = DiffieHellman.prototype.setKey;

DiffieHellman.prototype.computeSecret = function (otherpk) ***REMOVED***
	this._keyCheck(otherpk, true);
	if (!this._isPriv)
		throw (new Error('DH exchange has not been initialized with ' +
		    'a private key yet'));

	var pub;
	if (this._algo === 'dsa') ***REMOVED***
		return (this._dh.computeSecret(
		    otherpk.part.y.data));

	***REMOVED*** else if (this._algo === 'ecdsa') ***REMOVED***
		if (CRYPTO_HAVE_ECDH) ***REMOVED***
			return (this._dh.computeSecret(
			    otherpk.part.Q.data));
		***REMOVED*** else ***REMOVED***
			pub = new ECPublic(
			    this._ecParams, otherpk.part.Q.data);
			return (this._priv.deriveSharedSecret(pub));
		***REMOVED***

	***REMOVED*** else if (this._algo === 'curve25519') ***REMOVED***
		pub = otherpk.part.A.data;
		while (pub[0] === 0x00 && pub.length > 32)
			pub = pub.slice(1);
		var priv = this._priv;
		assert.strictEqual(pub.length, 32);
		assert.strictEqual(priv.length, 32);

		var secret = nacl.box.before(new Uint8Array(pub),
		    new Uint8Array(priv));

		return (Buffer.from(secret));
	***REMOVED***

	throw (new Error('Invalid algorithm: ' + this._algo));
***REMOVED***;

DiffieHellman.prototype.generateKey = function () ***REMOVED***
	var parts = [];
	var priv, pub;
	if (this._algo === 'dsa') ***REMOVED***
		this._dh.generateKeys();

		parts.push(***REMOVED***name: 'p', data: this._p.data***REMOVED***);
		parts.push(***REMOVED***name: 'q', data: this._key.part.q.data***REMOVED***);
		parts.push(***REMOVED***name: 'g', data: this._g.data***REMOVED***);
		parts.push(***REMOVED***name: 'y', data: this._dh.getPublicKey()***REMOVED***);
		parts.push(***REMOVED***name: 'x', data: this._dh.getPrivateKey()***REMOVED***);
		this._key = new PrivateKey(***REMOVED***
			type: 'dsa',
			parts: parts
		***REMOVED***);
		this._isPriv = true;
		return (this._key);

	***REMOVED*** else if (this._algo === 'ecdsa') ***REMOVED***
		if (CRYPTO_HAVE_ECDH) ***REMOVED***
			this._dh.generateKeys();

			parts.push(***REMOVED***name: 'curve',
			    data: Buffer.from(this._curve)***REMOVED***);
			parts.push(***REMOVED***name: 'Q', data: this._dh.getPublicKey()***REMOVED***);
			parts.push(***REMOVED***name: 'd', data: this._dh.getPrivateKey()***REMOVED***);
			this._key = new PrivateKey(***REMOVED***
				type: 'ecdsa',
				curve: this._curve,
				parts: parts
			***REMOVED***);
			this._isPriv = true;
			return (this._key);

		***REMOVED*** else ***REMOVED***
			var n = this._ecParams.getN();
			var r = new jsbn(crypto.randomBytes(n.bitLength()));
			var n1 = n.subtract(jsbn.ONE);
			priv = r.mod(n1).add(jsbn.ONE);
			pub = this._ecParams.getG().multiply(priv);

			priv = Buffer.from(priv.toByteArray());
			pub = Buffer.from(this._ecParams.getCurve().
			    encodePointHex(pub), 'hex');

			this._priv = new ECPrivate(this._ecParams, priv);

			parts.push(***REMOVED***name: 'curve',
			    data: Buffer.from(this._curve)***REMOVED***);
			parts.push(***REMOVED***name: 'Q', data: pub***REMOVED***);
			parts.push(***REMOVED***name: 'd', data: priv***REMOVED***);

			this._key = new PrivateKey(***REMOVED***
				type: 'ecdsa',
				curve: this._curve,
				parts: parts
			***REMOVED***);
			this._isPriv = true;
			return (this._key);
		***REMOVED***

	***REMOVED*** else if (this._algo === 'curve25519') ***REMOVED***
		var pair = nacl.box.keyPair();
		priv = Buffer.from(pair.secretKey);
		pub = Buffer.from(pair.publicKey);
		priv = Buffer.concat([priv, pub]);
		assert.strictEqual(priv.length, 64);
		assert.strictEqual(pub.length, 32);

		parts.push(***REMOVED***name: 'A', data: pub***REMOVED***);
		parts.push(***REMOVED***name: 'k', data: priv***REMOVED***);
		this._key = new PrivateKey(***REMOVED***
			type: 'curve25519',
			parts: parts
		***REMOVED***);
		this._isPriv = true;
		return (this._key);
	***REMOVED***

	throw (new Error('Invalid algorithm: ' + this._algo));
***REMOVED***;
DiffieHellman.prototype.generateKeys = DiffieHellman.prototype.generateKey;

/* These are helpers for using ecc-jsbn (for node 0.10 compatibility). */

function X9ECParameters(name) ***REMOVED***
	var params = algs.curves[name];
	assert.object(params);

	var p = new jsbn(params.p);
	var a = new jsbn(params.a);
	var b = new jsbn(params.b);
	var n = new jsbn(params.n);
	var h = jsbn.ONE;
	var curve = new ec.ECCurveFp(p, a, b);
	var G = curve.decodePointHex(params.G.toString('hex'));

	this.curve = curve;
	this.g = G;
	this.n = n;
	this.h = h;
***REMOVED***
X9ECParameters.prototype.getCurve = function () ***REMOVED*** return (this.curve); ***REMOVED***;
X9ECParameters.prototype.getG = function () ***REMOVED*** return (this.g); ***REMOVED***;
X9ECParameters.prototype.getN = function () ***REMOVED*** return (this.n); ***REMOVED***;
X9ECParameters.prototype.getH = function () ***REMOVED*** return (this.h); ***REMOVED***;

function ECPublic(params, buffer) ***REMOVED***
	this._params = params;
	if (buffer[0] === 0x00)
		buffer = buffer.slice(1);
	this._pub = params.getCurve().decodePointHex(buffer.toString('hex'));
***REMOVED***

function ECPrivate(params, buffer) ***REMOVED***
	this._params = params;
	this._priv = new jsbn(utils.mpNormalize(buffer));
***REMOVED***
ECPrivate.prototype.deriveSharedSecret = function (pubKey) ***REMOVED***
	assert.ok(pubKey instanceof ECPublic);
	var S = pubKey._pub.multiply(this._priv);
	return (Buffer.from(S.getX().toBigInteger().toByteArray()));
***REMOVED***;

function generateED25519() ***REMOVED***
	if (nacl === undefined)
		nacl = require('tweetnacl');

	var pair = nacl.sign.keyPair();
	var priv = Buffer.from(pair.secretKey);
	var pub = Buffer.from(pair.publicKey);
	assert.strictEqual(priv.length, 64);
	assert.strictEqual(pub.length, 32);

	var parts = [];
	parts.push(***REMOVED***name: 'A', data: pub***REMOVED***);
	parts.push(***REMOVED***name: 'k', data: priv.slice(0, 32)***REMOVED***);
	var key = new PrivateKey(***REMOVED***
		type: 'ed25519',
		parts: parts
	***REMOVED***);
	return (key);
***REMOVED***

/* Generates a new ECDSA private key on a given curve. */
function generateECDSA(curve) ***REMOVED***
	var parts = [];
	var key;

	if (CRYPTO_HAVE_ECDH) ***REMOVED***
		/*
		 * Node crypto doesn't expose key generation directly, but the
		 * ECDH instances can generate keys. It turns out this just
		 * calls into the OpenSSL generic key generator, and we can
		 * read its output happily without doing an actual DH. So we
		 * use that here.
		 */
		var osCurve = ***REMOVED***
			'nistp256': 'prime256v1',
			'nistp384': 'secp384r1',
			'nistp521': 'secp521r1'
		***REMOVED***[curve];

		var dh = crypto.createECDH(osCurve);
		dh.generateKeys();

		parts.push(***REMOVED***name: 'curve',
		    data: Buffer.from(curve)***REMOVED***);
		parts.push(***REMOVED***name: 'Q', data: dh.getPublicKey()***REMOVED***);
		parts.push(***REMOVED***name: 'd', data: dh.getPrivateKey()***REMOVED***);

		key = new PrivateKey(***REMOVED***
			type: 'ecdsa',
			curve: curve,
			parts: parts
		***REMOVED***);
		return (key);
	***REMOVED*** else ***REMOVED***
		if (ecdh === undefined)
			ecdh = require('ecc-jsbn');
		if (ec === undefined)
			ec = require('ecc-jsbn/lib/ec');
		if (jsbn === undefined)
			jsbn = require('jsbn').BigInteger;

		var ecParams = new X9ECParameters(curve);

		/* This algorithm taken from FIPS PUB 186-4 (section B.4.1) */
		var n = ecParams.getN();
		/*
		 * The crypto.randomBytes() function can only give us whole
		 * bytes, so taking a nod from X9.62, we round up.
		 */
		var cByteLen = Math.ceil((n.bitLength() + 64) / 8);
		var c = new jsbn(crypto.randomBytes(cByteLen));

		var n1 = n.subtract(jsbn.ONE);
		var priv = c.mod(n1).add(jsbn.ONE);
		var pub = ecParams.getG().multiply(priv);

		priv = Buffer.from(priv.toByteArray());
		pub = Buffer.from(ecParams.getCurve().
		    encodePointHex(pub), 'hex');

		parts.push(***REMOVED***name: 'curve', data: Buffer.from(curve)***REMOVED***);
		parts.push(***REMOVED***name: 'Q', data: pub***REMOVED***);
		parts.push(***REMOVED***name: 'd', data: priv***REMOVED***);

		key = new PrivateKey(***REMOVED***
			type: 'ecdsa',
			curve: curve,
			parts: parts
		***REMOVED***);
		return (key);
	***REMOVED***
***REMOVED***
