// sshpk/lib/ed-compat.js
// Copyright 2015 Joyent, Inc.

module.exports = ***REMOVED***
	Verifier: Verifier,
	Signer: Signer
***REMOVED***;

var nacl;
var stream = require('stream');
var util = require('util');
var assert = require('assert-plus');
var Signature = require('./signature');

function Verifier(key, hashAlgo) ***REMOVED***
	if (nacl === undefined)
		nacl = require('tweetnacl');

	if (hashAlgo.toLowerCase() !== 'sha512')
		throw (new Error('ED25519 only supports the use of ' +
		    'SHA-512 hashes'));

	this.key = key;
	this.chunks = [];

	stream.Writable.call(this, ***REMOVED******REMOVED***);
***REMOVED***
util.inherits(Verifier, stream.Writable);

Verifier.prototype._write = function (chunk, enc, cb) ***REMOVED***
	this.chunks.push(chunk);
	cb();
***REMOVED***;

Verifier.prototype.update = function (chunk) ***REMOVED***
	if (typeof (chunk) === 'string')
		chunk = new Buffer(chunk, 'binary');
	this.chunks.push(chunk);
***REMOVED***;

Verifier.prototype.verify = function (signature, fmt) ***REMOVED***
	var sig;
	if (Signature.isSignature(signature, [2, 0])) ***REMOVED***
		if (signature.type !== 'ed25519')
			return (false);
		sig = signature.toBuffer('raw');

	***REMOVED*** else if (typeof (signature) === 'string') ***REMOVED***
		sig = new Buffer(signature, 'base64');

	***REMOVED*** else if (Signature.isSignature(signature, [1, 0])) ***REMOVED***
		throw (new Error('signature was created by too old ' +
		    'a version of sshpk and cannot be verified'));
	***REMOVED***

	assert.buffer(sig);
	return (nacl.sign.detached.verify(
	    new Uint8Array(Buffer.concat(this.chunks)),
	    new Uint8Array(sig),
	    new Uint8Array(this.key.part.A.data)));
***REMOVED***;

function Signer(key, hashAlgo) ***REMOVED***
	if (nacl === undefined)
		nacl = require('tweetnacl');

	if (hashAlgo.toLowerCase() !== 'sha512')
		throw (new Error('ED25519 only supports the use of ' +
		    'SHA-512 hashes'));

	this.key = key;
	this.chunks = [];

	stream.Writable.call(this, ***REMOVED******REMOVED***);
***REMOVED***
util.inherits(Signer, stream.Writable);

Signer.prototype._write = function (chunk, enc, cb) ***REMOVED***
	this.chunks.push(chunk);
	cb();
***REMOVED***;

Signer.prototype.update = function (chunk) ***REMOVED***
	if (typeof (chunk) === 'string')
		chunk = new Buffer(chunk, 'binary');
	this.chunks.push(chunk);
***REMOVED***;

Signer.prototype.sign = function () ***REMOVED***
	var sig = nacl.sign.detached(
	    new Uint8Array(Buffer.concat(this.chunks)),
	    new Uint8Array(Buffer.concat([
		this.key.part.k.data, this.key.part.A.data])));
	var sigBuf = new Buffer(sig);
	var sigObj = Signature.parse(sigBuf, 'ed25519', 'raw');
	sigObj.hashAlgorithm = 'sha512';
	return (sigObj);
***REMOVED***;
