// sshpk/lib/fingerprint.js
// Copyright 2015 Joyent, Inc.

module.exports = Fingerprint;

var assert = require('assert-plus');
var algs = require('./algs');
var crypto = require('crypto');
var errs = require('./errors');
var Key = require('./key');
var Certificate = require('./certificate');
var utils = require('./utils');

var FingerprintFormatError = errs.FingerprintFormatError;
var InvalidAlgorithmError = errs.InvalidAlgorithmError;

function Fingerprint(opts) ***REMOVED***
	assert.object(opts, 'options');
	assert.string(opts.type, 'options.type');
	assert.buffer(opts.hash, 'options.hash');
	assert.string(opts.algorithm, 'options.algorithm');

	this.algorithm = opts.algorithm.toLowerCase();
	if (algs.hashAlgs[this.algorithm] !== true)
		throw (new InvalidAlgorithmError(this.algorithm));

	this.hash = opts.hash;
	this.type = opts.type;
***REMOVED***

Fingerprint.prototype.toString = function (format) ***REMOVED***
	if (format === undefined) ***REMOVED***
		if (this.algorithm === 'md5')
			format = 'hex';
		else
			format = 'base64';
	***REMOVED***
	assert.string(format);

	switch (format) ***REMOVED***
	case 'hex':
		return (addColons(this.hash.toString('hex')));
	case 'base64':
		return (sshBase64Format(this.algorithm,
		    this.hash.toString('base64')));
	default:
		throw (new FingerprintFormatError(undefined, format));
	***REMOVED***
***REMOVED***;

Fingerprint.prototype.matches = function (other) ***REMOVED***
	assert.object(other, 'key or certificate');
	if (this.type === 'key') ***REMOVED***
		utils.assertCompatible(other, Key, [1, 0], 'key');
	***REMOVED*** else ***REMOVED***
		utils.assertCompatible(other, Certificate, [1, 0],
		    'certificate');
	***REMOVED***

	var theirHash = other.hash(this.algorithm);
	var theirHash2 = crypto.createHash(this.algorithm).
	    update(theirHash).digest('base64');

	if (this.hash2 === undefined)
		this.hash2 = crypto.createHash(this.algorithm).
		    update(this.hash).digest('base64');

	return (this.hash2 === theirHash2);
***REMOVED***;

Fingerprint.parse = function (fp, options) ***REMOVED***
	assert.string(fp, 'fingerprint');

	var alg, hash, enAlgs;
	if (Array.isArray(options)) ***REMOVED***
		enAlgs = options;
		options = ***REMOVED******REMOVED***;
	***REMOVED***
	assert.optionalObject(options, 'options');
	if (options === undefined)
		options = ***REMOVED******REMOVED***;
	if (options.enAlgs !== undefined)
		enAlgs = options.enAlgs;
	assert.optionalArrayOfString(enAlgs, 'algorithms');

	var parts = fp.split(':');
	if (parts.length == 2) ***REMOVED***
		alg = parts[0].toLowerCase();
		/*JSSTYLED*/
		var base64RE = /^[A-Za-z0-9+\/=]+$/;
		if (!base64RE.test(parts[1]))
			throw (new FingerprintFormatError(fp));
		try ***REMOVED***
			hash = new Buffer(parts[1], 'base64');
		***REMOVED*** catch (e) ***REMOVED***
			throw (new FingerprintFormatError(fp));
		***REMOVED***
	***REMOVED*** else if (parts.length > 2) ***REMOVED***
		alg = 'md5';
		if (parts[0].toLowerCase() === 'md5')
			parts = parts.slice(1);
		parts = parts.join('');
		/*JSSTYLED*/
		var md5RE = /^[a-fA-F0-9]+$/;
		if (!md5RE.test(parts))
			throw (new FingerprintFormatError(fp));
		try ***REMOVED***
			hash = new Buffer(parts, 'hex');
		***REMOVED*** catch (e) ***REMOVED***
			throw (new FingerprintFormatError(fp));
		***REMOVED***
	***REMOVED***

	if (alg === undefined)
		throw (new FingerprintFormatError(fp));

	if (algs.hashAlgs[alg] === undefined)
		throw (new InvalidAlgorithmError(alg));

	if (enAlgs !== undefined) ***REMOVED***
		enAlgs = enAlgs.map(function (a) ***REMOVED*** return a.toLowerCase(); ***REMOVED***);
		if (enAlgs.indexOf(alg) === -1)
			throw (new InvalidAlgorithmError(alg));
	***REMOVED***

	return (new Fingerprint(***REMOVED***
		algorithm: alg,
		hash: hash,
		type: options.type || 'key'
	***REMOVED***));
***REMOVED***;

function addColons(s) ***REMOVED***
	/*JSSTYLED*/
	return (s.replace(/(.***REMOVED***2***REMOVED***)(?=.)/g, '$1:'));
***REMOVED***

function base64Strip(s) ***REMOVED***
	/*JSSTYLED*/
	return (s.replace(/=*$/, ''));
***REMOVED***

function sshBase64Format(alg, h) ***REMOVED***
	return (alg.toUpperCase() + ':' + base64Strip(h));
***REMOVED***

Fingerprint.isFingerprint = function (obj, ver) ***REMOVED***
	return (utils.isCompatible(obj, Fingerprint, ver));
***REMOVED***;

/*
 * API versions for Fingerprint:
 * [1,0] -- initial ver
 * [1,1] -- first tagged ver
 */
Fingerprint.prototype._sshpkApiVersion = [1, 1];

Fingerprint._oldVersionDetect = function (obj) ***REMOVED***
	assert.func(obj.toString);
	assert.func(obj.matches);
	return ([1, 0]);
***REMOVED***;
