// sshpk/lib/identity.js
// Copyright 2017 Joyent, Inc.

module.exports = Identity;

var assert = require('assert-plus');
var algs = require('./algs');
var crypto = require('crypto');
var Fingerprint = require('./fingerprint');
var Signature = require('./signature');
var errs = require('./errors');
var util = require('util');
var utils = require('./utils');
var asn1 = require('asn1');

/*JSSTYLED*/
var DNS_NAME_RE = /^([*]|[a-z0-9][a-z0-9\-]***REMOVED***0,62***REMOVED***)(?:\.([*]|[a-z0-9][a-z0-9\-]***REMOVED***0,62***REMOVED***))*$/i;

var oids = ***REMOVED******REMOVED***;
oids.cn = '2.5.4.3';
oids.o = '2.5.4.10';
oids.ou = '2.5.4.11';
oids.l = '2.5.4.7';
oids.s = '2.5.4.8';
oids.c = '2.5.4.6';
oids.sn = '2.5.4.4';
oids.dc = '0.9.2342.19200300.100.1.25';
oids.uid = '0.9.2342.19200300.100.1.1';
oids.mail = '0.9.2342.19200300.100.1.3';

var unoids = ***REMOVED******REMOVED***;
Object.keys(oids).forEach(function (k) ***REMOVED***
	unoids[oids[k]] = k;
***REMOVED***);

function Identity(opts) ***REMOVED***
	var self = this;
	assert.object(opts, 'options');
	assert.arrayOfObject(opts.components, 'options.components');
	this.components = opts.components;
	this.componentLookup = ***REMOVED******REMOVED***;
	this.components.forEach(function (c) ***REMOVED***
		if (c.name && !c.oid)
			c.oid = oids[c.name];
		if (c.oid && !c.name)
			c.name = unoids[c.oid];
		if (self.componentLookup[c.name] === undefined)
			self.componentLookup[c.name] = [];
		self.componentLookup[c.name].push(c);
	***REMOVED***);
	if (this.componentLookup.cn && this.componentLookup.cn.length > 0) ***REMOVED***
		this.cn = this.componentLookup.cn[0].value;
	***REMOVED***
	assert.optionalString(opts.type, 'options.type');
	if (opts.type === undefined) ***REMOVED***
		if (this.components.length === 1 &&
		    this.componentLookup.cn &&
		    this.componentLookup.cn.length === 1 &&
		    this.componentLookup.cn[0].value.match(DNS_NAME_RE)) ***REMOVED***
			this.type = 'host';
			this.hostname = this.componentLookup.cn[0].value;

		***REMOVED*** else if (this.componentLookup.dc &&
		    this.components.length === this.componentLookup.dc.length) ***REMOVED***
			this.type = 'host';
			this.hostname = this.componentLookup.dc.map(
			    function (c) ***REMOVED***
				return (c.value);
			***REMOVED***).join('.');

		***REMOVED*** else if (this.componentLookup.uid &&
		    this.components.length ===
		    this.componentLookup.uid.length) ***REMOVED***
			this.type = 'user';
			this.uid = this.componentLookup.uid[0].value;

		***REMOVED*** else if (this.componentLookup.cn &&
		    this.componentLookup.cn.length === 1 &&
		    this.componentLookup.cn[0].value.match(DNS_NAME_RE)) ***REMOVED***
			this.type = 'host';
			this.hostname = this.componentLookup.cn[0].value;

		***REMOVED*** else if (this.componentLookup.uid &&
		    this.componentLookup.uid.length === 1) ***REMOVED***
			this.type = 'user';
			this.uid = this.componentLookup.uid[0].value;

		***REMOVED*** else if (this.componentLookup.mail &&
		    this.componentLookup.mail.length === 1) ***REMOVED***
			this.type = 'email';
			this.email = this.componentLookup.mail[0].value;

		***REMOVED*** else if (this.componentLookup.cn &&
		    this.componentLookup.cn.length === 1) ***REMOVED***
			this.type = 'user';
			this.uid = this.componentLookup.cn[0].value;

		***REMOVED*** else ***REMOVED***
			this.type = 'unknown';
		***REMOVED***
	***REMOVED*** else ***REMOVED***
		this.type = opts.type;
		if (this.type === 'host')
			this.hostname = opts.hostname;
		else if (this.type === 'user')
			this.uid = opts.uid;
		else if (this.type === 'email')
			this.email = opts.email;
		else
			throw (new Error('Unknown type ' + this.type));
	***REMOVED***
***REMOVED***

Identity.prototype.toString = function () ***REMOVED***
	return (this.components.map(function (c) ***REMOVED***
		return (c.name.toUpperCase() + '=' + c.value);
	***REMOVED***).join(', '));
***REMOVED***;

/*
 * These are from X.680 -- PrintableString allowed chars are in section 37.4
 * table 8. Spec for IA5Strings is "1,6 + SPACE + DEL" where 1 refers to
 * ISO IR #001 (standard ASCII control characters) and 6 refers to ISO IR #006
 * (the basic ASCII character set).
 */
/* JSSTYLED */
var NOT_PRINTABLE = /[^a-zA-Z0-9 '(),+.\/:=?-]/;
/* JSSTYLED */
var NOT_IA5 = /[^\x00-\x7f]/;

Identity.prototype.toAsn1 = function (der, tag) ***REMOVED***
	der.startSequence(tag);
	this.components.forEach(function (c) ***REMOVED***
		der.startSequence(asn1.Ber.Constructor | asn1.Ber.Set);
		der.startSequence();
		der.writeOID(c.oid);
		/*
		 * If we fit in a PrintableString, use that. Otherwise use an
		 * IA5String or UTF8String.
		 *
		 * If this identity was parsed from a DN, use the ASN.1 types
		 * from the original representation (otherwise this might not
		 * be a full match for the original in some validators).
		 */
		if (c.asn1type === asn1.Ber.Utf8String ||
		    c.value.match(NOT_IA5)) ***REMOVED***
			var v = new Buffer(c.value, 'utf8');
			der.writeBuffer(v, asn1.Ber.Utf8String);

		***REMOVED*** else if (c.asn1type === asn1.Ber.IA5String ||
		    c.value.match(NOT_PRINTABLE)) ***REMOVED***
			der.writeString(c.value, asn1.Ber.IA5String);

		***REMOVED*** else ***REMOVED***
			var type = asn1.Ber.PrintableString;
			if (c.asn1type !== undefined)
				type = c.asn1type;
			der.writeString(c.value, type);
		***REMOVED***
		der.endSequence();
		der.endSequence();
	***REMOVED***);
	der.endSequence();
***REMOVED***;

function globMatch(a, b) ***REMOVED***
	if (a === '**' || b === '**')
		return (true);
	var aParts = a.split('.');
	var bParts = b.split('.');
	if (aParts.length !== bParts.length)
		return (false);
	for (var i = 0; i < aParts.length; ++i) ***REMOVED***
		if (aParts[i] === '*' || bParts[i] === '*')
			continue;
		if (aParts[i] !== bParts[i])
			return (false);
	***REMOVED***
	return (true);
***REMOVED***

Identity.prototype.equals = function (other) ***REMOVED***
	if (!Identity.isIdentity(other, [1, 0]))
		return (false);
	if (other.components.length !== this.components.length)
		return (false);
	for (var i = 0; i < this.components.length; ++i) ***REMOVED***
		if (this.components[i].oid !== other.components[i].oid)
			return (false);
		if (!globMatch(this.components[i].value,
		    other.components[i].value)) ***REMOVED***
			return (false);
		***REMOVED***
	***REMOVED***
	return (true);
***REMOVED***;

Identity.forHost = function (hostname) ***REMOVED***
	assert.string(hostname, 'hostname');
	return (new Identity(***REMOVED***
		type: 'host',
		hostname: hostname,
		components: [ ***REMOVED*** name: 'cn', value: hostname ***REMOVED*** ]
	***REMOVED***));
***REMOVED***;

Identity.forUser = function (uid) ***REMOVED***
	assert.string(uid, 'uid');
	return (new Identity(***REMOVED***
		type: 'user',
		uid: uid,
		components: [ ***REMOVED*** name: 'uid', value: uid ***REMOVED*** ]
	***REMOVED***));
***REMOVED***;

Identity.forEmail = function (email) ***REMOVED***
	assert.string(email, 'email');
	return (new Identity(***REMOVED***
		type: 'email',
		email: email,
		components: [ ***REMOVED*** name: 'mail', value: email ***REMOVED*** ]
	***REMOVED***));
***REMOVED***;

Identity.parseDN = function (dn) ***REMOVED***
	assert.string(dn, 'dn');
	var parts = dn.split(',');
	var cmps = parts.map(function (c) ***REMOVED***
		c = c.trim();
		var eqPos = c.indexOf('=');
		var name = c.slice(0, eqPos).toLowerCase();
		var value = c.slice(eqPos + 1);
		return (***REMOVED*** name: name, value: value ***REMOVED***);
	***REMOVED***);
	return (new Identity(***REMOVED*** components: cmps ***REMOVED***));
***REMOVED***;

Identity.parseAsn1 = function (der, top) ***REMOVED***
	var components = [];
	der.readSequence(top);
	var end = der.offset + der.length;
	while (der.offset < end) ***REMOVED***
		der.readSequence(asn1.Ber.Constructor | asn1.Ber.Set);
		var after = der.offset + der.length;
		der.readSequence();
		var oid = der.readOID();
		var type = der.peek();
		var value;
		switch (type) ***REMOVED***
		case asn1.Ber.PrintableString:
		case asn1.Ber.IA5String:
		case asn1.Ber.OctetString:
		case asn1.Ber.T61String:
			value = der.readString(type);
			break;
		case asn1.Ber.Utf8String:
			value = der.readString(type, true);
			value = value.toString('utf8');
			break;
		case asn1.Ber.CharacterString:
		case asn1.Ber.BMPString:
			value = der.readString(type, true);
			value = value.toString('utf16le');
			break;
		default:
			throw (new Error('Unknown asn1 type ' + type));
		***REMOVED***
		components.push(***REMOVED*** oid: oid, asn1type: type, value: value ***REMOVED***);
		der._offset = after;
	***REMOVED***
	der._offset = end;
	return (new Identity(***REMOVED***
		components: components
	***REMOVED***));
***REMOVED***;

Identity.isIdentity = function (obj, ver) ***REMOVED***
	return (utils.isCompatible(obj, Identity, ver));
***REMOVED***;

/*
 * API versions for Identity:
 * [1,0] -- initial ver
 */
Identity.prototype._sshpkApiVersion = [1, 0];

Identity._oldVersionDetect = function (obj) ***REMOVED***
	return ([1, 0]);
***REMOVED***;
