// sshpk/lib/ssh-buffer.js
// Copyright 2015 Joyent, Inc.

module.exports = SSHBuffer;

var assert = require('assert-plus');

function SSHBuffer(opts) ***REMOVED***
	assert.object(opts, 'options');
	if (opts.buffer !== undefined)
		assert.buffer(opts.buffer, 'options.buffer');

	this._size = opts.buffer ? opts.buffer.length : 1024;
	this._buffer = opts.buffer || (new Buffer(this._size));
	this._offset = 0;
***REMOVED***

SSHBuffer.prototype.toBuffer = function () ***REMOVED***
	return (this._buffer.slice(0, this._offset));
***REMOVED***;

SSHBuffer.prototype.atEnd = function () ***REMOVED***
	return (this._offset >= this._buffer.length);
***REMOVED***;

SSHBuffer.prototype.remainder = function () ***REMOVED***
	return (this._buffer.slice(this._offset));
***REMOVED***;

SSHBuffer.prototype.skip = function (n) ***REMOVED***
	this._offset += n;
***REMOVED***;

SSHBuffer.prototype.expand = function () ***REMOVED***
	this._size *= 2;
	var buf = new Buffer(this._size);
	this._buffer.copy(buf, 0);
	this._buffer = buf;
***REMOVED***;

SSHBuffer.prototype.readPart = function () ***REMOVED***
	return (***REMOVED***data: this.readBuffer()***REMOVED***);
***REMOVED***;

SSHBuffer.prototype.readBuffer = function () ***REMOVED***
	var len = this._buffer.readUInt32BE(this._offset);
	this._offset += 4;
	assert.ok(this._offset + len <= this._buffer.length,
	    'length out of bounds at +0x' + this._offset.toString(16) +
	    ' (data truncated?)');
	var buf = this._buffer.slice(this._offset, this._offset + len);
	this._offset += len;
	return (buf);
***REMOVED***;

SSHBuffer.prototype.readString = function () ***REMOVED***
	return (this.readBuffer().toString());
***REMOVED***;

SSHBuffer.prototype.readCString = function () ***REMOVED***
	var offset = this._offset;
	while (offset < this._buffer.length &&
	    this._buffer[offset] !== 0x00)
		offset++;
	assert.ok(offset < this._buffer.length, 'c string does not terminate');
	var str = this._buffer.slice(this._offset, offset).toString();
	this._offset = offset + 1;
	return (str);
***REMOVED***;

SSHBuffer.prototype.readInt = function () ***REMOVED***
	var v = this._buffer.readUInt32BE(this._offset);
	this._offset += 4;
	return (v);
***REMOVED***;

SSHBuffer.prototype.readInt64 = function () ***REMOVED***
	assert.ok(this._offset + 8 < this._buffer.length,
	    'buffer not long enough to read Int64');
	var v = this._buffer.slice(this._offset, this._offset + 8);
	this._offset += 8;
	return (v);
***REMOVED***;

SSHBuffer.prototype.readChar = function () ***REMOVED***
	var v = this._buffer[this._offset++];
	return (v);
***REMOVED***;

SSHBuffer.prototype.writeBuffer = function (buf) ***REMOVED***
	while (this._offset + 4 + buf.length > this._size)
		this.expand();
	this._buffer.writeUInt32BE(buf.length, this._offset);
	this._offset += 4;
	buf.copy(this._buffer, this._offset);
	this._offset += buf.length;
***REMOVED***;

SSHBuffer.prototype.writeString = function (str) ***REMOVED***
	this.writeBuffer(new Buffer(str, 'utf8'));
***REMOVED***;

SSHBuffer.prototype.writeCString = function (str) ***REMOVED***
	while (this._offset + 1 + str.length > this._size)
		this.expand();
	this._buffer.write(str, this._offset);
	this._offset += str.length;
	this._buffer[this._offset++] = 0;
***REMOVED***;

SSHBuffer.prototype.writeInt = function (v) ***REMOVED***
	while (this._offset + 4 > this._size)
		this.expand();
	this._buffer.writeUInt32BE(v, this._offset);
	this._offset += 4;
***REMOVED***;

SSHBuffer.prototype.writeInt64 = function (v) ***REMOVED***
	assert.buffer(v, 'value');
	if (v.length > 8) ***REMOVED***
		var lead = v.slice(0, v.length - 8);
		for (var i = 0; i < lead.length; ++i) ***REMOVED***
			assert.strictEqual(lead[i], 0,
			    'must fit in 64 bits of precision');
		***REMOVED***
		v = v.slice(v.length - 8, v.length);
	***REMOVED***
	while (this._offset + 8 > this._size)
		this.expand();
	v.copy(this._buffer, this._offset);
	this._offset += 8;
***REMOVED***;

SSHBuffer.prototype.writeChar = function (v) ***REMOVED***
	while (this._offset + 1 > this._size)
		this.expand();
	this._buffer[this._offset++] = v;
***REMOVED***;

SSHBuffer.prototype.writePart = function (p) ***REMOVED***
	this.writeBuffer(p.data);
***REMOVED***;

SSHBuffer.prototype.write = function (buf) ***REMOVED***
	while (this._offset + buf.length > this._size)
		this.expand();
	buf.copy(this._buffer, this._offset);
	this._offset += buf.length;
***REMOVED***;
