// sshpk/lib/utils.js
// Copyright 2015 Joyent, Inc.

module.exports = ***REMOVED***
	bufferSplit: bufferSplit,
	addRSAMissing: addRSAMissing,
	calculateDSAPublic: calculateDSAPublic,
	calculateED25519Public: calculateED25519Public,
	calculateX25519Public: calculateX25519Public,
	mpNormalize: mpNormalize,
	mpDenormalize: mpDenormalize,
	ecNormalize: ecNormalize,
	countZeros: countZeros,
	assertCompatible: assertCompatible,
	isCompatible: isCompatible,
	opensslKeyDeriv: opensslKeyDeriv,
	opensshCipherInfo: opensshCipherInfo,
	publicFromPrivateECDSA: publicFromPrivateECDSA,
	zeroPadToLength: zeroPadToLength,
	writeBitString: writeBitString,
	readBitString: readBitString
***REMOVED***;

var assert = require('assert-plus');
var PrivateKey = require('./private-key');
var Key = require('./key');
var crypto = require('crypto');
var algs = require('./algs');
var asn1 = require('asn1');

var ec, jsbn;
var nacl;

var MAX_CLASS_DEPTH = 3;

function isCompatible(obj, klass, needVer) ***REMOVED***
	if (obj === null || typeof (obj) !== 'object')
		return (false);
	if (needVer === undefined)
		needVer = klass.prototype._sshpkApiVersion;
	if (obj instanceof klass &&
	    klass.prototype._sshpkApiVersion[0] == needVer[0])
		return (true);
	var proto = Object.getPrototypeOf(obj);
	var depth = 0;
	while (proto.constructor.name !== klass.name) ***REMOVED***
		proto = Object.getPrototypeOf(proto);
		if (!proto || ++depth > MAX_CLASS_DEPTH)
			return (false);
	***REMOVED***
	if (proto.constructor.name !== klass.name)
		return (false);
	var ver = proto._sshpkApiVersion;
	if (ver === undefined)
		ver = klass._oldVersionDetect(obj);
	if (ver[0] != needVer[0] || ver[1] < needVer[1])
		return (false);
	return (true);
***REMOVED***

function assertCompatible(obj, klass, needVer, name) ***REMOVED***
	if (name === undefined)
		name = 'object';
	assert.ok(obj, name + ' must not be null');
	assert.object(obj, name + ' must be an object');
	if (needVer === undefined)
		needVer = klass.prototype._sshpkApiVersion;
	if (obj instanceof klass &&
	    klass.prototype._sshpkApiVersion[0] == needVer[0])
		return;
	var proto = Object.getPrototypeOf(obj);
	var depth = 0;
	while (proto.constructor.name !== klass.name) ***REMOVED***
		proto = Object.getPrototypeOf(proto);
		assert.ok(proto && ++depth <= MAX_CLASS_DEPTH,
		    name + ' must be a ' + klass.name + ' instance');
	***REMOVED***
	assert.strictEqual(proto.constructor.name, klass.name,
	    name + ' must be a ' + klass.name + ' instance');
	var ver = proto._sshpkApiVersion;
	if (ver === undefined)
		ver = klass._oldVersionDetect(obj);
	assert.ok(ver[0] == needVer[0] && ver[1] >= needVer[1],
	    name + ' must be compatible with ' + klass.name + ' klass ' +
	    'version ' + needVer[0] + '.' + needVer[1]);
***REMOVED***

var CIPHER_LEN = ***REMOVED***
	'des-ede3-cbc': ***REMOVED*** key: 7, iv: 8 ***REMOVED***,
	'aes-128-cbc': ***REMOVED*** key: 16, iv: 16 ***REMOVED***
***REMOVED***;
var PKCS5_SALT_LEN = 8;

function opensslKeyDeriv(cipher, salt, passphrase, count) ***REMOVED***
	assert.buffer(salt, 'salt');
	assert.buffer(passphrase, 'passphrase');
	assert.number(count, 'iteration count');

	var clen = CIPHER_LEN[cipher];
	assert.object(clen, 'supported cipher');

	salt = salt.slice(0, PKCS5_SALT_LEN);

	var D, D_prev, bufs;
	var material = new Buffer(0);
	while (material.length < clen.key + clen.iv) ***REMOVED***
		bufs = [];
		if (D_prev)
			bufs.push(D_prev);
		bufs.push(passphrase);
		bufs.push(salt);
		D = Buffer.concat(bufs);
		for (var j = 0; j < count; ++j)
			D = crypto.createHash('md5').update(D).digest();
		material = Buffer.concat([material, D]);
		D_prev = D;
	***REMOVED***

	return (***REMOVED***
	    key: material.slice(0, clen.key),
	    iv: material.slice(clen.key, clen.key + clen.iv)
	***REMOVED***);
***REMOVED***

/* Count leading zero bits on a buffer */
function countZeros(buf) ***REMOVED***
	var o = 0, obit = 8;
	while (o < buf.length) ***REMOVED***
		var mask = (1 << obit);
		if ((buf[o] & mask) === mask)
			break;
		obit--;
		if (obit < 0) ***REMOVED***
			o++;
			obit = 8;
		***REMOVED***
	***REMOVED***
	return (o*8 + (8 - obit) - 1);
***REMOVED***

function bufferSplit(buf, chr) ***REMOVED***
	assert.buffer(buf);
	assert.string(chr);

	var parts = [];
	var lastPart = 0;
	var matches = 0;
	for (var i = 0; i < buf.length; ++i) ***REMOVED***
		if (buf[i] === chr.charCodeAt(matches))
			++matches;
		else if (buf[i] === chr.charCodeAt(0))
			matches = 1;
		else
			matches = 0;

		if (matches >= chr.length) ***REMOVED***
			var newPart = i + 1;
			parts.push(buf.slice(lastPart, newPart - matches));
			lastPart = newPart;
			matches = 0;
		***REMOVED***
	***REMOVED***
	if (lastPart <= buf.length)
		parts.push(buf.slice(lastPart, buf.length));

	return (parts);
***REMOVED***

function ecNormalize(buf, addZero) ***REMOVED***
	assert.buffer(buf);
	if (buf[0] === 0x00 && buf[1] === 0x04) ***REMOVED***
		if (addZero)
			return (buf);
		return (buf.slice(1));
	***REMOVED*** else if (buf[0] === 0x04) ***REMOVED***
		if (!addZero)
			return (buf);
	***REMOVED*** else ***REMOVED***
		while (buf[0] === 0x00)
			buf = buf.slice(1);
		if (buf[0] === 0x02 || buf[0] === 0x03)
			throw (new Error('Compressed elliptic curve points ' +
			    'are not supported'));
		if (buf[0] !== 0x04)
			throw (new Error('Not a valid elliptic curve point'));
		if (!addZero)
			return (buf);
	***REMOVED***
	var b = new Buffer(buf.length + 1);
	b[0] = 0x0;
	buf.copy(b, 1);
	return (b);
***REMOVED***

function readBitString(der, tag) ***REMOVED***
	if (tag === undefined)
		tag = asn1.Ber.BitString;
	var buf = der.readString(tag, true);
	assert.strictEqual(buf[0], 0x00, 'bit strings with unused bits are ' +
	    'not supported (0x' + buf[0].toString(16) + ')');
	return (buf.slice(1));
***REMOVED***

function writeBitString(der, buf, tag) ***REMOVED***
	if (tag === undefined)
		tag = asn1.Ber.BitString;
	var b = new Buffer(buf.length + 1);
	b[0] = 0x00;
	buf.copy(b, 1);
	der.writeBuffer(b, tag);
***REMOVED***

function mpNormalize(buf) ***REMOVED***
	assert.buffer(buf);
	while (buf.length > 1 && buf[0] === 0x00 && (buf[1] & 0x80) === 0x00)
		buf = buf.slice(1);
	if ((buf[0] & 0x80) === 0x80) ***REMOVED***
		var b = new Buffer(buf.length + 1);
		b[0] = 0x00;
		buf.copy(b, 1);
		buf = b;
	***REMOVED***
	return (buf);
***REMOVED***

function mpDenormalize(buf) ***REMOVED***
	assert.buffer(buf);
	while (buf.length > 1 && buf[0] === 0x00)
		buf = buf.slice(1);
	return (buf);
***REMOVED***

function zeroPadToLength(buf, len) ***REMOVED***
	assert.buffer(buf);
	assert.number(len);
	while (buf.length > len) ***REMOVED***
		assert.equal(buf[0], 0x00);
		buf = buf.slice(1);
	***REMOVED***
	while (buf.length < len) ***REMOVED***
		var b = new Buffer(buf.length + 1);
		b[0] = 0x00;
		buf.copy(b, 1);
		buf = b;
	***REMOVED***
	return (buf);
***REMOVED***

function bigintToMpBuf(bigint) ***REMOVED***
	var buf = new Buffer(bigint.toByteArray());
	buf = mpNormalize(buf);
	return (buf);
***REMOVED***

function calculateDSAPublic(g, p, x) ***REMOVED***
	assert.buffer(g);
	assert.buffer(p);
	assert.buffer(x);
	try ***REMOVED***
		var bigInt = require('jsbn').BigInteger;
	***REMOVED*** catch (e) ***REMOVED***
		throw (new Error('To load a PKCS#8 format DSA private key, ' +
		    'the node jsbn library is required.'));
	***REMOVED***
	g = new bigInt(g);
	p = new bigInt(p);
	x = new bigInt(x);
	var y = g.modPow(x, p);
	var ybuf = bigintToMpBuf(y);
	return (ybuf);
***REMOVED***

function calculateED25519Public(k) ***REMOVED***
	assert.buffer(k);

	if (nacl === undefined)
		nacl = require('tweetnacl');

	var kp = nacl.sign.keyPair.fromSeed(new Uint8Array(k));
	return (new Buffer(kp.publicKey));
***REMOVED***

function calculateX25519Public(k) ***REMOVED***
	assert.buffer(k);

	if (nacl === undefined)
		nacl = require('tweetnacl');

	var kp = nacl.box.keyPair.fromSeed(new Uint8Array(k));
	return (new Buffer(kp.publicKey));
***REMOVED***

function addRSAMissing(key) ***REMOVED***
	assert.object(key);
	assertCompatible(key, PrivateKey, [1, 1]);
	try ***REMOVED***
		var bigInt = require('jsbn').BigInteger;
	***REMOVED*** catch (e) ***REMOVED***
		throw (new Error('To write a PEM private key from ' +
		    'this source, the node jsbn lib is required.'));
	***REMOVED***

	var d = new bigInt(key.part.d.data);
	var buf;

	if (!key.part.dmodp) ***REMOVED***
		var p = new bigInt(key.part.p.data);
		var dmodp = d.mod(p.subtract(1));

		buf = bigintToMpBuf(dmodp);
		key.part.dmodp = ***REMOVED***name: 'dmodp', data: buf***REMOVED***;
		key.parts.push(key.part.dmodp);
	***REMOVED***
	if (!key.part.dmodq) ***REMOVED***
		var q = new bigInt(key.part.q.data);
		var dmodq = d.mod(q.subtract(1));

		buf = bigintToMpBuf(dmodq);
		key.part.dmodq = ***REMOVED***name: 'dmodq', data: buf***REMOVED***;
		key.parts.push(key.part.dmodq);
	***REMOVED***
***REMOVED***

function publicFromPrivateECDSA(curveName, priv) ***REMOVED***
	assert.string(curveName, 'curveName');
	assert.buffer(priv);
	if (ec === undefined)
		ec = require('ecc-jsbn/lib/ec');
	if (jsbn === undefined)
		jsbn = require('jsbn').BigInteger;
	var params = algs.curves[curveName];
	var p = new jsbn(params.p);
	var a = new jsbn(params.a);
	var b = new jsbn(params.b);
	var curve = new ec.ECCurveFp(p, a, b);
	var G = curve.decodePointHex(params.G.toString('hex'));

	var d = new jsbn(mpNormalize(priv));
	var pub = G.multiply(d);
	pub = new Buffer(curve.encodePointHex(pub), 'hex');

	var parts = [];
	parts.push(***REMOVED***name: 'curve', data: new Buffer(curveName)***REMOVED***);
	parts.push(***REMOVED***name: 'Q', data: pub***REMOVED***);

	var key = new Key(***REMOVED***type: 'ecdsa', curve: curve, parts: parts***REMOVED***);
	return (key);
***REMOVED***

function opensshCipherInfo(cipher) ***REMOVED***
	var inf = ***REMOVED******REMOVED***;
	switch (cipher) ***REMOVED***
	case '3des-cbc':
		inf.keySize = 24;
		inf.blockSize = 8;
		inf.opensslName = 'des-ede3-cbc';
		break;
	case 'blowfish-cbc':
		inf.keySize = 16;
		inf.blockSize = 8;
		inf.opensslName = 'bf-cbc';
		break;
	case 'aes128-cbc':
	case 'aes128-ctr':
	case 'aes128-gcm@openssh.com':
		inf.keySize = 16;
		inf.blockSize = 16;
		inf.opensslName = 'aes-128-' + cipher.slice(7, 10);
		break;
	case 'aes192-cbc':
	case 'aes192-ctr':
	case 'aes192-gcm@openssh.com':
		inf.keySize = 24;
		inf.blockSize = 16;
		inf.opensslName = 'aes-192-' + cipher.slice(7, 10);
		break;
	case 'aes256-cbc':
	case 'aes256-ctr':
	case 'aes256-gcm@openssh.com':
		inf.keySize = 32;
		inf.blockSize = 16;
		inf.opensslName = 'aes-256-' + cipher.slice(7, 10);
		break;
	default:
		throw (new Error(
		    'Unsupported openssl cipher "' + cipher + '"'));
	***REMOVED***
	return (inf);
***REMOVED***
