// har-validator/lib/async.js
var Ajv = require('ajv')
var HARError = require('./error')
var schemas = require('har-schema')

var ajv

function validate (name, data, next) ***REMOVED***
  data = data || ***REMOVED******REMOVED***

  // validator config
  ajv = ajv || new Ajv(***REMOVED***
    allErrors: true,
    schemas: schemas
  ***REMOVED***)

  var validate = ajv.getSchema(name + '.json')

  var valid = validate(data)

  // callback?
  if (typeof next === 'function') ***REMOVED***
    return next(!valid ? new HARError(validate.errors) : null, valid)
  ***REMOVED***

  return valid
***REMOVED***

exports.afterRequest = function (data, next) ***REMOVED***
  return validate('afterRequest', data, next)
***REMOVED***

exports.beforeRequest = function (data, next) ***REMOVED***
  return validate('beforeRequest', data, next)
***REMOVED***

exports.browser = function (data, next) ***REMOVED***
  return validate('browser', data, next)
***REMOVED***

exports.cache = function (data, next) ***REMOVED***
  return validate('cache', data, next)
***REMOVED***

exports.content = function (data, next) ***REMOVED***
  return validate('content', data, next)
***REMOVED***

exports.cookie = function (data, next) ***REMOVED***
  return validate('cookie', data, next)
***REMOVED***

exports.creator = function (data, next) ***REMOVED***
  return validate('creator', data, next)
***REMOVED***

exports.entry = function (data, next) ***REMOVED***
  return validate('entry', data, next)
***REMOVED***

exports.har = function (data, next) ***REMOVED***
  return validate('har', data, next)
***REMOVED***

exports.header = function (data, next) ***REMOVED***
  return validate('header', data, next)
***REMOVED***

exports.log = function (data, next) ***REMOVED***
  return validate('log', data, next)
***REMOVED***

exports.page = function (data, next) ***REMOVED***
  return validate('page', data, next)
***REMOVED***

exports.pageTimings = function (data, next) ***REMOVED***
  return validate('pageTimings', data, next)
***REMOVED***

exports.postData = function (data, next) ***REMOVED***
  return validate('postData', data, next)
***REMOVED***

exports.query = function (data, next) ***REMOVED***
  return validate('query', data, next)
***REMOVED***

exports.request = function (data, next) ***REMOVED***
  return validate('request', data, next)
***REMOVED***

exports.response = function (data, next) ***REMOVED***
  return validate('response', data, next)
***REMOVED***

exports.timings = function (data, next) ***REMOVED***
  return validate('timings', data, next)
***REMOVED***
