// har-validator/lib/promise.js
var Ajv = require('ajv')
var HARError = require('./error')
var schemas = require('har-schema')

var ajv

function validate (name, data) ***REMOVED***
  data = data || ***REMOVED******REMOVED***

  // validator config
  ajv = ajv || new Ajv(***REMOVED***
    allErrors: true,
    schemas: schemas
  ***REMOVED***)

  var validate = ajv.getSchema(name + '.json')

  return new Promise(function (resolve, reject) ***REMOVED***
    var valid = validate(data)

    !valid ? reject(new HARError(validate.errors)) : resolve(data)
  ***REMOVED***)
***REMOVED***

exports.afterRequest = function (data) ***REMOVED***
  return validate('afterRequest', data)
***REMOVED***

exports.beforeRequest = function (data) ***REMOVED***
  return validate('beforeRequest', data)
***REMOVED***

exports.browser = function (data) ***REMOVED***
  return validate('browser', data)
***REMOVED***

exports.cache = function (data) ***REMOVED***
  return validate('cache', data)
***REMOVED***

exports.content = function (data) ***REMOVED***
  return validate('content', data)
***REMOVED***

exports.cookie = function (data) ***REMOVED***
  return validate('cookie', data)
***REMOVED***

exports.creator = function (data) ***REMOVED***
  return validate('creator', data)
***REMOVED***

exports.entry = function (data) ***REMOVED***
  return validate('entry', data)
***REMOVED***

exports.har = function (data) ***REMOVED***
  return validate('har', data)
***REMOVED***

exports.header = function (data) ***REMOVED***
  return validate('header', data)
***REMOVED***

exports.log = function (data) ***REMOVED***
  return validate('log', data)
***REMOVED***

exports.page = function (data) ***REMOVED***
  return validate('page', data)
***REMOVED***

exports.pageTimings = function (data) ***REMOVED***
  return validate('pageTimings', data)
***REMOVED***

exports.postData = function (data) ***REMOVED***
  return validate('postData', data)
***REMOVED***

exports.query = function (data) ***REMOVED***
  return validate('query', data)
***REMOVED***

exports.request = function (data) ***REMOVED***
  return validate('request', data)
***REMOVED***

exports.response = function (data) ***REMOVED***
  return validate('response', data)
***REMOVED***

exports.timings = function (data) ***REMOVED***
  return validate('timings', data)
***REMOVED***
