// bluebird/js/release/props.js
"use strict";
module.exports = function(
    Promise, PromiseArray, tryConvertToPromise, apiRejection) ***REMOVED***
var util = require("./util");
var isObject = util.isObject;
var es5 = require("./es5");
var Es6Map;
if (typeof Map === "function") Es6Map = Map;

var mapToEntries = (function() ***REMOVED***
    var index = 0;
    var size = 0;

    function extractEntry(value, key) ***REMOVED***
        this[index] = value;
        this[index + size] = key;
        index++;
    ***REMOVED***

    return function mapToEntries(map) ***REMOVED***
        size = map.size;
        index = 0;
        var ret = new Array(map.size * 2);
        map.forEach(extractEntry, ret);
        return ret;
    ***REMOVED***;
***REMOVED***)();

var entriesToMap = function(entries) ***REMOVED***
    var ret = new Es6Map();
    var length = entries.length / 2 | 0;
    for (var i = 0; i < length; ++i) ***REMOVED***
        var key = entries[length + i];
        var value = entries[i];
        ret.set(key, value);
    ***REMOVED***
    return ret;
***REMOVED***;

function PropertiesPromiseArray(obj) ***REMOVED***
    var isMap = false;
    var entries;
    if (Es6Map !== undefined && obj instanceof Es6Map) ***REMOVED***
        entries = mapToEntries(obj);
        isMap = true;
    ***REMOVED*** else ***REMOVED***
        var keys = es5.keys(obj);
        var len = keys.length;
        entries = new Array(len * 2);
        for (var i = 0; i < len; ++i) ***REMOVED***
            var key = keys[i];
            entries[i] = obj[key];
            entries[i + len] = key;
        ***REMOVED***
    ***REMOVED***
    this.constructor$(entries);
    this._isMap = isMap;
    this._init$(undefined, isMap ? -6 : -3);
***REMOVED***
util.inherits(PropertiesPromiseArray, PromiseArray);

PropertiesPromiseArray.prototype._init = function () ***REMOVED******REMOVED***;

PropertiesPromiseArray.prototype._promiseFulfilled = function (value, index) ***REMOVED***
    this._values[index] = value;
    var totalResolved = ++this._totalResolved;
    if (totalResolved >= this._length) ***REMOVED***
        var val;
        if (this._isMap) ***REMOVED***
            val = entriesToMap(this._values);
        ***REMOVED*** else ***REMOVED***
            val = ***REMOVED******REMOVED***;
            var keyOffset = this.length();
            for (var i = 0, len = this.length(); i < len; ++i) ***REMOVED***
                val[this._values[i + keyOffset]] = this._values[i];
            ***REMOVED***
        ***REMOVED***
        this._resolve(val);
        return true;
    ***REMOVED***
    return false;
***REMOVED***;

PropertiesPromiseArray.prototype.shouldCopyValues = function () ***REMOVED***
    return false;
***REMOVED***;

PropertiesPromiseArray.prototype.getActualLength = function (len) ***REMOVED***
    return len >> 1;
***REMOVED***;

function props(promises) ***REMOVED***
    var ret;
    var castValue = tryConvertToPromise(promises);

    if (!isObject(castValue)) ***REMOVED***
        return apiRejection("cannot await properties of a non-object\u000a\u000a    See http://goo.gl/MqrFmX\u000a");
    ***REMOVED*** else if (castValue instanceof Promise) ***REMOVED***
        ret = castValue._then(
            Promise.props, undefined, undefined, undefined, undefined);
    ***REMOVED*** else ***REMOVED***
        ret = new PropertiesPromiseArray(castValue).promise();
    ***REMOVED***

    if (castValue instanceof Promise) ***REMOVED***
        ret._propagateFrom(castValue, 2);
    ***REMOVED***
    return ret;
***REMOVED***

Promise.prototype.props = function () ***REMOVED***
    return props(this);
***REMOVED***;

Promise.props = function (promises) ***REMOVED***
    return props(promises);
***REMOVED***;
***REMOVED***;
