// bluebird/js/release/synchronous_inspection.js
"use strict";
module.exports = function(Promise) ***REMOVED***
function PromiseInspection(promise) ***REMOVED***
    if (promise !== undefined) ***REMOVED***
        promise = promise._target();
        this._bitField = promise._bitField;
        this._settledValueField = promise._isFateSealed()
            ? promise._settledValue() : undefined;
    ***REMOVED***
    else ***REMOVED***
        this._bitField = 0;
        this._settledValueField = undefined;
    ***REMOVED***
***REMOVED***

PromiseInspection.prototype._settledValue = function() ***REMOVED***
    return this._settledValueField;
***REMOVED***;

var value = PromiseInspection.prototype.value = function () ***REMOVED***
    if (!this.isFulfilled()) ***REMOVED***
        throw new TypeError("cannot get fulfillment value of a non-fulfilled promise\u000a\u000a    See http://goo.gl/MqrFmX\u000a");
    ***REMOVED***
    return this._settledValue();
***REMOVED***;

var reason = PromiseInspection.prototype.error =
PromiseInspection.prototype.reason = function () ***REMOVED***
    if (!this.isRejected()) ***REMOVED***
        throw new TypeError("cannot get rejection reason of a non-rejected promise\u000a\u000a    See http://goo.gl/MqrFmX\u000a");
    ***REMOVED***
    return this._settledValue();
***REMOVED***;

var isFulfilled = PromiseInspection.prototype.isFulfilled = function() ***REMOVED***
    return (this._bitField & 33554432) !== 0;
***REMOVED***;

var isRejected = PromiseInspection.prototype.isRejected = function () ***REMOVED***
    return (this._bitField & 16777216) !== 0;
***REMOVED***;

var isPending = PromiseInspection.prototype.isPending = function () ***REMOVED***
    return (this._bitField & 50397184) === 0;
***REMOVED***;

var isResolved = PromiseInspection.prototype.isResolved = function () ***REMOVED***
    return (this._bitField & 50331648) !== 0;
***REMOVED***;

PromiseInspection.prototype.isCancelled = function() ***REMOVED***
    return (this._bitField & 8454144) !== 0;
***REMOVED***;

Promise.prototype.__isCancelled = function() ***REMOVED***
    return (this._bitField & 65536) === 65536;
***REMOVED***;

Promise.prototype._isCancelled = function() ***REMOVED***
    return this._target().__isCancelled();
***REMOVED***;

Promise.prototype.isCancelled = function() ***REMOVED***
    return (this._target()._bitField & 8454144) !== 0;
***REMOVED***;

Promise.prototype.isPending = function() ***REMOVED***
    return isPending.call(this._target());
***REMOVED***;

Promise.prototype.isRejected = function() ***REMOVED***
    return isRejected.call(this._target());
***REMOVED***;

Promise.prototype.isFulfilled = function() ***REMOVED***
    return isFulfilled.call(this._target());
***REMOVED***;

Promise.prototype.isResolved = function() ***REMOVED***
    return isResolved.call(this._target());
***REMOVED***;

Promise.prototype.value = function() ***REMOVED***
    return value.call(this._target());
***REMOVED***;

Promise.prototype.reason = function() ***REMOVED***
    var target = this._target();
    target._unsetRejectionIsUnhandled();
    return reason.call(target);
***REMOVED***;

Promise.prototype._value = function() ***REMOVED***
    return this._settledValue();
***REMOVED***;

Promise.prototype._reason = function() ***REMOVED***
    this._unsetRejectionIsUnhandled();
    return this._settledValue();
***REMOVED***;

Promise.PromiseInspection = PromiseInspection;
***REMOVED***;
