// caseless/index.js
function Caseless (dict) ***REMOVED***
  this.dict = dict || ***REMOVED******REMOVED***
***REMOVED***
Caseless.prototype.set = function (name, value, clobber) ***REMOVED***
  if (typeof name === 'object') ***REMOVED***
    for (var i in name) ***REMOVED***
      this.set(i, name[i], value)
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    if (typeof clobber === 'undefined') clobber = true
    var has = this.has(name)

    if (!clobber && has) this.dict[has] = this.dict[has] + ',' + value
    else this.dict[has || name] = value
    return has
  ***REMOVED***
***REMOVED***
Caseless.prototype.has = function (name) ***REMOVED***
  var keys = Object.keys(this.dict)
    , name = name.toLowerCase()
    ;
  for (var i=0;i<keys.length;i++) ***REMOVED***
    if (keys[i].toLowerCase() === name) return keys[i]
  ***REMOVED***
  return false
***REMOVED***
Caseless.prototype.get = function (name) ***REMOVED***
  name = name.toLowerCase()
  var result, _key
  var headers = this.dict
  Object.keys(headers).forEach(function (key) ***REMOVED***
    _key = key.toLowerCase()
    if (name === _key) result = headers[key]
  ***REMOVED***)
  return result
***REMOVED***
Caseless.prototype.swap = function (name) ***REMOVED***
  var has = this.has(name)
  if (has === name) return
  if (!has) throw new Error('There is no header than matches "'+name+'"')
  this.dict[name] = this.dict[has]
  delete this.dict[has]
***REMOVED***
Caseless.prototype.del = function (name) ***REMOVED***
  var has = this.has(name)
  return delete this.dict[has || name]
***REMOVED***

module.exports = function (dict) ***REMOVED***return new Caseless(dict)***REMOVED***
module.exports.httpify = function (resp, headers) ***REMOVED***
  var c = new Caseless(headers)
  resp.setHeader = function (key, value, clobber) ***REMOVED***
    if (typeof value === 'undefined') return
    return c.set(key, value, clobber)
  ***REMOVED***
  resp.hasHeader = function (key) ***REMOVED***
    return c.has(key)
  ***REMOVED***
  resp.getHeader = function (key) ***REMOVED***
    return c.get(key)
  ***REMOVED***
  resp.removeHeader = function (key) ***REMOVED***
    return c.del(key)
  ***REMOVED***
  resp.headers = c.dict
  return c
***REMOVED***
