// request/lib/har.js
'use strict'

var fs = require('fs')
var qs = require('querystring')
var validate = require('har-validator')
var extend = require('extend')

function Har (request) ***REMOVED***
  this.request = request
***REMOVED***

Har.prototype.reducer = function (obj, pair) ***REMOVED***
  // new property ?
  if (obj[pair.name] === undefined) ***REMOVED***
    obj[pair.name] = pair.value
    return obj
  ***REMOVED***

  // existing? convert to array
  var arr = [
    obj[pair.name],
    pair.value
  ]

  obj[pair.name] = arr

  return obj
***REMOVED***

Har.prototype.prep = function (data) ***REMOVED***
  // construct utility properties
  data.queryObj = ***REMOVED******REMOVED***
  data.headersObj = ***REMOVED******REMOVED***
  data.postData.jsonObj = false
  data.postData.paramsObj = false

  // construct query objects
  if (data.queryString && data.queryString.length) ***REMOVED***
    data.queryObj = data.queryString.reduce(this.reducer, ***REMOVED******REMOVED***)
  ***REMOVED***

  // construct headers objects
  if (data.headers && data.headers.length) ***REMOVED***
    // loweCase header keys
    data.headersObj = data.headers.reduceRight(function (headers, header) ***REMOVED***
      headers[header.name] = header.value
      return headers
    ***REMOVED***, ***REMOVED******REMOVED***)
  ***REMOVED***

  // construct Cookie header
  if (data.cookies && data.cookies.length) ***REMOVED***
    var cookies = data.cookies.map(function (cookie) ***REMOVED***
      return cookie.name + '=' + cookie.value
    ***REMOVED***)

    if (cookies.length) ***REMOVED***
      data.headersObj.cookie = cookies.join('; ')
    ***REMOVED***
  ***REMOVED***

  // prep body
  function some (arr) ***REMOVED***
    return arr.some(function (type) ***REMOVED***
      return data.postData.mimeType.indexOf(type) === 0
    ***REMOVED***)
  ***REMOVED***

  if (some([
    'multipart/mixed',
    'multipart/related',
    'multipart/form-data',
    'multipart/alternative'])) ***REMOVED***
    // reset values
    data.postData.mimeType = 'multipart/form-data'
  ***REMOVED*** else if (some([
    'application/x-www-form-urlencoded'])) ***REMOVED***
    if (!data.postData.params) ***REMOVED***
      data.postData.text = ''
    ***REMOVED*** else ***REMOVED***
      data.postData.paramsObj = data.postData.params.reduce(this.reducer, ***REMOVED******REMOVED***)

      // always overwrite
      data.postData.text = qs.stringify(data.postData.paramsObj)
    ***REMOVED***
  ***REMOVED*** else if (some([
    'text/json',
    'text/x-json',
    'application/json',
    'application/x-json'])) ***REMOVED***
    data.postData.mimeType = 'application/json'

    if (data.postData.text) ***REMOVED***
      try ***REMOVED***
        data.postData.jsonObj = JSON.parse(data.postData.text)
      ***REMOVED*** catch (e) ***REMOVED***
        this.request.debug(e)

        // force back to text/plain
        data.postData.mimeType = 'text/plain'
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  return data
***REMOVED***

Har.prototype.options = function (options) ***REMOVED***
  // skip if no har property defined
  if (!options.har) ***REMOVED***
    return options
  ***REMOVED***

  var har = ***REMOVED******REMOVED***
  extend(har, options.har)

  // only process the first entry
  if (har.log && har.log.entries) ***REMOVED***
    har = har.log.entries[0]
  ***REMOVED***

  // add optional properties to make validation successful
  har.url = har.url || options.url || options.uri || options.baseUrl || '/'
  har.httpVersion = har.httpVersion || 'HTTP/1.1'
  har.queryString = har.queryString || []
  har.headers = har.headers || []
  har.cookies = har.cookies || []
  har.postData = har.postData || ***REMOVED******REMOVED***
  har.postData.mimeType = har.postData.mimeType || 'application/octet-stream'

  har.bodySize = 0
  har.headersSize = 0
  har.postData.size = 0

  if (!validate.request(har)) ***REMOVED***
    return options
  ***REMOVED***

  // clean up and get some utility properties
  var req = this.prep(har)

  // construct new options
  if (req.url) ***REMOVED***
    options.url = req.url
  ***REMOVED***

  if (req.method) ***REMOVED***
    options.method = req.method
  ***REMOVED***

  if (Object.keys(req.queryObj).length) ***REMOVED***
    options.qs = req.queryObj
  ***REMOVED***

  if (Object.keys(req.headersObj).length) ***REMOVED***
    options.headers = req.headersObj
  ***REMOVED***

  function test (type) ***REMOVED***
    return req.postData.mimeType.indexOf(type) === 0
  ***REMOVED***
  if (test('application/x-www-form-urlencoded')) ***REMOVED***
    options.form = req.postData.paramsObj
  ***REMOVED*** else if (test('application/json')) ***REMOVED***
    if (req.postData.jsonObj) ***REMOVED***
      options.body = req.postData.jsonObj
      options.json = true
    ***REMOVED***
  ***REMOVED*** else if (test('multipart/form-data')) ***REMOVED***
    options.formData = ***REMOVED******REMOVED***

    req.postData.params.forEach(function (param) ***REMOVED***
      var attachment = ***REMOVED******REMOVED***

      if (!param.fileName && !param.fileName && !param.contentType) ***REMOVED***
        options.formData[param.name] = param.value
        return
      ***REMOVED***

      // attempt to read from disk!
      if (param.fileName && !param.value) ***REMOVED***
        attachment.value = fs.createReadStream(param.fileName)
      ***REMOVED*** else if (param.value) ***REMOVED***
        attachment.value = param.value
      ***REMOVED***

      if (param.fileName) ***REMOVED***
        attachment.options = ***REMOVED***
          filename: param.fileName,
          contentType: param.contentType ? param.contentType : null
        ***REMOVED***
      ***REMOVED***

      options.formData[param.name] = attachment
    ***REMOVED***)
  ***REMOVED*** else ***REMOVED***
    if (req.postData.text) ***REMOVED***
      options.body = req.postData.text
    ***REMOVED***
  ***REMOVED***

  return options
***REMOVED***

exports.Har = Har
