// request/lib/multipart.js
'use strict'

var uuid = require('uuid')
var CombinedStream = require('combined-stream')
var isstream = require('isstream')
var Buffer = require('safe-buffer').Buffer

function Multipart (request) ***REMOVED***
  this.request = request
  this.boundary = uuid()
  this.chunked = false
  this.body = null
***REMOVED***

Multipart.prototype.isChunked = function (options) ***REMOVED***
  var self = this
  var chunked = false
  var parts = options.data || options

  if (!parts.forEach) ***REMOVED***
    self.request.emit('error', new Error('Argument error, options.multipart.'))
  ***REMOVED***

  if (options.chunked !== undefined) ***REMOVED***
    chunked = options.chunked
  ***REMOVED***

  if (self.request.getHeader('transfer-encoding') === 'chunked') ***REMOVED***
    chunked = true
  ***REMOVED***

  if (!chunked) ***REMOVED***
    parts.forEach(function (part) ***REMOVED***
      if (typeof part.body === 'undefined') ***REMOVED***
        self.request.emit('error', new Error('Body attribute missing in multipart.'))
      ***REMOVED***
      if (isstream(part.body)) ***REMOVED***
        chunked = true
      ***REMOVED***
    ***REMOVED***)
  ***REMOVED***

  return chunked
***REMOVED***

Multipart.prototype.setHeaders = function (chunked) ***REMOVED***
  var self = this

  if (chunked && !self.request.hasHeader('transfer-encoding')) ***REMOVED***
    self.request.setHeader('transfer-encoding', 'chunked')
  ***REMOVED***

  var header = self.request.getHeader('content-type')

  if (!header || header.indexOf('multipart') === -1) ***REMOVED***
    self.request.setHeader('content-type', 'multipart/related; boundary=' + self.boundary)
  ***REMOVED*** else ***REMOVED***
    if (header.indexOf('boundary') !== -1) ***REMOVED***
      self.boundary = header.replace(/.*boundary=([^\s;]+).*/, '$1')
    ***REMOVED*** else ***REMOVED***
      self.request.setHeader('content-type', header + '; boundary=' + self.boundary)
    ***REMOVED***
  ***REMOVED***
***REMOVED***

Multipart.prototype.build = function (parts, chunked) ***REMOVED***
  var self = this
  var body = chunked ? new CombinedStream() : []

  function add (part) ***REMOVED***
    if (typeof part === 'number') ***REMOVED***
      part = part.toString()
    ***REMOVED***
    return chunked ? body.append(part) : body.push(Buffer.from(part))
  ***REMOVED***

  if (self.request.preambleCRLF) ***REMOVED***
    add('\r\n')
  ***REMOVED***

  parts.forEach(function (part) ***REMOVED***
    var preamble = '--' + self.boundary + '\r\n'
    Object.keys(part).forEach(function (key) ***REMOVED***
      if (key === 'body') ***REMOVED*** return ***REMOVED***
      preamble += key + ': ' + part[key] + '\r\n'
    ***REMOVED***)
    preamble += '\r\n'
    add(preamble)
    add(part.body)
    add('\r\n')
  ***REMOVED***)
  add('--' + self.boundary + '--')

  if (self.request.postambleCRLF) ***REMOVED***
    add('\r\n')
  ***REMOVED***

  return body
***REMOVED***

Multipart.prototype.onRequest = function (options) ***REMOVED***
  var self = this

  var chunked = self.isChunked(options)
  var parts = options.data || options

  self.setHeaders(chunked)
  self.chunked = chunked
  self.body = self.build(parts, chunked)
***REMOVED***

exports.Multipart = Multipart
