// request/lib/redirect.js
'use strict'

var url = require('url')
var isUrl = /^https?:/

function Redirect (request) ***REMOVED***
  this.request = request
  this.followRedirect = true
  this.followRedirects = true
  this.followAllRedirects = false
  this.followOriginalHttpMethod = false
  this.allowRedirect = function () ***REMOVED*** return true ***REMOVED***
  this.maxRedirects = 10
  this.redirects = []
  this.redirectsFollowed = 0
  this.removeRefererHeader = false
***REMOVED***

Redirect.prototype.onRequest = function (options) ***REMOVED***
  var self = this

  if (options.maxRedirects !== undefined) ***REMOVED***
    self.maxRedirects = options.maxRedirects
  ***REMOVED***
  if (typeof options.followRedirect === 'function') ***REMOVED***
    self.allowRedirect = options.followRedirect
  ***REMOVED***
  if (options.followRedirect !== undefined) ***REMOVED***
    self.followRedirects = !!options.followRedirect
  ***REMOVED***
  if (options.followAllRedirects !== undefined) ***REMOVED***
    self.followAllRedirects = options.followAllRedirects
  ***REMOVED***
  if (self.followRedirects || self.followAllRedirects) ***REMOVED***
    self.redirects = self.redirects || []
  ***REMOVED***
  if (options.removeRefererHeader !== undefined) ***REMOVED***
    self.removeRefererHeader = options.removeRefererHeader
  ***REMOVED***
  if (options.followOriginalHttpMethod !== undefined) ***REMOVED***
    self.followOriginalHttpMethod = options.followOriginalHttpMethod
  ***REMOVED***
***REMOVED***

Redirect.prototype.redirectTo = function (response) ***REMOVED***
  var self = this
  var request = self.request

  var redirectTo = null
  if (response.statusCode >= 300 && response.statusCode < 400 && response.caseless.has('location')) ***REMOVED***
    var location = response.caseless.get('location')
    request.debug('redirect', location)

    if (self.followAllRedirects) ***REMOVED***
      redirectTo = location
    ***REMOVED*** else if (self.followRedirects) ***REMOVED***
      switch (request.method) ***REMOVED***
        case 'PATCH':
        case 'PUT':
        case 'POST':
        case 'DELETE':
          // Do not follow redirects
          break
        default:
          redirectTo = location
          break
      ***REMOVED***
    ***REMOVED***
  ***REMOVED*** else if (response.statusCode === 401) ***REMOVED***
    var authHeader = request._auth.onResponse(response)
    if (authHeader) ***REMOVED***
      request.setHeader('authorization', authHeader)
      redirectTo = request.uri
    ***REMOVED***
  ***REMOVED***
  return redirectTo
***REMOVED***

Redirect.prototype.onResponse = function (response) ***REMOVED***
  var self = this
  var request = self.request

  var redirectTo = self.redirectTo(response)
  if (!redirectTo || !self.allowRedirect.call(request, response)) ***REMOVED***
    return false
  ***REMOVED***

  request.debug('redirect to', redirectTo)

  // ignore any potential response body.  it cannot possibly be useful
  // to us at this point.
  // response.resume should be defined, but check anyway before calling. Workaround for browserify.
  if (response.resume) ***REMOVED***
    response.resume()
  ***REMOVED***

  if (self.redirectsFollowed >= self.maxRedirects) ***REMOVED***
    request.emit('error', new Error('Exceeded maxRedirects. Probably stuck in a redirect loop ' + request.uri.href))
    return false
  ***REMOVED***
  self.redirectsFollowed += 1

  if (!isUrl.test(redirectTo)) ***REMOVED***
    redirectTo = url.resolve(request.uri.href, redirectTo)
  ***REMOVED***

  var uriPrev = request.uri
  request.uri = url.parse(redirectTo)

  // handle the case where we change protocol from https to http or vice versa
  if (request.uri.protocol !== uriPrev.protocol) ***REMOVED***
    delete request.agent
  ***REMOVED***

  self.redirects.push(***REMOVED*** statusCode: response.statusCode, redirectUri: redirectTo ***REMOVED***)

  if (self.followAllRedirects && request.method !== 'HEAD' &&
    response.statusCode !== 401 && response.statusCode !== 307) ***REMOVED***
    request.method = self.followOriginalHttpMethod ? request.method : 'GET'
  ***REMOVED***
  // request.method = 'GET' // Force all redirects to use GET || commented out fixes #215
  delete request.src
  delete request.req
  delete request._started
  if (response.statusCode !== 401 && response.statusCode !== 307) ***REMOVED***
    // Remove parameters from the previous response, unless this is the second request
    // for a server that requires digest authentication.
    delete request.body
    delete request._form
    if (request.headers) ***REMOVED***
      request.removeHeader('host')
      request.removeHeader('content-type')
      request.removeHeader('content-length')
      if (request.uri.hostname !== request.originalHost.split(':')[0]) ***REMOVED***
        // Remove authorization if changing hostnames (but not if just
        // changing ports or protocols).  This matches the behavior of curl:
        // https://github.com/bagder/curl/blob/6beb0eee/lib/http.c#L710
        request.removeHeader('authorization')
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  if (!self.removeRefererHeader) ***REMOVED***
    request.setHeader('referer', uriPrev.href)
  ***REMOVED***

  request.emit('redirect')

  request.init()

  return true
***REMOVED***

exports.Redirect = Redirect
