// mongodb-core/lib/topologies/read_preference.js
'use strict';

/**
 * @fileOverview The **ReadPreference** class is a class that represents a MongoDB ReadPreference and is
 * used to construct connections.
 *
 * @example
 * const ReplSet = require('mongodb-core').ReplSet,
 *   ReadPreference = require('mongodb-core').ReadPreference,
 *   assert = require('assert');
 *
 * const server = new ReplSet([***REMOVED***host: 'localhost', port: 30000***REMOVED***], ***REMOVED***setName: 'rs'***REMOVED***);
 * // Wait for the connection event
 * server.on('connect', function(server) ***REMOVED***
 *   const cursor = server.cursor(
 *     'db.test',
 *     ***REMOVED*** find: 'db.test', query: ***REMOVED******REMOVED*** ***REMOVED***,
 *     ***REMOVED*** readPreference: new ReadPreference('secondary') ***REMOVED***
 *   );
 *
 *   cursor.next(function(err, doc) ***REMOVED***
 *     server.destroy();
 *   ***REMOVED***);
 * ***REMOVED***);
 *
 * // Start connecting
 * server.connect();
 */

/**
 * Creates a new Pool instance
 * @class
 * @param ***REMOVED***string***REMOVED*** mode A string describing the read preference mode (primary|primaryPreferred|secondary|secondaryPreferred|nearest)
 * @param ***REMOVED***array***REMOVED*** tags The tags object
 * @param ***REMOVED***object***REMOVED*** [options] Additional read preference options
 * @param ***REMOVED***number***REMOVED*** [options.maxStalenessSeconds] Max secondary read staleness in seconds, Minimum value is 90 seconds.
 * @property ***REMOVED***string***REMOVED*** mode The read preference mode (primary|primaryPreferred|secondary|secondaryPreferred|nearest)
 * @property ***REMOVED***array***REMOVED*** tags The tags object
 * @property ***REMOVED***object***REMOVED*** options Additional read preference options
 * @property ***REMOVED***number***REMOVED*** maxStalenessSeconds MaxStalenessSeconds value for the read preference
 * @return ***REMOVED***ReadPreference***REMOVED***
 */
const ReadPreference = function(mode, tags, options) ***REMOVED***
  this.mode = mode;
  this.tags = tags;
  this.options = options;

  // Add the maxStalenessSeconds value to the read Preference
  if (this.options && this.options.maxStalenessSeconds != null) ***REMOVED***
    this.options = options;
    this.maxStalenessSeconds =
      this.options.maxStalenessSeconds >= 0 ? this.options.maxStalenessSeconds : null;
  ***REMOVED*** else if (tags && typeof tags === 'object') ***REMOVED***
    (this.options = tags), (tags = null);
  ***REMOVED***
***REMOVED***;

// Support the deprecated `preference` property introduced in the porcelain layer
Object.defineProperty(ReadPreference.prototype, 'preference', ***REMOVED***
  enumerable: true,
  get: function() ***REMOVED***
    return this.mode;
  ***REMOVED***
***REMOVED***);

/**
 * Read preference mode constants
 */
ReadPreference.PRIMARY = 'primary';
ReadPreference.PRIMARY_PREFERRED = 'primaryPreferred';
ReadPreference.SECONDARY = 'secondary';
ReadPreference.SECONDARY_PREFERRED = 'secondaryPreferred';
ReadPreference.NEAREST = 'nearest';

const VALID_MODES = [
  ReadPreference.PRIMARY,
  ReadPreference.PRIMARY_PREFERRED,
  ReadPreference.SECONDARY,
  ReadPreference.SECONDARY_PREFERRED,
  ReadPreference.NEAREST,
  true,
  false,
  null
];

/**
 * Validate if a mode is legal
 *
 * @method
 * @param ***REMOVED***string***REMOVED*** mode The string representing the read preference mode.
 * @return ***REMOVED***boolean***REMOVED***
 */
ReadPreference.isValid = function(mode) ***REMOVED***
  return VALID_MODES.indexOf(mode) !== -1;
***REMOVED***;

/**
 * Validate if a mode is legal
 *
 * @method
 * @param ***REMOVED***string***REMOVED*** mode The string representing the read preference mode.
 * @return ***REMOVED***boolean***REMOVED***
 */
ReadPreference.prototype.isValid = function(mode) ***REMOVED***
  return ReadPreference.isValid(typeof mode === 'string' ? mode : this.mode);
***REMOVED***;

const needSlaveOk = ['primaryPreferred', 'secondary', 'secondaryPreferred', 'nearest'];

/**
 * This needs slaveOk bit set
 * @method
 * @return ***REMOVED***boolean***REMOVED***
 */
ReadPreference.prototype.slaveOk = function() ***REMOVED***
  return needSlaveOk.indexOf(this.mode) !== -1;
***REMOVED***;

/**
 * Are the two read preference equal
 * @method
 * @return ***REMOVED***boolean***REMOVED***
 */
ReadPreference.prototype.equals = function(readPreference) ***REMOVED***
  return readPreference.mode === this.mode;
***REMOVED***;

/**
 * Return JSON representation
 * @method
 * @return ***REMOVED***Object***REMOVED***
 */
ReadPreference.prototype.toJSON = function() ***REMOVED***
  const readPreference = ***REMOVED*** mode: this.mode ***REMOVED***;
  if (Array.isArray(this.tags)) readPreference.tags = this.tags;
  if (this.maxStalenessSeconds) readPreference.maxStalenessSeconds = this.maxStalenessSeconds;
  return readPreference;
***REMOVED***;

/**
 * Primary read preference
 * @method
 * @return ***REMOVED***ReadPreference***REMOVED***
 */
ReadPreference.primary = new ReadPreference('primary');
/**
 * Primary Preferred read preference
 * @method
 * @return ***REMOVED***ReadPreference***REMOVED***
 */
ReadPreference.primaryPreferred = new ReadPreference('primaryPreferred');
/**
 * Secondary read preference
 * @method
 * @return ***REMOVED***ReadPreference***REMOVED***
 */
ReadPreference.secondary = new ReadPreference('secondary');
/**
 * Secondary Preferred read preference
 * @method
 * @return ***REMOVED***ReadPreference***REMOVED***
 */
ReadPreference.secondaryPreferred = new ReadPreference('secondaryPreferred');
/**
 * Nearest read preference
 * @method
 * @return ***REMOVED***ReadPreference***REMOVED***
 */
ReadPreference.nearest = new ReadPreference('nearest');

module.exports = ReadPreference;
