// mongodb-core/lib/topologies/shared.js
'use strict';

const os = require('os');
const f = require('util').format;
const ReadPreference = require('./read_preference');

/**
 * Emit event if it exists
 * @method
 */
function emitSDAMEvent(self, event, description) ***REMOVED***
  if (self.listeners(event).length > 0) ***REMOVED***
    self.emit(event, description);
  ***REMOVED***
***REMOVED***

// Get package.json variable
var driverVersion = require('../../package.json').version;
var nodejsversion = f('Node.js %s, %s', process.version, os.endianness());
var type = os.type();
var name = process.platform;
var architecture = process.arch;
var release = os.release();

function createClientInfo(options) ***REMOVED***
  // Build default client information
  var clientInfo = options.clientInfo
    ? clone(options.clientInfo)
    : ***REMOVED***
        driver: ***REMOVED***
          name: 'nodejs-core',
          version: driverVersion
        ***REMOVED***,
        os: ***REMOVED***
          type: type,
          name: name,
          architecture: architecture,
          version: release
        ***REMOVED***
      ***REMOVED***;

  // Is platform specified
  if (clientInfo.platform && clientInfo.platform.indexOf('mongodb-core') === -1) ***REMOVED***
    clientInfo.platform = f('%s, mongodb-core: %s', clientInfo.platform, driverVersion);
  ***REMOVED*** else if (!clientInfo.platform) ***REMOVED***
    clientInfo.platform = nodejsversion;
  ***REMOVED***

  // Do we have an application specific string
  if (options.appname) ***REMOVED***
    // Cut at 128 bytes
    var buffer = new Buffer(options.appname);
    // Return the truncated appname
    var appname = buffer.length > 128 ? buffer.slice(0, 128).toString('utf8') : options.appname;
    // Add to the clientInfo
    clientInfo.application = ***REMOVED*** name: appname ***REMOVED***;
  ***REMOVED***

  return clientInfo;
***REMOVED***

function createCompressionInfo(options) ***REMOVED***
  if (!options.compression || !options.compression.compressors) ***REMOVED***
    return [];
  ***REMOVED***

  // Check that all supplied compressors are valid
  options.compression.compressors.forEach(function(compressor) ***REMOVED***
    if (compressor !== 'snappy' && compressor !== 'zlib') ***REMOVED***
      throw new Error('compressors must be at least one of snappy or zlib');
    ***REMOVED***
  ***REMOVED***);

  return options.compression.compressors;
***REMOVED***

function clone(object) ***REMOVED***
  return JSON.parse(JSON.stringify(object));
***REMOVED***

var getPreviousDescription = function(self) ***REMOVED***
  if (!self.s.serverDescription) ***REMOVED***
    self.s.serverDescription = ***REMOVED***
      address: self.name,
      arbiters: [],
      hosts: [],
      passives: [],
      type: 'Unknown'
    ***REMOVED***;
  ***REMOVED***

  return self.s.serverDescription;
***REMOVED***;

var emitServerDescriptionChanged = function(self, description) ***REMOVED***
  if (self.listeners('serverDescriptionChanged').length > 0) ***REMOVED***
    // Emit the server description changed events
    self.emit('serverDescriptionChanged', ***REMOVED***
      topologyId: self.s.topologyId !== -1 ? self.s.topologyId : self.id,
      address: self.name,
      previousDescription: getPreviousDescription(self),
      newDescription: description
    ***REMOVED***);

    self.s.serverDescription = description;
  ***REMOVED***
***REMOVED***;

var getPreviousTopologyDescription = function(self) ***REMOVED***
  if (!self.s.topologyDescription) ***REMOVED***
    self.s.topologyDescription = ***REMOVED***
      topologyType: 'Unknown',
      servers: [
        ***REMOVED***
          address: self.name,
          arbiters: [],
          hosts: [],
          passives: [],
          type: 'Unknown'
        ***REMOVED***
      ]
    ***REMOVED***;
  ***REMOVED***

  return self.s.topologyDescription;
***REMOVED***;

var emitTopologyDescriptionChanged = function(self, description) ***REMOVED***
  if (self.listeners('topologyDescriptionChanged').length > 0) ***REMOVED***
    // Emit the server description changed events
    self.emit('topologyDescriptionChanged', ***REMOVED***
      topologyId: self.s.topologyId !== -1 ? self.s.topologyId : self.id,
      address: self.name,
      previousDescription: getPreviousTopologyDescription(self),
      newDescription: description
    ***REMOVED***);

    self.s.serverDescription = description;
  ***REMOVED***
***REMOVED***;

var changedIsMaster = function(self, currentIsmaster, ismaster) ***REMOVED***
  var currentType = getTopologyType(self, currentIsmaster);
  var newType = getTopologyType(self, ismaster);
  if (newType !== currentType) return true;
  return false;
***REMOVED***;

var getTopologyType = function(self, ismaster) ***REMOVED***
  if (!ismaster) ***REMOVED***
    ismaster = self.ismaster;
  ***REMOVED***

  if (!ismaster) return 'Unknown';
  if (ismaster.ismaster && ismaster.msg === 'isdbgrid') return 'Mongos';
  if (ismaster.ismaster && !ismaster.hosts) return 'Standalone';
  if (ismaster.ismaster) return 'RSPrimary';
  if (ismaster.secondary) return 'RSSecondary';
  if (ismaster.arbiterOnly) return 'RSArbiter';
  return 'Unknown';
***REMOVED***;

var inquireServerState = function(self) ***REMOVED***
  return function(callback) ***REMOVED***
    if (self.s.state === 'destroyed') return;
    // Record response time
    var start = new Date().getTime();

    // emitSDAMEvent
    emitSDAMEvent(self, 'serverHeartbeatStarted', ***REMOVED*** connectionId: self.name ***REMOVED***);

    // Attempt to execute ismaster command
    self.command('admin.$cmd', ***REMOVED*** ismaster: true ***REMOVED***, ***REMOVED*** monitoring: true ***REMOVED***, function(err, r) ***REMOVED***
      if (!err) ***REMOVED***
        // Legacy event sender
        self.emit('ismaster', r, self);

        // Calculate latencyMS
        var latencyMS = new Date().getTime() - start;

        // Server heart beat event
        emitSDAMEvent(self, 'serverHeartbeatSucceeded', ***REMOVED***
          durationMS: latencyMS,
          reply: r.result,
          connectionId: self.name
        ***REMOVED***);

        // Did the server change
        if (changedIsMaster(self, self.s.ismaster, r.result)) ***REMOVED***
          // Emit server description changed if something listening
          emitServerDescriptionChanged(self, ***REMOVED***
            address: self.name,
            arbiters: [],
            hosts: [],
            passives: [],
            type: !self.s.inTopology ? 'Standalone' : getTopologyType(self)
          ***REMOVED***);
        ***REMOVED***

        // Updat ismaster view
        self.s.ismaster = r.result;

        // Set server response time
        self.s.isMasterLatencyMS = latencyMS;
      ***REMOVED*** else ***REMOVED***
        emitSDAMEvent(self, 'serverHeartbeatFailed', ***REMOVED***
          durationMS: latencyMS,
          failure: err,
          connectionId: self.name
        ***REMOVED***);
      ***REMOVED***

      // Peforming an ismaster monitoring callback operation
      if (typeof callback === 'function') ***REMOVED***
        return callback(err, r);
      ***REMOVED***

      // Perform another sweep
      self.s.inquireServerStateTimeout = setTimeout(inquireServerState(self), self.s.haInterval);
    ***REMOVED***);
  ***REMOVED***;
***REMOVED***;

//
// Clone the options
var cloneOptions = function(options) ***REMOVED***
  var opts = ***REMOVED******REMOVED***;
  for (var name in options) ***REMOVED***
    opts[name] = options[name];
  ***REMOVED***
  return opts;
***REMOVED***;

function Interval(fn, time) ***REMOVED***
  var timer = false;

  this.start = function() ***REMOVED***
    if (!this.isRunning()) ***REMOVED***
      timer = setInterval(fn, time);
    ***REMOVED***

    return this;
  ***REMOVED***;

  this.stop = function() ***REMOVED***
    clearInterval(timer);
    timer = false;
    return this;
  ***REMOVED***;

  this.isRunning = function() ***REMOVED***
    return timer !== false;
  ***REMOVED***;
***REMOVED***

function Timeout(fn, time) ***REMOVED***
  var timer = false;

  this.start = function() ***REMOVED***
    if (!this.isRunning()) ***REMOVED***
      timer = setTimeout(fn, time);
    ***REMOVED***
    return this;
  ***REMOVED***;

  this.stop = function() ***REMOVED***
    clearTimeout(timer);
    timer = false;
    return this;
  ***REMOVED***;

  this.isRunning = function() ***REMOVED***
    if (timer && timer._called) return false;
    return timer !== false;
  ***REMOVED***;
***REMOVED***

function diff(previous, current) ***REMOVED***
  // Difference document
  var diff = ***REMOVED***
    servers: []
  ***REMOVED***;

  // Previous entry
  if (!previous) ***REMOVED***
    previous = ***REMOVED*** servers: [] ***REMOVED***;
  ***REMOVED***

  // Check if we have any previous servers missing in the current ones
  for (var i = 0; i < previous.servers.length; i++) ***REMOVED***
    var found = false;

    for (var j = 0; j < current.servers.length; j++) ***REMOVED***
      if (current.servers[j].address.toLowerCase() === previous.servers[i].address.toLowerCase()) ***REMOVED***
        found = true;
        break;
      ***REMOVED***
    ***REMOVED***

    if (!found) ***REMOVED***
      // Add to the diff
      diff.servers.push(***REMOVED***
        address: previous.servers[i].address,
        from: previous.servers[i].type,
        to: 'Unknown'
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  // Check if there are any severs that don't exist
  for (j = 0; j < current.servers.length; j++) ***REMOVED***
    found = false;

    // Go over all the previous servers
    for (i = 0; i < previous.servers.length; i++) ***REMOVED***
      if (previous.servers[i].address.toLowerCase() === current.servers[j].address.toLowerCase()) ***REMOVED***
        found = true;
        break;
      ***REMOVED***
    ***REMOVED***

    // Add the server to the diff
    if (!found) ***REMOVED***
      diff.servers.push(***REMOVED***
        address: current.servers[j].address,
        from: 'Unknown',
        to: current.servers[j].type
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***

  // Got through all the servers
  for (i = 0; i < previous.servers.length; i++) ***REMOVED***
    var prevServer = previous.servers[i];

    // Go through all current servers
    for (j = 0; j < current.servers.length; j++) ***REMOVED***
      var currServer = current.servers[j];

      // Matching server
      if (prevServer.address.toLowerCase() === currServer.address.toLowerCase()) ***REMOVED***
        // We had a change in state
        if (prevServer.type !== currServer.type) ***REMOVED***
          diff.servers.push(***REMOVED***
            address: prevServer.address,
            from: prevServer.type,
            to: currServer.type
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  // Return difference
  return diff;
***REMOVED***

/**
 * Shared function to determine clusterTime for a given topology
 *
 * @param ***REMOVED*******REMOVED*** topology
 * @param ***REMOVED*******REMOVED*** clusterTime
 */
function resolveClusterTime(topology, $clusterTime) ***REMOVED***
  if (topology.clusterTime == null) ***REMOVED***
    topology.clusterTime = $clusterTime;
  ***REMOVED*** else ***REMOVED***
    if ($clusterTime.clusterTime.greaterThan(topology.clusterTime.clusterTime)) ***REMOVED***
      topology.clusterTime = $clusterTime;
    ***REMOVED***
  ***REMOVED***
***REMOVED***

// NOTE: this is a temporary move until the topologies can be more formally refactored
//       to share code.
const SessionMixins = ***REMOVED***
  endSessions: function(sessions, callback) ***REMOVED***
    if (!Array.isArray(sessions)) ***REMOVED***
      sessions = [sessions];
    ***REMOVED***

    // TODO:
    //   When connected to a sharded cluster the endSessions command
    //   can be sent to any mongos. When connected to a replica set the
    //   endSessions command MUST be sent to the primary if the primary
    //   is available, otherwise it MUST be sent to any available secondary.
    //   Is it enough to use: ReadPreference.primaryPreferred ?
    this.command(
      'admin.$cmd',
      ***REMOVED*** endSessions: sessions ***REMOVED***,
      ***REMOVED*** readPreference: ReadPreference.primaryPreferred ***REMOVED***,
      () => ***REMOVED***
        // intentionally ignored, per spec
        if (typeof callback === 'function') callback();
      ***REMOVED***
    );
  ***REMOVED***
***REMOVED***;

const RETRYABLE_WIRE_VERSION = 6;

/**
 * Determines whether the provided topology supports retryable writes
 *
 * @param ***REMOVED***Mongos|Replset***REMOVED*** topology
 */
const isRetryableWritesSupported = function(topology) ***REMOVED***
  const maxWireVersion = topology.lastIsMaster().maxWireVersion;
  if (maxWireVersion < RETRYABLE_WIRE_VERSION) ***REMOVED***
    return false;
  ***REMOVED***

  if (!topology.logicalSessionTimeoutMinutes) ***REMOVED***
    return false;
  ***REMOVED***

  return true;
***REMOVED***;

module.exports.SessionMixins = SessionMixins;
module.exports.resolveClusterTime = resolveClusterTime;
module.exports.inquireServerState = inquireServerState;
module.exports.getTopologyType = getTopologyType;
module.exports.emitServerDescriptionChanged = emitServerDescriptionChanged;
module.exports.emitTopologyDescriptionChanged = emitTopologyDescriptionChanged;
module.exports.cloneOptions = cloneOptions;
module.exports.createClientInfo = createClientInfo;
module.exports.createCompressionInfo = createCompressionInfo;
module.exports.clone = clone;
module.exports.diff = diff;
module.exports.Interval = Interval;
module.exports.Timeout = Timeout;
module.exports.isRetryableWritesSupported = isRetryableWritesSupported;
