// http-signature/lib/signer.js
// Copyright 2012 Joyent, Inc.  All rights reserved.

var assert = require('assert-plus');
var crypto = require('crypto');
var http = require('http');
var util = require('util');
var sshpk = require('sshpk');
var jsprim = require('jsprim');
var utils = require('./utils');

var sprintf = require('util').format;

var HASH_ALGOS = utils.HASH_ALGOS;
var PK_ALGOS = utils.PK_ALGOS;
var InvalidAlgorithmError = utils.InvalidAlgorithmError;
var HttpSignatureError = utils.HttpSignatureError;
var validateAlgorithm = utils.validateAlgorithm;

///--- Globals

var AUTHZ_FMT =
  'Signature keyId="%s",algorithm="%s",headers="%s",signature="%s"';

///--- Specific Errors

function MissingHeaderError(message) ***REMOVED***
  HttpSignatureError.call(this, message, MissingHeaderError);
***REMOVED***
util.inherits(MissingHeaderError, HttpSignatureError);

function StrictParsingError(message) ***REMOVED***
  HttpSignatureError.call(this, message, StrictParsingError);
***REMOVED***
util.inherits(StrictParsingError, HttpSignatureError);

/* See createSigner() */
function RequestSigner(options) ***REMOVED***
  assert.object(options, 'options');

  var alg = [];
  if (options.algorithm !== undefined) ***REMOVED***
    assert.string(options.algorithm, 'options.algorithm');
    alg = validateAlgorithm(options.algorithm);
  ***REMOVED***
  this.rs_alg = alg;

  /*
   * RequestSigners come in two varieties: ones with an rs_signFunc, and ones
   * with an rs_signer.
   *
   * rs_signFunc-based RequestSigners have to build up their entire signing
   * string within the rs_lines array and give it to rs_signFunc as a single
   * concat'd blob. rs_signer-based RequestSigners can add a line at a time to
   * their signing state by using rs_signer.update(), thus only needing to
   * buffer the hash function state and one line at a time.
   */
  if (options.sign !== undefined) ***REMOVED***
    assert.func(options.sign, 'options.sign');
    this.rs_signFunc = options.sign;

  ***REMOVED*** else if (alg[0] === 'hmac' && options.key !== undefined) ***REMOVED***
    assert.string(options.keyId, 'options.keyId');
    this.rs_keyId = options.keyId;

    if (typeof (options.key) !== 'string' && !Buffer.isBuffer(options.key))
      throw (new TypeError('options.key for HMAC must be a string or Buffer'));

    /*
     * Make an rs_signer for HMACs, not a rs_signFunc -- HMACs digest their
     * data in chunks rather than requiring it all to be given in one go
     * at the end, so they are more similar to signers than signFuncs.
     */
    this.rs_signer = crypto.createHmac(alg[1].toUpperCase(), options.key);
    this.rs_signer.sign = function () ***REMOVED***
      var digest = this.digest('base64');
      return (***REMOVED***
        hashAlgorithm: alg[1],
        toString: function () ***REMOVED*** return (digest); ***REMOVED***
      ***REMOVED***);
    ***REMOVED***;

  ***REMOVED*** else if (options.key !== undefined) ***REMOVED***
    var key = options.key;
    if (typeof (key) === 'string' || Buffer.isBuffer(key))
      key = sshpk.parsePrivateKey(key);

    assert.ok(sshpk.PrivateKey.isPrivateKey(key, [1, 2]),
      'options.key must be a sshpk.PrivateKey');
    this.rs_key = key;

    assert.string(options.keyId, 'options.keyId');
    this.rs_keyId = options.keyId;

    if (!PK_ALGOS[key.type]) ***REMOVED***
      throw (new InvalidAlgorithmError(key.type.toUpperCase() + ' type ' +
        'keys are not supported'));
    ***REMOVED***

    if (alg[0] !== undefined && key.type !== alg[0]) ***REMOVED***
      throw (new InvalidAlgorithmError('options.key must be a ' +
        alg[0].toUpperCase() + ' key, was given a ' +
        key.type.toUpperCase() + ' key instead'));
    ***REMOVED***

    this.rs_signer = key.createSign(alg[1]);

  ***REMOVED*** else ***REMOVED***
    throw (new TypeError('options.sign (func) or options.key is required'));
  ***REMOVED***

  this.rs_headers = [];
  this.rs_lines = [];
***REMOVED***

/**
 * Adds a header to be signed, with its value, into this signer.
 *
 * @param ***REMOVED***String***REMOVED*** header
 * @param ***REMOVED***String***REMOVED*** value
 * @return ***REMOVED***String***REMOVED*** value written
 */
RequestSigner.prototype.writeHeader = function (header, value) ***REMOVED***
  assert.string(header, 'header');
  header = header.toLowerCase();
  assert.string(value, 'value');

  this.rs_headers.push(header);

  if (this.rs_signFunc) ***REMOVED***
    this.rs_lines.push(header + ': ' + value);

  ***REMOVED*** else ***REMOVED***
    var line = header + ': ' + value;
    if (this.rs_headers.length > 0)
      line = '\n' + line;
    this.rs_signer.update(line);
  ***REMOVED***

  return (value);
***REMOVED***;

/**
 * Adds a default Date header, returning its value.
 *
 * @return ***REMOVED***String***REMOVED***
 */
RequestSigner.prototype.writeDateHeader = function () ***REMOVED***
  return (this.writeHeader('date', jsprim.rfc1123(new Date())));
***REMOVED***;

/**
 * Adds the request target line to be signed.
 *
 * @param ***REMOVED***String***REMOVED*** method, HTTP method (e.g. 'get', 'post', 'put')
 * @param ***REMOVED***String***REMOVED*** path
 */
RequestSigner.prototype.writeTarget = function (method, path) ***REMOVED***
  assert.string(method, 'method');
  assert.string(path, 'path');
  method = method.toLowerCase();
  this.writeHeader('(request-target)', method + ' ' + path);
***REMOVED***;

/**
 * Calculate the value for the Authorization header on this request
 * asynchronously.
 *
 * @param ***REMOVED***Func***REMOVED*** callback (err, authz)
 */
RequestSigner.prototype.sign = function (cb) ***REMOVED***
  assert.func(cb, 'callback');

  if (this.rs_headers.length < 1)
    throw (new Error('At least one header must be signed'));

  var alg, authz;
  if (this.rs_signFunc) ***REMOVED***
    var data = this.rs_lines.join('\n');
    var self = this;
    this.rs_signFunc(data, function (err, sig) ***REMOVED***
      if (err) ***REMOVED***
        cb(err);
        return;
      ***REMOVED***
      try ***REMOVED***
        assert.object(sig, 'signature');
        assert.string(sig.keyId, 'signature.keyId');
        assert.string(sig.algorithm, 'signature.algorithm');
        assert.string(sig.signature, 'signature.signature');
        alg = validateAlgorithm(sig.algorithm);

        authz = sprintf(AUTHZ_FMT,
          sig.keyId,
          sig.algorithm,
          self.rs_headers.join(' '),
          sig.signature);
      ***REMOVED*** catch (e) ***REMOVED***
        cb(e);
        return;
      ***REMOVED***
      cb(null, authz);
    ***REMOVED***);

  ***REMOVED*** else ***REMOVED***
    try ***REMOVED***
      var sigObj = this.rs_signer.sign();
    ***REMOVED*** catch (e) ***REMOVED***
      cb(e);
      return;
    ***REMOVED***
    alg = (this.rs_alg[0] || this.rs_key.type) + '-' + sigObj.hashAlgorithm;
    var signature = sigObj.toString();
    authz = sprintf(AUTHZ_FMT,
      this.rs_keyId,
      alg,
      this.rs_headers.join(' '),
      signature);
    cb(null, authz);
  ***REMOVED***
***REMOVED***;

///--- Exported API

module.exports = ***REMOVED***
  /**
   * Identifies whether a given object is a request signer or not.
   *
   * @param ***REMOVED***Object***REMOVED*** object, the object to identify
   * @returns ***REMOVED***Boolean***REMOVED***
   */
  isSigner: function (obj) ***REMOVED***
    if (typeof (obj) === 'object' && obj instanceof RequestSigner)
      return (true);
    return (false);
  ***REMOVED***,

  /**
   * Creates a request signer, used to asynchronously build a signature
   * for a request (does not have to be an http.ClientRequest).
   *
   * @param ***REMOVED***Object***REMOVED*** options, either:
   *                   - ***REMOVED***String***REMOVED*** keyId
   *                   - ***REMOVED***String|Buffer***REMOVED*** key
   *                   - ***REMOVED***String***REMOVED*** algorithm (optional, required for HMAC)
   *                 or:
   *                   - ***REMOVED***Func***REMOVED*** sign (data, cb)
   * @return ***REMOVED***RequestSigner***REMOVED***
   */
  createSigner: function createSigner(options) ***REMOVED***
    return (new RequestSigner(options));
  ***REMOVED***,

  /**
   * Adds an 'Authorization' header to an http.ClientRequest object.
   *
   * Note that this API will add a Date header if it's not already set. Any
   * other headers in the options.headers array MUST be present, or this
   * will throw.
   *
   * You shouldn't need to check the return type; it's just there if you want
   * to be pedantic.
   *
   * The optional flag indicates whether parsing should use strict enforcement
   * of the version draft-cavage-http-signatures-04 of the spec or beyond.
   * The default is to be loose and support
   * older versions for compatibility.
   *
   * @param ***REMOVED***Object***REMOVED*** request an instance of http.ClientRequest.
   * @param ***REMOVED***Object***REMOVED*** options signing parameters object:
   *                   - ***REMOVED***String***REMOVED*** keyId required.
   *                   - ***REMOVED***String***REMOVED*** key required (either a PEM or HMAC key).
   *                   - ***REMOVED***Array***REMOVED*** headers optional; defaults to ['date'].
   *                   - ***REMOVED***String***REMOVED*** algorithm optional (unless key is HMAC);
   *                              default is the same as the sshpk default
   *                              signing algorithm for the type of key given
   *                   - ***REMOVED***String***REMOVED*** httpVersion optional; defaults to '1.1'.
   *                   - ***REMOVED***Boolean***REMOVED*** strict optional; defaults to 'false'.
   * @return ***REMOVED***Boolean***REMOVED*** true if Authorization (and optionally Date) were added.
   * @throws ***REMOVED***TypeError***REMOVED*** on bad parameter types (input).
   * @throws ***REMOVED***InvalidAlgorithmError***REMOVED*** if algorithm was bad or incompatible with
   *                                 the given key.
   * @throws ***REMOVED***sshpk.KeyParseError***REMOVED*** if key was bad.
   * @throws ***REMOVED***MissingHeaderError***REMOVED*** if a header to be signed was specified but
   *                              was not present.
   */
  signRequest: function signRequest(request, options) ***REMOVED***
    assert.object(request, 'request');
    assert.object(options, 'options');
    assert.optionalString(options.algorithm, 'options.algorithm');
    assert.string(options.keyId, 'options.keyId');
    assert.optionalArrayOfString(options.headers, 'options.headers');
    assert.optionalString(options.httpVersion, 'options.httpVersion');

    if (!request.getHeader('Date'))
      request.setHeader('Date', jsprim.rfc1123(new Date()));
    if (!options.headers)
      options.headers = ['date'];
    if (!options.httpVersion)
      options.httpVersion = '1.1';

    var alg = [];
    if (options.algorithm) ***REMOVED***
      options.algorithm = options.algorithm.toLowerCase();
      alg = validateAlgorithm(options.algorithm);
    ***REMOVED***

    var i;
    var stringToSign = '';
    for (i = 0; i < options.headers.length; i++) ***REMOVED***
      if (typeof (options.headers[i]) !== 'string')
        throw new TypeError('options.headers must be an array of Strings');

      var h = options.headers[i].toLowerCase();

      if (h === 'request-line') ***REMOVED***
        if (!options.strict) ***REMOVED***
          /**
           * We allow headers from the older spec drafts if strict parsing isn't
           * specified in options.
           */
          stringToSign +=
            request.method + ' ' + request.path + ' HTTP/' +
            options.httpVersion;
        ***REMOVED*** else ***REMOVED***
          /* Strict parsing doesn't allow older draft headers. */
          throw (new StrictParsingError('request-line is not a valid header ' +
            'with strict parsing enabled.'));
        ***REMOVED***
      ***REMOVED*** else if (h === '(request-target)') ***REMOVED***
        stringToSign +=
          '(request-target): ' + request.method.toLowerCase() + ' ' +
          request.path;
      ***REMOVED*** else ***REMOVED***
        var value = request.getHeader(h);
        if (value === undefined || value === '') ***REMOVED***
          throw new MissingHeaderError(h + ' was not in the request');
        ***REMOVED***
        stringToSign += h + ': ' + value;
      ***REMOVED***

      if ((i + 1) < options.headers.length)
        stringToSign += '\n';
    ***REMOVED***

    /* This is just for unit tests. */
    if (request.hasOwnProperty('_stringToSign')) ***REMOVED***
      request._stringToSign = stringToSign;
    ***REMOVED***

    var signature;
    if (alg[0] === 'hmac') ***REMOVED***
      if (typeof (options.key) !== 'string' && !Buffer.isBuffer(options.key))
        throw (new TypeError('options.key must be a string or Buffer'));

      var hmac = crypto.createHmac(alg[1].toUpperCase(), options.key);
      hmac.update(stringToSign);
      signature = hmac.digest('base64');

    ***REMOVED*** else ***REMOVED***
      var key = options.key;
      if (typeof (key) === 'string' || Buffer.isBuffer(key))
        key = sshpk.parsePrivateKey(options.key);

      assert.ok(sshpk.PrivateKey.isPrivateKey(key, [1, 2]),
        'options.key must be a sshpk.PrivateKey');

      if (!PK_ALGOS[key.type]) ***REMOVED***
        throw (new InvalidAlgorithmError(key.type.toUpperCase() + ' type ' +
          'keys are not supported'));
      ***REMOVED***

      if (alg[0] !== undefined && key.type !== alg[0]) ***REMOVED***
        throw (new InvalidAlgorithmError('options.key must be a ' +
          alg[0].toUpperCase() + ' key, was given a ' +
          key.type.toUpperCase() + ' key instead'));
      ***REMOVED***

      var signer = key.createSign(alg[1]);
      signer.update(stringToSign);
      var sigObj = signer.sign();
      if (!HASH_ALGOS[sigObj.hashAlgorithm]) ***REMOVED***
        throw (new InvalidAlgorithmError(sigObj.hashAlgorithm.toUpperCase() +
          ' is not a supported hash algorithm'));
      ***REMOVED***
      options.algorithm = key.type + '-' + sigObj.hashAlgorithm;
      signature = sigObj.toString();
      assert.notStrictEqual(signature, '', 'empty signature produced');
    ***REMOVED***

    var authzHeaderName = options.authorizationHeaderName || 'Authorization';

    request.setHeader(authzHeaderName, sprintf(AUTHZ_FMT,
                                               options.keyId,
                                               options.algorithm,
                                               options.headers.join(' '),
                                               signature));

    return true;
  ***REMOVED***

***REMOVED***;
