// sshpk/lib/formats/pkcs1.js
// Copyright 2015 Joyent, Inc.

module.exports = ***REMOVED***
	read: read,
	readPkcs1: readPkcs1,
	write: write,
	writePkcs1: writePkcs1
***REMOVED***;

var assert = require('assert-plus');
var asn1 = require('asn1');
var Buffer = require('safer-buffer').Buffer;
var algs = require('../algs');
var utils = require('../utils');

var Key = require('../key');
var PrivateKey = require('../private-key');
var pem = require('./pem');

var pkcs8 = require('./pkcs8');
var readECDSACurve = pkcs8.readECDSACurve;

function read(buf, options) ***REMOVED***
	return (pem.read(buf, options, 'pkcs1'));
***REMOVED***

function write(key, options) ***REMOVED***
	return (pem.write(key, options, 'pkcs1'));
***REMOVED***

/* Helper to read in a single mpint */
function readMPInt(der, nm) ***REMOVED***
	assert.strictEqual(der.peek(), asn1.Ber.Integer,
	    nm + ' is not an Integer');
	return (utils.mpNormalize(der.readString(asn1.Ber.Integer, true)));
***REMOVED***

function readPkcs1(alg, type, der) ***REMOVED***
	switch (alg) ***REMOVED***
	case 'RSA':
		if (type === 'public')
			return (readPkcs1RSAPublic(der));
		else if (type === 'private')
			return (readPkcs1RSAPrivate(der));
		throw (new Error('Unknown key type: ' + type));
	case 'DSA':
		if (type === 'public')
			return (readPkcs1DSAPublic(der));
		else if (type === 'private')
			return (readPkcs1DSAPrivate(der));
		throw (new Error('Unknown key type: ' + type));
	case 'EC':
	case 'ECDSA':
		if (type === 'private')
			return (readPkcs1ECDSAPrivate(der));
		else if (type === 'public')
			return (readPkcs1ECDSAPublic(der));
		throw (new Error('Unknown key type: ' + type));
	case 'EDDSA':
	case 'EdDSA':
		if (type === 'private')
			return (readPkcs1EdDSAPrivate(der));
		throw (new Error(type + ' keys not supported with EdDSA'));
	default:
		throw (new Error('Unknown key algo: ' + alg));
	***REMOVED***
***REMOVED***

function readPkcs1RSAPublic(der) ***REMOVED***
	// modulus and exponent
	var n = readMPInt(der, 'modulus');
	var e = readMPInt(der, 'exponent');

	// now, make the key
	var key = ***REMOVED***
		type: 'rsa',
		parts: [
			***REMOVED*** name: 'e', data: e ***REMOVED***,
			***REMOVED*** name: 'n', data: n ***REMOVED***
		]
	***REMOVED***;

	return (new Key(key));
***REMOVED***

function readPkcs1RSAPrivate(der) ***REMOVED***
	var version = readMPInt(der, 'version');
	assert.strictEqual(version[0], 0);

	// modulus then public exponent
	var n = readMPInt(der, 'modulus');
	var e = readMPInt(der, 'public exponent');
	var d = readMPInt(der, 'private exponent');
	var p = readMPInt(der, 'prime1');
	var q = readMPInt(der, 'prime2');
	var dmodp = readMPInt(der, 'exponent1');
	var dmodq = readMPInt(der, 'exponent2');
	var iqmp = readMPInt(der, 'iqmp');

	// now, make the key
	var key = ***REMOVED***
		type: 'rsa',
		parts: [
			***REMOVED*** name: 'n', data: n ***REMOVED***,
			***REMOVED*** name: 'e', data: e ***REMOVED***,
			***REMOVED*** name: 'd', data: d ***REMOVED***,
			***REMOVED*** name: 'iqmp', data: iqmp ***REMOVED***,
			***REMOVED*** name: 'p', data: p ***REMOVED***,
			***REMOVED*** name: 'q', data: q ***REMOVED***,
			***REMOVED*** name: 'dmodp', data: dmodp ***REMOVED***,
			***REMOVED*** name: 'dmodq', data: dmodq ***REMOVED***
		]
	***REMOVED***;

	return (new PrivateKey(key));
***REMOVED***

function readPkcs1DSAPrivate(der) ***REMOVED***
	var version = readMPInt(der, 'version');
	assert.strictEqual(version.readUInt8(0), 0);

	var p = readMPInt(der, 'p');
	var q = readMPInt(der, 'q');
	var g = readMPInt(der, 'g');
	var y = readMPInt(der, 'y');
	var x = readMPInt(der, 'x');

	// now, make the key
	var key = ***REMOVED***
		type: 'dsa',
		parts: [
			***REMOVED*** name: 'p', data: p ***REMOVED***,
			***REMOVED*** name: 'q', data: q ***REMOVED***,
			***REMOVED*** name: 'g', data: g ***REMOVED***,
			***REMOVED*** name: 'y', data: y ***REMOVED***,
			***REMOVED*** name: 'x', data: x ***REMOVED***
		]
	***REMOVED***;

	return (new PrivateKey(key));
***REMOVED***

function readPkcs1EdDSAPrivate(der) ***REMOVED***
	var version = readMPInt(der, 'version');
	assert.strictEqual(version.readUInt8(0), 1);

	// private key
	var k = der.readString(asn1.Ber.OctetString, true);

	der.readSequence(0xa0);
	var oid = der.readOID();
	assert.strictEqual(oid, '1.3.101.112', 'the ed25519 curve identifier');

	der.readSequence(0xa1);
	var A = utils.readBitString(der);

	var key = ***REMOVED***
		type: 'ed25519',
		parts: [
			***REMOVED*** name: 'A', data: utils.zeroPadToLength(A, 32) ***REMOVED***,
			***REMOVED*** name: 'k', data: k ***REMOVED***
		]
	***REMOVED***;

	return (new PrivateKey(key));
***REMOVED***

function readPkcs1DSAPublic(der) ***REMOVED***
	var y = readMPInt(der, 'y');
	var p = readMPInt(der, 'p');
	var q = readMPInt(der, 'q');
	var g = readMPInt(der, 'g');

	var key = ***REMOVED***
		type: 'dsa',
		parts: [
			***REMOVED*** name: 'y', data: y ***REMOVED***,
			***REMOVED*** name: 'p', data: p ***REMOVED***,
			***REMOVED*** name: 'q', data: q ***REMOVED***,
			***REMOVED*** name: 'g', data: g ***REMOVED***
		]
	***REMOVED***;

	return (new Key(key));
***REMOVED***

function readPkcs1ECDSAPublic(der) ***REMOVED***
	der.readSequence();

	var oid = der.readOID();
	assert.strictEqual(oid, '1.2.840.10045.2.1', 'must be ecPublicKey');

	var curveOid = der.readOID();

	var curve;
	var curves = Object.keys(algs.curves);
	for (var j = 0; j < curves.length; ++j) ***REMOVED***
		var c = curves[j];
		var cd = algs.curves[c];
		if (cd.pkcs8oid === curveOid) ***REMOVED***
			curve = c;
			break;
		***REMOVED***
	***REMOVED***
	assert.string(curve, 'a known ECDSA named curve');

	var Q = der.readString(asn1.Ber.BitString, true);
	Q = utils.ecNormalize(Q);

	var key = ***REMOVED***
		type: 'ecdsa',
		parts: [
			***REMOVED*** name: 'curve', data: Buffer.from(curve) ***REMOVED***,
			***REMOVED*** name: 'Q', data: Q ***REMOVED***
		]
	***REMOVED***;

	return (new Key(key));
***REMOVED***

function readPkcs1ECDSAPrivate(der) ***REMOVED***
	var version = readMPInt(der, 'version');
	assert.strictEqual(version.readUInt8(0), 1);

	// private key
	var d = der.readString(asn1.Ber.OctetString, true);

	der.readSequence(0xa0);
	var curve = readECDSACurve(der);
	assert.string(curve, 'a known elliptic curve');

	der.readSequence(0xa1);
	var Q = der.readString(asn1.Ber.BitString, true);
	Q = utils.ecNormalize(Q);

	var key = ***REMOVED***
		type: 'ecdsa',
		parts: [
			***REMOVED*** name: 'curve', data: Buffer.from(curve) ***REMOVED***,
			***REMOVED*** name: 'Q', data: Q ***REMOVED***,
			***REMOVED*** name: 'd', data: d ***REMOVED***
		]
	***REMOVED***;

	return (new PrivateKey(key));
***REMOVED***

function writePkcs1(der, key) ***REMOVED***
	der.startSequence();

	switch (key.type) ***REMOVED***
	case 'rsa':
		if (PrivateKey.isPrivateKey(key))
			writePkcs1RSAPrivate(der, key);
		else
			writePkcs1RSAPublic(der, key);
		break;
	case 'dsa':
		if (PrivateKey.isPrivateKey(key))
			writePkcs1DSAPrivate(der, key);
		else
			writePkcs1DSAPublic(der, key);
		break;
	case 'ecdsa':
		if (PrivateKey.isPrivateKey(key))
			writePkcs1ECDSAPrivate(der, key);
		else
			writePkcs1ECDSAPublic(der, key);
		break;
	case 'ed25519':
		if (PrivateKey.isPrivateKey(key))
			writePkcs1EdDSAPrivate(der, key);
		else
			writePkcs1EdDSAPublic(der, key);
		break;
	default:
		throw (new Error('Unknown key algo: ' + key.type));
	***REMOVED***

	der.endSequence();
***REMOVED***

function writePkcs1RSAPublic(der, key) ***REMOVED***
	der.writeBuffer(key.part.n.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.e.data, asn1.Ber.Integer);
***REMOVED***

function writePkcs1RSAPrivate(der, key) ***REMOVED***
	var ver = Buffer.from([0]);
	der.writeBuffer(ver, asn1.Ber.Integer);

	der.writeBuffer(key.part.n.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.e.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.d.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.p.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.q.data, asn1.Ber.Integer);
	if (!key.part.dmodp || !key.part.dmodq)
		utils.addRSAMissing(key);
	der.writeBuffer(key.part.dmodp.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.dmodq.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.iqmp.data, asn1.Ber.Integer);
***REMOVED***

function writePkcs1DSAPrivate(der, key) ***REMOVED***
	var ver = Buffer.from([0]);
	der.writeBuffer(ver, asn1.Ber.Integer);

	der.writeBuffer(key.part.p.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.q.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.g.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.y.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.x.data, asn1.Ber.Integer);
***REMOVED***

function writePkcs1DSAPublic(der, key) ***REMOVED***
	der.writeBuffer(key.part.y.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.p.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.q.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.g.data, asn1.Ber.Integer);
***REMOVED***

function writePkcs1ECDSAPublic(der, key) ***REMOVED***
	der.startSequence();

	der.writeOID('1.2.840.10045.2.1'); /* ecPublicKey */
	var curve = key.part.curve.data.toString();
	var curveOid = algs.curves[curve].pkcs8oid;
	assert.string(curveOid, 'a known ECDSA named curve');
	der.writeOID(curveOid);

	der.endSequence();

	var Q = utils.ecNormalize(key.part.Q.data, true);
	der.writeBuffer(Q, asn1.Ber.BitString);
***REMOVED***

function writePkcs1ECDSAPrivate(der, key) ***REMOVED***
	var ver = Buffer.from([1]);
	der.writeBuffer(ver, asn1.Ber.Integer);

	der.writeBuffer(key.part.d.data, asn1.Ber.OctetString);

	der.startSequence(0xa0);
	var curve = key.part.curve.data.toString();
	var curveOid = algs.curves[curve].pkcs8oid;
	assert.string(curveOid, 'a known ECDSA named curve');
	der.writeOID(curveOid);
	der.endSequence();

	der.startSequence(0xa1);
	var Q = utils.ecNormalize(key.part.Q.data, true);
	der.writeBuffer(Q, asn1.Ber.BitString);
	der.endSequence();
***REMOVED***

function writePkcs1EdDSAPrivate(der, key) ***REMOVED***
	var ver = Buffer.from([1]);
	der.writeBuffer(ver, asn1.Ber.Integer);

	der.writeBuffer(key.part.k.data, asn1.Ber.OctetString);

	der.startSequence(0xa0);
	der.writeOID('1.3.101.112');
	der.endSequence();

	der.startSequence(0xa1);
	utils.writeBitString(der, key.part.A.data);
	der.endSequence();
***REMOVED***

function writePkcs1EdDSAPublic(der, key) ***REMOVED***
	throw (new Error('Public keys are not supported for EdDSA PKCS#1'));
***REMOVED***
