// sshpk/lib/formats/pkcs8.js
// Copyright 2015 Joyent, Inc.

module.exports = ***REMOVED***
	read: read,
	readPkcs8: readPkcs8,
	write: write,
	writePkcs8: writePkcs8,

	readECDSACurve: readECDSACurve,
	writeECDSACurve: writeECDSACurve
***REMOVED***;

var assert = require('assert-plus');
var asn1 = require('asn1');
var Buffer = require('safer-buffer').Buffer;
var algs = require('../algs');
var utils = require('../utils');
var Key = require('../key');
var PrivateKey = require('../private-key');
var pem = require('./pem');

function read(buf, options) ***REMOVED***
	return (pem.read(buf, options, 'pkcs8'));
***REMOVED***

function write(key, options) ***REMOVED***
	return (pem.write(key, options, 'pkcs8'));
***REMOVED***

/* Helper to read in a single mpint */
function readMPInt(der, nm) ***REMOVED***
	assert.strictEqual(der.peek(), asn1.Ber.Integer,
	    nm + ' is not an Integer');
	return (utils.mpNormalize(der.readString(asn1.Ber.Integer, true)));
***REMOVED***

function readPkcs8(alg, type, der) ***REMOVED***
	/* Private keys in pkcs#8 format have a weird extra int */
	if (der.peek() === asn1.Ber.Integer) ***REMOVED***
		assert.strictEqual(type, 'private',
		    'unexpected Integer at start of public key');
		der.readString(asn1.Ber.Integer, true);
	***REMOVED***

	der.readSequence();
	var next = der.offset + der.length;

	var oid = der.readOID();
	switch (oid) ***REMOVED***
	case '1.2.840.113549.1.1.1':
		der._offset = next;
		if (type === 'public')
			return (readPkcs8RSAPublic(der));
		else
			return (readPkcs8RSAPrivate(der));
	case '1.2.840.10040.4.1':
		if (type === 'public')
			return (readPkcs8DSAPublic(der));
		else
			return (readPkcs8DSAPrivate(der));
	case '1.2.840.10045.2.1':
		if (type === 'public')
			return (readPkcs8ECDSAPublic(der));
		else
			return (readPkcs8ECDSAPrivate(der));
	case '1.3.101.112':
		if (type === 'public') ***REMOVED***
			return (readPkcs8EdDSAPublic(der));
		***REMOVED*** else ***REMOVED***
			return (readPkcs8EdDSAPrivate(der));
		***REMOVED***
	case '1.3.101.110':
		if (type === 'public') ***REMOVED***
			return (readPkcs8X25519Public(der));
		***REMOVED*** else ***REMOVED***
			return (readPkcs8X25519Private(der));
		***REMOVED***
	default:
		throw (new Error('Unknown key type OID ' + oid));
	***REMOVED***
***REMOVED***

function readPkcs8RSAPublic(der) ***REMOVED***
	// bit string sequence
	der.readSequence(asn1.Ber.BitString);
	der.readByte();
	der.readSequence();

	// modulus
	var n = readMPInt(der, 'modulus');
	var e = readMPInt(der, 'exponent');

	// now, make the key
	var key = ***REMOVED***
		type: 'rsa',
		source: der.originalInput,
		parts: [
			***REMOVED*** name: 'e', data: e ***REMOVED***,
			***REMOVED*** name: 'n', data: n ***REMOVED***
		]
	***REMOVED***;

	return (new Key(key));
***REMOVED***

function readPkcs8RSAPrivate(der) ***REMOVED***
	der.readSequence(asn1.Ber.OctetString);
	der.readSequence();

	var ver = readMPInt(der, 'version');
	assert.equal(ver[0], 0x0, 'unknown RSA private key version');

	// modulus then public exponent
	var n = readMPInt(der, 'modulus');
	var e = readMPInt(der, 'public exponent');
	var d = readMPInt(der, 'private exponent');
	var p = readMPInt(der, 'prime1');
	var q = readMPInt(der, 'prime2');
	var dmodp = readMPInt(der, 'exponent1');
	var dmodq = readMPInt(der, 'exponent2');
	var iqmp = readMPInt(der, 'iqmp');

	// now, make the key
	var key = ***REMOVED***
		type: 'rsa',
		parts: [
			***REMOVED*** name: 'n', data: n ***REMOVED***,
			***REMOVED*** name: 'e', data: e ***REMOVED***,
			***REMOVED*** name: 'd', data: d ***REMOVED***,
			***REMOVED*** name: 'iqmp', data: iqmp ***REMOVED***,
			***REMOVED*** name: 'p', data: p ***REMOVED***,
			***REMOVED*** name: 'q', data: q ***REMOVED***,
			***REMOVED*** name: 'dmodp', data: dmodp ***REMOVED***,
			***REMOVED*** name: 'dmodq', data: dmodq ***REMOVED***
		]
	***REMOVED***;

	return (new PrivateKey(key));
***REMOVED***

function readPkcs8DSAPublic(der) ***REMOVED***
	der.readSequence();

	var p = readMPInt(der, 'p');
	var q = readMPInt(der, 'q');
	var g = readMPInt(der, 'g');

	// bit string sequence
	der.readSequence(asn1.Ber.BitString);
	der.readByte();

	var y = readMPInt(der, 'y');

	// now, make the key
	var key = ***REMOVED***
		type: 'dsa',
		parts: [
			***REMOVED*** name: 'p', data: p ***REMOVED***,
			***REMOVED*** name: 'q', data: q ***REMOVED***,
			***REMOVED*** name: 'g', data: g ***REMOVED***,
			***REMOVED*** name: 'y', data: y ***REMOVED***
		]
	***REMOVED***;

	return (new Key(key));
***REMOVED***

function readPkcs8DSAPrivate(der) ***REMOVED***
	der.readSequence();

	var p = readMPInt(der, 'p');
	var q = readMPInt(der, 'q');
	var g = readMPInt(der, 'g');

	der.readSequence(asn1.Ber.OctetString);
	var x = readMPInt(der, 'x');

	/* The pkcs#8 format does not include the public key */
	var y = utils.calculateDSAPublic(g, p, x);

	var key = ***REMOVED***
		type: 'dsa',
		parts: [
			***REMOVED*** name: 'p', data: p ***REMOVED***,
			***REMOVED*** name: 'q', data: q ***REMOVED***,
			***REMOVED*** name: 'g', data: g ***REMOVED***,
			***REMOVED*** name: 'y', data: y ***REMOVED***,
			***REMOVED*** name: 'x', data: x ***REMOVED***
		]
	***REMOVED***;

	return (new PrivateKey(key));
***REMOVED***

function readECDSACurve(der) ***REMOVED***
	var curveName, curveNames;
	var j, c, cd;

	if (der.peek() === asn1.Ber.OID) ***REMOVED***
		var oid = der.readOID();

		curveNames = Object.keys(algs.curves);
		for (j = 0; j < curveNames.length; ++j) ***REMOVED***
			c = curveNames[j];
			cd = algs.curves[c];
			if (cd.pkcs8oid === oid) ***REMOVED***
				curveName = c;
				break;
			***REMOVED***
		***REMOVED***

	***REMOVED*** else ***REMOVED***
		// ECParameters sequence
		der.readSequence();
		var version = der.readString(asn1.Ber.Integer, true);
		assert.strictEqual(version[0], 1, 'ECDSA key not version 1');

		var curve = ***REMOVED******REMOVED***;

		// FieldID sequence
		der.readSequence();
		var fieldTypeOid = der.readOID();
		assert.strictEqual(fieldTypeOid, '1.2.840.10045.1.1',
		    'ECDSA key is not from a prime-field');
		var p = curve.p = utils.mpNormalize(
		    der.readString(asn1.Ber.Integer, true));
		/*
		 * p always starts with a 1 bit, so count the zeros to get its
		 * real size.
		 */
		curve.size = p.length * 8 - utils.countZeros(p);

		// Curve sequence
		der.readSequence();
		curve.a = utils.mpNormalize(
		    der.readString(asn1.Ber.OctetString, true));
		curve.b = utils.mpNormalize(
		    der.readString(asn1.Ber.OctetString, true));
		if (der.peek() === asn1.Ber.BitString)
			curve.s = der.readString(asn1.Ber.BitString, true);

		// Combined Gx and Gy
		curve.G = der.readString(asn1.Ber.OctetString, true);
		assert.strictEqual(curve.G[0], 0x4,
		    'uncompressed G is required');

		curve.n = utils.mpNormalize(
		    der.readString(asn1.Ber.Integer, true));
		curve.h = utils.mpNormalize(
		    der.readString(asn1.Ber.Integer, true));
		assert.strictEqual(curve.h[0], 0x1, 'a cofactor=1 curve is ' +
		    'required');

		curveNames = Object.keys(algs.curves);
		var ks = Object.keys(curve);
		for (j = 0; j < curveNames.length; ++j) ***REMOVED***
			c = curveNames[j];
			cd = algs.curves[c];
			var equal = true;
			for (var i = 0; i < ks.length; ++i) ***REMOVED***
				var k = ks[i];
				if (cd[k] === undefined)
					continue;
				if (typeof (cd[k]) === 'object' &&
				    cd[k].equals !== undefined) ***REMOVED***
					if (!cd[k].equals(curve[k])) ***REMOVED***
						equal = false;
						break;
					***REMOVED***
				***REMOVED*** else if (Buffer.isBuffer(cd[k])) ***REMOVED***
					if (cd[k].toString('binary')
					    !== curve[k].toString('binary')) ***REMOVED***
						equal = false;
						break;
					***REMOVED***
				***REMOVED*** else ***REMOVED***
					if (cd[k] !== curve[k]) ***REMOVED***
						equal = false;
						break;
					***REMOVED***
				***REMOVED***
			***REMOVED***
			if (equal) ***REMOVED***
				curveName = c;
				break;
			***REMOVED***
		***REMOVED***
	***REMOVED***
	return (curveName);
***REMOVED***

function readPkcs8ECDSAPrivate(der) ***REMOVED***
	var curveName = readECDSACurve(der);
	assert.string(curveName, 'a known elliptic curve');

	der.readSequence(asn1.Ber.OctetString);
	der.readSequence();

	var version = readMPInt(der, 'version');
	assert.equal(version[0], 1, 'unknown version of ECDSA key');

	var d = der.readString(asn1.Ber.OctetString, true);
	der.readSequence(0xa1);

	var Q = der.readString(asn1.Ber.BitString, true);
	Q = utils.ecNormalize(Q);

	var key = ***REMOVED***
		type: 'ecdsa',
		parts: [
			***REMOVED*** name: 'curve', data: Buffer.from(curveName) ***REMOVED***,
			***REMOVED*** name: 'Q', data: Q ***REMOVED***,
			***REMOVED*** name: 'd', data: d ***REMOVED***
		]
	***REMOVED***;

	return (new PrivateKey(key));
***REMOVED***

function readPkcs8ECDSAPublic(der) ***REMOVED***
	var curveName = readECDSACurve(der);
	assert.string(curveName, 'a known elliptic curve');

	var Q = der.readString(asn1.Ber.BitString, true);
	Q = utils.ecNormalize(Q);

	var key = ***REMOVED***
		type: 'ecdsa',
		parts: [
			***REMOVED*** name: 'curve', data: Buffer.from(curveName) ***REMOVED***,
			***REMOVED*** name: 'Q', data: Q ***REMOVED***
		]
	***REMOVED***;

	return (new Key(key));
***REMOVED***

function readPkcs8EdDSAPublic(der) ***REMOVED***
	if (der.peek() === 0x00)
		der.readByte();

	var A = utils.readBitString(der);

	var key = ***REMOVED***
		type: 'ed25519',
		parts: [
			***REMOVED*** name: 'A', data: utils.zeroPadToLength(A, 32) ***REMOVED***
		]
	***REMOVED***;

	return (new Key(key));
***REMOVED***

function readPkcs8X25519Public(der) ***REMOVED***
	var A = utils.readBitString(der);

	var key = ***REMOVED***
		type: 'curve25519',
		parts: [
			***REMOVED*** name: 'A', data: utils.zeroPadToLength(A, 32) ***REMOVED***
		]
	***REMOVED***;

	return (new Key(key));
***REMOVED***

function readPkcs8EdDSAPrivate(der) ***REMOVED***
	if (der.peek() === 0x00)
		der.readByte();

	der.readSequence(asn1.Ber.OctetString);
	var k = der.readString(asn1.Ber.OctetString, true);
	k = utils.zeroPadToLength(k, 32);

	var A;
	if (der.peek() === asn1.Ber.BitString) ***REMOVED***
		A = utils.readBitString(der);
		A = utils.zeroPadToLength(A, 32);
	***REMOVED*** else ***REMOVED***
		A = utils.calculateED25519Public(k);
	***REMOVED***

	var key = ***REMOVED***
		type: 'ed25519',
		parts: [
			***REMOVED*** name: 'A', data: utils.zeroPadToLength(A, 32) ***REMOVED***,
			***REMOVED*** name: 'k', data: utils.zeroPadToLength(k, 32) ***REMOVED***
		]
	***REMOVED***;

	return (new PrivateKey(key));
***REMOVED***

function readPkcs8X25519Private(der) ***REMOVED***
	if (der.peek() === 0x00)
		der.readByte();

	der.readSequence(asn1.Ber.OctetString);
	var k = der.readString(asn1.Ber.OctetString, true);
	k = utils.zeroPadToLength(k, 32);

	var A = utils.calculateX25519Public(k);

	var key = ***REMOVED***
		type: 'curve25519',
		parts: [
			***REMOVED*** name: 'A', data: utils.zeroPadToLength(A, 32) ***REMOVED***,
			***REMOVED*** name: 'k', data: utils.zeroPadToLength(k, 32) ***REMOVED***
		]
	***REMOVED***;

	return (new PrivateKey(key));
***REMOVED***

function writePkcs8(der, key) ***REMOVED***
	der.startSequence();

	if (PrivateKey.isPrivateKey(key)) ***REMOVED***
		var sillyInt = Buffer.from([0]);
		der.writeBuffer(sillyInt, asn1.Ber.Integer);
	***REMOVED***

	der.startSequence();
	switch (key.type) ***REMOVED***
	case 'rsa':
		der.writeOID('1.2.840.113549.1.1.1');
		if (PrivateKey.isPrivateKey(key))
			writePkcs8RSAPrivate(key, der);
		else
			writePkcs8RSAPublic(key, der);
		break;
	case 'dsa':
		der.writeOID('1.2.840.10040.4.1');
		if (PrivateKey.isPrivateKey(key))
			writePkcs8DSAPrivate(key, der);
		else
			writePkcs8DSAPublic(key, der);
		break;
	case 'ecdsa':
		der.writeOID('1.2.840.10045.2.1');
		if (PrivateKey.isPrivateKey(key))
			writePkcs8ECDSAPrivate(key, der);
		else
			writePkcs8ECDSAPublic(key, der);
		break;
	case 'ed25519':
		der.writeOID('1.3.101.112');
		if (PrivateKey.isPrivateKey(key))
			throw (new Error('Ed25519 private keys in pkcs8 ' +
			    'format are not supported'));
		writePkcs8EdDSAPublic(key, der);
		break;
	default:
		throw (new Error('Unsupported key type: ' + key.type));
	***REMOVED***

	der.endSequence();
***REMOVED***

function writePkcs8RSAPrivate(key, der) ***REMOVED***
	der.writeNull();
	der.endSequence();

	der.startSequence(asn1.Ber.OctetString);
	der.startSequence();

	var version = Buffer.from([0]);
	der.writeBuffer(version, asn1.Ber.Integer);

	der.writeBuffer(key.part.n.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.e.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.d.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.p.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.q.data, asn1.Ber.Integer);
	if (!key.part.dmodp || !key.part.dmodq)
		utils.addRSAMissing(key);
	der.writeBuffer(key.part.dmodp.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.dmodq.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.iqmp.data, asn1.Ber.Integer);

	der.endSequence();
	der.endSequence();
***REMOVED***

function writePkcs8RSAPublic(key, der) ***REMOVED***
	der.writeNull();
	der.endSequence();

	der.startSequence(asn1.Ber.BitString);
	der.writeByte(0x00);

	der.startSequence();
	der.writeBuffer(key.part.n.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.e.data, asn1.Ber.Integer);
	der.endSequence();

	der.endSequence();
***REMOVED***

function writePkcs8DSAPrivate(key, der) ***REMOVED***
	der.startSequence();
	der.writeBuffer(key.part.p.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.q.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.g.data, asn1.Ber.Integer);
	der.endSequence();

	der.endSequence();

	der.startSequence(asn1.Ber.OctetString);
	der.writeBuffer(key.part.x.data, asn1.Ber.Integer);
	der.endSequence();
***REMOVED***

function writePkcs8DSAPublic(key, der) ***REMOVED***
	der.startSequence();
	der.writeBuffer(key.part.p.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.q.data, asn1.Ber.Integer);
	der.writeBuffer(key.part.g.data, asn1.Ber.Integer);
	der.endSequence();
	der.endSequence();

	der.startSequence(asn1.Ber.BitString);
	der.writeByte(0x00);
	der.writeBuffer(key.part.y.data, asn1.Ber.Integer);
	der.endSequence();
***REMOVED***

function writeECDSACurve(key, der) ***REMOVED***
	var curve = algs.curves[key.curve];
	if (curve.pkcs8oid) ***REMOVED***
		/* This one has a name in pkcs#8, so just write the oid */
		der.writeOID(curve.pkcs8oid);

	***REMOVED*** else ***REMOVED***
		// ECParameters sequence
		der.startSequence();

		var version = Buffer.from([1]);
		der.writeBuffer(version, asn1.Ber.Integer);

		// FieldID sequence
		der.startSequence();
		der.writeOID('1.2.840.10045.1.1'); // prime-field
		der.writeBuffer(curve.p, asn1.Ber.Integer);
		der.endSequence();

		// Curve sequence
		der.startSequence();
		var a = curve.p;
		if (a[0] === 0x0)
			a = a.slice(1);
		der.writeBuffer(a, asn1.Ber.OctetString);
		der.writeBuffer(curve.b, asn1.Ber.OctetString);
		der.writeBuffer(curve.s, asn1.Ber.BitString);
		der.endSequence();

		der.writeBuffer(curve.G, asn1.Ber.OctetString);
		der.writeBuffer(curve.n, asn1.Ber.Integer);
		var h = curve.h;
		if (!h) ***REMOVED***
			h = Buffer.from([1]);
		***REMOVED***
		der.writeBuffer(h, asn1.Ber.Integer);

		// ECParameters
		der.endSequence();
	***REMOVED***
***REMOVED***

function writePkcs8ECDSAPublic(key, der) ***REMOVED***
	writeECDSACurve(key, der);
	der.endSequence();

	var Q = utils.ecNormalize(key.part.Q.data, true);
	der.writeBuffer(Q, asn1.Ber.BitString);
***REMOVED***

function writePkcs8ECDSAPrivate(key, der) ***REMOVED***
	writeECDSACurve(key, der);
	der.endSequence();

	der.startSequence(asn1.Ber.OctetString);
	der.startSequence();

	var version = Buffer.from([1]);
	der.writeBuffer(version, asn1.Ber.Integer);

	der.writeBuffer(key.part.d.data, asn1.Ber.OctetString);

	der.startSequence(0xa1);
	var Q = utils.ecNormalize(key.part.Q.data, true);
	der.writeBuffer(Q, asn1.Ber.BitString);
	der.endSequence();

	der.endSequence();
	der.endSequence();
***REMOVED***

function writePkcs8EdDSAPublic(key, der) ***REMOVED***
	der.endSequence();

	utils.writeBitString(der, key.part.A.data);
***REMOVED***

function writePkcs8EdDSAPrivate(key, der) ***REMOVED***
	der.endSequence();

	var k = utils.mpNormalize(key.part.k.data, true);
	der.startSequence(asn1.Ber.OctetString);
	der.writeBuffer(k, asn1.Ber.OctetString);
	der.endSequence();
***REMOVED***
