// request-promise-core/lib/errors.js
'use strict';


function RequestError(cause, options, response) ***REMOVED***

    this.name = 'RequestError';
    this.message = String(cause);
    this.cause = cause;
    this.error = cause; // legacy attribute
    this.options = options;
    this.response = response;

    if (Error.captureStackTrace) ***REMOVED*** // required for non-V8 environments
        Error.captureStackTrace(this);
    ***REMOVED***

***REMOVED***
RequestError.prototype = Object.create(Error.prototype);
RequestError.prototype.constructor = RequestError;


function StatusCodeError(statusCode, body, options, response) ***REMOVED***

    this.name = 'StatusCodeError';
    this.statusCode = statusCode;
    this.message = statusCode + ' - ' + (JSON && JSON.stringify ? JSON.stringify(body) : body);
    this.error = body; // legacy attribute
    this.options = options;
    this.response = response;

    if (Error.captureStackTrace) ***REMOVED*** // required for non-V8 environments
        Error.captureStackTrace(this);
    ***REMOVED***

***REMOVED***
StatusCodeError.prototype = Object.create(Error.prototype);
StatusCodeError.prototype.constructor = StatusCodeError;


function TransformError(cause, options, response) ***REMOVED***

    this.name = 'TransformError';
    this.message = String(cause);
    this.cause = cause;
    this.error = cause; // legacy attribute
    this.options = options;
    this.response = response;

    if (Error.captureStackTrace) ***REMOVED*** // required for non-V8 environments
        Error.captureStackTrace(this);
    ***REMOVED***

***REMOVED***
TransformError.prototype = Object.create(Error.prototype);
TransformError.prototype.constructor = TransformError;


module.exports = ***REMOVED***
    RequestError: RequestError,
    StatusCodeError: StatusCodeError,
    TransformError: TransformError
***REMOVED***;
