// basic-auth/index.js
/*!
 * basic-auth
 * Copyright(c) 2013 TJ Holowaychuk
 * Copyright(c) 2014 Jonathan Ong
 * Copyright(c) 2015-2016 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var Buffer = require('safe-buffer').Buffer

/**
 * Module exports.
 * @public
 */

module.exports = auth
module.exports.parse = parse

/**
 * RegExp for basic auth credentials
 *
 * credentials = auth-scheme 1*SP token68
 * auth-scheme = "Basic" ; case insensitive
 * token68     = 1*( ALPHA / DIGIT / "-" / "." / "_" / "~" / "+" / "/" ) *"="
 * @private
 */

var CREDENTIALS_REGEXP = /^ *(?:[Bb][Aa][Ss][Ii][Cc]) +([A-Za-z0-9._~+/-]+=*) *$/

/**
 * RegExp for basic auth user/pass
 *
 * user-pass   = userid ":" password
 * userid      = *<TEXT excluding ":">
 * password    = *TEXT
 * @private
 */

var USER_PASS_REGEXP = /^([^:]*):(.*)$/

/**
 * Parse the Authorization header field of a request.
 *
 * @param ***REMOVED***object***REMOVED*** req
 * @return ***REMOVED***object***REMOVED*** with .name and .pass
 * @public
 */

function auth (req) ***REMOVED***
  if (!req) ***REMOVED***
    throw new TypeError('argument req is required')
  ***REMOVED***

  if (typeof req !== 'object') ***REMOVED***
    throw new TypeError('argument req is required to be an object')
  ***REMOVED***

  // get header
  var header = getAuthorization(req)

  // parse header
  return parse(header)
***REMOVED***

/**
 * Decode base64 string.
 * @private
 */

function decodeBase64 (str) ***REMOVED***
  return Buffer.from(str, 'base64').toString()
***REMOVED***

/**
 * Get the Authorization header from request object.
 * @private
 */

function getAuthorization (req) ***REMOVED***
  if (!req.headers || typeof req.headers !== 'object') ***REMOVED***
    throw new TypeError('argument req is required to have headers property')
  ***REMOVED***

  return req.headers.authorization
***REMOVED***

/**
 * Parse basic auth to object.
 *
 * @param ***REMOVED***string***REMOVED*** string
 * @return ***REMOVED***object***REMOVED***
 * @public
 */

function parse (string) ***REMOVED***
  if (typeof string !== 'string') ***REMOVED***
    return undefined
  ***REMOVED***

  // parse header
  var match = CREDENTIALS_REGEXP.exec(string)

  if (!match) ***REMOVED***
    return undefined
  ***REMOVED***

  // decode user pass
  var userPass = USER_PASS_REGEXP.exec(decodeBase64(match[1]))

  if (!userPass) ***REMOVED***
    return undefined
  ***REMOVED***

  // return credentials object
  return new Credentials(userPass[1], userPass[2])
***REMOVED***

/**
 * Object to represent user credentials.
 * @private
 */

function Credentials (name, pass) ***REMOVED***
  this.name = name
  this.pass = pass
***REMOVED***
