// mongoose/lib/schema/decimal128.js
/* eslint no-empty: 1 */

/*!
 * Module dependencies.
 */

var SchemaType = require('../schematype');
var CastError = SchemaType.CastError;
var Decimal128Type = require('../types/decimal128');
var utils = require('../utils');
var Document;

/**
 * Decimal128 SchemaType constructor.
 *
 * @param ***REMOVED***String***REMOVED*** key
 * @param ***REMOVED***Object***REMOVED*** options
 * @inherits SchemaType
 * @api public
 */

function Decimal128(key, options) ***REMOVED***
  SchemaType.call(this, key, options, 'Decimal128');
***REMOVED***

/**
 * This schema type's name, to defend against minifiers that mangle
 * function names.
 *
 * @api public
 */
Decimal128.schemaName = 'Decimal128';

/*!
 * Inherits from SchemaType.
 */
Decimal128.prototype = Object.create(SchemaType.prototype);
Decimal128.prototype.constructor = Decimal128;

/**
 * Check if the given value satisfies a required validator.
 *
 * @param ***REMOVED***Any***REMOVED*** value
 * @param ***REMOVED***Document***REMOVED*** doc
 * @return ***REMOVED***Boolean***REMOVED***
 * @api public
 */

Decimal128.prototype.checkRequired = function checkRequired(value, doc) ***REMOVED***
  if (SchemaType._isRef(this, value, doc, true)) ***REMOVED***
    return !!value;
  ***REMOVED***
  return value instanceof Decimal128Type;
***REMOVED***;

/**
 * Casts to Decimal128
 *
 * @param ***REMOVED***Object***REMOVED*** value
 * @param ***REMOVED***Object***REMOVED*** doc
 * @param ***REMOVED***Boolean***REMOVED*** init whether this is an initialization cast
 * @api private
 */

Decimal128.prototype.cast = function(value, doc, init) ***REMOVED***
  if (SchemaType._isRef(this, value, doc, init)) ***REMOVED***
    // wait! we may need to cast this to a document

    if (value === null || value === undefined) ***REMOVED***
      return value;
    ***REMOVED***

    // lazy load
    Document || (Document = require('./../document'));

    if (value instanceof Document) ***REMOVED***
      value.$__.wasPopulated = true;
      return value;
    ***REMOVED***

    // setting a populated path
    if (value instanceof Decimal128Type) ***REMOVED***
      return value;
    ***REMOVED*** else if (Buffer.isBuffer(value) || !utils.isObject(value)) ***REMOVED***
      throw new CastError('Decimal128', value, this.path);
    ***REMOVED***

    // Handle the case where user directly sets a populated
    // path to a plain object; cast to the Model used in
    // the population query.
    var path = doc.$__fullPath(this.path);
    var owner = doc.ownerDocument ? doc.ownerDocument() : doc;
    var pop = owner.populated(path, true);
    var ret = value;
    if (!doc.$__.populated ||
        !doc.$__.populated[path] ||
        !doc.$__.populated[path].options ||
        !doc.$__.populated[path].options.options ||
        !doc.$__.populated[path].options.options.lean) ***REMOVED***
      ret = new pop.options.model(value);
      ret.$__.wasPopulated = true;
    ***REMOVED***

    return ret;
  ***REMOVED***

  if (value == null) ***REMOVED***
    return value;
  ***REMOVED***

  if (typeof value === 'object' && typeof value.$numberDecimal === 'string') ***REMOVED***
    return Decimal128Type.fromString(value.$numberDecimal);
  ***REMOVED***

  if (value instanceof Decimal128Type) ***REMOVED***
    return value;
  ***REMOVED***

  if (typeof value === 'string') ***REMOVED***
    return Decimal128Type.fromString(value);
  ***REMOVED***

  if (Buffer.isBuffer(value)) ***REMOVED***
    return new Decimal128Type(value);
  ***REMOVED***

  if (typeof value === 'number') ***REMOVED***
    return Decimal128Type.fromString(String(value));
  ***REMOVED***

  if (typeof value.valueOf === 'function' && typeof value.valueOf() === 'string') ***REMOVED***
    return Decimal128Type.fromString(value.valueOf());
  ***REMOVED***

  throw new CastError('Decimal128', value, this.path);
***REMOVED***;

/*!
 * ignore
 */

function handleSingle(val) ***REMOVED***
  return this.cast(val);
***REMOVED***

Decimal128.prototype.$conditionalHandlers =
    utils.options(SchemaType.prototype.$conditionalHandlers, ***REMOVED***
      $gt: handleSingle,
      $gte: handleSingle,
      $lt: handleSingle,
      $lte: handleSingle
    ***REMOVED***);

/*!
 * Module exports.
 */

module.exports = Decimal128;
