// mongoose/lib/types/subdocument.js
'use strict';

var Document = require('../document');
var internalToObjectOptions = require('../options').internalToObjectOptions;
var utils = require('../utils');

module.exports = Subdocument;

/**
 * Subdocument constructor.
 *
 * @inherits Document
 * @api private
 */

function Subdocument(value, fields, parent, skipId, options) ***REMOVED***
  this.$isSingleNested = true;
  Document.call(this, value, fields, skipId, options);
***REMOVED***

Subdocument.prototype = Object.create(Document.prototype);

Subdocument.prototype.toBSON = function() ***REMOVED***
  return this.toObject(internalToObjectOptions);
***REMOVED***;

/**
 * Used as a stub for middleware
 *
 * ####NOTE:
 *
 * _This is a no-op. Does not actually save the doc to the db._
 *
 * @param ***REMOVED***Function***REMOVED*** [fn]
 * @return ***REMOVED***Promise***REMOVED*** resolved Promise
 * @api private
 */

Subdocument.prototype.save = function(options, fn) ***REMOVED***
  if (typeof options === 'function') ***REMOVED***
    fn = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  options = options || ***REMOVED******REMOVED***;

  if (!options.suppressWarning) ***REMOVED***
    console.warn('mongoose: calling `save()` on a subdoc does **not** save ' +
      'the document to MongoDB, it only runs save middleware. ' +
      'Use `subdoc.save(***REMOVED*** suppressWarning: true ***REMOVED***)` to hide this warning ' +
      'if you\'re sure this behavior is right for your app.');
  ***REMOVED***

  return utils.promiseOrCallback(fn, cb => ***REMOVED***
    this.$__save(cb);
  ***REMOVED***);
***REMOVED***;

/**
 * Used as a stub for middleware
 *
 * ####NOTE:
 *
 * _This is a no-op. Does not actually save the doc to the db._
 *
 * @param ***REMOVED***Function***REMOVED*** [fn]
 * @method $__save
 * @api private
 */

Subdocument.prototype.$__save = function(fn) ***REMOVED***
  return utils.immediate(() => fn(null, this));
***REMOVED***;

Subdocument.prototype.$isValid = function(path) ***REMOVED***
  if (this.$parent && this.$basePath) ***REMOVED***
    return this.$parent.$isValid([this.$basePath, path].join('.'));
  ***REMOVED***
  return Document.prototype.$isValid.call(this, path);
***REMOVED***;

Subdocument.prototype.markModified = function(path) ***REMOVED***
  Document.prototype.markModified.call(this, path);

  if (this.$parent && this.$basePath) ***REMOVED***
    if (this.$parent.isDirectModified(this.$basePath)) ***REMOVED***
      return;
    ***REMOVED***
    this.$parent.markModified([this.$basePath, path].join('.'), this);
  ***REMOVED***
***REMOVED***;

Subdocument.prototype.$markValid = function(path) ***REMOVED***
  Document.prototype.$markValid.call(this, path);
  if (this.$parent && this.$basePath) ***REMOVED***
    this.$parent.$markValid([this.$basePath, path].join('.'));
  ***REMOVED***
***REMOVED***;

/*!
 * ignore
 */

Subdocument.prototype.invalidate = function(path, err, val) ***REMOVED***
  // Hack: array subdocuments' validationError is equal to the owner doc's,
  // so validating an array subdoc gives the top-level doc back. Temporary
  // workaround for #5208 so we don't have circular errors.
  if (err !== this.ownerDocument().$__.validationError) ***REMOVED***
    Document.prototype.invalidate.call(this, path, err, val);
  ***REMOVED***

  if (this.$parent && this.$basePath) ***REMOVED***
    this.$parent.invalidate([this.$basePath, path].join('.'), err, val);
  ***REMOVED*** else if (err.kind === 'cast' || err.name === 'CastError') ***REMOVED***
    throw err;
  ***REMOVED***
***REMOVED***;

/*!
 * ignore
 */

Subdocument.prototype.$ignore = function(path) ***REMOVED***
  Document.prototype.$ignore.call(this, path);
  if (this.$parent && this.$basePath) ***REMOVED***
    this.$parent.$ignore([this.$basePath, path].join('.'));
  ***REMOVED***
***REMOVED***;

/**
 * Returns the top level document of this sub-document.
 *
 * @return ***REMOVED***Document***REMOVED***
 */

Subdocument.prototype.ownerDocument = function() ***REMOVED***
  if (this.$__.ownerDocument) ***REMOVED***
    return this.$__.ownerDocument;
  ***REMOVED***

  var parent = this.$parent;
  if (!parent) ***REMOVED***
    return this;
  ***REMOVED***

  while (parent.$parent || parent.__parent) ***REMOVED***
    parent = parent.$parent || parent.__parent;
  ***REMOVED***
  this.$__.ownerDocument = parent;
  return this.$__.ownerDocument;
***REMOVED***;

/**
 * Returns this sub-documents parent document.
 *
 * @api public
 */

Subdocument.prototype.parent = function() ***REMOVED***
  return this.$parent;
***REMOVED***;

/*!
 * no-op for hooks
 */

Subdocument.prototype.$__remove = function(cb) ***REMOVED***
  return cb(null, this);
***REMOVED***;

/**
 * Null-out this subdoc
 *
 * @param ***REMOVED***Object***REMOVED*** [options]
 * @param ***REMOVED***Function***REMOVED*** [callback] optional callback for compatibility with Document.prototype.remove
 */

Subdocument.prototype.remove = function(options, callback) ***REMOVED***
  if (typeof options === 'function') ***REMOVED***
    callback = options;
    options = null;
  ***REMOVED***

  registerRemoveListener(this);

  // If removing entire doc, no need to remove subdoc
  if (!options || !options.noop) ***REMOVED***
    this.$parent.set(this.$basePath, null);
  ***REMOVED***

  if (typeof callback === 'function') ***REMOVED***
    callback(null);
  ***REMOVED***
***REMOVED***;

/*!
 * ignore
 */

Subdocument.prototype.populate = function() ***REMOVED***
  throw new Error('Mongoose does not support calling populate() on nested ' +
    'docs. Instead of `doc.nested.populate("path")`, use ' +
    '`doc.populate("nested.path")`');
***REMOVED***;

/*!
 * Registers remove event listeners for triggering
 * on subdocuments.
 *
 * @param ***REMOVED***EmbeddedDocument***REMOVED*** sub
 * @api private
 */

function registerRemoveListener(sub) ***REMOVED***
  var owner = sub.ownerDocument();

  function emitRemove() ***REMOVED***
    owner.removeListener('save', emitRemove);
    owner.removeListener('remove', emitRemove);
    sub.emit('remove', sub);
    sub.constructor.emit('remove', sub);
    owner = sub = null;
  ***REMOVED***

  owner.on('save', emitRemove);
  owner.on('remove', emitRemove);
***REMOVED***
