// walk/lib/node-type-emitter.js
/*jshint strict:true node:true es5:true onevar:true laxcomma:true laxbreak:true*/
(function () ***REMOVED***
  "use strict";

  // "FIFO" isn't easy to convert to camelCase and back reliably
  var isFnodeTypes = [
      "isFile", "isDirectory",  "isSymbolicLink", "isBlockDevice",  "isCharacterDevice", "isFIFO", "isSocket"
    ],
    fnodeTypes = [
      "file",   "directory",    "symbolicLink",   "blockDevice",    "characterDevice",    "FIFO",   "socket"
    ],
    fnodeTypesPlural = [
      "files",  "directories",  "symbolicLinks",  "blockDevices",   "characterDevices",   "FIFOs",  "sockets"
    ];


  // 
  function createNodeGroups() ***REMOVED***
    var nodeGroups = ***REMOVED******REMOVED***;
    fnodeTypesPlural.concat("nodes", "errors").forEach(function (fnodeTypePlural) ***REMOVED***
      nodeGroups[fnodeTypePlural] = [];
    ***REMOVED***);
    return nodeGroups;
  ***REMOVED***


  // Determine each file node's type
  // 
  function sortFnodesByType(stat, fnodes) ***REMOVED***
    var i, isType;

    for (i = 0; i < isFnodeTypes.length; i += 1) ***REMOVED***
      isType = isFnodeTypes[i];
      if (stat[isType]()) ***REMOVED***
        stat.type = fnodeTypes[i];
        fnodes[fnodeTypesPlural[i]].push(stat);
        return;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***


  // Get the current number of listeners (which may change)
  // Emit events to each listener
  // Wait for all listeners to `next()` before continueing
  // (in theory this may avoid disk thrashing)
  function emitSingleEvents(emitter, path, stats, next, self) ***REMOVED***
    var num = 1 + emitter.listeners(stats.type).length + emitter.listeners("node").length;

    function nextWhenReady() ***REMOVED***
      num -= 1;
      if (0 === num) ***REMOVED*** next.call(self); ***REMOVED***
    ***REMOVED***

    emitter.emit(stats.type, path, stats, nextWhenReady);
    emitter.emit("node", path, stats, nextWhenReady);
    nextWhenReady();
  ***REMOVED***


  // Since the risk for disk thrashing among anything
  // other than files is relatively low, all types are
  // emitted at once, but all must complete before advancing
  function emitPluralEvents(emitter, path, nodes, next, self) ***REMOVED***
    var num = 1;

    function nextWhenReady() ***REMOVED***
      num -= 1;
      if (0 === num) ***REMOVED*** next.call(self); ***REMOVED***
    ***REMOVED***

    fnodeTypesPlural.concat(["nodes", "errors"]).forEach(function (fnodeType) ***REMOVED***
      if (0 === nodes[fnodeType].length) ***REMOVED*** return; ***REMOVED***
      num += emitter.listeners(fnodeType).length;
      emitter.emit(fnodeType, path, nodes[fnodeType], nextWhenReady);
    ***REMOVED***);
    nextWhenReady();
  ***REMOVED***

  module.exports = ***REMOVED***
    emitNodeType: emitSingleEvents,
    emitNodeTypeGroups: emitPluralEvents,
    isFnodeTypes: isFnodeTypes,
    fnodeTypes: fnodeTypes,
    fnodeTypesPlural: fnodeTypesPlural,
    sortFnodesByType: sortFnodesByType,
    createNodeGroups: createNodeGroups
  ***REMOVED***;
***REMOVED***());
